{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "06e5ee54-4659-4353-b6ad-de92dd9ab319",
   "metadata": {},
   "source": [
    "# Data download and processing from Swissprot\n",
    "\n",
    "Downloaded the data on 13th of May 2024. 571,282 results filtering for reviewed \"Swiss-Prot\". (https://www.uniprot.org/uniprotkb?query=*&facets=reviewed%3Atrue)\n",
    "\n",
    "Selected download TSV and the columns, Seqeunce (under Sequences tab), EC number (under Function).\n",
    "\n",
    "Unzipped the downloaded file.\n",
    "\n",
    "## 1. Protein to EC processing\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4f961afc-aedd-41b9-8635-4c6eabd017ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>269</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>796</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>3423</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>248</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>471</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571277</th>\n",
       "      <td>V9HVX0</td>\n",
       "      <td>YPAA_ECOLI</td>\n",
       "      <td>61</td>\n",
       "      <td>MTIAERLRQEGHQIGWQEGKLEGLHEQAIKIALRMLEQGFDRDQVL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571278</th>\n",
       "      <td>W5X2N3</td>\n",
       "      <td>AP_FRAAN</td>\n",
       "      <td>193</td>\n",
       "      <td>MGWVWKDDDEQGGHVNPSAADISPRLDGDRCSTRKVVRTQCKTEEV...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571279</th>\n",
       "      <td>W6JGV7</td>\n",
       "      <td>REP_CPBDV</td>\n",
       "      <td>484</td>\n",
       "      <td>MSAPLSIEQDDLLTDDLKSWLSDIDFSNDNEEAIEMEPSDIEMSSP...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571280</th>\n",
       "      <td>W6JHZ8</td>\n",
       "      <td>VP1_CPBDV</td>\n",
       "      <td>210</td>\n",
       "      <td>MNVKGASDKAQLAMQAEWEEVLAPEGALAVEEASSVLKISDDERRS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571281</th>\n",
       "      <td>W6JQK2</td>\n",
       "      <td>YBW7_SCHPO</td>\n",
       "      <td>268</td>\n",
       "      <td>MIVIEPTILYEAAALYTRYTANGGDSNDVVALLVGPIVQTLSIPSS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>571282 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9     269   \n",
       "1       A0A023I7E1   ENG1_RHIMI     796   \n",
       "2       A0A024B7W1   POLG_ZIKVF    3423   \n",
       "3       A0A024SC78  CUTI1_HYPJR     248   \n",
       "4       A0A024SH76   GUX2_HYPJR     471   \n",
       "...            ...          ...     ...   \n",
       "571277      V9HVX0   YPAA_ECOLI      61   \n",
       "571278      W5X2N3     AP_FRAAN     193   \n",
       "571279      W6JGV7    REP_CPBDV     484   \n",
       "571280      W6JHZ8    VP1_CPBDV     210   \n",
       "571281      W6JQK2   YBW7_SCHPO     268   \n",
       "\n",
       "                                                 Sequence  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...   \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   \n",
       "...                                                   ...   \n",
       "571277  MTIAERLRQEGHQIGWQEGKLEGLHEQAIKIALRMLEQGFDRDQVL...   \n",
       "571278  MGWVWKDDDEQGGHVNPSAADISPRLDGDRCSTRKVVRTQCKTEEV...   \n",
       "571279  MSAPLSIEQDDLLTDDLKSWLSDIDFSNDNEEAIEMEPSDIEMSSP...   \n",
       "571280  MNVKGASDKAQLAMQAEWEEVLAPEGALAVEEASSVLKISDDERRS...   \n",
       "571281  MIVIEPTILYEAAALYTRYTANGGDSNDVVALLVGPIVQTLSIPSS...   \n",
       "\n",
       "                                                EC number  \n",
       "0                                        3.2.2.-; 3.2.2.6  \n",
       "1                                                3.2.1.39  \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...  \n",
       "3                                                3.1.1.74  \n",
       "4                                                3.2.1.91  \n",
       "...                                                   ...  \n",
       "571277                                                NaN  \n",
       "571278                                                NaN  \n",
       "571279                                                NaN  \n",
       "571280                                                NaN  \n",
       "571281                                                NaN  \n",
       "\n",
       "[571282 rows x 5 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "swissprot = pd.read_csv('raw_data/uniprotkb_AND_reviewed_true_2024_05_13.tsv', sep='\\t')\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee69eeba-0d87-4f0e-92f8-8d950cc616f3",
   "metadata": {},
   "source": [
    "## Explode out EC number so each EC has it's own row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "724377c7-9690-4cb2-ad79-e081cc7f9fb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "275960\n",
      "275960\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_23877/1897238878.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  swissprot['EC All'] = swissprot['EC number'].values\n",
      "/tmp/ipykernel_23877/1897238878.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  swissprot['EC number'] = [ec.split(';') for ec in swissprot['EC number']]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18356\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>EC All</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>269</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>269</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>796</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "      <td>3.2.1.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>3423</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.1.1.56</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>3423</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.1.1.57</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570995</th>\n",
       "      <td>Q9Z7Q0</td>\n",
       "      <td>Y654_CHLPN</td>\n",
       "      <td>155</td>\n",
       "      <td>MLKKKPVSFSCIDGHIYKIFPNDLNANNTVFGGLLMSLLDRLALVV...</td>\n",
       "      <td>3.1.2.-</td>\n",
       "      <td>3.1.2.-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571093</th>\n",
       "      <td>Q9ZD84</td>\n",
       "      <td>Y459_RICPR</td>\n",
       "      <td>226</td>\n",
       "      <td>MIDKSSADILEISAKCGYLTGLLKNVYRSANIIVTDMSPLLLDSFD...</td>\n",
       "      <td>2.1.1.-</td>\n",
       "      <td>2.1.1.-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571104</th>\n",
       "      <td>Q9ZDD5</td>\n",
       "      <td>Y396_RICPR</td>\n",
       "      <td>208</td>\n",
       "      <td>MKKIIFCFLYLYTLNVFGASKIDHNKLNSCSITRNIFNNYEPKVFE...</td>\n",
       "      <td>1.13.11.-</td>\n",
       "      <td>1.13.11.-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571116</th>\n",
       "      <td>Q9ZDI9</td>\n",
       "      <td>Y339_RICPR</td>\n",
       "      <td>318</td>\n",
       "      <td>MKQNIYSPLVSIIIPVYNGANYMREAIDSALAQTYKNIEIIVVNDG...</td>\n",
       "      <td>2.4.-.-</td>\n",
       "      <td>2.4.-.-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571212</th>\n",
       "      <td>Q9ZLA7</td>\n",
       "      <td>Y736_HELPJ</td>\n",
       "      <td>369</td>\n",
       "      <td>MLLFTPGPVAINEEMRSSFSQPMPHHRTKDFEKIFQSVRENLKKMT...</td>\n",
       "      <td>2.6.1.-</td>\n",
       "      <td>2.6.1.-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>299631 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9     269   \n",
       "0       A0A009IHW8  ABTIR_ACIB9     269   \n",
       "1       A0A023I7E1   ENG1_RHIMI     796   \n",
       "2       A0A024B7W1   POLG_ZIKVF    3423   \n",
       "2       A0A024B7W1   POLG_ZIKVF    3423   \n",
       "...            ...          ...     ...   \n",
       "570995      Q9Z7Q0   Y654_CHLPN     155   \n",
       "571093      Q9ZD84   Y459_RICPR     226   \n",
       "571104      Q9ZDD5   Y396_RICPR     208   \n",
       "571116      Q9ZDI9   Y339_RICPR     318   \n",
       "571212      Q9ZLA7   Y736_HELPJ     369   \n",
       "\n",
       "                                                 Sequence  EC number  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.-   \n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.6   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...   3.2.1.39   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...   2.1.1.56   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...   2.1.1.57   \n",
       "...                                                   ...        ...   \n",
       "570995  MLKKKPVSFSCIDGHIYKIFPNDLNANNTVFGGLLMSLLDRLALVV...    3.1.2.-   \n",
       "571093  MIDKSSADILEISAKCGYLTGLLKNVYRSANIIVTDMSPLLLDSFD...    2.1.1.-   \n",
       "571104  MKKIIFCFLYLYTLNVFGASKIDHNKLNSCSITRNIFNNYEPKVFE...  1.13.11.-   \n",
       "571116  MKQNIYSPLVSIIIPVYNGANYMREAIDSALAQTYKNIEIIVVNDG...    2.4.-.-   \n",
       "571212  MLLFTPGPVAINEEMRSSFSQPMPHHRTKDFEKIFQSVRENLKKMT...    2.6.1.-   \n",
       "\n",
       "                                                   EC All  \n",
       "0                                        3.2.2.-; 3.2.2.6  \n",
       "0                                        3.2.2.-; 3.2.2.6  \n",
       "1                                                3.2.1.39  \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...  \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...  \n",
       "...                                                   ...  \n",
       "570995                                            3.1.2.-  \n",
       "571093                                            2.1.1.-  \n",
       "571104                                          1.13.11.-  \n",
       "571116                                            2.4.-.-  \n",
       "571212                                            2.6.1.-  \n",
       "\n",
       "[299631 rows x 6 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop rows that don't have an ec number\n",
    "swissprot = swissprot[swissprot['EC number'].notna()]\n",
    "print(len(swissprot))\n",
    "swissprot['EC All'] = swissprot['EC number'].values\n",
    "print(len(swissprot))\n",
    "# Now expand out the ones we have left\n",
    "swissprot['EC number'] = [ec.split(';') for ec in swissprot['EC number']]\n",
    "# Then check how many have mulitple\n",
    "print(sum([1 if len(ec) > 1 else 0 for ec in swissprot['EC number'].values]))\n",
    "swissprot = swissprot.explode('EC number')\n",
    "# Clean the EC numbers\n",
    "swissprot['EC number'] = [ec.replace(' ', '') for ec in swissprot['EC number']]\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4b81d7a",
   "metadata": {},
   "source": [
    "## Filtering steps \n",
    "\n",
    "* Filter to a certain sequence length\n",
    "* Drop duplicates\n",
    "* Remove EC numbers with dash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e42c854f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Partially filled EC numbers:  253572\n",
      "215372\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>EC All</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>269</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>796</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "      <td>3.2.1.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>3423</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.1.1.56</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>3423</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.1.1.57</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>3423</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564731</th>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>Y3032_BACAN</td>\n",
       "      <td>584</td>\n",
       "      <td>MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>3.5.4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564754</th>\n",
       "      <td>Q6L032</td>\n",
       "      <td>Y1085_PICTO</td>\n",
       "      <td>573</td>\n",
       "      <td>MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>3.5.4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565753</th>\n",
       "      <td>Q7TDB6</td>\n",
       "      <td>RDRP_MYRV9</td>\n",
       "      <td>1354</td>\n",
       "      <td>MSLTSRYTHFVPDSTITEILNDSNTPQILLHYANIVNGSTPVHFTS...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>2.7.7.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568711</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>172</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>3.6.1.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569733</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>124</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>3.1.1.29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>215372 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9     269   \n",
       "1       A0A023I7E1   ENG1_RHIMI     796   \n",
       "2       A0A024B7W1   POLG_ZIKVF    3423   \n",
       "2       A0A024B7W1   POLG_ZIKVF    3423   \n",
       "2       A0A024B7W1   POLG_ZIKVF    3423   \n",
       "...            ...          ...     ...   \n",
       "564731      Q6HX62  Y3032_BACAN     584   \n",
       "564754      Q6L032  Y1085_PICTO     573   \n",
       "565753      Q7TDB6   RDRP_MYRV9    1354   \n",
       "568711      Q94MV8   VG56_BPLZ5     172   \n",
       "569733      Q9J5H2   V029_FOWPN     124   \n",
       "\n",
       "                                                 Sequence EC number  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.6   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...  3.2.1.39   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...  2.1.1.56   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...  2.1.1.57   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...  2.7.7.48   \n",
       "...                                                   ...       ...   \n",
       "564731  MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...   3.5.4.2   \n",
       "564754  MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...   3.5.4.2   \n",
       "565753  MSLTSRYTHFVPDSTITEILNDSNTPQILLHYANIVNGSTPVHFTS...  2.7.7.48   \n",
       "568711  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...  3.6.1.12   \n",
       "569733  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...  3.1.1.29   \n",
       "\n",
       "                                                   EC All  \n",
       "0                                        3.2.2.-; 3.2.2.6  \n",
       "1                                                3.2.1.39  \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...  \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...  \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...  \n",
       "...                                                   ...  \n",
       "564731                                            3.5.4.2  \n",
       "564754                                            3.5.4.2  \n",
       "565753                                           2.7.7.48  \n",
       "568711                                           3.6.1.12  \n",
       "569733                                           3.1.1.29  \n",
       "\n",
       "[215372 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "swissprot = swissprot[~swissprot['EC number'].str.contains('-')]\n",
    "print('Partially filled EC numbers: ', len(swissprot))\n",
    "swissprot = swissprot.drop_duplicates(subset=['Sequence', 'EC number'])\n",
    "print(len(swissprot))\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4607689a",
   "metadata": {},
   "source": [
    "### Comparison to the dataset used in CLEAN\n",
    "\n",
    "CLEAN: https://www.science.org/doi/10.1126/science.adf2465 \n",
    "\n",
    "\n",
    "Dataset was downloaded from the CLEA repository: https://github.com/tttianhao/CLEAN/tree/main/app/data specifically the split 100 dataset, alternatively they can be downloaded from: https://drive.google.com/file/d/1MzYQ--CRK253eAKOmZrD1nzF-mXHfXtq/view\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9d4113b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P51451</td>\n",
       "      <td>2.7.10.2</td>\n",
       "      <td>MGLVSSKKPDKEKPIKEKDKGQWSPLKVSAQDKDAPPLPPLVVFNH...</td>\n",
       "      <td>505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C7C422</td>\n",
       "      <td>3.5.2.6</td>\n",
       "      <td>MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...</td>\n",
       "      <td>270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q07457</td>\n",
       "      <td>2.3.2.27</td>\n",
       "      <td>MTAEPATKKIKLELSDPSEPLTQSDVIAFQKEALFRCINRRRVDFE...</td>\n",
       "      <td>700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q07009</td>\n",
       "      <td>3.4.22.53</td>\n",
       "      <td>MAGIAMKLAKDREAAEGLGSHERAIKYLNQDYETLRNECLEAGALF...</td>\n",
       "      <td>700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q9TTH8</td>\n",
       "      <td>3.4.22.54</td>\n",
       "      <td>MPTVISASVAPRTGAEPRSPGPIAQAAQGKGTEAGGGNPSGIYSAI...</td>\n",
       "      <td>822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227357</th>\n",
       "      <td>Q97HD1</td>\n",
       "      <td>3.1.11.6</td>\n",
       "      <td>MPSKKESYESMIKELEKIVSSMENEELPLEEAMKNYEDGVKLCDKL...</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227358</th>\n",
       "      <td>A0A140CUL9</td>\n",
       "      <td>1.1.99.13</td>\n",
       "      <td>MAENNTYDAIVIGSGISGGWATKELTEKGLKTLMLERGRNIEHVKD...</td>\n",
       "      <td>562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227359</th>\n",
       "      <td>W6JQJ6</td>\n",
       "      <td>1.13.12.2</td>\n",
       "      <td>MNKNNRHPADGKKPITIFGPDFPFAFDDWLEHPAGLGSIPAARHGE...</td>\n",
       "      <td>560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227360</th>\n",
       "      <td>B3IVI7</td>\n",
       "      <td>3.5.1.30</td>\n",
       "      <td>MRIALYQGAPKPLDVPGNLQRLRHQAQLAAERGAQLLVCPEMFLTG...</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227361</th>\n",
       "      <td>Q88QV2</td>\n",
       "      <td>3.5.1.30</td>\n",
       "      <td>MRIALYQGAPKPLDVPGNLQRLRHQAQLAADRGAQLLVCPEMFLSG...</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>191488 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry  EC number  \\\n",
       "0           P51451   2.7.10.2   \n",
       "1           C7C422    3.5.2.6   \n",
       "2           Q07457   2.3.2.27   \n",
       "3           Q07009  3.4.22.53   \n",
       "4           Q9TTH8  3.4.22.54   \n",
       "...            ...        ...   \n",
       "227357      Q97HD1   3.1.11.6   \n",
       "227358  A0A140CUL9  1.1.99.13   \n",
       "227359      W6JQJ6  1.13.12.2   \n",
       "227360      B3IVI7   3.5.1.30   \n",
       "227361      Q88QV2   3.5.1.30   \n",
       "\n",
       "                                                 Sequence  Length  \n",
       "0       MGLVSSKKPDKEKPIKEKDKGQWSPLKVSAQDKDAPPLPPLVVFNH...     505  \n",
       "1       MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...     270  \n",
       "2       MTAEPATKKIKLELSDPSEPLTQSDVIAFQKEALFRCINRRRVDFE...     700  \n",
       "3       MAGIAMKLAKDREAAEGLGSHERAIKYLNQDYETLRNECLEAGALF...     700  \n",
       "4       MPTVISASVAPRTGAEPRSPGPIAQAAQGKGTEAGGGNPSGIYSAI...     822  \n",
       "...                                                   ...     ...  \n",
       "227357  MPSKKESYESMIKELEKIVSSMENEELPLEEAMKNYEDGVKLCDKL...      77  \n",
       "227358  MAENNTYDAIVIGSGISGGWATKELTEKGLKTLMLERGRNIEHVKD...     562  \n",
       "227359  MNKNNRHPADGKKPITIFGPDFPFAFDDWLEHPAGLGSIPAARHGE...     560  \n",
       "227360  MRIALYQGAPKPLDVPGNLQRLRHQAQLAAERGAQLLVCPEMFLTG...     264  \n",
       "227361  MRIALYQGAPKPLDVPGNLQRLRHQAQLAADRGAQLLVCPEMFLSG...     264  \n",
       "\n",
       "[191488 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_df = pd.read_csv('raw_data/CLEAN_split100.tsv', sep='\\t')\n",
    "clean_df = clean_df.drop_duplicates(subset=['Sequence', 'EC number'])\n",
    "clean_df['Length'] = [len(seq) for seq in clean_df['Sequence']]\n",
    "clean_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8d17ad2",
   "metadata": {},
   "source": [
    "## 2. Reaction to EC processing\n",
    "\n",
    "EC and reaction data were downloaded from multiple sources, the first being ECReact.\n",
    "\n",
    "1. ECReact: https://www.nature.com/articles/s41467-022-28536-w, with the data downloaded from https://github.com/rxn4chemistry/biocatalysis-model specifically, https://github.com/rxn4chemistry/biocatalysis-model/blob/main/data/ecreact-1.0.csv\n",
    "2. EnzymeMap: https://chemrxiv.org/engage/chemrxiv/article-details/643e9f4034f4ff2ea4f4f84e https://zenodo.org/records/8254726 and the data https://github.com/hesther/enzymemap, using the data file https://github.com/hesther/enzymemap/blob/main/data/processed_reactions.csv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b6d71562",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>1.2.1.8</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C=C(C)CCOP(=O)([O-])OP(=O)([O-])[O-].CC(C)=CCO...</td>\n",
       "      <td>2.5.1.85</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O...</td>\n",
       "      <td>1.4.1.4</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>1.1.1.3</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cc1cn([C@H]2C[C@H](O)[C@@H](COP(=O)(O)OP(=O)(O...</td>\n",
       "      <td>2.7.7.9</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62217</th>\n",
       "      <td>CCCCC/C=C\\C/C=C\\C/C=C\\C/C=C\\C/C=C\\CCC(=O)O[C@@...</td>\n",
       "      <td>2.3.1.76</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62218</th>\n",
       "      <td>CCCCCCCCCCCC/C=C\\CCCCCCCC(=O)OC[C@H](COP(=O)(O...</td>\n",
       "      <td>2.7.8.41</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62219</th>\n",
       "      <td>CCCCCCCCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCCCCCCC...</td>\n",
       "      <td>2.7.8.41</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62220</th>\n",
       "      <td>CCCCCC/C=C\\CCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCC...</td>\n",
       "      <td>2.7.8.41</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62221</th>\n",
       "      <td>CCCCCCCC/C=C\\CCCCCCCC(=O)O[C@H](COC(=O)CCCCCCC...</td>\n",
       "      <td>4.1.1.65</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62222 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Reaction EC number  \\\n",
       "0      NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   1.2.1.8   \n",
       "1      C=C(C)CCOP(=O)([O-])OP(=O)([O-])[O-].CC(C)=CCO...  2.5.1.85   \n",
       "2      N.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O...   1.4.1.4   \n",
       "3      NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...   1.1.1.3   \n",
       "4      Cc1cn([C@H]2C[C@H](O)[C@@H](COP(=O)(O)OP(=O)(O...   2.7.7.9   \n",
       "...                                                  ...       ...   \n",
       "62217  CCCCC/C=C\\C/C=C\\C/C=C\\C/C=C\\C/C=C\\CCC(=O)O[C@@...  2.3.1.76   \n",
       "62218  CCCCCCCCCCCC/C=C\\CCCCCCCC(=O)OC[C@H](COP(=O)(O...  2.7.8.41   \n",
       "62219  CCCCCCCCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCCCCCCC...  2.7.8.41   \n",
       "62220  CCCCCC/C=C\\CCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCC...  2.7.8.41   \n",
       "62221  CCCCCCCC/C=C\\CCCCCCCC(=O)O[C@H](COC(=O)CCCCCCC...  4.1.1.65   \n",
       "\n",
       "                         Source  \n",
       "0        brenda_reaction_smiles  \n",
       "1        brenda_reaction_smiles  \n",
       "2        brenda_reaction_smiles  \n",
       "3        brenda_reaction_smiles  \n",
       "4        brenda_reaction_smiles  \n",
       "...                         ...  \n",
       "62217  pathbank_reaction_smiles  \n",
       "62218  pathbank_reaction_smiles  \n",
       "62219  pathbank_reaction_smiles  \n",
       "62220  pathbank_reaction_smiles  \n",
       "62221  pathbank_reaction_smiles  \n",
       "\n",
       "[62222 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ECreact = pd.read_csv('raw_data/ECReact.csv')\n",
    "ECreact['rxn_smiles'] = ECreact['rxn_smiles'].str.split('|').str[0] + '>>' + ECreact['rxn_smiles'].str.split('>>').str[1]\n",
    "ECreact.rename(columns={'rxn_smiles': 'Reaction', 'ec': 'EC number', 'source':'Source'}, inplace=True)\n",
    "ECreact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "21ac8cce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "349458\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...</td>\n",
       "      <td>[CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349453</th>\n",
       "      <td>CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + dimethylallyl diphosphate = dimethylally...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349454</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl diphosphate = geranyl-triphospho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349455</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl triphosphate = geranyl-tetraphos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349456</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl diphosphate = adenosine 5'-i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349457</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl triphosphate = isopentenyl-t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62896 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Reaction  \\\n",
       "0       CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1       CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...   \n",
       "25      CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...   \n",
       "26      CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...   \n",
       "27      NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                   ...   \n",
       "349453  CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349454  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...   \n",
       "349455  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...   \n",
       "349456  C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349457  C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...   \n",
       "\n",
       "                                          Mapped Reaction EC number  \\\n",
       "0       [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1       [CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...   1.1.1.1   \n",
       "25      [CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...   1.1.1.1   \n",
       "26      [CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...   1.1.1.1   \n",
       "27      [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                   ...       ...   \n",
       "349453  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...   6.5.1.3   \n",
       "349454  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349455  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349456  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "349457  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "\n",
       "                                            Reaction Text  \n",
       "0          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "25          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "26          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "27      2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...  \n",
       "...                                                   ...  \n",
       "349453  ATP + dimethylallyl diphosphate = dimethylally...  \n",
       "349454  ATP + geranyl diphosphate = geranyl-triphospho...  \n",
       "349455  ATP + geranyl triphosphate = geranyl-tetraphos...  \n",
       "349456  ATP + isopentenyl diphosphate = adenosine 5'-i...  \n",
       "349457  ATP + isopentenyl triphosphate = isopentenyl-t...  \n",
       "\n",
       "[62896 rows x 4 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enzymemap = pd.read_csv('raw_data/EnzymeMap.csv')\n",
    "print(len(enzymemap))\n",
    "enzymemap.rename(columns={'ec_num': 'EC number', 'unmapped': 'Reaction', 'mapped': 'Mapped Reaction', 'orig_rxn_text':'Reaction Text'}, inplace=True)\n",
    "enzymemap = enzymemap[['Reaction', 'Mapped Reaction', 'EC number', 'Reaction Text']]\n",
    "enzymemap.drop_duplicates(subset=['Reaction', 'EC number'], inplace=True)\n",
    "enzymemap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a24320d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...</td>\n",
       "      <td>[CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349453</th>\n",
       "      <td>CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + dimethylallyl diphosphate = dimethylally...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349454</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl diphosphate = geranyl-triphospho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349455</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl triphosphate = geranyl-tetraphos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349456</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl diphosphate = adenosine 5'-i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349457</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl triphosphate = isopentenyl-t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62896 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Reaction  \\\n",
       "0       CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1       CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...   \n",
       "25      CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...   \n",
       "26      CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...   \n",
       "27      NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                   ...   \n",
       "349453  CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349454  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...   \n",
       "349455  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...   \n",
       "349456  C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349457  C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...   \n",
       "\n",
       "                                          Mapped Reaction EC number  \\\n",
       "0       [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1       [CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...   1.1.1.1   \n",
       "25      [CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...   1.1.1.1   \n",
       "26      [CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...   1.1.1.1   \n",
       "27      [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                   ...       ...   \n",
       "349453  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...   6.5.1.3   \n",
       "349454  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349455  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349456  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "349457  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "\n",
       "                                            Reaction Text  \n",
       "0          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "25          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "26          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "27      2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...  \n",
       "...                                                   ...  \n",
       "349453  ATP + dimethylallyl diphosphate = dimethylally...  \n",
       "349454  ATP + geranyl diphosphate = geranyl-triphospho...  \n",
       "349455  ATP + geranyl triphosphate = geranyl-tetraphos...  \n",
       "349456  ATP + isopentenyl diphosphate = adenosine 5'-i...  \n",
       "349457  ATP + isopentenyl triphosphate = isopentenyl-t...  \n",
       "\n",
       "[62896 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enzymemap.dropna(subset=['Mapped Reaction'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b244d31b",
   "metadata": {},
   "source": [
    "### Append EC numbers that are covered by ECreact but not by EnzymeMap, to EnzymeMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "94ce2b25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2353"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#find the EC numbers covered by ECreact but not by enzymemap\n",
    "ECreact = ECreact[~ECreact['EC number'].str.contains('-')]\n",
    "not_covered = ECreact[~ECreact['EC number'].isin(enzymemap['EC number'].unique())]\n",
    "not_covered['EC number'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "89e5d4dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1172</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1173</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1176</th>\n",
       "      <td>NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15228</th>\n",
       "      <td>Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.j</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21396</th>\n",
       "      <td>C[C@@]1(O)OC[C@H](O)C1(O)O.Nc1ncnc2c1ncn2[C@@H...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.l</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34012</th>\n",
       "      <td>*CS(=O)(=O)[O-].Nc1ncnc2c1ncn2[C@@H]1O[C@H](CO...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.o</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15541</th>\n",
       "      <td>Cc1ncc(C[n+]2csc(CCO)c2C)c(N)n1.Nc1ncnc2c1ncn2...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.p</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33659</th>\n",
       "      <td>Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.q</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72490 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Reaction  \\\n",
       "0      CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1172   CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...   \n",
       "1173   CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...   \n",
       "1176   NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...   \n",
       "1177   NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                  ...   \n",
       "15228  Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...   \n",
       "21396  C[C@@]1(O)OC[C@H](O)C1(O)O.Nc1ncnc2c1ncn2[C@@H...   \n",
       "34012  *CS(=O)(=O)[O-].Nc1ncnc2c1ncn2[C@@H]1O[C@H](CO...   \n",
       "15541  Cc1ncc(C[n+]2csc(CCO)c2C)c(N)n1.Nc1ncnc2c1ncn2...   \n",
       "33659  Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...   \n",
       "\n",
       "                                         Mapped Reaction EC number  \\\n",
       "0      [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1172   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1173   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1176   [H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...   1.1.1.1   \n",
       "1177   [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                  ...       ...   \n",
       "15228                                                NaN   7.6.2.j   \n",
       "21396                                                NaN   7.6.2.l   \n",
       "34012                                                NaN   7.6.2.o   \n",
       "15541                                                NaN   7.6.2.p   \n",
       "33659                                                NaN   7.6.2.q   \n",
       "\n",
       "                                           Reaction Text  \n",
       "0         acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1172   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1173   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1176   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "1177   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "...                                                  ...  \n",
       "15228                                                NaN  \n",
       "21396                                                NaN  \n",
       "34012                                                NaN  \n",
       "15541                                                NaN  \n",
       "33659                                                NaN  \n",
       "\n",
       "[72490 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reaction2EC = pd.concat((enzymemap, not_covered[['Reaction', 'EC number']])).sort_values(by='EC number')\n",
    "reaction2EC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac489eea",
   "metadata": {},
   "source": [
    "## 3. Filter protein2EC and reaction2EC to intersection of their EC numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "83eb77eb-dcb7-423d-ad4d-b0246c207fd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>EC All</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A0A061I403</td>\n",
       "      <td>FICD_CRIGR</td>\n",
       "      <td>455</td>\n",
       "      <td>MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...</td>\n",
       "      <td>2.7.7.108</td>\n",
       "      <td>2.7.7.108; 3.1.4.-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A0A067XR63</td>\n",
       "      <td>XTH7_DIOKA</td>\n",
       "      <td>268</td>\n",
       "      <td>MNAEGGNLHREFEITWGDGRARIHNNGGLLTLSLDRASGSGFRSKN...</td>\n",
       "      <td>2.4.1.207</td>\n",
       "      <td>2.4.1.207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A0A067XRK9</td>\n",
       "      <td>XTH6_DIOKA</td>\n",
       "      <td>299</td>\n",
       "      <td>MASSLTLPMAMAFTLLALSFASAMGGSMNSSRFDELFQPSWAFDHF...</td>\n",
       "      <td>2.4.1.207</td>\n",
       "      <td>2.4.1.207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A0A067YMX8</td>\n",
       "      <td>XTH8_DIOKA</td>\n",
       "      <td>288</td>\n",
       "      <td>MAASPYSIFAVQLLLLASWMLSSSSSNFNQDFNIAWGGGRARILNN...</td>\n",
       "      <td>2.4.1.207</td>\n",
       "      <td>2.4.1.207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>A0A0B7P3V8</td>\n",
       "      <td>YP41B_YEAST</td>\n",
       "      <td>1104</td>\n",
       "      <td>MATPVRDETRNVIDDNISARIQSKVKTNDTVRQTPSSLRKVSIKDE...</td>\n",
       "      <td>3.1.26.4</td>\n",
       "      <td>2.7.7.49; 2.7.7.7; 3.1.26.4; 3.4.23.-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534459</th>\n",
       "      <td>Q9ZNA5</td>\n",
       "      <td>SAHH_ROSDO</td>\n",
       "      <td>462</td>\n",
       "      <td>MTKDFIVKDIALAEFGRKELDIAETEMPGLMALRAEYGDSKPLAGA...</td>\n",
       "      <td>3.13.2.1</td>\n",
       "      <td>3.13.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534465</th>\n",
       "      <td>Q9ZNJ9</td>\n",
       "      <td>MSRB_HATHI</td>\n",
       "      <td>159</td>\n",
       "      <td>MNHKKEKEYKKLDTEKLKENLTELQYNVTQRNATEKPFLNKYDKHF...</td>\n",
       "      <td>1.8.4.12</td>\n",
       "      <td>1.8.4.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534723</th>\n",
       "      <td>V5TD18</td>\n",
       "      <td>AMBI3_FISAU</td>\n",
       "      <td>273</td>\n",
       "      <td>MIVSTSVEQSAQFSVKSLTPFGALLEATEDHSDIQQLSIEQLCQLT...</td>\n",
       "      <td>1.14.20.11</td>\n",
       "      <td>1.14.20.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534724</th>\n",
       "      <td>V5TES5</td>\n",
       "      <td>AMBI2_FISAU</td>\n",
       "      <td>330</td>\n",
       "      <td>MTQIINITQSKVISEQILRHVFRHRRLISDTEPCVHQPCSLCLAPH...</td>\n",
       "      <td>4.1.99.25</td>\n",
       "      <td>4.1.99.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534725</th>\n",
       "      <td>V5TF65</td>\n",
       "      <td>AMBI1_FISAU</td>\n",
       "      <td>319</td>\n",
       "      <td>MISEKILRHIFQYRRLLSDTEPCAKEPCSICLAPHLPKIQSFIENN...</td>\n",
       "      <td>4.1.99.25</td>\n",
       "      <td>4.1.99.25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15713 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "11      A0A061I403   FICD_CRIGR     455   \n",
       "15      A0A067XR63   XTH7_DIOKA     268   \n",
       "16      A0A067XRK9   XTH6_DIOKA     299   \n",
       "17      A0A067YMX8   XTH8_DIOKA     288   \n",
       "86      A0A0B7P3V8  YP41B_YEAST    1104   \n",
       "...            ...          ...     ...   \n",
       "534459      Q9ZNA5   SAHH_ROSDO     462   \n",
       "534465      Q9ZNJ9   MSRB_HATHI     159   \n",
       "534723      V5TD18  AMBI3_FISAU     273   \n",
       "534724      V5TES5  AMBI2_FISAU     330   \n",
       "534725      V5TF65  AMBI1_FISAU     319   \n",
       "\n",
       "                                                 Sequence   EC number  \\\n",
       "11      MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...   2.7.7.108   \n",
       "15      MNAEGGNLHREFEITWGDGRARIHNNGGLLTLSLDRASGSGFRSKN...   2.4.1.207   \n",
       "16      MASSLTLPMAMAFTLLALSFASAMGGSMNSSRFDELFQPSWAFDHF...   2.4.1.207   \n",
       "17      MAASPYSIFAVQLLLLASWMLSSSSSNFNQDFNIAWGGGRARILNN...   2.4.1.207   \n",
       "86      MATPVRDETRNVIDDNISARIQSKVKTNDTVRQTPSSLRKVSIKDE...    3.1.26.4   \n",
       "...                                                   ...         ...   \n",
       "534459  MTKDFIVKDIALAEFGRKELDIAETEMPGLMALRAEYGDSKPLAGA...    3.13.2.1   \n",
       "534465  MNHKKEKEYKKLDTEKLKENLTELQYNVTQRNATEKPFLNKYDKHF...    1.8.4.12   \n",
       "534723  MIVSTSVEQSAQFSVKSLTPFGALLEATEDHSDIQQLSIEQLCQLT...  1.14.20.11   \n",
       "534724  MTQIINITQSKVISEQILRHVFRHRRLISDTEPCVHQPCSLCLAPH...   4.1.99.25   \n",
       "534725  MISEKILRHIFQYRRLLSDTEPCAKEPCSICLAPHLPKIQSFIENN...   4.1.99.25   \n",
       "\n",
       "                                       EC All  \n",
       "11                         2.7.7.108; 3.1.4.-  \n",
       "15                                  2.4.1.207  \n",
       "16                                  2.4.1.207  \n",
       "17                                  2.4.1.207  \n",
       "86      2.7.7.49; 2.7.7.7; 3.1.26.4; 3.4.23.-  \n",
       "...                                       ...  \n",
       "534459                               3.13.2.1  \n",
       "534465                               1.8.4.12  \n",
       "534723                             1.14.20.11  \n",
       "534724                              4.1.99.25  \n",
       "534725                              4.1.99.25  \n",
       "\n",
       "[15713 rows x 6 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ns_swissprot = swissprot[~swissprot['EC number'].isin(reaction2EC['EC number'].unique())]\n",
    "ns_swissprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8f2dc451-4b77-4da7-99dd-3d69bbbd964d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "456"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ns_swissprot['EC number'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "624fb47e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>EC All</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>269</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>796</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "      <td>3.2.1.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>3423</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.1.1.56</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>3423</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.1.1.57</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>3423</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564731</th>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>Y3032_BACAN</td>\n",
       "      <td>584</td>\n",
       "      <td>MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>3.5.4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564754</th>\n",
       "      <td>Q6L032</td>\n",
       "      <td>Y1085_PICTO</td>\n",
       "      <td>573</td>\n",
       "      <td>MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>3.5.4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565753</th>\n",
       "      <td>Q7TDB6</td>\n",
       "      <td>RDRP_MYRV9</td>\n",
       "      <td>1354</td>\n",
       "      <td>MSLTSRYTHFVPDSTITEILNDSNTPQILLHYANIVNGSTPVHFTS...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>2.7.7.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568711</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>172</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>3.6.1.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569733</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>124</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>3.1.1.29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>199659 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9     269   \n",
       "1       A0A023I7E1   ENG1_RHIMI     796   \n",
       "2       A0A024B7W1   POLG_ZIKVF    3423   \n",
       "2       A0A024B7W1   POLG_ZIKVF    3423   \n",
       "2       A0A024B7W1   POLG_ZIKVF    3423   \n",
       "...            ...          ...     ...   \n",
       "564731      Q6HX62  Y3032_BACAN     584   \n",
       "564754      Q6L032  Y1085_PICTO     573   \n",
       "565753      Q7TDB6   RDRP_MYRV9    1354   \n",
       "568711      Q94MV8   VG56_BPLZ5     172   \n",
       "569733      Q9J5H2   V029_FOWPN     124   \n",
       "\n",
       "                                                 Sequence EC number  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.6   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...  3.2.1.39   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...  2.1.1.56   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...  2.1.1.57   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...  2.7.7.48   \n",
       "...                                                   ...       ...   \n",
       "564731  MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...   3.5.4.2   \n",
       "564754  MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...   3.5.4.2   \n",
       "565753  MSLTSRYTHFVPDSTITEILNDSNTPQILLHYANIVNGSTPVHFTS...  2.7.7.48   \n",
       "568711  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...  3.6.1.12   \n",
       "569733  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...  3.1.1.29   \n",
       "\n",
       "                                                   EC All  \n",
       "0                                        3.2.2.-; 3.2.2.6  \n",
       "1                                                3.2.1.39  \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...  \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...  \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...  \n",
       "...                                                   ...  \n",
       "564731                                            3.5.4.2  \n",
       "564754                                            3.5.4.2  \n",
       "565753                                           2.7.7.48  \n",
       "568711                                           3.6.1.12  \n",
       "569733                                           3.1.1.29  \n",
       "\n",
       "[199659 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "swissprot = swissprot[swissprot['EC number'].isin(reaction2EC['EC number'].unique())]\n",
    "swissprot # 149788"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b15e8b1d-d035-49a9-a524-d79453b5609e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5088"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(swissprot['EC number'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "443b6cdb-fe81-40ba-86a6-a52fe0103c29",
   "metadata": {},
   "source": [
    "## Save unique proteins\n",
    "\n",
    "#save the indices corresponding to unique proteins, for downstream use\n",
    "unique_protein_indices = swissprot.reset_index().drop_duplicates(subset=['Sequence']).index\n",
    "with open('unique_protein_indices.txt', 'w') as f:\n",
    "    for index in unique_protein_indices:\n",
    "        f.write(str(index) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "407b09c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1172</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1173</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1176</th>\n",
       "      <td>NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34226</th>\n",
       "      <td>CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@@H]8[C@H...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34223</th>\n",
       "      <td>CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@H]([C@H]...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15471</th>\n",
       "      <td>C[N+]1(C)CCC[C@H]1C(=O)[O-].Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)(O)OP(=O)(O...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8665</th>\n",
       "      <td>*[N+](*)(*)*.Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63735 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Reaction  \\\n",
       "0      CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1172   CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...   \n",
       "1173   CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...   \n",
       "1176   NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...   \n",
       "1177   NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                  ...   \n",
       "34226  CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@@H]8[C@H...   \n",
       "34223  CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@H]([C@H]...   \n",
       "15471  C[N+]1(C)CCC[C@H]1C(=O)[O-].Nc1ncnc2c1ncn2[C@@...   \n",
       "644    Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)(O)OP(=O)(O...   \n",
       "8665   *[N+](*)(*)*.Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=...   \n",
       "\n",
       "                                         Mapped Reaction EC number  \\\n",
       "0      [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1172   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1173   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1176   [H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...   1.1.1.1   \n",
       "1177   [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                  ...       ...   \n",
       "34226                                                NaN   7.6.2.8   \n",
       "34223                                                NaN   7.6.2.8   \n",
       "15471                                                NaN   7.6.2.9   \n",
       "644                                                  NaN   7.6.2.9   \n",
       "8665                                                 NaN   7.6.2.9   \n",
       "\n",
       "                                           Reaction Text  \n",
       "0         acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1172   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1173   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1176   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "1177   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "...                                                  ...  \n",
       "34226                                                NaN  \n",
       "34223                                                NaN  \n",
       "15471                                                NaN  \n",
       "644                                                  NaN  \n",
       "8665                                                 NaN  \n",
       "\n",
       "[63735 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reaction2EC = reaction2EC[reaction2EC['EC number'].isin(swissprot['EC number'].unique())]\n",
    "reaction2EC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5721bdf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "reaction2EC.to_csv('../processed_data/reaction2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5250f2c8",
   "metadata": {},
   "source": [
    "## 4. Perform protein clustering using mmseqs\n",
    "\n",
    "Use mmseqs to identify clusters, https://github.com/soedinglab/MMseqs2\n",
    "\n",
    "Optional parameters:\n",
    "\n",
    "```\n",
    " --seed-sub-mat TWIN              Substitution matrix file for k-mer generation [aa:VTML80.out,nucl:nucleotide.out]\n",
    " -s FLOAT                         Sensitivity: 1.0 faster; 4.0 fast; 7.5 sensitive [4.000]\n",
    " -k INT                           k-mer length (0: automatically set to optimum) [0]\n",
    " --target-search-mode INT         target search mode (0: regular k-mer, 1: similar k-mer) [0]\n",
    " --k-score TWIN                   k-mer threshold for generating similar k-mer lists [seq:2147483647,prof:2147483647]\n",
    " --alph-size TWIN                 Alphabet size (range 2-21) [aa:21,nucl:5]\n",
    " --max-seqs INT                   Maximum results per query sequence allowed to pass the prefilter (affects sensitivity) [20]\n",
    " --split INT                      Split input into N equally distributed chunks. 0: set the best split automatically [0]\n",
    " --split-mode INT                 0: split target db; 1: split query db; 2: auto, depending on main memory [2]\n",
    " --split-memory-limit BYTE        Set max memory per split. E.g. 800B, 5K, 10M, 1G. Default (0) to all available system memory [0]\n",
    " --comp-bias-corr INT             Correct for locally biased amino acid composition (range 0-1) [1]\n",
    " --comp-bias-corr-scale FLOAT     Correct for locally biased amino acid composition (range 0-1) [1.000]\n",
    " --diag-score BOOL                Use ungapped diagonal scoring during prefilter [1]\n",
    " --exact-kmer-matching INT        Extract only exact k-mers for matching (range 0-1) [0]\n",
    " --mask INT                       Mask sequences in k-mer stage: 0: w/o low complexity masking, 1: with low complexity masking [1]\n",
    " --mask-prob FLOAT                Mask sequences is probablity is above threshold [0.900]\n",
    " --mask-lower-case INT            Lowercase letters will be excluded from k-mer search 0: include region, 1: exclude region [0]\n",
    " --min-ungapped-score INT         Accept only matches with ungapped alignment score above threshold [15]\n",
    " --add-self-matches BOOL          Artificially add entries of queries with themselves (for clustering) [0]\n",
    " --spaced-kmer-mode INT           0: use consecutive positions in k-mers; 1: use spaced k-mers [1]\n",
    " --spaced-kmer-pattern STR        User-specified spaced k-mer pattern []\n",
    " --local-tmp STR                  Path where some of the temporary files will be created []\n",
    "align:\n",
    " -c FLOAT                         List matches above this fraction of aligned (covered) residues (see --cov-mode) [0.800]\n",
    " --cov-mode INT                   0: coverage of query and target\n",
    "                                  1: coverage of target\n",
    "                                  2: coverage of query\n",
    "                                  3: target seq. length has to be at least x% of query length\n",
    "                                  4: query seq. length has to be at least x% of target length\n",
    "                                  5: short seq. needs to be at least x% of the other seq. length [0]\n",
    " -a BOOL                          Add backtrace string (convert to alignments with mmseqs convertalis module) [0]\n",
    " --alignment-mode INT             How to compute the alignment:\n",
    "                                  0: automatic\n",
    "                                  1: only score and end_pos\n",
    "                                  2: also start_pos and cov\n",
    "                                  3: also seq.id\n",
    "                                  4: only ungapped alignment [3]\n",
    " --alignment-output-mode INT      How to compute the alignment:\n",
    "                                  0: automatic\n",
    "                                  1: only score and end_pos\n",
    "                                  2: also start_pos and cov\n",
    "                                  3: also seq.id\n",
    "                                  4: only ungapped alignment\n",
    "                                  5: score only (output) cluster format [0]\n",
    " --wrapped-scoring BOOL           Double the (nucleotide) query sequence during the scoring process to allow wrapped diagonal scoring around end and start [0]\n",
    " -e DOUBLE                        List matches below this E-value (range 0.0-inf) [1.000E-03]\n",
    " --min-seq-id FLOAT               List matches above this sequence identity (for clustering) (range 0.0-1.0) [0.000]\n",
    " --min-aln-len INT                Minimum alignment length (range 0-INT_MAX) [0]\n",
    " --seq-id-mode INT                0: alignment length 1: shorter, 2: longer sequence [0]\n",
    " --alt-ali INT                    Show up to this many alternative alignments [0]\n",
    " --max-rejected INT               Maximum rejected alignments before alignment calculation for a query is stopped [2147483647]\n",
    " --max-accept INT                 Maximum accepted alignments before alignment calculation for a query is stopped [2147483647]\n",
    " --score-bias FLOAT               Score bias when computing SW alignment (in bits) [0.000]\n",
    " --realign BOOL                   Compute more conservative, shorter alignments (scores and E-values not changed) [0]\n",
    " --realign-score-bias FLOAT       Additional bias when computing realignment [-0.200]\n",
    " --realign-max-seqs INT           Maximum number of results to return in realignment [2147483647]\n",
    " --corr-score-weight FLOAT        Weight of backtrace correlation score that is added to the alignment score [0.000]\n",
    " --gap-open TWIN                  Gap open cost [aa:11,nucl:5]\n",
    " --gap-extend TWIN                Gap extension cost [aa:1,nucl:2]\n",
    " --zdrop INT                      Maximal allowed difference between score values before alignment is truncated  (nucleotide alignment only) [40]\n",
    "clust:\n",
    " --cluster-mode INT               0: Set-Cover (greedy)\n",
    "                                  1: Connected component (BLASTclust)\n",
    "                                  2,3: Greedy clustering by sequence length (CDHIT) [0]\n",
    " --max-iterations INT             Maximum depth of breadth first search in connected component clustering [1000]\n",
    " --similarity-type INT            Type of score used for clustering. 1: alignment score 2: sequence identity [2]\n",
    " --single-step-clustering BOOL    Switch from cascaded to simple clustering workflow [0]\n",
    " --cluster-steps INT              Cascaded clustering steps from 1 to -s [3]\n",
    " --cluster-reassign BOOL          Cascaded clustering can cluster sequence that do not fulfill the clustering criteria.\n",
    "                                  Cluster reassignment corrects these errors [0]\n",
    "kmermatcher:\n",
    " --weights STR                    Weights used for cluster priorization []\n",
    " --cluster-weight-threshold FLOAT Weight threshold used for cluster priorization [0.900]\n",
    " --kmer-per-seq INT               k-mers per sequence [21]\n",
    " --kmer-per-seq-scale TWIN        Scale k-mer per sequence based on sequence length as kmer-per-seq val + scale x seqlen [aa:0.000,nucl:0.200]\n",
    " --adjust-kmer-len BOOL           Adjust k-mer length based on specificity (only for nucleotides) [0]\n",
    " --hash-shift INT                 Shift k-mer hash initialization [67]\n",
    " --include-only-extendable BOOL   Include only extendable [0]\n",
    " --ignore-multi-kmer BOOL         Skip k-mers occurring multiple times (>=2) [0]\n",
    "profile:\n",
    " --pca                            Pseudo count admixture strength []\n",
    " --pcb                            Pseudo counts: Neff at half of maximum admixture (range 0.0-inf) []\n",
    "misc:\n",
    " --taxon-list STR                 Taxonomy ID, possibly multiple values separated by ',' []\n",
    " --rescore-mode INT               Rescore diagonals with:\n",
    "                                  0: Hamming distance\n",
    "                                  1: local alignment (score only)\n",
    "                                  2: local alignment\n",
    "                                  3: global alignment\n",
    "                                  4: longest alignment fulfilling window quality criterion [0]\n",
    " --dbtype INT                     Database type 0: auto, 1: amino acid 2: nucleotides [0]\n",
    " --shuffle BOOL                   Shuffle input database [1]\n",
    " --createdb-mode INT              Createdb mode 0: copy data, 1: soft link data and write new index (works only with single line fasta/q) [1]\n",
    " --id-offset INT                  Numeric ids in index file are offset by this value [0]\n",
    "common:\n",
    " --sub-mat TWIN                   Substitution matrix file [aa:blosum62.out,nucl:nucleotide.out]\n",
    " --max-seq-len INT                Maximum sequence length [65535]\n",
    " --db-load-mode INT               Database preload mode 0: auto, 1: fread, 2: mmap, 3: mmap+touch [0]\n",
    " --threads INT                    Number of CPU-cores used (all by default) [384]\n",
    " --compressed INT                 Write compressed output [0]\n",
    " -v INT                           Verbosity level: 0: quiet, 1: +errors, 2: +warnings, 3: +info [3]\n",
    " --remove-tmp-files BOOL          Delete temporary files [1]\n",
    " --force-reuse BOOL               Reuse tmp filse in tmp/latest folder ignoring parameters and version changes [0]\n",
    " --mpi-runner STR                 Use MPI on compute cluster with this MPI command (e.g. \"mpirun -np 42\") []\n",
    "expert:\n",
    " --filter-hits BOOL               Filter hits by seq.id. and coverage [0]\n",
    " --sort-results INT               Sort results: 0: no sorting, 1: sort by E-value (Alignment) or seq.id. (Hamming) [0]\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "19acec52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate a fasta file as input to mmseqs\n",
    "with open('clustering/swissprot.fasta', 'w') as f:\n",
    "    for index, row in swissprot.iterrows():\n",
    "        f.write(f'>{row[\"Entry\"]}\\n{row[\"Sequence\"]}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5e6cdd85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes50 clustering/mmseqs_tmp --min-seq-id 0.5 -c 0.5 --cov-mode 0 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.5\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.5\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/12727037950097592956/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[199577] 0s 42ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 43ms\n",
      "Create directory clustering/mmseqs_tmp/12727037950097592956/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/12727037950097592956/input clustering/mmseqs_tmp/12727037950097592956/clu clustering/mmseqs_tmp/12727037950097592956/clu_tmp --max-seqs 20 -c 0.5 --cov-mode 0 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.5 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 3.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 3\n",
      "linclust clustering/mmseqs_tmp/12727037950097592956/input clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/clu_redundancy clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.5 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/12727037950097592956/input clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.5 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/12727037950097592956/input clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.5 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 199659 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=======================================>                         ] 60.26% 120.32K eta 0s                                                          ] 6.71% 13.41K eta 0s       K eta 0s                                        ] 38.76% 77.39K eta 0s       =>                                ] 49.30% 98.43K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 69.60K 0s 166ms    \n",
      "Time for merging to pref_filter2: 0h 0m 0s 30ms\n",
      "Time for processing: 0h 0m 0s 447ms\n",
      "rescorediagonal clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/input_step_redundancy clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/input_step_redundancy clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/pref_filter2 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.5 -a 0 --cov-mode 0 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 69.60K 0s 226ms    ==========>                                       ] 39.00% 27.15K eta 0s         ] 62.48% 43.49K eta 0s       ====================================>             ] 79.22% 55.14K eta 0s       \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 20ms=================>] 99.75% 69.43K eta 0s       \n",
      "Time for processing: 0h 0m 0s 471ms\n",
      "align clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/input_step_redundancy clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/input_step_redundancy clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/pref_rescore2 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.5 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 69603 type: Aminoacid\n",
      "Target database size: 69603 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 69.60K 1s 391ms    \n",
      "Time for merging to aln: 0h 0m 0s 23ms\n",
      "126208 alignments calculated\n",
      "102955 sequence pairs passed the thresholds (0.815756 of overall calculated)\n",
      "1.479175 hits per query sequence\n",
      "Time for processing: 0h 0m 1s 631ms\n",
      "clust clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/input_step_redundancy clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/aln clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=============================>                                   ] 43.02% 29.95K eta 0s       0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 199.66K 0s 527ms          ] 80.78% 161.29K eta 0s       \n",
      "Time for merging to clu_redundancy: 0h 0m 0s 24ms\n",
      "Time for processing: 0h 0m 0s 606ms\n",
      "rmdb clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/linclust/9466570530660296739/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/clu_redundancy clustering/mmseqs_tmp/12727037950097592956/input clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 14ms\n",
      "prefilter clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step_redundancy clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step_redundancy clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.5 --cov-mode 0 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 50068 type: Aminoacid\n",
      "Estimated memory consumption: 2G\n",
      "Target database size: 50068 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=========================================>                       ] 68.19% 34.14K eta 0s                                     ] 19.57% 9.80K eta 0s                         ] 41.54% 20.80K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "align clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step_redundancy clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step_redundancy clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/pref_step0 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.5 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 50068 type: Aminoacid\n",
      "Target database size: 50068 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 50.07K 2s 451ms    \n",
      "Time for merging to aln_step0: 0h 0m 0s 19ms\n",
      "580637 alignments calculated\n",
      "154313 sequence pairs passed the thresholds (0.265765 of overall calculated)\n",
      "3.082068 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 639ms\n",
      "clust clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step_redundancy clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/aln_step0 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 50.07K 0s 203ms                                                    ] 19.01% 9.52K eta 0s       =====================================>            ] 80.44% 40.28K eta 0s       \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 6355 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 50.07K 0s 199ms    \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 50.07K 0s 6ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 529ms\n",
      "Total time: 0h 0m 0s 573ms\n",
      "\n",
      "Size of the sequence database: 50068\n",
      "Size of the alignment database: 50068\n",
      "Number of clusters: 31440\n",
      "\n",
      "Writing results 0h 0m 0s 4ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 588ms\n",
      "createsubdb clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/clu_step0 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step_redundancy clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "prefilter clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step1 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step1 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 2 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.5 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 31440 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 31440 type: Aminoacid\n",
      "Index table k-mer threshold: 145 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[============>                                                    ] 19.37% 6.09K eta 0s       \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 31.44K 1s 826ms    \n",
      "Time for merging to aln_step1: 0h 0m 0s 16ms\n",
      "329239 alignments calculated\n",
      "33499 sequence pairs passed the thresholds (0.101747 of overall calculated)\n",
      "1.065490 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 51ms\n",
      "clust clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step1 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/aln_step1 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/clu_step1 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 31.44K 0s 97ms     ===>                                            ] 31.54% 9.92K eta 0s       ==========================>       ] 87.93% 27.65K eta 0s       \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 477 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 31.44K 0s 123ms                            ] 28.83% 9.07K eta 0s         ] 80.52% 25.32K eta 0s       \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 31.44K 0s 0ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 306ms\n",
      "Total time: 0h 0m 0s 349ms\n",
      "\n",
      "Size of the sequence database: 31440\n",
      "Size of the alignment database: 31440\n",
      "Number of clusters: 30320\n",
      "\n",
      "Writing results 0h 0m 0s 4ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 362ms\n",
      "createsubdb clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/clu_step1 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step1 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "prefilter clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step2 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step2 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 3 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.5 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 30320 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 30320 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[====================================>                            ] 56.14% 17.02K eta 0s         ] 9.09% 2.76K eta 0s       K eta 0s       \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 30.32K 1s 901ms    \n",
      "Time for merging to aln_step2: 0h 0m 0s 16ms\n",
      "404915 alignments calculated\n",
      "30519 sequence pairs passed the thresholds (0.075371 of overall calculated)\n",
      "1.006563 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 107ms\n",
      "clust clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/input_step2 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/aln_step2 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/clu_step2 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 30.32K 0s 105ms                                     ] 39.20% 11.89K eta 0s       \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 47 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 30.32K 0s 117ms    ta 0s       \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 30.32K 0s 0ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 306ms\n",
      "Total time: 0h 0m 0s 344ms\n",
      "\n",
      "Size of the sequence database: 30320\n",
      "Size of the alignment database: 30320\n",
      "Number of clusters: 30214\n",
      "\n",
      "Writing results 0h 0m 0s 4ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 357ms\n",
      "mergeclusters clustering/mmseqs_tmp/12727037950097592956/input clustering/mmseqs_tmp/12727037950097592956/clu clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/clu_redundancy clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/clu_step0 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/clu_step1 clustering/mmseqs_tmp/12727037950097592956/clu_tmp/12600948275945122152/clu_step2 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 50.07K 0s 178ms    ================>                                 ] 47.75% 23.91K eta 0s       K eta 0s       \n",
      "Clustering step 2\n",
      "[===================>                                             ] 28.36% 8.92K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster clustering/swissprot.fasta clustering/clusterRes50 clustering/mmseqs_tmp --min-seq-id 0.5 -c 0.5 --cov-mode 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "286e6959",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes30 clustering/mmseqs_tmp --min-seq-id 0.3 -c 0.3 --cov-mode 0 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.3\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.3\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/3880434244763828342/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[199577] 0s 44ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 44ms\n",
      "Create directory clustering/mmseqs_tmp/3880434244763828342/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/3880434244763828342/input clustering/mmseqs_tmp/3880434244763828342/clu clustering/mmseqs_tmp/3880434244763828342/clu_tmp --max-seqs 20 -c 0.3 --cov-mode 0 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.3 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 5.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 3\n",
      "linclust clustering/mmseqs_tmp/3880434244763828342/input clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/clu_redundancy clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.3 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/3880434244763828342/input clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/linclust/17132673411897700559/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.3 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/3880434244763828342/input clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/linclust/17132673411897700559/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.3 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 199659 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=====================================>                           ] 57.99% 115.78K eta 0s                                          ] 3.33% 6.65K eta 1s                                                         ] 15.54% 31.03K eta 0s                                         ] 26.04% 51.99K eta 0s       =========>                                        ] 37.00% 73.87K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for merging to pref_filter2: 0h 0m 0s 28ms\n",
      "Time for processing: 0h 0m 0s 421ms\n",
      "rescorediagonal clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/linclust/17132673411897700559/input_step_redundancy clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/linclust/17132673411897700559/input_step_redundancy clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/linclust/17132673411897700559/pref_filter2 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/linclust/17132673411897700559/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.3 -a 0 --cov-mode 0 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 69.60K 0s 248ms      ] 24.37% 16.97K eta 0s       ================>                                 ] 47.86% 33.31K eta 0s       ===============================>  ] 96.73% 67.32K eta 0s       \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 21ms=================>] 99.27% 69.09K eta 0s        \n",
      "Time for processing: 0h 0m 0s 526ms\n",
      "align clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/linclust/17132673411897700559/input_step_redundancy clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/linclust/17132673411897700559/input_step_redundancy clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/linclust/17132673411897700559/pref_rescore2 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/linclust/17132673411897700559/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.3 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 69601 type: Aminoacid\n",
      "Target database size: 69601 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 69.60K 3s 862ms    \n",
      "Time for merging to aln: 0h 0m 0s 23ms\n",
      "139721 alignments calculated\n",
      "136819 sequence pairs passed the thresholds (0.979230 of overall calculated)\n",
      "1.965762 hits per query sequence\n",
      "Time for processing: 0h 0m 4s 87ms\n",
      "clust clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/linclust/17132673411897700559/input_step_redundancy clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/linclust/17132673411897700559/aln clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/linclust/17132673411897700559/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 69.60K 0s 180ms    \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 67218 new connections.\n",
      "Reconstruct initial order\n",
      "[=========================================>                       ] 63.66% 44.31K eta 0s       0s                         ] 35.67% 24.83K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "1.755706 k-mers per position\n",
      "415 DB matches per sequence\n",
      "0 overflows\n",
      "9 sequences passed prefiltering per query sequence\n",
      "8 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 16ms\n",
      "Time for processing: 0h 0m 1s 920ms\n",
      "align clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step_redundancy clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step_redundancy clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/pref_step0 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.3 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 38302 type: Aminoacid\n",
      "Target database size: 38302 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 38.30K 2s 0ms      \n",
      "Time for merging to aln_step0: 0h 0m 0s 18ms\n",
      "365646 alignments calculated\n",
      "315560 sequence pairs passed the thresholds (0.863021 of overall calculated)\n",
      "8.238734 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 377ms\n",
      "clust clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step_redundancy clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/aln_step0 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 38.30K 0s 213ms                                                    ] 3.16% 1.21K eta 0s                                         ] 33.13% 12.69K eta 0s       ==========================>                       ] 64.25% 24.61K eta 0s       \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 39194 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 38.30K 0s 232ms                                               ] 31.27% 11.98K eta 0s         ] 73.51% 28.16K eta 0s       ==============================================>   ] 94.23% 36.09K eta 0s       \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 38.30K 0s 10ms     \n",
      "\n",
      "Time for read in: 0h 0m 0s 570ms\n",
      "Total time: 0h 0m 0s 612ms\n",
      "\n",
      "Size of the sequence database: 38302\n",
      "Size of the alignment database: 38302\n",
      "Number of clusters: 13251\n",
      "\n",
      "Writing results 0h 0m 0s 4ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 626ms\n",
      "createsubdb clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/clu_step0 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step_redundancy clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "prefilter clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step1 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step1 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 3 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.3 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 13251 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 13251 type: Aminoacid\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clust clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step1 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/aln_step1 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/clu_step1 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 13.25K 0s 45ms                    ] 70.26% 9.31K eta 0s       \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 2804 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 13.25K 0s 34ms     \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 13.25K 0s 1ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 147ms\n",
      "Total time: 0h 0m 0s 184ms\n",
      "\n",
      "Size of the sequence database: 13251\n",
      "Size of the alignment database: 13251\n",
      "Number of clusters: 8420\n",
      "\n",
      "Writing results 0h 0m 0s 1ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 193ms\n",
      "createsubdb clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/clu_step1 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step1 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "prefilter clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step2 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step2 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 5 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.3 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 8420 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 8420 type: Aminoacid\n",
      "Index table k-mer threshold: 118 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 8.42K 0s 79ms     4.49K eta 0s       \n",
      "Index table: Masked residues: 59392\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 8.42K 0s 87ms     \n",
      "Index statistics\n",
      "Entries:          3741770\n",
      "DB size:          509 MB\n",
      "Avg k-mer size:   0.058465\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t44\n",
      "    VHDKNI\t31\n",
      "    LGGFVY\t24\n",
      "    IGGFVY\t18\n",
      "    TGGTYL\t13\n",
      "    RFNSPN\t13\n",
      "    IHDKNV\t13\n",
      "    IGGYVY\t13\n",
      "    DIAGLY\t12\n",
      "    VLGPGK\t10\n",
      "Time for index table init: 0h 0m 0s 517ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 118\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 8420\n",
      "Target db start 1 to 8420\n",
      "[=================================================================] 100.00% 8.42K 0s 366ms    \n",
      "[===============================================================> ] 97.61% 8.22K eta 0s       \n",
      "172.923112 k-mers per position\n",
      "4606 DB matches per sequence\n",
      "0 overflows\n",
      "16 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 12ms\n",
      "Time for processing: 0h 0m 1s 431ms\n",
      "align clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step2 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step2 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/pref_step2 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.3 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 8420 type: Aminoacid\n",
      "Target database size: 8420 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 8.42K 0s 757ms    \n",
      "Time for merging to aln_step2: 0h 0m 0s 13ms\n",
      "110582 alignments calculated\n",
      "11152 sequence pairs passed the thresholds (0.100848 of overall calculated)\n",
      "1.324466 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 947ms\n",
      "clust clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/input_step2 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/aln_step2 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/clu_step2 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 8.42K 0s 18ms     \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 426 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 8.42K 0s 37ms     \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 8.42K 0s 0ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 116ms\n",
      "Total time: 0h 0m 0s 151ms\n",
      "\n",
      "Size of the sequence database: 8420\n",
      "Size of the alignment database: 8420\n",
      "Number of clusters: 7297\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 158ms\n",
      "mergeclusters clustering/mmseqs_tmp/3880434244763828342/input clustering/mmseqs_tmp/3880434244763828342/clu clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/clu_redundancy clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/clu_step0 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/clu_step1 clustering/mmseqs_tmp/3880434244763828342/clu_tmp/17372756696026749192/clu_step2 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 38.30K 0s 159ms     ] 46.35% 17.75K eta 0s       K eta 0s       \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 13.25K 0s 260ms                   ] 51.98% 6.89K eta 0s       \n",
      "Clustering step 3\n",
      "[=================================================================] 100.00% 8.42K 0s 318ms    \n",
      "Clustering step 4\n",
      "[=================================================================] 100.00% 7.30K 0s 372ms    \n",
      "Write merged clustering\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for merging to cluster.tsv: 0h 0m 0s 13ms\n",
      "Time for processing: 0h 0m 0s 225ms\n",
      "result2repseq clustering/mmseqs_tmp/3880434244763828342/input clustering/mmseqs_tmp/3880434244763828342/clu clustering/mmseqs_tmp/3880434244763828342/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 7.30K 0s 34ms     \n",
      "Time for merging to clu_rep: 0h 0m 0s 13ms\n",
      "Time for processing: 0h 0m 0s 221ms\n",
      "result2flat clustering/mmseqs_tmp/3880434244763828342/input clustering/mmseqs_tmp/3880434244763828342/input clustering/mmseqs_tmp/3880434244763828342/clu_rep clustering/mmseqs_tmp/3880434244763828342/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 40ms\n",
      "createseqfiledb clustering/mmseqs_tmp/3880434244763828342/input clustering/mmseqs_tmp/3880434244763828342/clu clustering/mmseqs_tmp/3880434244763828342/clu_seqs --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 7.30K 0s 53ms     \n",
      "Time for merging to clu_seqs: 0h 0m 0s 9ms\n",
      "Time for processing: 0h 0m 0s 279ms\n",
      "result2flat clustering/mmseqs_tmp/3880434244763828342/input clustering/mmseqs_tmp/3880434244763828342/input clustering/mmseqs_tmp/3880434244763828342/clu_seqs clustering/mmseqs_tmp/3880434244763828342/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 227ms\n",
      "rmdb clustering/mmseqs_tmp/3880434244763828342/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/3880434244763828342/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/3880434244763828342/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 12ms\n",
      "rmdb clustering/mmseqs_tmp/3880434244763828342/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/3880434244763828342/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster clustering/swissprot.fasta clustering/clusterRes30 clustering/mmseqs_tmp --min-seq-id 0.3 -c 0.3 --cov-mode 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c87d03a8-8931-46fe-93ea-adfd02f557b9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes70 clustering/mmseqs_tmp --min-seq-id 0.7 -c 0.7 --cov-mode 0 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.7\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.7\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/3126897581721805954/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[199577] 0s 41ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 41ms\n",
      "Create directory clustering/mmseqs_tmp/3126897581721805954/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/3126897581721805954/input clustering/mmseqs_tmp/3126897581721805954/clu clustering/mmseqs_tmp/3126897581721805954/clu_tmp --max-seqs 20 -c 0.7 --cov-mode 0 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.7 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 1.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 1\n",
      "linclust clustering/mmseqs_tmp/3126897581721805954/input clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/clu_redundancy clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.7 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.7 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/3126897581721805954/input clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/linclust/7513959034062082419/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.7 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.7 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/3126897581721805954/input clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/linclust/7513959034062082419/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.7 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.7 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 199659 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[======================================>                          ] 59.62% 119.04K eta 0s       a 1s                         ] 16.70% 33.35K eta 0s                         ] 27.80% 55.50K eta 0s         ] 38.86% 77.58K eta 0s         ] 49.09% 98.01K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for merging to pref_filter2: 0h 0m 0s 29ms\n",
      "Time for processing: 0h 0m 0s 454ms\n",
      "rescorediagonal clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/linclust/7513959034062082419/input_step_redundancy clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/linclust/7513959034062082419/input_step_redundancy clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/linclust/7513959034062082419/pref_filter2 clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/linclust/7513959034062082419/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.7 -a 0 --cov-mode 0 --min-seq-id 0.7 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 91.01K 0s 314ms    >                                                 ] 24.37% 22.18K eta 0s         ] 44.90% 40.87K eta 0s         ] 61.47% 55.94K eta 0s       K eta 0s       ==============================>   ] 94.61% 86.10K eta 0s       \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 24ms===============>  ] 96.79% 88.09K eta 0s       \n",
      "Time for processing: 0h 0m 0s 602ms\n",
      "align clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/linclust/7513959034062082419/input_step_redundancy clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/linclust/7513959034062082419/input_step_redundancy clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/linclust/7513959034062082419/pref_rescore2 clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/linclust/7513959034062082419/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.7 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.7 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 91010 type: Aminoacid\n",
      "Target database size: 91010 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 91.01K 0s 871ms    \n",
      "Time for merging to aln: 0h 0m 0s 27ms\n",
      "177929 alignments calculated\n",
      "106313 sequence pairs passed the thresholds (0.597502 of overall calculated)\n",
      "1.168146 hits per query sequence\n",
      "Time for processing: 0h 0m 1s 209ms\n",
      "clust clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/linclust/7513959034062082419/input_step_redundancy clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/linclust/7513959034062082419/aln clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/linclust/7513959034062082419/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=======================================================>         ] 86.15% 68.44K eta 0s                                          ] 30.24% 24.02K eta 0s       K eta 0s       K eta 0s       ======================>           ] 82.15% 65.26K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 79.44K 3s 64ms     \n",
      "Time for merging to aln_step0: 0h 0m 0s 24ms\n",
      "1124397 alignments calculated\n",
      "100672 sequence pairs passed the thresholds (0.089534 of overall calculated)\n",
      "1.267207 hits per query sequence\n",
      "Time for processing: 0h 0m 3s 293ms\n",
      "clust clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/input_step_redundancy clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/aln_step0 clustering/mmseqs_tmp/3126897581721805954/clu_tmp/9737502746166104508/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 79.44K 0s 196ms     eta 0s       K eta 0s       ==>               ] 76.25% 60.58K eta 0s       \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 158 new connections.\n",
      "Reconstruct initial order\n",
      "[====================================================>            ] 80.23% 63.74K eta 0s       \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for processing: 0h 0m 0s 269ms\n",
      "rmdb clustering/mmseqs_tmp/3126897581721805954/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/3126897581721805954/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/3126897581721805954/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "rmdb clustering/mmseqs_tmp/3126897581721805954/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/3126897581721805954/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster clustering/swissprot.fasta clustering/clusterRes70 clustering/mmseqs_tmp --min-seq-id 0.7 -c 0.7 --cov-mode 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5665a723-85b5-458d-bc4a-9bbae4731d61",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clustRes90 clustering/mmseqs_tmp --min-seq-id 0.9 -c 0.9 --cov-mode 0 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.9\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.9\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/5862581231708670875/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[199577] 0s 42ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 42ms\n",
      "Create directory clustering/mmseqs_tmp/5862581231708670875/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/5862581231708670875/input clustering/mmseqs_tmp/5862581231708670875/clu clustering/mmseqs_tmp/5862581231708670875/clu_tmp --max-seqs 20 -c 0.9 --cov-mode 0 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.9 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 1.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 1\n",
      "linclust clustering/mmseqs_tmp/5862581231708670875/input clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/clu_redundancy clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.9 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/5862581231708670875/input clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.9 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.9 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/5862581231708670875/input clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.9 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.9 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 199659 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=======================================>                         ] 61.17% 122.13K eta 0s       4.10% 8.20K eta 1s         ] 17.28% 34.50K eta 0s                                         ] 28.94% 57.78K eta 0s       K eta 0s               ] 50.85% 101.54K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for processing: 0h 0m 0s 473ms\n",
      "rescorediagonal clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/input_step_redundancy clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/input_step_redundancy clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/pref_filter2 clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.9 -a 0 --cov-mode 0 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[===========================================================>     ] 90.98% 116.92K eta 0s                                         ] 13.29% 17.08K eta 0s                         ] 30.36% 39.02K eta 0s                         ] 45.41% 58.36K eta 0s       00.00K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for merging to clu_redundancy: 0h 0m 0s 37ms\n",
      "Time for processing: 0h 0m 0s 755ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/linclust/14760282331649708590/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/clu_redundancy clustering/mmseqs_tmp/5862581231708670875/input clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 22ms\n",
      "prefilter clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/input_step_redundancy clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/input_step_redundancy clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.9 --cov-mode 0 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 126008 type: Aminoacid\n",
      "Estimated memory consumption: 3G\n",
      "Target database size: 126008 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=======================================================>         ] 85.77% 108.08K eta 0s                                                     ] 13.88% 17.49K eta 0s       ==>                                              ] 29.10% 36.67K eta 0s         ] 53.02% 66.81K eta 0s       K eta 0s       K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 126.01K 4s 3ms      \n",
      "Time for merging to aln_step0: 0h 0m 0s 32ms\n",
      "1860433 alignments calculated\n",
      "131302 sequence pairs passed the thresholds (0.070576 of overall calculated)\n",
      "1.042013 hits per query sequence\n",
      "Time for processing: 0h 0m 4s 242ms\n",
      "clust clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/input_step_redundancy clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/aln_step0 clustering/mmseqs_tmp/5862581231708670875/clu_tmp/6290754456147120823/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 126.01K 0s 261ms     ] 4.64% 5.84K eta 0s       ======>                                           ] 32.99% 41.57K eta 0s       K eta 0s       =================================>                ] 74.28% 93.60K eta 0s       \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 58 new connections.\n",
      "Reconstruct initial order\n",
      "[==================================>                              ] 53.06% 66.87K eta 0s       \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 124.11K 0s 414ms    \n",
      "Time for merging to clu_seqs: 0h 0m 0s 34ms\n",
      "Time for processing: 0h 0m 0s 695ms\n",
      "result2flat clustering/mmseqs_tmp/5862581231708670875/input clustering/mmseqs_tmp/5862581231708670875/input clustering/mmseqs_tmp/5862581231708670875/clu_seqs clustering/mmseqs_tmp/5862581231708670875/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 293ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 17ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/5862581231708670875/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster clustering/swissprot.fasta clustering/clustRes90 clustering/mmseqs_tmp --min-seq-id 0.9 -c 0.9 --cov-mode 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "02322b64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30214\n",
      "7297\n",
      "72419\n",
      "124110\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "      <th>EC All</th>\n",
       "      <th>clusterRes50</th>\n",
       "      <th>clusterRes30</th>\n",
       "      <th>clusterRes70</th>\n",
       "      <th>clustRes90</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "      <td>796</td>\n",
       "      <td>3.2.1.39</td>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>D4AZ24</td>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>A0A023I7E1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.1.1.56</td>\n",
       "      <td>3423</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "      <td>Q32ZD5</td>\n",
       "      <td>P33515</td>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>A0A024B7W1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.1.1.57</td>\n",
       "      <td>3423</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "      <td>Q32ZD5</td>\n",
       "      <td>P33515</td>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>A0A024B7W1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>3423</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "      <td>Q32ZD5</td>\n",
       "      <td>P33515</td>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>A0A024B7W1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199654</th>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>Y3032_BACAN</td>\n",
       "      <td>MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>584</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>O34909</td>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>Q6HX62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199655</th>\n",
       "      <td>Q6L032</td>\n",
       "      <td>Y1085_PICTO</td>\n",
       "      <td>MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>573</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>Q6L032</td>\n",
       "      <td>O34909</td>\n",
       "      <td>Q6L032</td>\n",
       "      <td>Q6L032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199656</th>\n",
       "      <td>Q7TDB6</td>\n",
       "      <td>RDRP_MYRV9</td>\n",
       "      <td>MSLTSRYTHFVPDSTITEILNDSNTPQILLHYANIVNGSTPVHFTS...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>1354</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>Q7TDB6</td>\n",
       "      <td>Q7TDB6</td>\n",
       "      <td>Q7TDB6</td>\n",
       "      <td>Q7TDB6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199657</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>P39262</td>\n",
       "      <td>P39262</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199658</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>P61235</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q9J5H2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>199659 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "1       A0A023I7E1   ENG1_RHIMI   \n",
       "2       A0A024B7W1   POLG_ZIKVF   \n",
       "3       A0A024B7W1   POLG_ZIKVF   \n",
       "4       A0A024B7W1   POLG_ZIKVF   \n",
       "...            ...          ...   \n",
       "199654      Q6HX62  Y3032_BACAN   \n",
       "199655      Q6L032  Y1085_PICTO   \n",
       "199656      Q7TDB6   RDRP_MYRV9   \n",
       "199657      Q94MV8   VG56_BPLZ5   \n",
       "199658      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence EC number  Length  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.6     269   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...  3.2.1.39     796   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...  2.1.1.56    3423   \n",
       "3       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...  2.1.1.57    3423   \n",
       "4       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...  2.7.7.48    3423   \n",
       "...                                                   ...       ...     ...   \n",
       "199654  MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...   3.5.4.2     584   \n",
       "199655  MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...   3.5.4.2     573   \n",
       "199656  MSLTSRYTHFVPDSTITEILNDSNTPQILLHYANIVNGSTPVHFTS...  2.7.7.48    1354   \n",
       "199657  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...  3.6.1.12     172   \n",
       "199658  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...  3.1.1.29     124   \n",
       "\n",
       "                                                   EC All clusterRes50  \\\n",
       "0                                        3.2.2.-; 3.2.2.6   A0A009IHW8   \n",
       "1                                                3.2.1.39   A0A023I7E1   \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...       Q32ZD5   \n",
       "3       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...       Q32ZD5   \n",
       "4       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...       Q32ZD5   \n",
       "...                                                   ...          ...   \n",
       "199654                                            3.5.4.2       Q6HX62   \n",
       "199655                                            3.5.4.2       Q6L032   \n",
       "199656                                           2.7.7.48       Q7TDB6   \n",
       "199657                                           3.6.1.12       P39262   \n",
       "199658                                           3.1.1.29       Q9J5H2   \n",
       "\n",
       "       clusterRes30 clusterRes70  clustRes90  \n",
       "0        A0A009IHW8   A0A009IHW8  A0A009IHW8  \n",
       "1            D4AZ24   A0A023I7E1  A0A023I7E1  \n",
       "2            P33515   A0A024B7W1  A0A024B7W1  \n",
       "3            P33515   A0A024B7W1  A0A024B7W1  \n",
       "4            P33515   A0A024B7W1  A0A024B7W1  \n",
       "...             ...          ...         ...  \n",
       "199654       O34909       Q6HX62      Q6HX62  \n",
       "199655       O34909       Q6L032      Q6L032  \n",
       "199656       Q7TDB6       Q7TDB6      Q7TDB6  \n",
       "199657       P39262       Q94MV8      Q94MV8  \n",
       "199658       P61235       Q9J5H2      Q9J5H2  \n",
       "\n",
       "[199659 rows x 10 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkIAAAGdCAYAAAD+JxxnAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuNSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/xnp5ZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAz4klEQVR4nO3de3RU9b3+8ScJuSK5AIeE1AA5rYeL3BQ0BtGjh5jQUmsspYKppjaFapMWzPmBghgBxQgcKaAIpT1quwqKrFNSihiZEypRGQMEolwEaUvFJZ3QFsJwkcmQfH9/tNmHIQik7hDC9/1ai7WYvT+z93c/ifKs2TNJmDHGCAAAwELhbb0AAACAtkIRAgAA1qIIAQAAa1GEAACAtShCAADAWhQhAABgLYoQAACwFkUIAABYq0NbL+By1tjYqIMHD6pTp04KCwtr6+UAAICLYIzRsWPHlJqaqvDw87/mQxE6j4MHDyotLa2tlwEAAP4Jn3zyia6++urzzlCEzqNTp06S/h5kfHy8a8cNBoNav369srOzFRkZ6dpxbUeu7iNT95Gp+8i0dbTnXP1+v9LS0px/x8+HInQeTbfD4uPjXS9CcXFxio+Pb3ffXJczcnUfmbqPTN1Hpq3jSsj1Yt7WwpulAQCAtShCAADAWhQhAABgLYoQAACwFkUIAABYiyIEAACsRRECAADWoggBAABrUYQAAIC1KEIAAMBaFCEAAGAtihAAALAWRQgAAFiLIgQAAKzV4iJUWVmpO++8U6mpqQoLC1NZWZmzLxgM6pFHHtGAAQPUsWNHpaam6v7779fBgwdDjnH48GHl5eUpPj5eiYmJKigo0PHjx0NmPvjgA91yyy2KiYlRWlqa5s6d22wtq1atUp8+fRQTE6MBAwZo3bp1IfuNMSopKVH37t0VGxurrKws7du3r6WX3Gr2Dr1BH/bp29bLAADAWi0uQidOnNCgQYO0ePHiZvtOnjypbdu26fHHH9e2bdv061//Wnv37tU3vvGNkLm8vDzt2rVLHo9Ha9euVWVlpSZMmODs9/v9ys7OVs+ePVVdXa158+ZpxowZWrZsmTOzadMmjRs3TgUFBdq+fbtyc3OVm5urnTt3OjNz587VokWLtHTpUlVVValjx47KycnRqVOnWnrZAADgSmS+AElm9erV553ZvHmzkWQ+/vhjY4wxu3fvNpLMli1bnJk33njDhIWFmU8//dQYY8wLL7xgkpKSTCAQcGYeeeQR07t3b+fxt7/9bTNq1KiQc2VkZJgf/OAHxhhjGhsbTUpKipk3b56zv66uzkRHR5tXXnnloq7v6NGjRpI5evToRc1frPr6elNWVmZ2DBxkdvfu4+qxbdaUa319fVsv5YpBpu4jU/eRaetoz7m25N/vDq1dtI4ePaqwsDAlJiZKkrxerxITEzV06FBnJisrS+Hh4aqqqtLdd98tr9erW2+9VVFRUc5MTk6O5syZoyNHjigpKUler1fFxcUh58rJyXFu1e3fv18+n09ZWVnO/oSEBGVkZMjr9Wrs2LHN1hoIBBQIBJzHfr9f0t9v+QWDwS+cRZOmYzVGR4c8xhfTlCN5uodM3Uem7iPT1tGec23Jmlu1CJ06dUqPPPKIxo0bp/j4eEmSz+dTt27dQhfRoYM6d+4sn8/nzKSnp4fMJCcnO/uSkpLk8/mcbWfOnHmMM593rpmzlZaWaubMmc22r1+/XnFxcRd1zS2x/7FpkqQ/nPXeJnwxHo+nrZdwxSFT95Gp+8i0dbTHXE+ePHnRs61WhILBoL797W/LGKMlS5a01mlcNXXq1JBXmfx+v9LS0pSdne0UOTcEg0F5PB6lz35a4YGAem/d4tqxbdaU6x133KHIyMi2Xs4VgUzdR6buI9PW0Z5zbbqjczFapQg1laCPP/5YGzZsCCkRKSkpOnToUMj86dOndfjwYaWkpDgztbW1ITNNjy80c+b+pm3du3cPmRk8ePA51x0dHa3of9yuOlNkZGSrfBOEBwKKCATa3TfY5a61vl42I1P3kan7yLR1tMdcW7Je13+OUFMJ2rdvn/73f/9XXbp0CdmfmZmpuro6VVdXO9s2bNigxsZGZWRkODOVlZUh9/g8Ho969+6tpKQkZ6aioiLk2B6PR5mZmZKk9PR0paSkhMz4/X5VVVU5MwAAwG4tLkLHjx9XTU2NampqJP39Tck1NTU6cOCAgsGgvvWtb2nr1q1avny5Ghoa5PP55PP5VF9fL0nq27evRo4cqfHjx2vz5s169913VVRUpLFjxyo1NVWSdO+99yoqKkoFBQXatWuXVq5cqYULF4bctpo4caLKy8v17LPPas+ePZoxY4a2bt2qoqIiSVJYWJgmTZqkp556SmvWrNGOHTt0//33KzU1Vbm5uV8wNgAAcCVo8a2xrVu36vbbb3ceN5WT/Px8zZgxQ2vWrJGkZreffve73+m2226TJC1fvlxFRUUaMWKEwsPDNXr0aC1atMiZTUhI0Pr161VYWKghQ4aoa9euKikpCflZQ8OGDdOKFSs0ffp0TZs2Tddcc43KysrUv39/Z2bKlCk6ceKEJkyYoLq6Og0fPlzl5eWKiYlp6WUDAIArUIuL0G233SZjzOfuP9++Jp07d9aKFSvOOzNw4EC9/fbb550ZM2aMxowZ87n7w8LCNGvWLM2aNeuCawIAAPbhd40BAABrUYQAAIC1KEIAAMBaFCEAAGAtihAAALAWRQgAAFiLIgQAAKxFEQIAANaiCAEAAGtRhAAAgLUoQgAAwFoUIQAAYC2KEAAAsBZFCAAAWIsiBAAArEURAgAA1qIIAQAAa1GEAACAtShCAADAWhQhAABgLYoQAACwFkUIAABYiyIEAACsRRECAADWoggBAABrUYQAAIC1KEIAAMBaFCEAAGAtihAAALAWRQgAAFiLIgQAAKxFEQIAANaiCAEAAGtRhAAAgLUoQgAAwFoUIQAAYC2KEAAAsBZFCAAAWIsiBAAArEURAgAA1qIIAQAAa1GEAACAtShCAADAWhQhAABgLYoQAACwFkUIAABYiyIEAACs1eIiVFlZqTvvvFOpqakKCwtTWVlZyH5jjEpKStS9e3fFxsYqKytL+/btC5k5fPiw8vLyFB8fr8TERBUUFOj48eMhMx988IFuueUWxcTEKC0tTXPnzm22llWrVqlPnz6KiYnRgAEDtG7duhavBQAA2KvFRejEiRMaNGiQFi9efM79c+fO1aJFi7R06VJVVVWpY8eOysnJ0alTp5yZvLw87dq1Sx6PR2vXrlVlZaUmTJjg7Pf7/crOzlbPnj1VXV2tefPmacaMGVq2bJkzs2nTJo0bN04FBQXavn27cnNzlZubq507d7ZoLQAAwGLmC5BkVq9e7TxubGw0KSkpZt68ec62uro6Ex0dbV555RVjjDG7d+82ksyWLVucmTfeeMOEhYWZTz/91BhjzAsvvGCSkpJMIBBwZh555BHTu3dv5/G3v/1tM2rUqJD1ZGRkmB/84AcXvZYLOXr0qJFkjh49elHzF6u+vt6UlZWZHQMHmd29+7h6bJs15VpfX9/WS7likKn7yNR9ZNo62nOuLfn3u4ObpWr//v3y+XzKyspytiUkJCgjI0Ner1djx46V1+tVYmKihg4d6sxkZWUpPDxcVVVVuvvuu+X1enXrrbcqKirKmcnJydGcOXN05MgRJSUlyev1qri4OOT8OTk5zq26i1nL2QKBgAKBgPPY7/dLkoLBoILB4BcL5wxNx2qMjg55jC+mKUfydA+Zuo9M3UemraM959qSNbtahHw+nyQpOTk5ZHtycrKzz+fzqVu3bqGL6NBBnTt3DplJT09vdoymfUlJSfL5fBc8z4XWcrbS0lLNnDmz2fb169crLi7uc676n7f/sWmSpD+c9d4mfDEej6etl3DFIVP3kan7yLR1tMdcT548edGzrhah9m7q1KkhrzL5/X6lpaUpOztb8fHxrp0nGAzK4/EoffbTCg8E1HvrFteObbOmXO+44w5FRka29XKuCGTqPjJ1H5m2jvaca9MdnYvhahFKSUmRJNXW1qp79+7O9traWg0ePNiZOXToUMjzTp8+rcOHDzvPT0lJUW1tbchM0+MLzZy5/0JrOVt0dLSi/3G76kyRkZGt8k0QHggoIhBod99gl7vW+nrZjEzdR6buI9PW0R5zbcl6Xf05Qunp6UpJSVFFRYWzze/3q6qqSpmZmZKkzMxM1dXVqbq62pnZsGGDGhsblZGR4cxUVlaG3OPzeDzq3bu3kpKSnJkzz9M003Sei1kLAACwW4uL0PHjx1VTU6OamhpJf39Tck1NjQ4cOKCwsDBNmjRJTz31lNasWaMdO3bo/vvvV2pqqnJzcyVJffv21ciRIzV+/Hht3rxZ7777roqKijR27FilpqZKku69915FRUWpoKBAu3bt0sqVK7Vw4cKQ21YTJ05UeXm5nn32We3Zs0czZszQ1q1bVVRUJEkXtRYAAGC3Ft8a27p1q26//XbncVM5yc/P18svv6wpU6boxIkTmjBhgurq6jR8+HCVl5crJibGec7y5ctVVFSkESNGKDw8XKNHj9aiRYuc/QkJCVq/fr0KCws1ZMgQde3aVSUlJSE/a2jYsGFasWKFpk+frmnTpumaa65RWVmZ+vfv78xczFoAAIC9WlyEbrvtNhljPnd/WFiYZs2apVmzZn3uTOfOnbVixYrznmfgwIF6++23zzszZswYjRkz5gutBQAA2IvfNQYAAKxFEQIAANaiCAEAAGtRhAAAgLUoQgAAwFoUIQAAYC2KEAAAsBZFCAAAWIsiBAAArEURAgAA1qIIAQAAa1GEAACAtShCAADAWhQhAABgLYoQAACwFkUIAABYiyIEAACsRRECAADWoggBAABrUYQAAIC1KEIAAMBaFCEAAGAtihAAALAWRQgAAFiLIgQAAKxFEQIAANaiCAEAAGtRhAAAgLUoQgAAwFoUIQAAYC2KEAAAsBZFCAAAWIsiBAAArEURAgAA1qIIAQAAa1GEAACAtShCAADAWhQhAABgLYoQAACwFkUIAABYiyIEAACsRRECAADWoggBAABrUYQAAIC1KEIAAMBaFCEAAGAtihAAALCW60WooaFBjz/+uNLT0xUbG6svf/nLevLJJ2WMcWaMMSopKVH37t0VGxurrKws7du3L+Q4hw8fVl5enuLj45WYmKiCggIdP348ZOaDDz7QLbfcopiYGKWlpWnu3LnN1rNq1Sr16dNHMTExGjBggNatW+f2JQMAgHbK9SI0Z84cLVmyRM8//7w+/PBDzZkzR3PnztVzzz3nzMydO1eLFi3S0qVLVVVVpY4dOyonJ0enTp1yZvLy8rRr1y55PB6tXbtWlZWVmjBhgrPf7/crOztbPXv2VHV1tebNm6cZM2Zo2bJlzsymTZs0btw4FRQUaPv27crNzVVubq527tzp9mUDAIB2yPUitGnTJt11110aNWqUevXqpW9961vKzs7W5s2bJf391aAFCxZo+vTpuuuuuzRw4ED98pe/1MGDB1VWViZJ+vDDD1VeXq6f//znysjI0PDhw/Xcc8/p1Vdf1cGDByVJy5cvV319vV588UVde+21Gjt2rH784x9r/vz5zloWLlyokSNHavLkyerbt6+efPJJXX/99Xr++efdvmwAANAOdXD7gMOGDdOyZcv00Ucf6d/+7d/0/vvv65133nEKyv79++Xz+ZSVleU8JyEhQRkZGfJ6vRo7dqy8Xq8SExM1dOhQZyYrK0vh4eGqqqrS3XffLa/Xq1tvvVVRUVHOTE5OjubMmaMjR44oKSlJXq9XxcXFIevLyclxCtfZAoGAAoGA89jv90uSgsGggsHgF86mSdOxGqOjQx7ji2nKkTzdQ6buI1P3kWnraM+5tmTNrhehRx99VH6/X3369FFERIQaGho0e/Zs5eXlSZJ8Pp8kKTk5OeR5ycnJzj6fz6du3bqFLrRDB3Xu3DlkJj09vdkxmvYlJSXJ5/Od9zxnKy0t1cyZM5ttX79+veLi4i7q+lti/2PTJEl/4H1LrvJ4PG29hCsOmbqPTN1Hpq2jPeZ68uTJi551vQi99tprWr58uVasWKFrr71WNTU1mjRpklJTU5Wfn+/26Vw1derUkFeQ/H6/0tLSlJ2drfj4eNfOEwwG5fF4lD77aYUHAuq9dYtrx7ZZU6533HGHIiMj23o5VwQydR+Zuo9MW0d7zrXpjs7FcL0ITZ48WY8++qjGjh0rSRowYIA+/vhjlZaWKj8/XykpKZKk2tpade/e3XlebW2tBg8eLElKSUnRoUOHQo57+vRpHT582Hl+SkqKamtrQ2aaHl9opmn/2aKjoxX9j9tVZ4qMjGyVb4LwQEARgUC7+wa73LXW18tmZOo+MnUfmbaO9phrS9br+pulT548qfDw0MNGRESosbFRkpSenq6UlBRVVFQ4+/1+v6qqqpSZmSlJyszMVF1dnaqrq52ZDRs2qLGxURkZGc5MZWVlyH1Aj8ej3r17KykpyZk58zxNM03nAQAAdnO9CN15552aPXu2Xn/9df3pT3/S6tWrNX/+fN19992SpLCwME2aNElPPfWU1qxZox07duj+++9XamqqcnNzJUl9+/bVyJEjNX78eG3evFnvvvuuioqKNHbsWKWmpkqS7r33XkVFRamgoEC7du3SypUrtXDhwpBbWxMnTlR5ebmeffZZ7dmzRzNmzNDWrVtVVFTk9mUDAIB2yPVbY88995wef/xx/fCHP9ShQ4eUmpqqH/zgByopKXFmpkyZohMnTmjChAmqq6vT8OHDVV5erpiYGGdm+fLlKioq0ogRIxQeHq7Ro0dr0aJFzv6EhAStX79ehYWFGjJkiLp27aqSkpKQnzU0bNgwrVixQtOnT9e0adN0zTXXqKysTP3793f7sgEAQDvkehHq1KmTFixYoAULFnzuTFhYmGbNmqVZs2Z97kznzp21YsWK855r4MCBevvtt887M2bMGI0ZM+a8MwAAwE78rjEAAGAtihAAALAWRQgAAFiLIgQAAKxFEQIAANaiCAEAAGtRhAAAgLUoQgAAwFoUIQAAYC2KEAAAsBZFCAAAWIsiBAAArEURAgAA1qIIAQAAa1GEAACAtShCAADAWhQhAABgLYoQAACwFkUIAABYiyIEAACsRRECAADWoggBAABrUYQAAIC1KEIAAMBaFCEAAGAtihAAALAWRQgAAFiLIgQAAKxFEQIAANaiCAEAAGtRhAAAgLUoQgAAwFoUIQAAYC2KEAAAsBZFCAAAWIsiBAAArEURAgAA1qIIAQAAa1GEAACAtShCAADAWhQhAABgLYoQAACwFkUIAABYiyIEAACsRRECAADWoggBAABrUYQAAIC1WqUIffrpp/rOd76jLl26KDY2VgMGDNDWrVud/cYYlZSUqHv37oqNjVVWVpb27dsXcozDhw8rLy9P8fHxSkxMVEFBgY4fPx4y88EHH+iWW25RTEyM0tLSNHfu3GZrWbVqlfr06aOYmBgNGDBA69ata41LBgAA7ZDrRejIkSO6+eabFRkZqTfeeEO7d+/Ws88+q6SkJGdm7ty5WrRokZYuXaqqqip17NhROTk5OnXqlDOTl5enXbt2yePxaO3ataqsrNSECROc/X6/X9nZ2erZs6eqq6s1b948zZgxQ8uWLXNmNm3apHHjxqmgoEDbt29Xbm6ucnNztXPnTrcvGwAAtEMd3D7gnDlzlJaWppdeesnZlp6e7vzdGKMFCxZo+vTpuuuuuyRJv/zlL5WcnKyysjKNHTtWH374ocrLy7VlyxYNHTpUkvTcc8/pa1/7mv7rv/5LqampWr58uerr6/Xiiy8qKipK1157rWpqajR//nynMC1cuFAjR47U5MmTJUlPPvmkPB6Pnn/+eS1dutTtSwcAAO2M60VozZo1ysnJ0ZgxY7Rx40Z96Utf0g9/+EONHz9ekrR//375fD5lZWU5z0lISFBGRoa8Xq/Gjh0rr9erxMREpwRJUlZWlsLDw1VVVaW7775bXq9Xt956q6KiopyZnJwczZkzR0eOHFFSUpK8Xq+Ki4tD1peTk6OysrJzrj0QCCgQCDiP/X6/JCkYDCoYDH7hbJo0HasxOjrkMb6YphzJ0z1k6j4ydR+Zto72nGtL1ux6EfrjH/+oJUuWqLi4WNOmTdOWLVv04x//WFFRUcrPz5fP55MkJScnhzwvOTnZ2efz+dStW7fQhXbooM6dO4fMnPlK05nH9Pl8SkpKks/nO+95zlZaWqqZM2c2275+/XrFxcVdbAQXbf9j0yRJf+B9S67yeDxtvYQrDpm6j0zdR6atoz3mevLkyYuedb0INTY2aujQoXr66aclSdddd5127typpUuXKj8/3+3TuWrq1KkhryD5/X6lpaUpOztb8fHxrp0nGAzK4/EoffbTCg8E1HvrFteObbOmXO+44w5FRka29XKuCGTqPjJ1H5m2jvaca9MdnYvhehHq3r27+vXrF7Ktb9+++p//+R9JUkpKiiSptrZW3bt3d2Zqa2s1ePBgZ+bQoUMhxzh9+rQOHz7sPD8lJUW1tbUhM02PLzTTtP9s0dHRiv7H7aozRUZGtso3QXggoIhAoN19g13uWuvrZTMydR+Zuo9MW0d7zLUl63X9U2M333yz9u7dG7Lto48+Us+ePSX9/Y3TKSkpqqiocPb7/X5VVVUpMzNTkpSZmam6ujpVV1c7Mxs2bFBjY6MyMjKcmcrKypD7gB6PR71793Y+oZaZmRlynqaZpvMAAAC7uV6EHn74Yb333nt6+umn9fvf/14rVqzQsmXLVFhYKEkKCwvTpEmT9NRTT2nNmjXasWOH7r//fqWmpio3N1fS319BGjlypMaPH6/Nmzfr3XffVVFRkcaOHavU1FRJ0r333quoqCgVFBRo165dWrlypRYuXBhya2vixIkqLy/Xs88+qz179mjGjBnaunWrioqK3L5sAADQDrl+a+yGG27Q6tWrNXXqVM2aNUvp6elasGCB8vLynJkpU6boxIkTmjBhgurq6jR8+HCVl5crJibGmVm+fLmKioo0YsQIhYeHa/To0Vq0aJGzPyEhQevXr1dhYaGGDBmirl27qqSkJORnDQ0bNkwrVqzQ9OnTNW3aNF1zzTUqKytT//793b5sAADQDrlehCTp61//ur7+9a9/7v6wsDDNmjVLs2bN+tyZzp07a8WKFec9z8CBA/X222+fd2bMmDEaM2bM+RcMAACsxO8aAwAA1qIIAQAAa1GEAACAtShCAADAWhQhAABgLYoQAACwFkUIAABYiyIEAACsRRECAADWoggBAABrUYQAAIC1KEIAAMBaFCEAAGAtihAAALAWRQgAAFiLIgQAAKxFEQIAANaiCAEAAGtRhAAAgLUoQgAAwFoUIQAAYC2KEAAAsBZFCAAAWIsiBAAArEURAgAA1qIIAQAAa1GEAACAtShCAADAWhQhAABgLYoQAACwFkUIAABYiyIEAACsRRECAADWoggBAABrUYQAAIC1KEIAAMBaFCEAAGAtihAAALAWRQgAAFiLIgQAAKxFEQIAANaiCAEAAGtRhAAAgLUoQgAAwFoUIQAAYC2KEAAAsBZFCAAAWKvVi9AzzzyjsLAwTZo0ydl26tQpFRYWqkuXLrrqqqs0evRo1dbWhjzvwIEDGjVqlOLi4tStWzdNnjxZp0+fDpl56623dP311ys6Olpf+cpX9PLLLzc7/+LFi9WrVy/FxMQoIyNDmzdvbo3LBAAA7VCrFqEtW7bopz/9qQYOHBiy/eGHH9Zvf/tbrVq1Shs3btTBgwf1zW9+09nf0NCgUaNGqb6+Xps2bdIvfvELvfzyyyopKXFm9u/fr1GjRun2229XTU2NJk2apO9///t68803nZmVK1equLhYTzzxhLZt26ZBgwYpJydHhw4das3LBgAA7USrFaHjx48rLy9PP/vZz5SUlORsP3r0qP77v/9b8+fP13/8x39oyJAheumll7Rp0ya99957kqT169dr9+7d+tWvfqXBgwfrq1/9qp588kktXrxY9fX1kqSlS5cqPT1dzz77rPr27auioiJ961vf0k9+8hPnXPPnz9f48eP1wAMPqF+/flq6dKni4uL04osvttZlAwCAdqRDax24sLBQo0aNUlZWlp566ilne3V1tYLBoLKyspxtffr0UY8ePeT1enXTTTfJ6/VqwIABSk5OdmZycnL00EMPadeuXbruuuvk9XpDjtE003QLrr6+XtXV1Zo6daqzPzw8XFlZWfJ6vedccyAQUCAQcB77/X5JUjAYVDAY/OfDOEvTsRqjo0Me44tpypE83UOm7iNT95Fp62jPubZkza1ShF599VVt27ZNW7ZsabbP5/MpKipKiYmJIduTk5Pl8/mcmTNLUNP+pn3nm/H7/frss8905MgRNTQ0nHNmz54951x3aWmpZs6c2Wz7+vXrFRcXd54r/ufsf2yaJOkP69a5fmybeTyetl7CFYdM3Uem7iPT1tEecz158uRFz7pehD755BNNnDhRHo9HMTExbh++VU2dOlXFxcXOY7/fr7S0NGVnZys+Pt618wSDQXk8HqXPflrhgYB6b21eGNFyTbnecccdioyMbOvlXBHI1H1k6j4ybR3tOdemOzoXw/UiVF1drUOHDun66693tjU0NKiyslLPP/+83nzzTdXX16uuri7kVaHa2lqlpKRIklJSUpp9uqvpU2Vnzpz9SbPa2lrFx8crNjZWERERioiIOOdM0zHOFh0dreh/3K46U2RkZKt8E4QHAooIBNrdN9jlrrW+XjYjU/eRqfvItHW0x1xbsl7X3yw9YsQI7dixQzU1Nc6foUOHKi8vz/l7ZGSkKioqnOfs3btXBw4cUGZmpiQpMzNTO3bsCPl0l8fjUXx8vPr16+fMnHmMppmmY0RFRWnIkCEhM42NjaqoqHBmAACA3Vx/RahTp07q379/yLaOHTuqS5cuzvaCggIVFxerc+fOio+P149+9CNlZmbqpptukiRlZ2erX79+uu+++zR37lz5fD5Nnz5dhYWFzis2Dz74oJ5//nlNmTJF3/ve97Rhwwa99tprev31153zFhcXKz8/X0OHDtWNN96oBQsW6MSJE3rggQfcvmwAANAOtdqnxs7nJz/5icLDwzV69GgFAgHl5OTohRdecPZHRERo7dq1euihh5SZmamOHTsqPz9fs2bNcmbS09P1+uuv6+GHH9bChQt19dVX6+c//7lycnKcmXvuuUd/+ctfVFJSIp/Pp8GDB6u8vLzZG6gBAICdLkkReuutt0Iex8TEaPHixVq8ePHnPqdnz55ad4FPU912223avn37eWeKiopUVFR00WsFAAD24HeNAQAAa1GEAACAtShCAADAWhQhAABgLYoQAACwFkUIAABYiyIEAACsRRECAADWoggBAABrUYQAAIC1KEIAAMBaFCEAAGAtihAAALAWRQgAAFiLIgQAAKxFEQIAANaiCAEAAGtRhAAAgLUoQgAAwFoUIQAAYC2KEAAAsBZFCAAAWIsiBAAArEURAgAA1qIIAQAAa1GEAACAtShCAADAWhQhAABgLYoQAACwFkUIAABYiyIEAACsRRG6DPR69PW2XgIAAFaiCAEAAGtRhAAAgLUoQgAAwFoUIQAAYC2KEAAAsBZFCAAAWIsiBAAArEURAgAA1qIIAQAAa1GEAACAtShCAADAWhQhAABgLYoQAACwFkUIAABYiyIEAACs5XoRKi0t1Q033KBOnTqpW7duys3N1d69e0NmTp06pcLCQnXp0kVXXXWVRo8erdra2pCZAwcOaNSoUYqLi1O3bt00efJknT59OmTmrbfe0vXXX6/o6Gh95Stf0csvv9xsPYsXL1avXr0UExOjjIwMbd682e1LBgAA7ZTrRWjjxo0qLCzUe++9J4/Ho2AwqOzsbJ04ccKZefjhh/Xb3/5Wq1at0saNG3Xw4EF985vfdPY3NDRo1KhRqq+v16ZNm/SLX/xCL7/8skpKSpyZ/fv3a9SoUbr99ttVU1OjSZMm6fvf/77efPNNZ2blypUqLi7WE088oW3btmnQoEHKycnRoUOH3L5sAADQHplWdujQISPJbNy40RhjTF1dnYmMjDSrVq1yZj788EMjyXi9XmOMMevWrTPh4eHG5/M5M0uWLDHx8fEmEAgYY4yZMmWKufbaa0POdc8995icnBzn8Y033mgKCwudxw0NDSY1NdWUlpZe1NqPHj1qJJmjR4+28KrPr76+3pSVlZkdAweZ3b37mJ6PrHX1+LZqyrW+vr6tl3LFIFP3kan7yLR1tOdcW/Lvd4fWLlpHjx6VJHXu3FmSVF1drWAwqKysLGemT58+6tGjh7xer2666SZ5vV4NGDBAycnJzkxOTo4eeugh7dq1S9ddd528Xm/IMZpmJk2aJEmqr69XdXW1pk6d6uwPDw9XVlaWvF7vOdcaCAQUCAScx36/X5IUDAYVDAa/QAqhmo7VGB0tSYqOMK4e31ZNGZKle8jUfWTqPjJtHe0515asuVWLUGNjoyZNmqSbb75Z/fv3lyT5fD5FRUUpMTExZDY5OVk+n8+ZObMENe1v2ne+Gb/fr88++0xHjhxRQ0PDOWf27NlzzvWWlpZq5syZzbavX79ecXFxF3nVF2//Y9MkSXPVoHXr1rl+fFt5PJ62XsIVh0zdR6buI9PW0R5zPXny5EXPtmoRKiws1M6dO/XOO++05mlcM3XqVBUXFzuP/X6/0tLSlJ2drfj4eNfOEwwG5fF4lD77aYUHAhr99ae0c0aOa8e3VVOud9xxhyIjI9t6OVcEMnUfmbqPTFtHe8616Y7OxWi1IlRUVKS1a9eqsrJSV199tbM9JSVF9fX1qqurC3lVqLa2VikpKc7M2Z/uavpU2ZkzZ3/SrLa2VvHx8YqNjVVERIQiIiLOOdN0jLNFR0cr+h+3q84UGRnZKt8E4YGAIgIBBRrC2t032eWstb5eNiNT95Gp+8i0dbTHXFuyXtc/NWaMUVFRkVavXq0NGzYoPT09ZP+QIUMUGRmpiooKZ9vevXt14MABZWZmSpIyMzO1Y8eOkE93eTwexcfHq1+/fs7Mmcdommk6RlRUlIYMGRIy09jYqIqKCmcGAADYzfVXhAoLC7VixQr95je/UadOnZz39CQkJCg2NlYJCQkqKChQcXGxOnfurPj4eP3oRz9SZmambrrpJklSdna2+vXrp/vuu09z586Vz+fT9OnTVVhY6Lxi8+CDD+r555/XlClT9L3vfU8bNmzQa6+9ptdff91ZS3FxsfLz8zV06FDdeOONWrBggU6cOKEHHnjA7csGAADtkOtFaMmSJZKk2267LWT7Sy+9pO9+97uSpJ/85CcKDw/X6NGjFQgElJOToxdeeMGZjYiI0Nq1a/XQQw8pMzNTHTt2VH5+vmbNmuXMpKen6/XXX9fDDz+shQsX6uqrr9bPf/5z5eT833tt7rnnHv3lL39RSUmJfD6fBg8erPLy8mZvoAYAAHZyvQgZYy44ExMTo8WLF2vx4sWfO9OzZ88LfpLqtttu0/bt2887U1RUpKKioguuCQAA2IffNQYAAKxFEQIAANaiCAEAAGtRhAAAgLUoQgAAwFoUIQAAYC2KEAAAsBZFCAAAWIsiBAAArEURAgAA1qIIAQAAa1GEAACAtShCAADAWhQhAABgLYoQAACwFkUIAABYiyIEAACsRRECAADWoggBAABrUYQAAIC1KEIAAMBaFCEAAGAtihAAALAWRQgAAFiLIgQAAKxFEQIAANaiCAEAAGtRhAAAgLUoQgAAwFoUIQAAYC2KEAAAsBZFCAAAWIsiBAAArEURAgAA1qIIAQAAa1GELgN/irlXmpHQ1ssAAMA6FCEAAGAtihAAALAWRQgAAFiLIgQAAKxFEQIAANaiCAEAAGtRhAAAgLU6tPUCIA1I7yFJ2tHG6wAAwDa8IgQAAKxFEQIAANbi1thl4LXS03//S37brgMAANtY8YrQ4sWL1atXL8XExCgjI0ObN29u6yWdU69HX3f+AACA1nfFF6GVK1equLhYTzzxhLZt26ZBgwYpJydHhw4dauulAQCANnbFF6H58+dr/PjxeuCBB9SvXz8tXbpUcXFxevHFF9t6ac106vuo80czEviN9AAAtLIr+j1C9fX1qq6u1tSpU51t4eHhysrKktfrbTYfCAQUCAScx0ePHpUkHT58WMFg0LV1BYNBnTx5Uv4OHRTe0OBs/9k86aGiv39Jrkv5V0nSsf/3WrPnV00d4dpariRNuf7tb39TZGRkWy/nikCm7iNT95Fp62jPuR47dkySZIy54OwVXYT++te/qqGhQcnJySHbk5OTtWfPnmbzpaWlmjlzZrPt6enprbbGZh46e8M9zUa6PntJVgIAQLt27NgxJSSc/+7KFV2EWmrq1KkqLi52Hjc2Nurw4cPq0qWLwsLCXDuP3+9XWlqaPvnkE8XHx7t2XNuRq/vI1H1k6j4ybR3tOVdjjI4dO6bU1NQLzl7RRahr166KiIhQbW1tyPba2lqlpKQ0m4+OjlZ0dHTItsTExFZbX3x8fLv75moPyNV9ZOo+MnUfmbaO9prrhV4JanJFv1k6KipKQ4YMUUVFhbOtsbFRFRUVyszMbMOVAQCAy8EV/YqQJBUXFys/P19Dhw7VjTfeqAULFujEiRN64IEH2nppAACgjV3xReiee+7RX/7yF5WUlMjn82nw4MEqLy9v9gbqSyk6OlpPPPFEs9tw+GLI1X1k6j4ydR+Ztg5bcg0zF/PZMgAAgCvQFf0eIQAAgPOhCAEAAGtRhAAAgLUoQgAAwFoUoTawePFi9erVSzExMcrIyNDmzZvbekmXrdLSUt1www3q1KmTunXrptzcXO3duzdk5tSpUyosLFSXLl101VVXafTo0c1+iOaBAwc0atQoxcXFqVu3bpo8ebJOnz59KS/lsvXMM88oLCxMkyZNcraRact9+umn+s53vqMuXbooNjZWAwYM0NatW539xhiVlJSoe/fuio2NVVZWlvbt2xdyjMOHDysvL0/x8fFKTExUQUGBjh8/fqkv5bLQ0NCgxx9/XOnp6YqNjdWXv/xlPfnkkyG/O4pML6yyslJ33nmnUlNTFRYWprKyspD9bmX4wQcf6JZbblFMTIzS0tI0d+7c1r409xhcUq+++qqJiooyL774otm1a5cZP368SUxMNLW1tW29tMtSTk6Oeemll8zOnTtNTU2N+drXvmZ69Ohhjh8/7sw8+OCDJi0tzVRUVJitW7eam266yQwbNszZf/r0adO/f3+TlZVltm/fbtatW2e6du1qpk6d2haXdFnZvHmz6dWrlxk4cKCZOHGis51MW+bw4cOmZ8+e5rvf/a6pqqoyf/zjH82bb75pfv/73zszzzzzjElISDBlZWXm/fffN9/4xjdMenq6+eyzz5yZkSNHmkGDBpn33nvPvP322+YrX/mKGTduXFtcUpubPXu26dKli1m7dq3Zv3+/WbVqlbnqqqvMwoULnRkyvbB169aZxx57zPz61782kszq1atD9ruR4dGjR01ycrLJy8szO3fuNK+88oqJjY01P/3pTy/VZX4hFKFL7MYbbzSFhYXO44aGBpOammpKS0vbcFXtx6FDh4wks3HjRmOMMXV1dSYyMtKsWrXKmfnwww+NJOP1eo0xf/8fQXh4uPH5fM7MkiVLTHx8vAkEApf2Ai4jx44dM9dcc43xeDzm3//9350iRKYt98gjj5jhw4d/7v7GxkaTkpJi5s2b52yrq6sz0dHR5pVXXjHGGLN7924jyWzZssWZeeONN0xYWJj59NNPW2/xl6lRo0aZ733veyHbvvnNb5q8vDxjDJn+M84uQm5l+MILL5ikpKSQ//YfeeQR07t371a+Indwa+wSqq+vV3V1tbKyspxt4eHhysrKktfrbcOVtR9Hjx6VJHXu3FmSVF1drWAwGJJpnz591KNHDydTr9erAQMGhPwQzZycHPn9fu3atesSrv7yUlhYqFGjRoVkJ5HpP2PNmjUaOnSoxowZo27duum6667Tz372M2f//v375fP5QjJNSEhQRkZGSKaJiYkaOnSoM5OVlaXw8HBVVVVduou5TAwbNkwVFRX66KOPJEnvv/++3nnnHX31q1+VRKZucCtDr9erW2+9VVFRUc5MTk6O9u7dqyNHjlyiq/nnXfE/Wfpy8te//lUNDQ3Nfqp1cnKy9uzZ00araj8aGxs1adIk3Xzzzerfv78kyefzKSoqqtkvx01OTpbP53NmzpV50z4bvfrqq9q2bZu2bNnSbB+Zttwf//hHLVmyRMXFxZo2bZq2bNmiH//4x4qKilJ+fr6TybkyOzPTbt26hezv0KGDOnfubGWmjz76qPx+v/r06aOIiAg1NDRo9uzZysvLkyQydYFbGfp8PqWnpzc7RtO+pKSkVlm/WyhCaDcKCwu1c+dOvfPOO229lHbtk08+0cSJE+XxeBQTE9PWy7kiNDY2aujQoXr66aclSdddd5127typpUuXKj8/v41X1z699tprWr58uVasWKFrr71WNTU1mjRpklJTU8kUruLW2CXUtWtXRURENPv0TW1trVJSUtpoVe1DUVGR1q5dq9/97ne6+uqrne0pKSmqr69XXV1dyPyZmaakpJwz86Z9tqmurtahQ4d0/fXXq0OHDurQoYM2btyoRYsWqUOHDkpOTibTFurevbv69esXsq1v3746cOCApP/L5Hz/7aekpOjQoUMh+0+fPq3Dhw9bmenkyZP16KOPauzYsRowYIDuu+8+PfzwwyotLZVEpm5wK8P2/v8DitAlFBUVpSFDhqiiosLZ1tjYqIqKCmVmZrbhyi5fxhgVFRVp9erV2rBhQ7OXX4cMGaLIyMiQTPfu3asDBw44mWZmZmrHjh0h/zF7PB7Fx8c3+8fLBiNGjNCOHTtUU1Pj/Bk6dKjy8vKcv5Npy9x8883NfqzDRx99pJ49e0qS0tPTlZKSEpKp3+9XVVVVSKZ1dXWqrq52ZjZs2KDGxkZlZGRcgqu4vJw8eVLh4aH/REVERKixsVESmbrBrQwzMzNVWVmpYDDozHg8HvXu3fuyvy0miY/PX2qvvvqqiY6ONi+//LLZvXu3mTBhgklMTAz59A3+z0MPPWQSEhLMW2+9Zf785z87f06ePOnMPPjgg6ZHjx5mw4YNZuvWrSYzM9NkZmY6+5s+6p2dnW1qampMeXm5+Zd/+RdrP+p9Lmd+aswYMm2pzZs3mw4dOpjZs2ebffv2meXLl5u4uDjzq1/9ypl55plnTGJiovnNb35jPvjgA3PXXXed82PK1113namqqjLvvPOOueaaa6z6qPeZ8vPzzZe+9CXn4/O//vWvTdeuXc2UKVOcGTK9sGPHjpnt27eb7du3G0lm/vz5Zvv27ebjjz82xriTYV1dnUlOTjb33Xef2blzp3n11VdNXFwcH5/H53vuuedMjx49TFRUlLnxxhvNe++919ZLumxJOuefl156yZn57LPPzA9/+EOTlJRk4uLizN13323+/Oc/hxznT3/6k/nqV79qYmNjTdeuXc1//ud/mmAweImv5vJ1dhEi05b77W9/a/r372+io6NNnz59zLJly0L2NzY2mscff9wkJyeb6OhoM2LECLN3796Qmb/97W9m3Lhx5qqrrjLx8fHmgQceMMeOHbuUl3HZ8Pv9ZuLEiaZHjx4mJibG/Ou//qt57LHHQj6iTaYX9rvf/e6c/w/Nz883xriX4fvvv2+GDx9uoqOjzZe+9CXzzDPPXKpL/MLCjDnjx3QCAABYhPcIAQAAa1GEAACAtShCAADAWhQhAABgLYoQAACwFkUIAABYiyIEAACsRRECAADWoggBAABrUYQAAIC1KEIAAMBaFCEAAGCt/w8PpBsb+LcOtAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def process_clustering(name):\n",
    "    clustering = pd.read_csv('clustering/{}_cluster.tsv'.format(name), delimiter='\\t', header=None)\n",
    "    #rename heading as cluster reference and id\n",
    "    clustering.columns = [name, 'Entry']\n",
    "    clustering.drop_duplicates(subset='Entry', keep='first', inplace=True)\n",
    "    print(clustering[name].nunique())\n",
    "    clustering[name].value_counts().hist(bins=100)\n",
    "    return clustering\n",
    "\n",
    "swissprot = swissprot[['Entry', 'Entry Name', 'Sequence', 'EC number', 'Length', 'EC All']]\n",
    "\n",
    "cluster50 = process_clustering('clusterRes50')\n",
    "cluster30 = process_clustering('clusterRes30')\n",
    "cluster70 = process_clustering('clusterRes70')\n",
    "cluster90 = process_clustering('clustRes90')\n",
    "\n",
    "swissprot = pd.merge(swissprot, cluster50, on='Entry', how='left')\n",
    "swissprot = pd.merge(swissprot, cluster30, on='Entry', how='left')\n",
    "swissprot = pd.merge(swissprot, cluster70, on='Entry', how='left')\n",
    "swissprot = pd.merge(swissprot, cluster90, on='Entry', how='left')\n",
    "\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "63ca5f5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "swissprot['EC3'] = swissprot['EC number'].str.split('.').str[:3].str.join('.')\n",
    "swissprot['EC2'] = swissprot['EC number'].str.split('.').str[:2].str.join('.')\n",
    "swissprot['EC1'] = swissprot['EC number'].str.split('.').str[:1].str.join('.')\n",
    "swissprot.to_csv('../processed_data/protein2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "756c5423",
   "metadata": {},
   "source": [
    "# Finally process textual description of EC numbers\n",
    "This is optionally used in CREEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b5f94ed4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EC number</th>\n",
       "      <th>Text</th>\n",
       "      <th>Text Incomplete</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2679</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835</th>\n",
       "      <td>1.1.1.100</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2241</th>\n",
       "      <td>1.1.1.101</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2212</th>\n",
       "      <td>1.1.1.102</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>7.6.2.5</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1812</th>\n",
       "      <td>7.6.2.6</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4813</th>\n",
       "      <td>7.6.2.7</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1019</th>\n",
       "      <td>7.6.2.8</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3281</th>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5088 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      EC number                                               Text  \\\n",
       "375     1.1.1.1  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "2679   1.1.1.10  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "835   1.1.1.100  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "2241  1.1.1.101  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "2212  1.1.1.102  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "...         ...                                                ...   \n",
       "102     7.6.2.5  primary active transmembrane transporter; ABC-...   \n",
       "1812    7.6.2.6  primary active transmembrane transporter; ABC-...   \n",
       "4813    7.6.2.7  primary active transmembrane transporter; ABC-...   \n",
       "1019    7.6.2.8  primary active transmembrane transporter; ABC-...   \n",
       "3281    7.6.2.9  primary active transmembrane transporter; ABC-...   \n",
       "\n",
       "      Text Incomplete  \n",
       "375             False  \n",
       "2679            False  \n",
       "835             False  \n",
       "2241            False  \n",
       "2212            False  \n",
       "...               ...  \n",
       "102             False  \n",
       "1812            False  \n",
       "4813            False  \n",
       "1019            False  \n",
       "3281            False  \n",
       "\n",
       "[5088 rows x 3 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read ECtoGO.txt line by line\n",
    "ECtoGO = open(\"raw_data/ECtoGO_raw.txt\", \"r\")\n",
    "ECtoGO_lines = ECtoGO.readlines()\n",
    "ECtoGO.close()\n",
    "\n",
    "#skip the first two lines\n",
    "ECtoGO_lines = ECtoGO_lines[2:]\n",
    "EC2desc = {}\n",
    "for line in ECtoGO_lines:\n",
    "    line = line.strip().split(\">\")\n",
    "    EC = line[0][3:-1]\n",
    "    desc = line[1].split(\";\")[0][4:-1]\n",
    "    EC2desc[EC] = desc\n",
    "\n",
    "def get_EC_desc(EC):\n",
    "    num_dashes = EC.count('-')\n",
    "\n",
    "    description_missing = False\n",
    "    EC1 = '.'.join(EC.split('.')[:1])\n",
    "    if EC1 + '.-.-.-' in EC2desc and num_dashes < 3:\n",
    "        desc1 = EC2desc[EC1 + '.-.-.-']\n",
    "    else:\n",
    "         desc1 = ''\n",
    "    EC2 = '.'.join(EC.split('.')[:2]) \n",
    "    if EC2 + '.-.-' in EC2desc and num_dashes < 2:\n",
    "        desc2 = EC2desc[EC2 + '.-.-']\n",
    "    else:\n",
    "        desc2 = ''\n",
    "    EC3 = '.'.join(EC.split('.')[:3])\n",
    "    if EC3 + '.-' in EC2desc and num_dashes < 1:\n",
    "        desc3 = EC2desc[EC3 + '.-']\n",
    "    else:\n",
    "        desc3 = ''\n",
    "        \n",
    "    if EC in EC2desc:\n",
    "        desc4 = EC2desc[EC]\n",
    "    else:\n",
    "        description_missing = True\n",
    "        desc4 = ''\n",
    "\n",
    "  \n",
    "    description = desc1 + '; ' + desc2 + '; ' + desc3 + '; ' + desc4\n",
    "    description = description.replace(' ; ', ' ')\n",
    "    description = description.replace(' ; ', ' ')\n",
    "    description = description.replace(' ; ', ' ')\n",
    "    description = description.replace(' activity', '')\n",
    "    #if string starts with ;, replace with space\n",
    "    if description[0] == ';':\n",
    "        description = description[2:]\n",
    "    if description[-2:] == '; ':\n",
    "        description = description[:-2]\n",
    "    return description\n",
    "\n",
    "#subset to the EC numbers in swissprot\n",
    "EC2desc_df = pd.DataFrame(swissprot['EC number'].unique(), columns=['EC number'])\n",
    "EC2desc_df['Text'] = EC2desc_df['EC number'].apply(get_EC_desc)\n",
    "EC2desc_df['Text Incomplete'] = ~EC2desc_df['EC number'].isin(EC2desc.keys())\n",
    "EC2desc_df = EC2desc_df.sort_values(by='EC number')\n",
    "EC2desc_df = EC2desc_df[EC2desc_df['EC number'].isin(swissprot['EC number'].unique())]\n",
    "EC2desc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "bda4ac06",
   "metadata": {},
   "outputs": [],
   "source": [
    "EC2desc_df.to_csv('../processed_data/text2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c71ca7a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save ECs to a txt as the order of the cluster centers for downstream tasks\n",
    "with open('../processed_data/EC_list.txt', 'w') as f:\n",
    "    for ec in EC2desc_df['EC number']:\n",
    "        f.write(ec + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "401a1818-e84f-4b05-acb2-8dc734cd8c8a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70269638-6b8d-4ab4-8da2-2b1245085f81",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edde741a-fc23-4c5a-b543-4d7e5e2ab329",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc38c1de-37c5-4b8c-8950-5779d9548908",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
