{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "06e5ee54-4659-4353-b6ad-de92dd9ab319",
   "metadata": {},
   "source": [
    "# Data download and processing from Swissprot\n",
    "\n",
    "Downloaded the data on 13th of May 2024. 571,282 results filtering for reviewed \"Swiss-Prot\". (https://www.uniprot.org/uniprotkb?query=*&facets=reviewed%3Atrue)\n",
    "\n",
    "Selected download TSV and the columns, Seqeunce (under Sequences tab), EC number (under Function).\n",
    "\n",
    "Unzipped the downloaded file.\n",
    "\n",
    "## 1. Protein to EC processing\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "4f961afc-aedd-41b9-8635-4c6eabd017ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "      <td>796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "      <td>3423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254924</th>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>Y3032_BACAN</td>\n",
       "      <td>MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254925</th>\n",
       "      <td>Q6L032</td>\n",
       "      <td>Y1085_PICTO</td>\n",
       "      <td>MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254926</th>\n",
       "      <td>Q7TDB6</td>\n",
       "      <td>RDRP_MYRV9</td>\n",
       "      <td>MSLTSRYTHFVPDSTITEILNDSNTPQILLHYANIVNGSTPVHFTS...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>1354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254927</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254928</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>254929 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "1       A0A023I7E1   ENG1_RHIMI   \n",
       "2       A0A024B7W1   POLG_ZIKVF   \n",
       "3       A0A024SC78  CUTI1_HYPJR   \n",
       "4       A0A024SH76   GUX2_HYPJR   \n",
       "...            ...          ...   \n",
       "254924      Q6HX62  Y3032_BACAN   \n",
       "254925      Q6L032  Y1085_PICTO   \n",
       "254926      Q7TDB6   RDRP_MYRV9   \n",
       "254927      Q94MV8   VG56_BPLZ5   \n",
       "254928      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...   \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   \n",
       "...                                                   ...   \n",
       "254924  MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...   \n",
       "254925  MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...   \n",
       "254926  MSLTSRYTHFVPDSTITEILNDSNTPQILLHYANIVNGSTPVHFTS...   \n",
       "254927  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...   \n",
       "254928  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...   \n",
       "\n",
       "                                                EC number  Length  \n",
       "0                                        3.2.2.-; 3.2.2.6     269  \n",
       "1                                                3.2.1.39     796  \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...    3423  \n",
       "3                                                3.1.1.74     248  \n",
       "4                                                3.2.1.91     471  \n",
       "...                                                   ...     ...  \n",
       "254924                                            3.5.4.2     584  \n",
       "254925                                            3.5.4.2     573  \n",
       "254926                                           2.7.7.48    1354  \n",
       "254927                                           3.6.1.12     172  \n",
       "254928                                           3.1.1.29     124  \n",
       "\n",
       "[254929 rows x 5 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "swissprot = pd.read_csv('raw_data/uniprotkb_AND_reviewed_true_AND_protein_2024_05_01.tsv', sep='\\t')\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "160aa2c9-277e-427f-8e52-58d77299d79e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A0A061I403</td>\n",
       "      <td>FICD_CRIGR</td>\n",
       "      <td>MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...</td>\n",
       "      <td>2.7.7.108; 3.1.4.-</td>\n",
       "      <td>455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254913</th>\n",
       "      <td>P36352</td>\n",
       "      <td>POLR_PHMV</td>\n",
       "      <td>VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254914</th>\n",
       "      <td>P39262</td>\n",
       "      <td>VG56_BPT4</td>\n",
       "      <td>MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254923</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254927</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254928</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>198392 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "3       A0A024SC78  CUTI1_HYPJR   \n",
       "4       A0A024SH76   GUX2_HYPJR   \n",
       "6       A0A059TC02   CCR1_PETHY   \n",
       "8       A0A061I403   FICD_CRIGR   \n",
       "...            ...          ...   \n",
       "254913      P36352    POLR_PHMV   \n",
       "254914      P39262    VG56_BPT4   \n",
       "254923      Q05115   AMDA_BORBO   \n",
       "254927      Q94MV8   VG56_BPLZ5   \n",
       "254928      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence           EC number  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.-; 3.2.2.6   \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...            3.1.1.74   \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...            3.2.1.91   \n",
       "6       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...            1.2.1.44   \n",
       "8       MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...  2.7.7.108; 3.1.4.-   \n",
       "...                                                   ...                 ...   \n",
       "254913  VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...            2.7.7.48   \n",
       "254914  MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...            3.6.1.12   \n",
       "254923  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...            4.1.1.76   \n",
       "254927  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...            3.6.1.12   \n",
       "254928  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...            3.1.1.29   \n",
       "\n",
       "        Length  \n",
       "0          269  \n",
       "3          248  \n",
       "4          471  \n",
       "6          333  \n",
       "8          455  \n",
       "...        ...  \n",
       "254913     178  \n",
       "254914     171  \n",
       "254923     240  \n",
       "254927     172  \n",
       "254928     124  \n",
       "\n",
       "[198392 rows x 5 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter to only include sequneces < 512\n",
    "swissprot = swissprot[swissprot['Length'] < 512]\n",
    "swissprot = swissprot[swissprot['Length'] >= 100 ]\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee69eeba-0d87-4f0e-92f8-8d950cc616f3",
   "metadata": {},
   "source": [
    "## Explode out EC number so each EC has it's own row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "724377c7-9690-4cb2-ad79-e081cc7f9fb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "195265\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3559200/662554174.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  swissprot['EC number'] = [ec.split(';') for ec in swissprot['EC number']]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254913</th>\n",
       "      <td>P36352</td>\n",
       "      <td>POLR_PHMV</td>\n",
       "      <td>VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254914</th>\n",
       "      <td>P39262</td>\n",
       "      <td>VG56_BPT4</td>\n",
       "      <td>MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254923</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254927</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254928</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>208984 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "3       A0A024SC78  CUTI1_HYPJR   \n",
       "4       A0A024SH76   GUX2_HYPJR   \n",
       "6       A0A059TC02   CCR1_PETHY   \n",
       "...            ...          ...   \n",
       "254913      P36352    POLR_PHMV   \n",
       "254914      P39262    VG56_BPT4   \n",
       "254923      Q05115   AMDA_BORBO   \n",
       "254927      Q94MV8   VG56_BPLZ5   \n",
       "254928      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence EC number  Length  \n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.-     269  \n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.6     269  \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...  3.1.1.74     248  \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...  3.2.1.91     471  \n",
       "6       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...  1.2.1.44     333  \n",
       "...                                                   ...       ...     ...  \n",
       "254913  VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...  2.7.7.48     178  \n",
       "254914  MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...  3.6.1.12     171  \n",
       "254923  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...  4.1.1.76     240  \n",
       "254927  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...  3.6.1.12     172  \n",
       "254928  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...  3.1.1.29     124  \n",
       "\n",
       "[208984 rows x 5 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop rows that don't have an ec number\n",
    "swissprot = swissprot[swissprot['EC number'].notna()]\n",
    "print(len(swissprot))\n",
    "# Now expand out the ones we have left\n",
    "swissprot['EC number'] = [ec.split(';') for ec in swissprot['EC number']]\n",
    "swissprot = swissprot.explode('EC number')\n",
    "# Clean the EC numbers\n",
    "swissprot['EC number'] = [ec.replace(' ', '') for ec in swissprot['EC number']]\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4b81d7a",
   "metadata": {},
   "source": [
    "## Filtering steps \n",
    "\n",
    "* Filter to a certain sequence length\n",
    "* Drop duplicates\n",
    "* Remove EC numbers with dash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e42c854f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A0A061I403</td>\n",
       "      <td>FICD_CRIGR</td>\n",
       "      <td>MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...</td>\n",
       "      <td>2.7.7.108</td>\n",
       "      <td>455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254913</th>\n",
       "      <td>P36352</td>\n",
       "      <td>POLR_PHMV</td>\n",
       "      <td>VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254914</th>\n",
       "      <td>P39262</td>\n",
       "      <td>VG56_BPT4</td>\n",
       "      <td>MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254923</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254927</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254928</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161738 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "3       A0A024SC78  CUTI1_HYPJR   \n",
       "4       A0A024SH76   GUX2_HYPJR   \n",
       "6       A0A059TC02   CCR1_PETHY   \n",
       "8       A0A061I403   FICD_CRIGR   \n",
       "...            ...          ...   \n",
       "254913      P36352    POLR_PHMV   \n",
       "254914      P39262    VG56_BPT4   \n",
       "254923      Q05115   AMDA_BORBO   \n",
       "254927      Q94MV8   VG56_BPLZ5   \n",
       "254928      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence  EC number  Length  \n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.6     269  \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   3.1.1.74     248  \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   3.2.1.91     471  \n",
       "6       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...   1.2.1.44     333  \n",
       "8       MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...  2.7.7.108     455  \n",
       "...                                                   ...        ...     ...  \n",
       "254913  VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...   2.7.7.48     178  \n",
       "254914  MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...   3.6.1.12     171  \n",
       "254923  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...   4.1.1.76     240  \n",
       "254927  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...   3.6.1.12     172  \n",
       "254928  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...   3.1.1.29     124  \n",
       "\n",
       "[161738 rows x 5 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "swissprot = swissprot[~swissprot['EC number'].str.contains('-')]\n",
    "swissprot = swissprot.drop_duplicates(subset=['Sequence', 'EC number'])\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4607689a",
   "metadata": {},
   "source": [
    "### Comparison to the dataset used in CLEAN\n",
    "\n",
    "CLEAN: https://www.science.org/doi/10.1126/science.adf2465 \n",
    "\n",
    "\n",
    "Dataset was downloaded from the CLEA repository: https://github.com/tttianhao/CLEAN/tree/main/app/data specifically the split 100 dataset, alternatively they can be downloaded from: https://drive.google.com/file/d/1MzYQ--CRK253eAKOmZrD1nzF-mXHfXtq/view\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9d4113b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P51451</td>\n",
       "      <td>2.7.10.2</td>\n",
       "      <td>MGLVSSKKPDKEKPIKEKDKGQWSPLKVSAQDKDAPPLPPLVVFNH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C7C422</td>\n",
       "      <td>3.5.2.6</td>\n",
       "      <td>MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q07457</td>\n",
       "      <td>2.3.2.27</td>\n",
       "      <td>MTAEPATKKIKLELSDPSEPLTQSDVIAFQKEALFRCINRRRVDFE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q07009</td>\n",
       "      <td>3.4.22.53</td>\n",
       "      <td>MAGIAMKLAKDREAAEGLGSHERAIKYLNQDYETLRNECLEAGALF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q9TTH8</td>\n",
       "      <td>3.4.22.54</td>\n",
       "      <td>MPTVISASVAPRTGAEPRSPGPIAQAAQGKGTEAGGGNPSGIYSAI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227357</th>\n",
       "      <td>Q97HD1</td>\n",
       "      <td>3.1.11.6</td>\n",
       "      <td>MPSKKESYESMIKELEKIVSSMENEELPLEEAMKNYEDGVKLCDKL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227358</th>\n",
       "      <td>A0A140CUL9</td>\n",
       "      <td>1.1.99.13</td>\n",
       "      <td>MAENNTYDAIVIGSGISGGWATKELTEKGLKTLMLERGRNIEHVKD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227359</th>\n",
       "      <td>W6JQJ6</td>\n",
       "      <td>1.13.12.2</td>\n",
       "      <td>MNKNNRHPADGKKPITIFGPDFPFAFDDWLEHPAGLGSIPAARHGE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227360</th>\n",
       "      <td>B3IVI7</td>\n",
       "      <td>3.5.1.30</td>\n",
       "      <td>MRIALYQGAPKPLDVPGNLQRLRHQAQLAAERGAQLLVCPEMFLTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227361</th>\n",
       "      <td>Q88QV2</td>\n",
       "      <td>3.5.1.30</td>\n",
       "      <td>MRIALYQGAPKPLDVPGNLQRLRHQAQLAADRGAQLLVCPEMFLSG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>191488 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry  EC number  \\\n",
       "0           P51451   2.7.10.2   \n",
       "1           C7C422    3.5.2.6   \n",
       "2           Q07457   2.3.2.27   \n",
       "3           Q07009  3.4.22.53   \n",
       "4           Q9TTH8  3.4.22.54   \n",
       "...            ...        ...   \n",
       "227357      Q97HD1   3.1.11.6   \n",
       "227358  A0A140CUL9  1.1.99.13   \n",
       "227359      W6JQJ6  1.13.12.2   \n",
       "227360      B3IVI7   3.5.1.30   \n",
       "227361      Q88QV2   3.5.1.30   \n",
       "\n",
       "                                                 Sequence  \n",
       "0       MGLVSSKKPDKEKPIKEKDKGQWSPLKVSAQDKDAPPLPPLVVFNH...  \n",
       "1       MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...  \n",
       "2       MTAEPATKKIKLELSDPSEPLTQSDVIAFQKEALFRCINRRRVDFE...  \n",
       "3       MAGIAMKLAKDREAAEGLGSHERAIKYLNQDYETLRNECLEAGALF...  \n",
       "4       MPTVISASVAPRTGAEPRSPGPIAQAAQGKGTEAGGGNPSGIYSAI...  \n",
       "...                                                   ...  \n",
       "227357  MPSKKESYESMIKELEKIVSSMENEELPLEEAMKNYEDGVKLCDKL...  \n",
       "227358  MAENNTYDAIVIGSGISGGWATKELTEKGLKTLMLERGRNIEHVKD...  \n",
       "227359  MNKNNRHPADGKKPITIFGPDFPFAFDDWLEHPAGLGSIPAARHGE...  \n",
       "227360  MRIALYQGAPKPLDVPGNLQRLRHQAQLAAERGAQLLVCPEMFLTG...  \n",
       "227361  MRIALYQGAPKPLDVPGNLQRLRHQAQLAADRGAQLLVCPEMFLSG...  \n",
       "\n",
       "[191488 rows x 3 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_df = pd.read_csv('raw_data/CLEAN_split100.tsv', sep='\\t')\n",
    "clean_df = clean_df.drop_duplicates(subset=['Sequence', 'EC number'])\n",
    "clean_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "3cdeac1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P51451</td>\n",
       "      <td>2.7.10.2</td>\n",
       "      <td>MGLVSSKKPDKEKPIKEKDKGQWSPLKVSAQDKDAPPLPPLVVFNH...</td>\n",
       "      <td>505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C7C422</td>\n",
       "      <td>3.5.2.6</td>\n",
       "      <td>MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...</td>\n",
       "      <td>270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P00917</td>\n",
       "      <td>4.2.1.1</td>\n",
       "      <td>MAHSDWGYDSPNGPZEWVKLYPIANGNNQSPIDIKTSETKHDTSLK...</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q9ULX7</td>\n",
       "      <td>4.2.1.1</td>\n",
       "      <td>MLFSALLLEVIWILAADGGQHWTYEGPHGQDHWPASYPECGNNAQS...</td>\n",
       "      <td>337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>P10506</td>\n",
       "      <td>2.7.12.2</td>\n",
       "      <td>MFKRRRNPKGLVLNPNASVKSSDNDHKEELINNQKSFESNVEAFME...</td>\n",
       "      <td>340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227341</th>\n",
       "      <td>Q9KQH9</td>\n",
       "      <td>2.3.1.179</td>\n",
       "      <td>MSKRRVVVTGMGMLSPVGNTVESSWKALLAGQSGIVNIEHFDTTNF...</td>\n",
       "      <td>414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227352</th>\n",
       "      <td>Q11WK3</td>\n",
       "      <td>6.3.5.2</td>\n",
       "      <td>MTEKILILDFGSQYTQLIARRVRELNVYCEIHPYNKAPQIDGTVKG...</td>\n",
       "      <td>509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227354</th>\n",
       "      <td>B1KNT2</td>\n",
       "      <td>4.2.1.59</td>\n",
       "      <td>MSEQLNTMDIKEIMSFLPHRYPFLLIDRVLDYTPGETLHAIKNVTI...</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227360</th>\n",
       "      <td>B3IVI7</td>\n",
       "      <td>3.5.1.30</td>\n",
       "      <td>MRIALYQGAPKPLDVPGNLQRLRHQAQLAAERGAQLLVCPEMFLTG...</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227361</th>\n",
       "      <td>Q88QV2</td>\n",
       "      <td>3.5.1.30</td>\n",
       "      <td>MRIALYQGAPKPLDVPGNLQRLRHQAQLAADRGAQLLVCPEMFLSG...</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>151801 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Entry  EC number                                           Sequence  \\\n",
       "0       P51451   2.7.10.2  MGLVSSKKPDKEKPIKEKDKGQWSPLKVSAQDKDAPPLPPLVVFNH...   \n",
       "1       C7C422    3.5.2.6  MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...   \n",
       "6       P00917    4.2.1.1  MAHSDWGYDSPNGPZEWVKLYPIANGNNQSPIDIKTSETKHDTSLK...   \n",
       "7       Q9ULX7    4.2.1.1  MLFSALLLEVIWILAADGGQHWTYEGPHGQDHWPASYPECGNNAQS...   \n",
       "11      P10506   2.7.12.2  MFKRRRNPKGLVLNPNASVKSSDNDHKEELINNQKSFESNVEAFME...   \n",
       "...        ...        ...                                                ...   \n",
       "227341  Q9KQH9  2.3.1.179  MSKRRVVVTGMGMLSPVGNTVESSWKALLAGQSGIVNIEHFDTTNF...   \n",
       "227352  Q11WK3    6.3.5.2  MTEKILILDFGSQYTQLIARRVRELNVYCEIHPYNKAPQIDGTVKG...   \n",
       "227354  B1KNT2   4.2.1.59  MSEQLNTMDIKEIMSFLPHRYPFLLIDRVLDYTPGETLHAIKNVTI...   \n",
       "227360  B3IVI7   3.5.1.30  MRIALYQGAPKPLDVPGNLQRLRHQAQLAAERGAQLLVCPEMFLTG...   \n",
       "227361  Q88QV2   3.5.1.30  MRIALYQGAPKPLDVPGNLQRLRHQAQLAADRGAQLLVCPEMFLSG...   \n",
       "\n",
       "        Length  \n",
       "0          505  \n",
       "1          270  \n",
       "6          261  \n",
       "7          337  \n",
       "11         340  \n",
       "...        ...  \n",
       "227341     414  \n",
       "227352     509  \n",
       "227354     152  \n",
       "227360     264  \n",
       "227361     264  \n",
       "\n",
       "[151801 rows x 4 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_df['Length'] = [len(seq) for seq in clean_df['Sequence']]\n",
    "clean_df = clean_df[clean_df['Length'] < 512]\n",
    "clean_df = clean_df[clean_df['Length'] >= 100]\n",
    "clean_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8d17ad2",
   "metadata": {},
   "source": [
    "## 2. Reaction to EC processing\n",
    "\n",
    "EC and reaction data were downloaded from multiple sources, the first being ECReact.\n",
    "\n",
    "1. ECReact: https://www.nature.com/articles/s41467-022-28536-w, with the data downloaded from https://github.com/rxn4chemistry/biocatalysis-model specifically, https://github.com/rxn4chemistry/biocatalysis-model/blob/main/data/ecreact-1.0.csv\n",
    "2. EnzymeMap: https://chemrxiv.org/engage/chemrxiv/article-details/643e9f4034f4ff2ea4f4f84e https://zenodo.org/records/8254726 and the data https://github.com/hesther/enzymemap, using the data file https://github.com/hesther/enzymemap/blob/main/data/processed_reactions.csv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "b6d71562",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>1.2.1.8</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C=C(C)CCOP(=O)([O-])OP(=O)([O-])[O-].CC(C)=CCO...</td>\n",
       "      <td>2.5.1.85</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O...</td>\n",
       "      <td>1.4.1.4</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>1.1.1.3</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cc1cn([C@H]2C[C@H](O)[C@@H](COP(=O)(O)OP(=O)(O...</td>\n",
       "      <td>2.7.7.9</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62217</th>\n",
       "      <td>CCCCC/C=C\\C/C=C\\C/C=C\\C/C=C\\C/C=C\\CCC(=O)O[C@@...</td>\n",
       "      <td>2.3.1.76</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62218</th>\n",
       "      <td>CCCCCCCCCCCC/C=C\\CCCCCCCC(=O)OC[C@H](COP(=O)(O...</td>\n",
       "      <td>2.7.8.41</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62219</th>\n",
       "      <td>CCCCCCCCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCCCCCCC...</td>\n",
       "      <td>2.7.8.41</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62220</th>\n",
       "      <td>CCCCCC/C=C\\CCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCC...</td>\n",
       "      <td>2.7.8.41</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62221</th>\n",
       "      <td>CCCCCCCC/C=C\\CCCCCCCC(=O)O[C@H](COC(=O)CCCCCCC...</td>\n",
       "      <td>4.1.1.65</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62222 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Reaction EC number  \\\n",
       "0      NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   1.2.1.8   \n",
       "1      C=C(C)CCOP(=O)([O-])OP(=O)([O-])[O-].CC(C)=CCO...  2.5.1.85   \n",
       "2      N.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O...   1.4.1.4   \n",
       "3      NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...   1.1.1.3   \n",
       "4      Cc1cn([C@H]2C[C@H](O)[C@@H](COP(=O)(O)OP(=O)(O...   2.7.7.9   \n",
       "...                                                  ...       ...   \n",
       "62217  CCCCC/C=C\\C/C=C\\C/C=C\\C/C=C\\C/C=C\\CCC(=O)O[C@@...  2.3.1.76   \n",
       "62218  CCCCCCCCCCCC/C=C\\CCCCCCCC(=O)OC[C@H](COP(=O)(O...  2.7.8.41   \n",
       "62219  CCCCCCCCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCCCCCCC...  2.7.8.41   \n",
       "62220  CCCCCC/C=C\\CCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCC...  2.7.8.41   \n",
       "62221  CCCCCCCC/C=C\\CCCCCCCC(=O)O[C@H](COC(=O)CCCCCCC...  4.1.1.65   \n",
       "\n",
       "                         Source  \n",
       "0        brenda_reaction_smiles  \n",
       "1        brenda_reaction_smiles  \n",
       "2        brenda_reaction_smiles  \n",
       "3        brenda_reaction_smiles  \n",
       "4        brenda_reaction_smiles  \n",
       "...                         ...  \n",
       "62217  pathbank_reaction_smiles  \n",
       "62218  pathbank_reaction_smiles  \n",
       "62219  pathbank_reaction_smiles  \n",
       "62220  pathbank_reaction_smiles  \n",
       "62221  pathbank_reaction_smiles  \n",
       "\n",
       "[62222 rows x 3 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ECreact = pd.read_csv('raw_data/ECReact.csv')\n",
    "ECreact['rxn_smiles'] = ECreact['rxn_smiles'].str.split('|').str[0] + '>>' + ECreact['rxn_smiles'].str.split('>>').str[1]\n",
    "ECreact.rename(columns={'rxn_smiles': 'Reaction', 'ec': 'EC number', 'source':'Source'}, inplace=True)\n",
    "ECreact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "21ac8cce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...</td>\n",
       "      <td>[CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349453</th>\n",
       "      <td>CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + dimethylallyl diphosphate = dimethylally...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349454</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl diphosphate = geranyl-triphospho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349455</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl triphosphate = geranyl-tetraphos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349456</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl diphosphate = adenosine 5'-i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349457</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl triphosphate = isopentenyl-t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62896 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Reaction  \\\n",
       "0       CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1       CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...   \n",
       "25      CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...   \n",
       "26      CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...   \n",
       "27      NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                   ...   \n",
       "349453  CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349454  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...   \n",
       "349455  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...   \n",
       "349456  C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349457  C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...   \n",
       "\n",
       "                                          Mapped Reaction EC number  \\\n",
       "0       [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1       [CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...   1.1.1.1   \n",
       "25      [CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...   1.1.1.1   \n",
       "26      [CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...   1.1.1.1   \n",
       "27      [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                   ...       ...   \n",
       "349453  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...   6.5.1.3   \n",
       "349454  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349455  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349456  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "349457  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "\n",
       "                                            Reaction Text  \n",
       "0          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "25          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "26          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "27      2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...  \n",
       "...                                                   ...  \n",
       "349453  ATP + dimethylallyl diphosphate = dimethylally...  \n",
       "349454  ATP + geranyl diphosphate = geranyl-triphospho...  \n",
       "349455  ATP + geranyl triphosphate = geranyl-tetraphos...  \n",
       "349456  ATP + isopentenyl diphosphate = adenosine 5'-i...  \n",
       "349457  ATP + isopentenyl triphosphate = isopentenyl-t...  \n",
       "\n",
       "[62896 rows x 4 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enzymemap = pd.read_csv('raw_data/EnzymeMap.csv')\n",
    "enzymemap.rename(columns={'ec_num': 'EC number', 'unmapped': 'Reaction', 'mapped': 'Mapped Reaction', 'orig_rxn_text':'Reaction Text'}, inplace=True)\n",
    "enzymemap = enzymemap[['Reaction', 'Mapped Reaction', 'EC number', 'Reaction Text']]\n",
    "enzymemap.drop_duplicates(subset=['Reaction', 'EC number'], inplace=True)\n",
    "enzymemap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "a24320d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...</td>\n",
       "      <td>[CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349453</th>\n",
       "      <td>CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + dimethylallyl diphosphate = dimethylally...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349454</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl diphosphate = geranyl-triphospho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349455</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl triphosphate = geranyl-tetraphos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349456</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl diphosphate = adenosine 5'-i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349457</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl triphosphate = isopentenyl-t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62896 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Reaction  \\\n",
       "0       CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1       CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...   \n",
       "25      CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...   \n",
       "26      CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...   \n",
       "27      NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                   ...   \n",
       "349453  CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349454  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...   \n",
       "349455  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...   \n",
       "349456  C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349457  C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...   \n",
       "\n",
       "                                          Mapped Reaction EC number  \\\n",
       "0       [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1       [CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...   1.1.1.1   \n",
       "25      [CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...   1.1.1.1   \n",
       "26      [CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...   1.1.1.1   \n",
       "27      [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                   ...       ...   \n",
       "349453  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...   6.5.1.3   \n",
       "349454  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349455  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349456  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "349457  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "\n",
       "                                            Reaction Text  \n",
       "0          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "25          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "26          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "27      2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...  \n",
       "...                                                   ...  \n",
       "349453  ATP + dimethylallyl diphosphate = dimethylally...  \n",
       "349454  ATP + geranyl diphosphate = geranyl-triphospho...  \n",
       "349455  ATP + geranyl triphosphate = geranyl-tetraphos...  \n",
       "349456  ATP + isopentenyl diphosphate = adenosine 5'-i...  \n",
       "349457  ATP + isopentenyl triphosphate = isopentenyl-t...  \n",
       "\n",
       "[62896 rows x 4 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enzymemap.dropna(subset=['Mapped Reaction'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b244d31b",
   "metadata": {},
   "source": [
    "### Append EC numbers that are covered by ECreact but not by EnzymeMap, to EnzymeMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "94ce2b25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2353"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#find the EC numbers covered by ECreact but not by enzymemap\n",
    "ECreact = ECreact[~ECreact['EC number'].str.contains('-')]\n",
    "not_covered = ECreact[~ECreact['EC number'].isin(enzymemap['EC number'].unique())]\n",
    "not_covered['EC number'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "89e5d4dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1172</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1173</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1176</th>\n",
       "      <td>NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15228</th>\n",
       "      <td>Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.j</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21396</th>\n",
       "      <td>C[C@@]1(O)OC[C@H](O)C1(O)O.Nc1ncnc2c1ncn2[C@@H...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.l</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34012</th>\n",
       "      <td>*CS(=O)(=O)[O-].Nc1ncnc2c1ncn2[C@@H]1O[C@H](CO...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.o</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15541</th>\n",
       "      <td>Cc1ncc(C[n+]2csc(CCO)c2C)c(N)n1.Nc1ncnc2c1ncn2...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.p</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33659</th>\n",
       "      <td>Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.q</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72490 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Reaction  \\\n",
       "0      CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1172   CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...   \n",
       "1173   CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...   \n",
       "1176   NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...   \n",
       "1177   NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                  ...   \n",
       "15228  Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...   \n",
       "21396  C[C@@]1(O)OC[C@H](O)C1(O)O.Nc1ncnc2c1ncn2[C@@H...   \n",
       "34012  *CS(=O)(=O)[O-].Nc1ncnc2c1ncn2[C@@H]1O[C@H](CO...   \n",
       "15541  Cc1ncc(C[n+]2csc(CCO)c2C)c(N)n1.Nc1ncnc2c1ncn2...   \n",
       "33659  Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...   \n",
       "\n",
       "                                         Mapped Reaction EC number  \\\n",
       "0      [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1172   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1173   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1176   [H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...   1.1.1.1   \n",
       "1177   [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                  ...       ...   \n",
       "15228                                                NaN   7.6.2.j   \n",
       "21396                                                NaN   7.6.2.l   \n",
       "34012                                                NaN   7.6.2.o   \n",
       "15541                                                NaN   7.6.2.p   \n",
       "33659                                                NaN   7.6.2.q   \n",
       "\n",
       "                                           Reaction Text  \n",
       "0         acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1172   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1173   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1176   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "1177   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "...                                                  ...  \n",
       "15228                                                NaN  \n",
       "21396                                                NaN  \n",
       "34012                                                NaN  \n",
       "15541                                                NaN  \n",
       "33659                                                NaN  \n",
       "\n",
       "[72490 rows x 4 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reaction2EC = pd.concat((enzymemap, not_covered[['Reaction', 'EC number']])).sort_values(by='EC number')\n",
    "reaction2EC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac489eea",
   "metadata": {},
   "source": [
    "## 3. Filter protein2EC and reaction2EC to intersection of their EC numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "624fb47e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A0A068J840</td>\n",
       "      <td>UGT1_PANGI</td>\n",
       "      <td>MKSELIFLPAPAIGHLVGMVEMAKLFISRHENLSVTVLIAKFYMDT...</td>\n",
       "      <td>2.4.1.363</td>\n",
       "      <td>475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254913</th>\n",
       "      <td>P36352</td>\n",
       "      <td>POLR_PHMV</td>\n",
       "      <td>VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254914</th>\n",
       "      <td>P39262</td>\n",
       "      <td>VG56_BPT4</td>\n",
       "      <td>MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254923</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254927</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254928</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>149788 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "3       A0A024SC78  CUTI1_HYPJR   \n",
       "4       A0A024SH76   GUX2_HYPJR   \n",
       "6       A0A059TC02   CCR1_PETHY   \n",
       "14      A0A068J840   UGT1_PANGI   \n",
       "...            ...          ...   \n",
       "254913      P36352    POLR_PHMV   \n",
       "254914      P39262    VG56_BPT4   \n",
       "254923      Q05115   AMDA_BORBO   \n",
       "254927      Q94MV8   VG56_BPLZ5   \n",
       "254928      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence  EC number  Length  \n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.6     269  \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   3.1.1.74     248  \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   3.2.1.91     471  \n",
       "6       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...   1.2.1.44     333  \n",
       "14      MKSELIFLPAPAIGHLVGMVEMAKLFISRHENLSVTVLIAKFYMDT...  2.4.1.363     475  \n",
       "...                                                   ...        ...     ...  \n",
       "254913  VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...   2.7.7.48     178  \n",
       "254914  MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...   3.6.1.12     171  \n",
       "254923  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...   4.1.1.76     240  \n",
       "254927  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...   3.6.1.12     172  \n",
       "254928  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...   3.1.1.29     124  \n",
       "\n",
       "[149788 rows x 5 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "swissprot = swissprot[swissprot['EC number'].isin(reaction2EC['EC number'].unique())]\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "443b6cdb-fe81-40ba-86a6-a52fe0103c29",
   "metadata": {},
   "source": [
    "## Save unique proteins\n",
    "\n",
    "#save the indices corresponding to unique proteins, for downstream use\n",
    "unique_protein_indices = swissprot.reset_index().drop_duplicates(subset=['Sequence']).index\n",
    "with open('unique_protein_indices.txt', 'w') as f:\n",
    "    for index in unique_protein_indices:\n",
    "        f.write(str(index) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "407b09c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1172</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1173</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1176</th>\n",
       "      <td>NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34226</th>\n",
       "      <td>CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@@H]8[C@H...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34223</th>\n",
       "      <td>CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@H]([C@H]...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15471</th>\n",
       "      <td>C[N+]1(C)CCC[C@H]1C(=O)[O-].Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)(O)OP(=O)(O...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8665</th>\n",
       "      <td>*[N+](*)(*)*.Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53594 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Reaction  \\\n",
       "0      CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1172   CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...   \n",
       "1173   CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...   \n",
       "1176   NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...   \n",
       "1177   NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                  ...   \n",
       "34226  CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@@H]8[C@H...   \n",
       "34223  CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@H]([C@H]...   \n",
       "15471  C[N+]1(C)CCC[C@H]1C(=O)[O-].Nc1ncnc2c1ncn2[C@@...   \n",
       "644    Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)(O)OP(=O)(O...   \n",
       "8665   *[N+](*)(*)*.Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=...   \n",
       "\n",
       "                                         Mapped Reaction EC number  \\\n",
       "0      [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1172   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1173   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1176   [H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...   1.1.1.1   \n",
       "1177   [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                  ...       ...   \n",
       "34226                                                NaN   7.6.2.8   \n",
       "34223                                                NaN   7.6.2.8   \n",
       "15471                                                NaN   7.6.2.9   \n",
       "644                                                  NaN   7.6.2.9   \n",
       "8665                                                 NaN   7.6.2.9   \n",
       "\n",
       "                                           Reaction Text  \n",
       "0         acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1172   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1173   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1176   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "1177   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "...                                                  ...  \n",
       "34226                                                NaN  \n",
       "34223                                                NaN  \n",
       "15471                                                NaN  \n",
       "644                                                  NaN  \n",
       "8665                                                 NaN  \n",
       "\n",
       "[53594 rows x 4 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reaction2EC = reaction2EC[reaction2EC['EC number'].isin(swissprot['EC number'].unique())]\n",
    "reaction2EC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "5721bdf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "reaction2EC.to_csv('../processed_data/reaction2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5250f2c8",
   "metadata": {},
   "source": [
    "## 4. Perform protein clustering using mmseqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "19acec52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate a fasta file as input to mmseqs\n",
    "with open('clustering/swissprot.fasta', 'w') as f:\n",
    "    for index, row in swissprot.iterrows():\n",
    "        f.write(f'>{row[\"Entry\"]}\\n{row[\"Sequence\"]}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "5e6cdd85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes50 clustering/mmseqs_tmp --min-seq-id 0.5 -c 0.8 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.5\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/18009605123812206450/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[149784] 0s 27ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 27ms\n",
      "Create directory clustering/mmseqs_tmp/18009605123812206450/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/18009605123812206450/input clustering/mmseqs_tmp/18009605123812206450/clu clustering/mmseqs_tmp/18009605123812206450/clu_tmp --max-seqs 20 -c 0.8 --cov-mode 1 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.5 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 3.000000\n",
      "Set cluster mode GREEDY MEM\n",
      "Set cluster iterations to 3\n",
      "linclust clustering/mmseqs_tmp/18009605123812206450/input clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/clu_redundancy clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/linclust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/18009605123812206450/input clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/linclust/5404717882821860712/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/18009605123812206450/input clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/linclust/5404717882821860712/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 149788 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[======================================>                          ] 59.75% 89.50K eta 0s       % 15.97K eta 0s       >                                                 ] 24.11% 36.12K eta 0s                                         ] 35.86% 53.72K eta 0s       K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 51.14K 0s 96ms     \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 36.44K 0s 202ms    =====================>                                           ] 32.46% 11.83K eta 0s       \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 149.79K 0s 462ms    \n",
      "Time for merging to clu_redundancy: 0h 0m 0s 27ms\n",
      "Time for processing: 0h 0m 0s 538ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/linclust/5404717882821860712/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/linclust/5404717882821860712/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/linclust/5404717882821860712/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/linclust/5404717882821860712/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/linclust/5404717882821860712/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/linclust/5404717882821860712/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/linclust/5404717882821860712/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/linclust/5404717882821860712/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/linclust/5404717882821860712/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/linclust/5404717882821860712/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/clu_redundancy clustering/mmseqs_tmp/18009605123812206450/input clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 11ms\n",
      "prefilter clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/input_step_redundancy clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/input_step_redundancy clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 36442 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 36442 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[==============================================>                  ] 71.05% 25.89K eta 0s         ] 14.00% 5.10K eta 0s       K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 24.66K 0s 127ms    \n",
      "Index table: Masked residues: 31940\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 24.66K 0s 135ms    ========>                         ] 60.14% 14.83K eta 0s       \n",
      "Index statistics\n",
      "Entries:          5369908\n",
      "DB size:          519 MB\n",
      "Avg k-mer size:   0.083905\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t89\n",
      "    GQQVAR\t77\n",
      "    STYLGG\t53\n",
      "    GSELFN\t53\n",
      "    AYPWGT\t52\n",
      "    FNLPNP\t50\n",
      "    FPHNAQ\t49\n",
      "    FREPGR\t49\n",
      "    GRGHCS\t48\n",
      "    TGNKTT\t44\n",
      "Time for index table init: 0h 0m 0s 605ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 145\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 24660\n",
      "Target db start 1 to 24660\n",
      "[=================================================================] 100.00% 24.66K 0s 215ms    K eta 0s       ==============================>   ] 94.47% 23.30K eta 0s       \n",
      "[================================================================>] 99.79% 24.61K eta 0s       \n",
      "7.214568 k-mers per position\n",
      "474 DB matches per sequence\n",
      "0 overflows\n",
      "11 sequences passed prefiltering per query sequence\n",
      "10 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 15ms\n",
      "Time for processing: 0h 0m 1s 388ms\n",
      "align clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/input_step1 clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/input_step1 clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/pref_step1 clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 24660 type: Aminoacid\n",
      "Target database size: 24660 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[==========================================>                      ] 65.68% 16.20K eta 0s       \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 15ms\n",
      "Time for processing: 0h 0m 1s 403ms\n",
      "align clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/input_step2 clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/input_step2 clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/pref_step2 clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 23723 type: Aminoacid\n",
      "Target database size: 23723 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 23.72K 0s 415ms    \n",
      "Time for merging to aln_step2: 0h 0m 0s 14ms\n",
      "308155 alignments calculated\n",
      "23874 sequence pairs passed the thresholds (0.077474 of overall calculated)\n",
      "1.006365 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 628ms\n",
      "clust clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/input_step2 clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/aln_step2 clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/clu_step2 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 46ms\n",
      "\n",
      "Size of the sequence database: 23723\n",
      "Size of the alignment database: 23723\n",
      "Number of clusters: 23633\n",
      "\n",
      "Writing results 0h 0m 0s 4ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 58ms\n",
      "mergeclusters clustering/mmseqs_tmp/18009605123812206450/input clustering/mmseqs_tmp/18009605123812206450/clu clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/clu_redundancy clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/clu_step0 clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/clu_step1 clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/clu_step2 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 36.44K 0s 114ms    \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 24.66K 0s 206ms                                           ] 12.58% 3.10K eta 1s       \n",
      "Clustering step 3\n",
      "[=================================================================] 100.00% 23.72K 0s 275ms    =========================================>        ] 87.18% 20.68K eta 0s       \n",
      "Clustering step 4\n",
      "[=================================================================] 100.00% 23.63K 0s 329ms      ] 61.20% 14.46K eta 0s       \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 149.79K 0s 541ms           ] 5.14% 7.71K eta 8s       \n",
      "Time for merging to clu: 0h 0m 0s 19ms\n",
      "Time for processing: 0h 0m 0s 607ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/pref_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/aln_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/clu_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/pref_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/aln_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/clu_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/input_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/input_step1_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/input_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/18009605123812206450/clu_tmp/5777700569579494009/input_step2_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering/mmseqs_tmp/18009605123812206450/input clustering/mmseqs_tmp/18009605123812206450/input clustering/mmseqs_tmp/18009605123812206450/clu clustering/mmseqs_tmp/18009605123812206450/cluster.tsv --threads 384 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 15ms\n",
      "Time for processing: 0h 0m 0s 247ms\n",
      "result2repseq clustering/mmseqs_tmp/18009605123812206450/input clustering/mmseqs_tmp/18009605123812206450/clu clustering/mmseqs_tmp/18009605123812206450/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=============================>                                   ] 45.61% 10.78K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster clustering/swissprot.fasta clustering/clusterRes50 clustering/mmseqs_tmp --min-seq-id 0.5 -c 0.8 --cov-mode 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "286e6959",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes30 clustering/mmseqs_tmp --min-seq-id 0.3 -c 0.8 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.3\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/1342346348063921006/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[149784] 0s 27ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 27ms\n",
      "Create directory clustering/mmseqs_tmp/1342346348063921006/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/1342346348063921006/input clustering/mmseqs_tmp/1342346348063921006/clu clustering/mmseqs_tmp/1342346348063921006/clu_tmp --max-seqs 20 -c 0.8 --cov-mode 1 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.3 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 5.000000\n",
      "Set cluster mode GREEDY MEM\n",
      "Set cluster iterations to 3\n",
      "linclust clustering/mmseqs_tmp/1342346348063921006/input clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/clu_redundancy clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/linclust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/1342346348063921006/input clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/linclust/6919856536007606354/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/1342346348063921006/input clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/linclust/6919856536007606354/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 149788 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=======================================>                         ] 60.02% 89.90K eta 0s       ======>                                                         ] 11.42% 17.10K eta 0s         ] 23.78% 35.63K eta 0s                                         ] 35.84% 53.69K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 51.14K 0s 99ms     \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 28.29K 0s 205ms      ] 83.03% 23.49K eta 0s       \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 149.79K 0s 412ms                                                      ] 5.01% 7.51K eta 6s       \n",
      "Time for merging to clu_redundancy: 0h 0m 0s 19ms\n",
      "Time for processing: 0h 0m 0s 478ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/linclust/6919856536007606354/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/linclust/6919856536007606354/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/linclust/6919856536007606354/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/linclust/6919856536007606354/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/linclust/6919856536007606354/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/linclust/6919856536007606354/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/linclust/6919856536007606354/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/linclust/6919856536007606354/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/linclust/6919856536007606354/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/linclust/6919856536007606354/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/clu_redundancy clustering/mmseqs_tmp/1342346348063921006/input clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 10ms\n",
      "prefilter clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/input_step_redundancy clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/input_step_redundancy clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 28287 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 28287 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 28.29K 0s 149ms    .69K eta 0s       =============================>                    ] 68.24% 19.30K eta 0s       \n",
      "Index table: Masked residues: 32510\n",
      "Index table: fill\n",
      "[=======================>                                         ] 35.90% 10.16K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index statistics\n",
      "Entries:          3561501\n",
      "DB size:          508 MB\n",
      "Avg k-mer size:   0.055648\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t43\n",
      "    GPGGTL\t39\n",
      "    GLGIGK\t24\n",
      "    RDKELL\t22\n",
      "    AVGGGL\t20\n",
      "    LVGGGL\t20\n",
      "    VVGGGL\t19\n",
      "    KKGARN\t19\n",
      "    GQQVAR\t18\n",
      "    VHDKNL\t15\n",
      "Time for index table init: 0h 0m 0s 508ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 12230\n",
      "Target db start 1 to 12230\n",
      "[=================================================================] 100.00% 12.23K 0s 137ms    =====================================>                      ] 65.01% 7.95K eta 0s       \n",
      "[===============================================================> ] 97.72% 11.95K eta 0s       \n",
      "22.120738 k-mers per position\n",
      "617 DB matches per sequence\n",
      "0 overflows\n",
      "7 sequences passed prefiltering per query sequence\n",
      "6 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 12ms\n",
      "Time for processing: 0h 0m 1s 129ms\n",
      "align clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/input_step1 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/input_step1 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/pref_step1 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 12230 type: Aminoacid\n",
      "Target database size: 12230 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 12.23K 0s 208ms    \n",
      "Time for merging to aln_step1: 0h 0m 0s 13ms\n",
      "74942 alignments calculated\n",
      "34824 sequence pairs passed the thresholds (0.464679 of overall calculated)\n",
      "2.847425 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 387ms\n",
      "clust clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/input_step1 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/aln_step1 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/clu_step1 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 45ms\n",
      "\n",
      "Size of the sequence database: 12230\n",
      "Size of the alignment database: 12230\n",
      "Number of clusters: 8018\n",
      "\n",
      "Writing results 0h 0m 0s 1ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 58ms\n",
      "createsubdb clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/clu_step1 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/input_step1 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "prefilter clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/input_step2 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/input_step2 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 5 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 8018 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 8018 type: Aminoacid\n",
      "Index table k-mer threshold: 118 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 8.02K 0s 46ms     >                                            ] 31.62% 2.54K eta 0s       \n",
      "Index table: Masked residues: 16896\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 8.02K 0s 42ms     \n",
      "Index statistics\n",
      "Entries:          2569770\n",
      "DB size:          502 MB\n",
      "Avg k-mer size:   0.040153\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t42\n",
      "    VHDKNI\t22\n",
      "    RDKELL\t21\n",
      "    GPGGTL\t17\n",
      "    HRIPIL\t15\n",
      "    HRLPIL\t14\n",
      "    VHDKNL\t14\n",
      "    QCGGGP\t13\n",
      "    RVGLSR\t13\n",
      "    LGGFVY\t13\n",
      "Time for index table init: 0h 0m 0s 426ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 118\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 8018\n",
      "Target db start 1 to 8018\n",
      "[=================================================================] 100.00% 8.02K 0s 165ms    \n",
      "[===============================================================> ] 97.99% 7.86K eta 0s       \n",
      "176.908029 k-mers per position\n",
      "2492 DB matches per sequence\n",
      "0 overflows\n",
      "16 sequences passed prefiltering per query sequence\n",
      "19 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 12ms\n",
      "Time for processing: 0h 0m 0s 973ms\n",
      "align clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/input_step2 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/input_step2 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/pref_step2 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 8018 type: Aminoacid\n",
      "Target database size: 8018 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 8.02K 0s 191ms    \n",
      "Time for merging to aln_step2: 0h 0m 0s 14ms\n",
      "97441 alignments calculated\n",
      "10952 sequence pairs passed the thresholds (0.112396 of overall calculated)\n",
      "1.365927 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 407ms\n",
      "clust clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/input_step2 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/aln_step2 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/clu_step2 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 44ms\n",
      "\n",
      "Size of the sequence database: 8018\n",
      "Size of the alignment database: 8018\n",
      "Number of clusters: 6958\n",
      "\n",
      "Writing results 0h 0m 0s 3ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 55ms\n",
      "mergeclusters clustering/mmseqs_tmp/1342346348063921006/input clustering/mmseqs_tmp/1342346348063921006/clu clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/clu_redundancy clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/clu_step0 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/clu_step1 clustering/mmseqs_tmp/1342346348063921006/clu_tmp/1588136530516962783/clu_step2 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 28.29K 0s 85ms     a 0s       \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 12.23K 0s 157ms    \n",
      "Clustering step 3\n",
      "[=================================================================] 100.00% 8.02K 0s 219ms    \n",
      "Clustering step 4\n",
      "[=================================================================] 100.00% 6.96K 0s 263ms    \n",
      "Write merged clustering\n",
      "[>                                                                ] 1.01% 1.51K eta 46s       \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for merging to cluster.tsv: 0h 0m 0s 13ms\n",
      "Time for processing: 0h 0m 0s 188ms\n",
      "result2repseq clustering/mmseqs_tmp/1342346348063921006/input clustering/mmseqs_tmp/1342346348063921006/clu clustering/mmseqs_tmp/1342346348063921006/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 6.96K 0s 35ms     \n",
      "Time for merging to clu_rep: 0h 0m 0s 14ms\n",
      "Time for processing: 0h 0m 0s 225ms\n",
      "result2flat clustering/mmseqs_tmp/1342346348063921006/input clustering/mmseqs_tmp/1342346348063921006/input clustering/mmseqs_tmp/1342346348063921006/clu_rep clustering/mmseqs_tmp/1342346348063921006/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 27ms\n",
      "createseqfiledb clustering/mmseqs_tmp/1342346348063921006/input clustering/mmseqs_tmp/1342346348063921006/clu clustering/mmseqs_tmp/1342346348063921006/clu_seqs --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 6.96K 0s 47ms     \n",
      "Time for merging to clu_seqs: 0h 0m 0s 8ms\n",
      "Time for processing: 0h 0m 0s 263ms\n",
      "result2flat clustering/mmseqs_tmp/1342346348063921006/input clustering/mmseqs_tmp/1342346348063921006/input clustering/mmseqs_tmp/1342346348063921006/clu_seqs clustering/mmseqs_tmp/1342346348063921006/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 124ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/1342346348063921006/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster clustering/swissprot.fasta clustering/clusterRes30 clustering/mmseqs_tmp --min-seq-id 0.3 -c 0.8 --cov-mode 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "02322b64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23633\n",
      "6958\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "      <th>clusterRes50</th>\n",
       "      <th>clusterRes30</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A1AY86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>A0A024SC78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "      <td>G4MM92</td>\n",
       "      <td>G4MM92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "      <td>Q9S9N9</td>\n",
       "      <td>P14721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A068J840</td>\n",
       "      <td>UGT1_PANGI</td>\n",
       "      <td>MKSELIFLPAPAIGHLVGMVEMAKLFISRHENLSVTVLIAKFYMDT...</td>\n",
       "      <td>2.4.1.363</td>\n",
       "      <td>475</td>\n",
       "      <td>Q2V6K0</td>\n",
       "      <td>Q40287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149783</th>\n",
       "      <td>P36352</td>\n",
       "      <td>POLR_PHMV</td>\n",
       "      <td>VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>178</td>\n",
       "      <td>P36352</td>\n",
       "      <td>P36352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149784</th>\n",
       "      <td>P39262</td>\n",
       "      <td>VG56_BPT4</td>\n",
       "      <td>MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>171</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149785</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149786</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149787</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q54ZD0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>149788 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "1       A0A024SC78  CUTI1_HYPJR   \n",
       "2       A0A024SH76   GUX2_HYPJR   \n",
       "3       A0A059TC02   CCR1_PETHY   \n",
       "4       A0A068J840   UGT1_PANGI   \n",
       "...            ...          ...   \n",
       "149783      P36352    POLR_PHMV   \n",
       "149784      P39262    VG56_BPT4   \n",
       "149785      Q05115   AMDA_BORBO   \n",
       "149786      Q94MV8   VG56_BPLZ5   \n",
       "149787      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence  EC number  Length  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.6     269   \n",
       "1       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   3.1.1.74     248   \n",
       "2       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   3.2.1.91     471   \n",
       "3       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...   1.2.1.44     333   \n",
       "4       MKSELIFLPAPAIGHLVGMVEMAKLFISRHENLSVTVLIAKFYMDT...  2.4.1.363     475   \n",
       "...                                                   ...        ...     ...   \n",
       "149783  VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...   2.7.7.48     178   \n",
       "149784  MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...   3.6.1.12     171   \n",
       "149785  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...   4.1.1.76     240   \n",
       "149786  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...   3.6.1.12     172   \n",
       "149787  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...   3.1.1.29     124   \n",
       "\n",
       "       clusterRes50 clusterRes30  \n",
       "0        A0A009IHW8       A1AY86  \n",
       "1        A0A024SC78   A0A024SC78  \n",
       "2            G4MM92       G4MM92  \n",
       "3            Q9S9N9       P14721  \n",
       "4            Q2V6K0       Q40287  \n",
       "...             ...          ...  \n",
       "149783       P36352       P36352  \n",
       "149784       Q94MV8       Q94MV8  \n",
       "149785       Q05115       Q05115  \n",
       "149786       Q94MV8       Q94MV8  \n",
       "149787       Q9J5H2       Q54ZD0  \n",
       "\n",
       "[149788 rows x 7 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjkAAAGdCAYAAADwjmIIAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuNSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/xnp5ZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA6YklEQVR4nO3de3RU5aH+8ScJmQlRQgKYWw0houUi4V7jeKFQQgJm2aIcjgIqKkrxBAvEH2IsYiC1QS0orSjLo4BnCUXpUqpAIUMQkDKARCIEC/UC0lYmtCKMgE6GZP/+cGWXMeESO2OY1+9nrazFfvc7e7/PzDQ+nT0zibIsyxIAAIBholt6AQAAAOFAyQEAAEai5AAAACNRcgAAgJEoOQAAwEiUHAAAYCRKDgAAMBIlBwAAGKlVSy+gJdXX1+vTTz9VmzZtFBUV1dLLAQAA58GyLH3xxRdKT09XdPSZX6/5XpecTz/9VBkZGS29DAAA8C387W9/06WXXnrG/d/rktOmTRtJX99JCQkJITtuIBBQeXm58vLyFBsbG7LjXihMzyeR0QSm55PIaArTM4Yjn8/nU0ZGhv3f8TP5XpechktUCQkJIS858fHxSkhIMPYJa3I+iYwmMD2fREZTmJ4xnPnO9VYT3ngMAACMRMkBAABGouQAAAAjUXIAAICRKDkAAMBIlBwAAGAkSg4AADASJQcAABiJkgMAAIxEyQEAAEai5AAAACNRcgAAgJEoOQAAwEiUHAAAYKRWLb0Ak/UoWSt/3b//DPyB2QUtuBoAAL5feCUHAAAYiZIDAACMRMkBAABGouQAAAAjUXIAAICRmlVyysrK9KMf/Uht2rRRcnKyhg8frn379gXN+eqrr1RYWKj27dvr4osv1ogRI1RTUxM05+DBgyooKFB8fLySk5M1depUnTp1KmjOhg0b1LdvXzmdTl1++eVavHhxo/XMnz9fnTp1UlxcnHJycrR9+/bmxAEAAAZrVsnZuHGjCgsLtXXrVrndbgUCAeXl5enEiRP2nClTpujNN9/U8uXLtXHjRn366ae6+eab7f11dXUqKChQbW2ttmzZopdeekmLFy/WjBkz7Dn79+9XQUGBBg0apKqqKk2ePFn33HOP1q5da8955ZVXVFRUpEcffVTvvvuuevXqpfz8fB0+fPg/uT8AAIAhmvU9OWvWrAnaXrx4sZKTk1VZWakBAwbo2LFjevHFF7V06VL95Cc/kSQtWrRI3bp109atW3X11VervLxc77//vtatW6eUlBT17t1bpaWlmjZtmkpKSuRwOLRgwQJlZWVpzpw5kqRu3bpp8+bNeuqpp5Sfny9Jmjt3ru69917dddddkqQFCxZo1apVWrhwoR566KH/+I4BAACR7T/6MsBjx45Jktq1aydJqqysVCAQUG5urj2na9eu6tixozwej66++mp5PB5lZ2crJSXFnpOfn6/77rtPe/bsUZ8+feTxeIKO0TBn8uTJkqTa2lpVVlaquLjY3h8dHa3c3Fx5PJ4zrtfv98vv99vbPp9PkhQIBBQIBL7lvdBYw7Gc0VaT45GuIYcpeZpCxshnej6JjKYwPWM48p3vsb51yamvr9fkyZN17bXXqkePHpIkr9crh8OhxMTEoLkpKSnyer32nNMLTsP+hn1nm+Pz+fTll1/q888/V11dXZNz9u7de8Y1l5WVaebMmY3Gy8vLFR8ffx6pm6e0f33Q9urVq0N+jpbkdrtbeglhR8bIZ3o+iYymMD1jKPOdPHnyvOZ965JTWFio6upqbd68+dse4jtXXFysoqIie9vn8ykjI0N5eXlKSEgI2XkCgYDcbrce2REtf/2//6xDdUl+yM7RkhryDRkyRLGxsS29nLAgY+QzPZ9ERlOYnjEc+RquxJzLtyo5EydO1MqVK7Vp0yZdeuml9nhqaqpqa2t19OjRoFdzampqlJqaas/55qegGj59dfqcb34iq6amRgkJCWrdurViYmIUExPT5JyGYzTF6XTK6XQ2Go+NjQ3LE8tfHxX0t6tMe/KG6367kJAx8pmeTyKjKUzPGMp853ucZn26yrIsTZw4Ua+//rrWr1+vrKysoP39+vVTbGysKioq7LF9+/bp4MGDcrlckiSXy6Xdu3cHfQrK7XYrISFB3bt3t+ecfoyGOQ3HcDgc6tevX9Cc+vp6VVRU2HMAAMD3W7NeySksLNTSpUv1xz/+UW3atLHfQ9O2bVu1bt1abdu21bhx41RUVKR27dopISFB999/v1wul66++mpJUl5enrp3767bb79dTzzxhLxer6ZPn67CwkL7VZYJEybomWee0YMPPqi7775b69ev16uvvqpVq1bZaykqKtLYsWPVv39/XXXVVXr66ad14sQJ+9NWAADg+61ZJee5556TJA0cODBofNGiRbrzzjslSU899ZSio6M1YsQI+f1+5efn69lnn7XnxsTEaOXKlbrvvvvkcrl00UUXaezYsZo1a5Y9JysrS6tWrdKUKVM0b948XXrppXrhhRfsj49L0i233KJ//vOfmjFjhrxer3r37q01a9Y0ejMyAAD4fmpWybEs65xz4uLiNH/+fM2fP/+MczIzM8/5SaOBAwdq586dZ50zceJETZw48ZxrAgAA3z/87SoAAGAkSg4AADASJQcAABiJkgMAAIxEyQEAAEai5AAAACNRcgAAgJEoOQAAwEiUHAAAYCRKDgAAMBIlBwAAGImSAwAAjETJAQAARqLkAAAAI1FyAACAkSg5AADASJQcAABgJEoOAAAwEiUHAAAYiZIDAACMRMkBAABGouQAAAAjUXIAAICRKDkAAMBIlBwAAGAkSg4AADASJQcAABiJkgMAAIxEyQEAAEai5AAAACNRcgAAgJEoOQAAwEjNLjmbNm3SjTfeqPT0dEVFRWnFihVB+6Oiopr8efLJJ+05nTp1arR/9uzZQcfZtWuXrr/+esXFxSkjI0NPPPFEo7UsX75cXbt2VVxcnLKzs7V69ermxgEAAIZqdsk5ceKEevXqpfnz5ze5/9ChQ0E/CxcuVFRUlEaMGBE0b9asWUHz7r//fnufz+dTXl6eMjMzVVlZqSeffFIlJSV6/vnn7TlbtmzRqFGjNG7cOO3cuVPDhw/X8OHDVV1d3dxIAADAQK2ae4Nhw4Zp2LBhZ9yfmpoatP3HP/5RgwYN0mWXXRY03qZNm0ZzGyxZskS1tbVauHChHA6HrrzySlVVVWnu3LkaP368JGnevHkaOnSopk6dKkkqLS2V2+3WM888owULFjQ3FgAAMEyzS05z1NTUaNWqVXrppZca7Zs9e7ZKS0vVsWNHjR49WlOmTFGrVl8vx+PxaMCAAXI4HPb8/Px8Pf744/r888+VlJQkj8ejoqKioGPm5+c3unx2Or/fL7/fb2/7fD5JUiAQUCAQ+E+iBmk4ljPaanI80jXkMCVPU8gY+UzPJ5HRFKZnDEe+8z1WWEvOSy+9pDZt2ujmm28OGv/FL36hvn37ql27dtqyZYuKi4t16NAhzZ07V5Lk9XqVlZUVdJuUlBR7X1JSkrxerz12+hyv13vG9ZSVlWnmzJmNxsvLyxUfH/+tMp5Naf/6oG3T3jPkdrtbeglhR8bIZ3o+iYymMD1jKPOdPHnyvOaFteQsXLhQY8aMUVxcXND46a/A9OzZUw6HQz//+c9VVlYmp9MZtvUUFxcHndvn8ykjI0N5eXlKSEgI2XkCgYDcbrce2REtf32UPV5dkh+yc7SkhnxDhgxRbGxsSy8nLMgY+UzPJ5HRFKZnDEe+hisx5xK2kvP2229r3759euWVV845NycnR6dOndKBAwfUpUsXpaamqqamJmhOw3bD+3jONOdM7/ORJKfT2WSJio2NDcsTy18fJX/dv0uOaU/ecN1vFxIyRj7T80lkNIXpGUOZ73yPE7bvyXnxxRfVr18/9erV65xzq6qqFB0dreTkZEmSy+XSpk2bgq65ud1udenSRUlJSfacioqKoOO43W65XK4QpgAAAJGq2SXn+PHjqqqqUlVVlSRp//79qqqq0sGDB+05Pp9Py5cv1z333NPo9h6PR08//bTee+89ffzxx1qyZImmTJmi2267zS4wo0ePlsPh0Lhx47Rnzx698sormjdvXtClpkmTJmnNmjWaM2eO9u7dq5KSEu3YsUMTJ05sbiQAAGCgZl+u2rFjhwYNGmRvNxSPsWPHavHixZKkZcuWybIsjRo1qtHtnU6nli1bppKSEvn9fmVlZWnKlClBBaZt27YqLy9XYWGh+vXrpw4dOmjGjBn2x8cl6ZprrtHSpUs1ffp0Pfzww7riiiu0YsUK9ejRo7mRAACAgZpdcgYOHCjLss46Z/z48UGF5HR9+/bV1q1bz3menj176u233z7rnJEjR2rkyJHnPBYAAPj+4W9XAQAAI1FyAACAkSg5AADASJQcAABgJEoOAAAwEiUHAAAYiZIDAACMRMkBAABGouQAAAAjUXIAAICRKDkAAMBIlBwAAGAkSg4AADASJQcAABiJkgMAAIxEyQEAAEai5AAAACNRcgAAgJEoOQAAwEiUHAAAYCRKDgAAMBIlBwAAGImSAwAAjETJAQAARqLkAAAAI1FyAACAkSg5AADASJQcAABgJEoOAAAwEiUHAAAYiZIDAACMRMkBAABGanbJ2bRpk2688Ualp6crKipKK1asCNp/5513KioqKuhn6NChQXOOHDmiMWPGKCEhQYmJiRo3bpyOHz8eNGfXrl26/vrrFRcXp4yMDD3xxBON1rJ8+XJ17dpVcXFxys7O1urVq5sbBwAAGKrZJefEiRPq1auX5s+ff8Y5Q4cO1aFDh+yf3//+90H7x4wZoz179sjtdmvlypXatGmTxo8fb+/3+XzKy8tTZmamKisr9eSTT6qkpETPP/+8PWfLli0aNWqUxo0bp507d2r48OEaPny4qqurmxsJAAAYqFVzbzBs2DANGzbsrHOcTqdSU1Ob3PeXv/xFa9as0TvvvKP+/ftLkn73u9/phhtu0G9+8xulp6dryZIlqq2t1cKFC+VwOHTllVeqqqpKc+fOtcvQvHnzNHToUE2dOlWSVFpaKrfbrWeeeUYLFixobiwAAGCYZpec87FhwwYlJycrKSlJP/nJT/SrX/1K7du3lyR5PB4lJibaBUeScnNzFR0drW3btummm26Sx+PRgAED5HA47Dn5+fl6/PHH9fnnnyspKUkej0dFRUVB583Pz290+ex0fr9ffr/f3vb5fJKkQCCgQCAQiuj28STJGW01OR7pGnKYkqcpZIx8pueTyGgK0zOGI9/5HivkJWfo0KG6+eablZWVpY8++kgPP/ywhg0bJo/Ho5iYGHm9XiUnJwcvolUrtWvXTl6vV5Lk9XqVlZUVNCclJcXel5SUJK/Xa4+dPqfhGE0pKyvTzJkzG42Xl5crPj7+W+U9m9L+9UHbpr1nyO12t/QSwo6Mkc/0fBIZTWF6xlDmO3ny5HnNC3nJufXWW+1/Z2dnq2fPnurcubM2bNigwYMHh/p0zVJcXBz06o/P51NGRoby8vKUkJAQsvMEAgG53W49siNa/vooe7y6JD9k52hJDfmGDBmi2NjYll5OWJAx8pmeTyKjKUzPGI58DVdiziUsl6tOd9lll6lDhw768MMPNXjwYKWmpurw4cNBc06dOqUjR47Y7+NJTU1VTU1N0JyG7XPNOdN7gaSv3yvkdDobjcfGxoblieWvj5K/7t8lx7Qnb7jutwsJGSOf6fkkMprC9IyhzHe+xwn79+T8/e9/12effaa0tDRJksvl0tGjR1VZWWnPWb9+verr65WTk2PP2bRpU9A1N7fbrS5duigpKcmeU1FREXQut9stl8sV7kgAACACNLvkHD9+XFVVVaqqqpIk7d+/X1VVVTp48KCOHz+uqVOnauvWrTpw4IAqKir0s5/9TJdffrny87++VNOtWzcNHTpU9957r7Zv364///nPmjhxom699Valp6dLkkaPHi2Hw6Fx48Zpz549euWVVzRv3rygS02TJk3SmjVrNGfOHO3du1clJSXasWOHJk6cGIK7BQAARLpml5wdO3aoT58+6tOnjySpqKhIffr00YwZMxQTE6Ndu3bppz/9qX74wx9q3Lhx6tevn95+++2gy0RLlixR165dNXjwYN1www267rrrgr4Dp23btiovL9f+/fvVr18/PfDAA5oxY0bQd+lcc801Wrp0qZ5//nn16tVLf/jDH7RixQr16NHjP7k/AACAIZr9npyBAwfKsqwz7l+7du05j9GuXTstXbr0rHN69uypt99++6xzRo4cqZEjR57zfAAA4PuHv10FAACMRMkBAABGouQAAAAjUXIAAICRKDkAAMBIlBwAAGAkSg4AADASJQcAABiJkgMAAIxEyQEAAEai5AAAACNRcgAAgJEoOQAAwEiUHAAAYCRKDgAAMBIlBwAAGImSAwAAjETJAQAARqLkAAAAI1FyAACAkSg5AADASJQcAABgJEoOAAAwEiUHAAAYiZIDAACMRMkBAABGouQAAAAjUXIAAICRKDkAAMBIlBwAAGAkSg4AADBSs0vOpk2bdOONNyo9PV1RUVFasWKFvS8QCGjatGnKzs7WRRddpPT0dN1xxx369NNPg47RqVMnRUVFBf3Mnj07aM6uXbt0/fXXKy4uThkZGXriiScarWX58uXq2rWr4uLilJ2drdWrVzc3DgAAMFSzS86JEyfUq1cvzZ8/v9G+kydP6t1339Ujjzyid999V6+99pr27dunn/70p43mzpo1S4cOHbJ/7r//fnufz+dTXl6eMjMzVVlZqSeffFIlJSV6/vnn7TlbtmzRqFGjNG7cOO3cuVPDhw/X8OHDVV1d3dxIAADAQK2ae4Nhw4Zp2LBhTe5r27at3G530Ngzzzyjq666SgcPHlTHjh3t8TZt2ig1NbXJ4yxZskS1tbVauHChHA6HrrzySlVVVWnu3LkaP368JGnevHkaOnSopk6dKkkqLS2V2+3WM888owULFjQ3FgAAMEyzS05zHTt2TFFRUUpMTAwanz17tkpLS9WxY0eNHj1aU6ZMUatWXy/H4/FowIABcjgc9vz8/Hw9/vjj+vzzz5WUlCSPx6OioqKgY+bn5wddPvsmv98vv99vb/t8PklfX2YLBAL/YdJ/aziWM9pqcjzSNeQwJU9TyBj5TM8nkdEUpmcMR77zPVZYS85XX32ladOmadSoUUpISLDHf/GLX6hv375q166dtmzZouLiYh06dEhz586VJHm9XmVlZQUdKyUlxd6XlJQkr9drj50+x+v1nnE9ZWVlmjlzZqPx8vJyxcfHf+ucZ1Lavz5o27T3DH3zVTsTkTHymZ5PIqMpTM8YynwnT548r3lhKzmBQED//d//Lcuy9NxzzwXtO/0VmJ49e8rhcOjnP/+5ysrK5HQ6w7UkFRcXB53b5/MpIyNDeXl5QSXsPxUIBOR2u/XIjmj566Ps8eqS/JCdoyU15BsyZIhiY2NbejlhQcbIZ3o+iYymMD1jOPI1XIk5l7CUnIaC88knn2j9+vXnLBA5OTk6deqUDhw4oC5duig1NVU1NTVBcxq2G97Hc6Y5Z3qfjyQ5nc4mS1RsbGxYnlj++ij56/5dckx78obrfruQkDHymZ5PIqMpTM8Yynzne5yQf09OQ8H54IMPtG7dOrVv3/6ct6mqqlJ0dLSSk5MlSS6XS5s2bQq65uZ2u9WlSxclJSXZcyoqKoKO43a75XK5QpgGAABEqma/knP8+HF9+OGH9vb+/ftVVVWldu3aKS0tTf/1X/+ld999VytXrlRdXZ39Hpl27drJ4XDI4/Fo27ZtGjRokNq0aSOPx6MpU6botttuswvM6NGjNXPmTI0bN07Tpk1TdXW15s2bp6eeeso+76RJk/TjH/9Yc+bMUUFBgZYtW6YdO3YEfcwcAAB8fzW75OzYsUODBg2ytxve4zJ27FiVlJTojTfekCT17t076HZvvfWWBg4cKKfTqWXLlqmkpER+v19ZWVmaMmVK0Htl2rZtq/LychUWFqpfv37q0KGDZsyYYX98XJKuueYaLV26VNOnT9fDDz+sK664QitWrFCPHj2aGwkAABio2SVn4MCBsizrjPvPtk+S+vbtq61bt57zPD179tTbb7991jkjR47UyJEjz3ksAADw/cPfrgIAAEai5AAAACNRcgAAgJEoOQAAwEiUHAAAYCRKDgAAMBIlBwAAGImSAwAAjETJAQAARqLkAAAAI1FyAACAkSg5AADASJQcAABgJEoOAAAwEiUHAAAYiZIDAACMRMkBAABGouQAAAAjUXIAAICRKDkAAMBIlBwAAGAkSg4AADASJQcAABiJkgMAAIxEyQEAAEai5AAAACNRcgAAgJEoOQAAwEiUHAAAYCRKDgAAMBIlBwAAGImSAwAAjNTskrNp0ybdeOONSk9PV1RUlFasWBG037IszZgxQ2lpaWrdurVyc3P1wQcfBM05cuSIxowZo4SEBCUmJmrcuHE6fvx40Jxdu3bp+uuvV1xcnDIyMvTEE080Wsvy5cvVtWtXxcXFKTs7W6tXr25uHAAAYKhml5wTJ06oV69emj9/fpP7n3jiCf32t7/VggULtG3bNl100UXKz8/XV199Zc8ZM2aM9uzZI7fbrZUrV2rTpk0aP368vd/n8ykvL0+ZmZmqrKzUk08+qZKSEj3//PP2nC1btmjUqFEaN26cdu7cqeHDh2v48OGqrq5ubiQAAGCgVs29wbBhwzRs2LAm91mWpaefflrTp0/Xz372M0nS//3f/yklJUUrVqzQrbfeqr/85S9as2aN3nnnHfXv31+S9Lvf/U433HCDfvOb3yg9PV1LlixRbW2tFi5cKIfDoSuvvFJVVVWaO3euXYbmzZunoUOHaurUqZKk0tJSud1uPfPMM1qwYMG3ujMAAIA5ml1yzmb//v3yer3Kzc21x9q2baucnBx5PB7deuut8ng8SkxMtAuOJOXm5io6Olrbtm3TTTfdJI/HowEDBsjhcNhz8vPz9fjjj+vzzz9XUlKSPB6PioqKgs6fn5/f6PLZ6fx+v/x+v73t8/kkSYFAQIFA4D+Nb2s4ljPaanI80jXkMCVPU8gY+UzPJ5HRFKZnDEe+8z1WSEuO1+uVJKWkpASNp6Sk2Pu8Xq+Sk5ODF9Gqldq1axc0Jysrq9ExGvYlJSXJ6/We9TxNKSsr08yZMxuNl5eXKz4+/nwiNktp//qgbdPeM+R2u1t6CWFHxshnej6JjKYwPWMo8508efK85oW05FzoiouLg1798fl8ysjIUF5enhISEkJ2nkAgILfbrUd2RMtfH2WPV5fkh+wcLakh35AhQxQbG9vSywkLMkY+0/NJZDSF6RnDka/hSsy5hLTkpKamSpJqamqUlpZmj9fU1Kh37972nMOHDwfd7tSpUzpy5Ih9+9TUVNXU1ATNadg+15yG/U1xOp1yOp2NxmNjY8PyxPLXR8lf9++SY9qTN1z324WEjJHP9HwSGU1hesZQ5jvf44T0e3KysrKUmpqqiooKe8zn82nbtm1yuVySJJfLpaNHj6qystKes379etXX1ysnJ8ees2nTpqBrbm63W126dFFSUpI95/TzNMxpOA8AAPh+a3bJOX78uKqqqlRVVSXp6zcbV1VV6eDBg4qKitLkyZP1q1/9Sm+88YZ2796tO+64Q+np6Ro+fLgkqVu3bho6dKjuvfdebd++XX/+8581ceJE3XrrrUpPT5ckjR49Wg6HQ+PGjdOePXv0yiuvaN68eUGXmiZNmqQ1a9Zozpw52rt3r0pKSrRjxw5NnDjxP79XAABAxGv25aodO3Zo0KBB9nZD8Rg7dqwWL16sBx98UCdOnND48eN19OhRXXfddVqzZo3i4uLs2yxZskQTJ07U4MGDFR0drREjRui3v/2tvb9t27YqLy9XYWGh+vXrpw4dOmjGjBlB36VzzTXXaOnSpZo+fboefvhhXXHFFVqxYoV69Ojxre4IAABglmaXnIEDB8qyrDPuj4qK0qxZszRr1qwzzmnXrp2WLl161vP07NlTb7/99lnnjBw5UiNHjjz7ggEAwPcSf7sKAAAYiZIDAACMRMkBAABGouQAAAAjUXIAAICRKDkAAMBIlBwAAGAkSg4AADASJQcAABiJkgMAAIxEyQEAAEai5AAAACNRcgAAgJEoOQAAwEiUHAAAYCRKDgAAMBIlBwAAGImSAwAAjETJAQAARqLkAAAAI1FyAACAkSg5AADASJQcAABgJEoOAAAwEiUHAAAYiZIDAACMRMkBAABGouQAAAAjUXIAAICRKDkAAMBIlBwAAGAkSg4AADBSyEtOp06dFBUV1einsLBQkjRw4MBG+yZMmBB0jIMHD6qgoEDx8fFKTk7W1KlTderUqaA5GzZsUN++feV0OnX55Zdr8eLFoY4CAAAiWKtQH/Cdd95RXV2dvV1dXa0hQ4Zo5MiR9ti9996rWbNm2dvx8fH2v+vq6lRQUKDU1FRt2bJFhw4d0h133KHY2Fj9+te/liTt379fBQUFmjBhgpYsWaKKigrdc889SktLU35+fqgjAQCACBTyknPJJZcEbc+ePVudO3fWj3/8Y3ssPj5eqampTd6+vLxc77//vtatW6eUlBT17t1bpaWlmjZtmkpKSuRwOLRgwQJlZWVpzpw5kqRu3bpp8+bNeuqppyg5AABAUhhKzulqa2v18ssvq6ioSFFRUfb4kiVL9PLLLys1NVU33nijHnnkEfvVHI/Ho+zsbKWkpNjz8/Pzdd9992nPnj3q06ePPB6PcnNzg86Vn5+vyZMnn3U9fr9ffr/f3vb5fJKkQCCgQCDwn8a1NRzLGW01OR7pGnKYkqcpZIx8pueTyGgK0zOGI9/5HiusJWfFihU6evSo7rzzTnts9OjRyszMVHp6unbt2qVp06Zp3759eu211yRJXq83qOBIsre9Xu9Z5/h8Pn355Zdq3bp1k+spKyvTzJkzG42Xl5cHXTILldL+9UHbq1evDvk5WpLb7W7pJYQdGSOf6fkkMprC9IyhzHfy5MnzmhfWkvPiiy9q2LBhSk9Pt8fGjx9v/zs7O1tpaWkaPHiwPvroI3Xu3Dmcy1FxcbGKiorsbZ/Pp4yMDOXl5SkhISFk5wkEAnK73XpkR7T89f9+Bau6xIxLaQ35hgwZotjY2JZeTliQMfKZnk8ioylMzxiOfA1XYs4lbCXnk08+0bp16+xXaM4kJydHkvThhx+qc+fOSk1N1fbt24Pm1NTUSJL9Pp7U1FR77PQ5CQkJZ3wVR5KcTqecTmej8djY2LA8sfz1UfLX/bvkmPbkDdf9diEhY+QzPZ9ERlOYnjGU+c73OGH7npxFixYpOTlZBQUFZ51XVVUlSUpLS5MkuVwu7d69W4cPH7bnuN1uJSQkqHv37vacioqKoOO43W65XK4QJgAAAJEsLCWnvr5eixYt0tixY9Wq1b9fLProo49UWlqqyspKHThwQG+88YbuuOMODRgwQD179pQk5eXlqXv37rr99tv13nvvae3atZo+fboKCwvtV2EmTJigjz/+WA8++KD27t2rZ599Vq+++qqmTJkSjjgAACAChaXkrFu3TgcPHtTdd98dNO5wOLRu3Trl5eWpa9eueuCBBzRixAi9+eab9pyYmBitXLlSMTExcrlcuu2223THHXcEfa9OVlaWVq1aJbfbrV69emnOnDl64YUX+Pg4AACwheU9OXl5ebIsq9F4RkaGNm7ceM7bZ2ZmnvOTSAMHDtTOnTu/9RoBAIDZ+NtVAADASJQcAABgJEoOAAAwEiUHAAAYiZIDAACMRMkBAABGouQAAAAjUXIAAICRKDkAAMBIlBwAAGAkSg4AADASJQcAABiJkgMAAIxEyQEAAEai5AAAACNRcgAAgJEoOQAAwEiUHAAAYCRKDgAAMBIlBwAAGImSAwAAjETJAQAARqLkAAAAI1FyAACAkSg5AADASJQcAABgJEoOAAAwEiUHAAAYiZIDAACMRMkBAABGouQAAAAjUXIAAICRQl5ySkpKFBUVFfTTtWtXe/9XX32lwsJCtW/fXhdffLFGjBihmpqaoGMcPHhQBQUFio+PV3JysqZOnapTp04FzdmwYYP69u0rp9Opyy+/XIsXLw51FAAAEMHC8krOlVdeqUOHDtk/mzdvtvdNmTJFb775ppYvX66NGzfq008/1c0332zvr6urU0FBgWpra7Vlyxa99NJLWrx4sWbMmGHP2b9/vwoKCjRo0CBVVVVp8uTJuueee7R27dpwxAEAABGoVVgO2qqVUlNTG40fO3ZML774opYuXaqf/OQnkqRFixapW7du2rp1q66++mqVl5fr/fff17p165SSkqLevXurtLRU06ZNU0lJiRwOhxYsWKCsrCzNmTNHktStWzdt3rxZTz31lPLz88MRCQAARJiwlJwPPvhA6enpiouLk8vlUllZmTp27KjKykoFAgHl5ubac7t27aqOHTvK4/Ho6quvlsfjUXZ2tlJSUuw5+fn5uu+++7Rnzx716dNHHo8n6BgNcyZPnnzWdfn9fvn9fnvb5/NJkgKBgAKBQAiSyz6eJDmjrSbHI11DDlPyNIWMkc/0fBIZTWF6xnDkO99jhbzk5OTkaPHixerSpYsOHTqkmTNn6vrrr1d1dbW8Xq8cDocSExODbpOSkiKv1ytJ8nq9QQWnYX/DvrPN8fl8+vLLL9W6desm11ZWVqaZM2c2Gi8vL1d8fPy3yns2pf3rg7ZXr14d8nO0JLfb3dJLCDsyRj7T80lkNIXpGUOZ7+TJk+c1L+QlZ9iwYfa/e/bsqZycHGVmZurVV189Y/n4rhQXF6uoqMje9vl8ysjIUF5enhISEkJ2nkAgILfbrSG7f6HY+q9OW8DfQ3aOlmTnGzJEsbGxLb2csCBj5DM9n0RGU5ieMRz5Gq7EnEtYLledLjExUT/84Q/14YcfasiQIaqtrdXRo0eDXs2pqamx38OTmpqq7du3Bx2j4dNXp8/55ieyampqlJCQcNYi5XQ65XQ6G43HxsaG5YkVW/9VcMkx7MkbrvvtQkLGyGd6PomMpjA9Yyjzne9xwv49OcePH9dHH32ktLQ09evXT7GxsaqoqLD379u3TwcPHpTL5ZIkuVwu7d69W4cPH7bnuN1uJSQkqHv37vac04/RMKfhGAAAACEvOf/v//0/bdy4UQcOHNCWLVt00003KSYmRqNGjVLbtm01btw4FRUV6a233lJlZaXuuusuuVwuXX311ZKkvLw8de/eXbfffrvee+89rV27VtOnT1dhYaH9KsyECRP08ccf68EHH9TevXv17LPP6tVXX9WUKVNCHQcAAESokF+u+vvf/65Ro0bps88+0yWXXKLrrrtOW7du1SWXXCJJeuqppxQdHa0RI0bI7/crPz9fzz77rH37mJgYrVy5Uvfdd59cLpcuuugijR07VrNmzbLnZGVladWqVZoyZYrmzZunSy+9VC+88AIfHwcAALaQl5xly5addX9cXJzmz5+v+fPnn3FOZmbmOT+JNHDgQO3cufNbrREAAJiPv10FAACMRMkBAABGouQAAAAjUXIAAICRKDkAAMBIlBwAAGAkSg4AADASJQcAABiJkgMAAIxEyQEAAEai5AAAACNRcgAAgJEoOQAAwEiUHAAAYCRKDgAAMBIlBwAAGImSAwAAjETJAQAARqLkAAAAI1FyAACAkSg5AADASJQcAABgJEoOAAAwEiUHAAAYiZIDAACMRMkBAABGouQAAAAjUXIAAICRKDkAAMBIlBwAAGAkSg4AADBSyEtOWVmZfvSjH6lNmzZKTk7W8OHDtW/fvqA5AwcOVFRUVNDPhAkTguYcPHhQBQUFio+PV3JysqZOnapTp04FzdmwYYP69u0rp9Opyy+/XIsXLw51HAAAEKFCXnI2btyowsJCbd26VW63W4FAQHl5eTpx4kTQvHvvvVeHDh2yf5544gl7X11dnQoKClRbW6stW7bopZde0uLFizVjxgx7zv79+1VQUKBBgwapqqpKkydP1j333KO1a9eGOhIAAIhArUJ9wDVr1gRtL168WMnJyaqsrNSAAQPs8fj4eKWmpjZ5jPLycr3//vtat26dUlJS1Lt3b5WWlmratGkqKSmRw+HQggULlJWVpTlz5kiSunXrps2bN+upp55Sfn5+qGMBAIAIE/b35Bw7dkyS1K5du6DxJUuWqEOHDurRo4eKi4t18uRJe5/H41F2drZSUlLssfz8fPl8Pu3Zs8eek5ubG3TM/Px8eTyecEUBAAARJOSv5Jyuvr5ekydP1rXXXqsePXrY46NHj1ZmZqbS09O1a9cuTZs2Tfv27dNrr70mSfJ6vUEFR5K97fV6zzrH5/Ppyy+/VOvWrRutx+/3y+/329s+n0+SFAgEFAgEQpBY9vEkKRAd980dITtHS7LzGZKnKWSMfKbnk8hoCtMzhiPf+R4rrCWnsLBQ1dXV2rx5c9D4+PHj7X9nZ2crLS1NgwcP1kcffaTOnTuHbT1lZWWaOXNmo/Hy8nLFx8eH/Hzu7N8GD6xeHfJztCS3293SSwg7MkY+0/NJZDSF6RlDme/0qz9nE7aSM3HiRK1cuVKbNm3SpZdeeta5OTk5kqQPP/xQnTt3VmpqqrZv3x40p6amRpLs9/GkpqbaY6fPSUhIaPJVHEkqLi5WUVGRve3z+ZSRkaG8vDwlJCQ0L+BZBAIBud1uDdn9C8XWf3XaAv4esnO0JDvfkCGKjY1t6eWEBRkjn+n5JDKawvSM4cjXcCXmXEJecizL0v3336/XX39dGzZsUFZW1jlvU1VVJUlKS0uTJLlcLj322GM6fPiwkpOTJX3dABMSEtS9e3d7zupvvDLidrvlcrnOeB6n0ymn09loPDY2NixPrNj6r4JLjmFP3nDdbxcSMkY+0/NJZDSF6RlDme98jxPyNx4XFhbq5Zdf1tKlS9WmTRt5vV55vV59+eWXkqSPPvpIpaWlqqys1IEDB/TGG2/ojjvu0IABA9SzZ09JUl5enrp3767bb79d7733ntauXavp06ersLDQLikTJkzQxx9/rAcffFB79+7Vs88+q1dffVVTpkwJdSQAABCBQl5ynnvuOR07dkwDBw5UWlqa/fPKK69IkhwOh9atW6e8vDx17dpVDzzwgEaMGKE333zTPkZMTIxWrlypmJgYuVwu3Xbbbbrjjjs0a9Yse05WVpZWrVolt9utXr16ac6cOXrhhRf4+DgAAJAUpstVZ5ORkaGNGzee8ziZmZmNLkd908CBA7Vz585mrQ8AAHw/8LerAACAkSg5AADASJQcAABgJEoOAAAwEiUHAAAYiZIDAACMRMkBAABGouQAAAAjUXIAAICRKDkAAMBIlBwAAGCkkP/tKpxZp4dWBW0fmF3QQisBAMB8vJIDAACMRMkBAABGouQAAAAjUXIAAICRKDkAAMBIlBwAAGAkSg4AADASJQcAABiJkgMAAIxEyQEAAEai5AAAACNRcgAAgJH4A53foQNxo78xcqxF1gEAwPcBr+QAAAAjUXIAAICRuFzVgjo9tKrR2IHZBS2wEgAAzMMrOQAAwEiUHAAAYCRKDgAAMBLvyWlBjT9SLvGxcgAAQiPiX8mZP3++OnXqpLi4OOXk5Gj79u0tvaT/SKeHVgX9AACAbyeiX8l55ZVXVFRUpAULFignJ0dPP/208vPztW/fPiUnJ7f08r6Vb7660+mhpY3n8AksAADOKaJLzty5c3XvvffqrrvukiQtWLBAq1at0sKFC/XQQw+18OpCo6lLWt8sPpQeAAAai9iSU1tbq8rKShUXF9tj0dHRys3NlcfjafI2fr9ffr/f3j527Ov3vxw5ckSBQCBkawsEAjp58qQ+q3Uotr4+ZMdtUBl9Z9D2Zw9/ywM9sPdb3czO99lnio2N/ZYnv7CRMfKZnk8ioylMzxiOfF988YUkybKss86L2JLzr3/9S3V1dUpJSQkaT0lJ0d69Tf/Hu6ysTDNnzmw0npWVFZY1XvDKOrT0CgAA+Na++OILtW3b9oz7I7bkfBvFxcUqKiqyt+vr63XkyBG1b99eUVFRITuPz+dTRkaG/va3vykhISFkx71QmJ5PIqMJTM8nkdEUpmcMRz7LsvTFF18oPT39rPMituR06NBBMTExqqmpCRqvqalRampqk7dxOp1yOp1BY4mJieFaohISEox8wjYwPZ9ERhOYnk8ioylMzxjqfGd7BadBxH6E3OFwqF+/fqqoqLDH6uvrVVFRIZfL1YIrAwAAF4KIfSVHkoqKijR27Fj1799fV111lZ5++mmdOHHC/rQVAAD4/oroknPLLbfon//8p2bMmCGv16vevXtrzZo1jd6M/F1zOp169NFHG10aM4Xp+SQymsD0fBIZTWF6xpbMF2Wd6/NXAAAAEShi35MDAABwNpQcAABgJEoOAAAwEiUHAAAYiZITYvPnz1enTp0UFxennJwcbd++vaWXdN42bdqkG2+8Uenp6YqKitKKFSuC9luWpRkzZigtLU2tW7dWbm6uPvjgg6A5R44c0ZgxY5SQkKDExESNGzdOx48f/w5TnFlZWZl+9KMfqU2bNkpOTtbw4cO1b9++oDlfffWVCgsL1b59e1188cUaMWJEoy+cPHjwoAoKChQfH6/k5GRNnTpVp06d+i6jnNFzzz2nnj172l+65XK59Kc//cneH+n5vmn27NmKiorS5MmT7bFIz1hSUqKoqKign65du9r7Iz1fg3/84x+67bbb1L59e7Vu3VrZ2dnasWOHvT/Sf9906tSp0eMYFRWlwsJCSZH/ONbV1emRRx5RVlaWWrdurc6dO6u0tDTob0ldEI+hhZBZtmyZ5XA4rIULF1p79uyx7r33XisxMdGqqalp6aWdl9WrV1u//OUvrddee82SZL3++utB+2fPnm21bdvWWrFihfXee+9ZP/3pT62srCzryy+/tOcMHTrU6tWrl7V161br7bffti6//HJr1KhR33GSpuXn51uLFi2yqqurraqqKuuGG26wOnbsaB0/ftyeM2HCBCsjI8OqqKiwduzYYV199dXWNddcY+8/deqU1aNHDys3N9fauXOntXr1aqtDhw5WcXFxS0Rq5I033rBWrVpl/fWvf7X27dtnPfzww1ZsbKxVXV1tWVbk5zvd9u3brU6dOlk9e/a0Jk2aZI9HesZHH33UuvLKK61Dhw7ZP//85z/t/ZGez7Is68iRI1ZmZqZ15513Wtu2bbM+/vhja+3atdaHH35oz4n03zeHDx8OegzdbrclyXrrrbcsy4r8x/Gxxx6z2rdvb61cudLav3+/tXz5cuviiy+25s2bZ8+5EB5DSk4IXXXVVVZhYaG9XVdXZ6Wnp1tlZWUtuKpv55slp76+3kpNTbWefPJJe+zo0aOW0+m0fv/731uWZVnvv/++Jcl655137Dl/+tOfrKioKOsf//jHd7b283X48GFLkrVx40bLsr7OExsbay1fvtye85e//MWSZHk8Hsuyvi6C0dHRltfrtec899xzVkJCguX3+7/bAOcpKSnJeuGFF4zK98UXX1hXXHGF5Xa7rR//+Md2yTEh46OPPmr16tWryX0m5LMsy5o2bZp13XXXnXG/ib9vJk2aZHXu3Nmqr6834nEsKCiw7r777qCxm2++2RozZoxlWRfOY8jlqhCpra1VZWWlcnNz7bHo6Gjl5ubK4/G04MpCY//+/fJ6vUH52rZtq5ycHDufx+NRYmKi+vfvb8/Jzc1VdHS0tm3b9p2v+VyOHTsmSWrXrp0kqbKyUoFAIChj165d1bFjx6CM2dnZQV84mZ+fL5/Ppz179nyHqz+3uro6LVu2TCdOnJDL5TIqX2FhoQoKCoKySOY8hh988IHS09N12WWXacyYMTp48KAkc/K98cYb6t+/v0aOHKnk5GT16dNH//u//2vvN+33TW1trV5++WXdfffdioqKMuJxvOaaa1RRUaG//vWvkqT33ntPmzdv1rBhwyRdOI9hRH/j8YXkX//6l+rq6hp923JKSor27t3bQqsKHa/XK0lN5mvY5/V6lZycHLS/VatWateunT3nQlFfX6/Jkyfr2muvVY8ePSR9vX6Hw9Hoj7Z+M2NT90HDvgvB7t275XK59NVXX+niiy/W66+/ru7du6uqqsqIfMuWLdO7776rd955p9E+Ex7DnJwcLV68WF26dNGhQ4c0c+ZMXX/99aqurjYinyR9/PHHeu6551RUVKSHH35Y77zzjn7xi1/I4XBo7Nixxv2+WbFihY4ePao777xTkhnP04ceekg+n09du3ZVTEyM6urq9Nhjj2nMmDGSLpz/ZlBy8L1UWFio6upqbd68uaWXEnJdunRRVVWVjh07pj/84Q8aO3asNm7c2NLLCom//e1vmjRpktxut+Li4lp6OWHR8P+EJalnz57KyclRZmamXn31VbVu3boFVxY69fX16t+/v379619Lkvr06aPq6motWLBAY8eObeHVhd6LL76oYcOGKT09vaWXEjKvvvqqlixZoqVLl+rKK69UVVWVJk+erPT09AvqMeRyVYh06NBBMTExjd4dX1NTo9TU1BZaVeg0ZDhbvtTUVB0+fDho/6lTp3TkyJEL6j6YOHGiVq5cqbfeekuXXnqpPZ6amqra2lodPXo0aP43MzZ1HzTsuxA4HA5dfvnl6tevn8rKytSrVy/NmzfPiHyVlZU6fPiw+vbtq1atWqlVq1bauHGjfvvb36pVq1ZKSUmJ+IzflJiYqB/+8If68MMPjXgMJSktLU3du3cPGuvWrZt9Wc6k3zeffPKJ1q1bp3vuucceM+FxnDp1qh566CHdeuutys7O1u23364pU6aorKxM0oXzGFJyQsThcKhfv36qqKiwx+rr61VRUSGXy9WCKwuNrKwspaamBuXz+Xzatm2bnc/lcuno0aOqrKy056xfv1719fXKycn5ztf8TZZlaeLEiXr99de1fv16ZWVlBe3v16+fYmNjgzLu27dPBw8eDMq4e/fuoP9hut1uJSQkNPqlfaGor6+X3+83It/gwYO1e/duVVVV2T/9+/fXmDFj7H9HesZvOn78uD766COlpaUZ8RhK0rXXXtvo6xv++te/KjMzU5IZv28aLFq0SMnJySooKLDHTHgcT548qejo4AoRExOj+vp6SRfQYxiSty/DsqyvP0LudDqtxYsXW++//741fvx4KzExMejd8ReyL774wtq5c6e1c+dOS5I1d+5ca+fOndYnn3xiWdbXHwdMTEy0/vjHP1q7du2yfvaznzX5ccA+ffpY27ZtszZv3mxdccUVF8xHOu+77z6rbdu21oYNG4I+2nny5El7zoQJE6yOHTta69evt3bs2GG5XC7L5XLZ+xs+1pmXl2dVVVVZa9assS655JIL5mOdDz30kLVx40Zr//791q5du6yHHnrIioqKssrLyy3Livx8TTn901WWFfkZH3jgAWvDhg3W/v37rT//+c9Wbm6u1aFDB+vw4cOWZUV+Psv6+uP/rVq1sh577DHrgw8+sJYsWWLFx8dbL7/8sj0n0n/fWNbXn7Dt2LGjNW3atEb7Iv1xHDt2rPWDH/zA/gj5a6+9ZnXo0MF68MEH7TkXwmNIyQmx3/3ud1bHjh0th8NhXXXVVdbWrVtbeknn7a233rIkNfoZO3asZVlffyTwkUcesVJSUiyn02kNHjzY2rdvX9AxPvvsM2vUqFHWxRdfbCUkJFh33XWX9cUXX7RAmsaayibJWrRokT3nyy+/tP7nf/7HSkpKsuLj462bbrrJOnToUNBxDhw4YA0bNsxq3bq11aFDB+uBBx6wAoHAd5ymaXfffbeVmZlpORwO65JLLrEGDx5sFxzLivx8TflmyYn0jLfccouVlpZmORwO6wc/+IF1yy23BH1/TKTna/Dmm29aPXr0sJxOp9W1a1fr+eefD9of6b9vLMuy1q5da0lqtG7LivzH0efzWZMmTbI6duxoxcXFWZdddpn1y1/+Mujj7RfCYxhlWad9PSEAAIAheE8OAAAwEiUHAAAYiZIDAACMRMkBAABGouQAAAAjUXIAAICRKDkAAMBIlBwAAGAkSg4AADASJQcAABiJkgMAAIxEyQEAAEb6/++OK+Wcx2owAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def process_clustering(name):\n",
    "    clustering = pd.read_csv('clustering/{}_cluster.tsv'.format(name), delimiter='\\t', header=None)\n",
    "    #rename heading as cluster reference and id\n",
    "    clustering.columns = [name, 'Entry']\n",
    "    clustering.drop_duplicates(subset='Entry', keep='first', inplace=True)\n",
    "    print(clustering[name].nunique())\n",
    "    clustering[name].value_counts().hist(bins=100)\n",
    "    return clustering\n",
    "\n",
    "swissprot = swissprot[['Entry', 'Entry Name', 'Sequence', 'EC number', 'Length']]\n",
    "\n",
    "cluster50 = process_clustering('clusterRes50')\n",
    "cluster30 = process_clustering('clusterRes30')\n",
    "\n",
    "swissprot = pd.merge(swissprot, cluster50, on='Entry', how='left')\n",
    "swissprot = pd.merge(swissprot, cluster30, on='Entry', how='left')\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "63ca5f5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "swissprot.to_csv('../processed_data/protein2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "756c5423",
   "metadata": {},
   "source": [
    "# Finally process textual description of EC numbers\n",
    "This is optionally used in CREEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "b5f94ed4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EC number</th>\n",
       "      <th>Text</th>\n",
       "      <th>Text Incomplete</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2008</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1231</th>\n",
       "      <td>1.1.1.100</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1646</th>\n",
       "      <td>1.1.1.101</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1616</th>\n",
       "      <td>1.1.1.102</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1535</th>\n",
       "      <td>7.6.2.16</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2973</th>\n",
       "      <td>7.6.2.5</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3871</th>\n",
       "      <td>7.6.2.7</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1043</th>\n",
       "      <td>7.6.2.8</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2502</th>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4146 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      EC number                                               Text  \\\n",
       "216     1.1.1.1  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "2008   1.1.1.10  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "1231  1.1.1.100  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "1646  1.1.1.101  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "1616  1.1.1.102  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "...         ...                                                ...   \n",
       "1535   7.6.2.16  primary active transmembrane transporter; ABC-...   \n",
       "2973    7.6.2.5  primary active transmembrane transporter; ABC-...   \n",
       "3871    7.6.2.7  primary active transmembrane transporter; ABC-...   \n",
       "1043    7.6.2.8  primary active transmembrane transporter; ABC-...   \n",
       "2502    7.6.2.9  primary active transmembrane transporter; ABC-...   \n",
       "\n",
       "      Text Incomplete  \n",
       "216             False  \n",
       "2008            False  \n",
       "1231            False  \n",
       "1646            False  \n",
       "1616            False  \n",
       "...               ...  \n",
       "1535            False  \n",
       "2973            False  \n",
       "3871            False  \n",
       "1043            False  \n",
       "2502            False  \n",
       "\n",
       "[4146 rows x 3 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read ECtoGO.txt line by line\n",
    "ECtoGO = open(\"raw_data/ECtoGO_raw.txt\", \"r\")\n",
    "ECtoGO_lines = ECtoGO.readlines()\n",
    "ECtoGO.close()\n",
    "\n",
    "#skip the first two lines\n",
    "ECtoGO_lines = ECtoGO_lines[2:]\n",
    "EC2desc = {}\n",
    "for line in ECtoGO_lines:\n",
    "    line = line.strip().split(\">\")\n",
    "    EC = line[0][3:-1]\n",
    "    desc = line[1].split(\";\")[0][4:-1]\n",
    "    EC2desc[EC] = desc\n",
    "\n",
    "def get_EC_desc(EC):\n",
    "    num_dashes = EC.count('-')\n",
    "\n",
    "    description_missing = False\n",
    "    EC1 = '.'.join(EC.split('.')[:1])\n",
    "    if EC1 + '.-.-.-' in EC2desc and num_dashes < 3:\n",
    "        desc1 = EC2desc[EC1 + '.-.-.-']\n",
    "    else:\n",
    "         desc1 = ''\n",
    "    EC2 = '.'.join(EC.split('.')[:2]) \n",
    "    if EC2 + '.-.-' in EC2desc and num_dashes < 2:\n",
    "        desc2 = EC2desc[EC2 + '.-.-']\n",
    "    else:\n",
    "        desc2 = ''\n",
    "    EC3 = '.'.join(EC.split('.')[:3])\n",
    "    if EC3 + '.-' in EC2desc and num_dashes < 1:\n",
    "        desc3 = EC2desc[EC3 + '.-']\n",
    "    else:\n",
    "        desc3 = ''\n",
    "        \n",
    "    if EC in EC2desc:\n",
    "        desc4 = EC2desc[EC]\n",
    "    else:\n",
    "        description_missing = True\n",
    "        desc4 = ''\n",
    "\n",
    "  \n",
    "    description = desc1 + '; ' + desc2 + '; ' + desc3 + '; ' + desc4\n",
    "    description = description.replace(' ; ', ' ')\n",
    "    description = description.replace(' ; ', ' ')\n",
    "    description = description.replace(' ; ', ' ')\n",
    "    description = description.replace(' activity', '')\n",
    "    #if string starts with ;, replace with space\n",
    "    if description[0] == ';':\n",
    "        description = description[2:]\n",
    "    if description[-2:] == '; ':\n",
    "        description = description[:-2]\n",
    "    return description\n",
    "\n",
    "#subset to the EC numbers in swissprot\n",
    "EC2desc_df = pd.DataFrame(swissprot['EC number'].unique(), columns=['EC number'])\n",
    "EC2desc_df['Text'] = EC2desc_df['EC number'].apply(get_EC_desc)\n",
    "EC2desc_df['Text Incomplete'] = ~EC2desc_df['EC number'].isin(EC2desc.keys())\n",
    "EC2desc_df = EC2desc_df.sort_values(by='EC number')\n",
    "EC2desc_df = EC2desc_df[EC2desc_df['EC number'].isin(swissprot['EC number'].unique())]\n",
    "EC2desc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "bda4ac06",
   "metadata": {},
   "outputs": [],
   "source": [
    "EC2desc_df.to_csv('../processed_data/text2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "c71ca7a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save ECs to a txt as the order of the cluster centers for downstream tasks\n",
    "with open('../processed_data/EC_list.txt', 'w') as f:\n",
    "    for ec in EC2desc_df['EC number']:\n",
    "        f.write(ec + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "401a1818-e84f-4b05-acb2-8dc734cd8c8a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70269638-6b8d-4ab4-8da2-2b1245085f81",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edde741a-fc23-4c5a-b543-4d7e5e2ab329",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc38c1de-37c5-4b8c-8950-5779d9548908",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
