{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "06e5ee54-4659-4353-b6ad-de92dd9ab319",
   "metadata": {},
   "source": [
    "# Data download and processing from Swissprot\n",
    "\n",
    "Downloaded the data on 13th of May 2024. 571,282 results filtering for reviewed \"Swiss-Prot\". (https://www.uniprot.org/uniprotkb?query=*&facets=reviewed%3Atrue)\n",
    "\n",
    "Selected download TSV and the columns, Seqeunce (under Sequences tab), EC number (under Function).\n",
    "\n",
    "Unzipped the downloaded file.\n",
    "\n",
    "## 1. Protein to EC processing\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4f961afc-aedd-41b9-8635-4c6eabd017ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>269</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>796</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>3423</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>248</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>471</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571277</th>\n",
       "      <td>V9HVX0</td>\n",
       "      <td>YPAA_ECOLI</td>\n",
       "      <td>61</td>\n",
       "      <td>MTIAERLRQEGHQIGWQEGKLEGLHEQAIKIALRMLEQGFDRDQVL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571278</th>\n",
       "      <td>W5X2N3</td>\n",
       "      <td>AP_FRAAN</td>\n",
       "      <td>193</td>\n",
       "      <td>MGWVWKDDDEQGGHVNPSAADISPRLDGDRCSTRKVVRTQCKTEEV...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571279</th>\n",
       "      <td>W6JGV7</td>\n",
       "      <td>REP_CPBDV</td>\n",
       "      <td>484</td>\n",
       "      <td>MSAPLSIEQDDLLTDDLKSWLSDIDFSNDNEEAIEMEPSDIEMSSP...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571280</th>\n",
       "      <td>W6JHZ8</td>\n",
       "      <td>VP1_CPBDV</td>\n",
       "      <td>210</td>\n",
       "      <td>MNVKGASDKAQLAMQAEWEEVLAPEGALAVEEASSVLKISDDERRS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571281</th>\n",
       "      <td>W6JQK2</td>\n",
       "      <td>YBW7_SCHPO</td>\n",
       "      <td>268</td>\n",
       "      <td>MIVIEPTILYEAAALYTRYTANGGDSNDVVALLVGPIVQTLSIPSS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>571282 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9     269   \n",
       "1       A0A023I7E1   ENG1_RHIMI     796   \n",
       "2       A0A024B7W1   POLG_ZIKVF    3423   \n",
       "3       A0A024SC78  CUTI1_HYPJR     248   \n",
       "4       A0A024SH76   GUX2_HYPJR     471   \n",
       "...            ...          ...     ...   \n",
       "571277      V9HVX0   YPAA_ECOLI      61   \n",
       "571278      W5X2N3     AP_FRAAN     193   \n",
       "571279      W6JGV7    REP_CPBDV     484   \n",
       "571280      W6JHZ8    VP1_CPBDV     210   \n",
       "571281      W6JQK2   YBW7_SCHPO     268   \n",
       "\n",
       "                                                 Sequence  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...   \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   \n",
       "...                                                   ...   \n",
       "571277  MTIAERLRQEGHQIGWQEGKLEGLHEQAIKIALRMLEQGFDRDQVL...   \n",
       "571278  MGWVWKDDDEQGGHVNPSAADISPRLDGDRCSTRKVVRTQCKTEEV...   \n",
       "571279  MSAPLSIEQDDLLTDDLKSWLSDIDFSNDNEEAIEMEPSDIEMSSP...   \n",
       "571280  MNVKGASDKAQLAMQAEWEEVLAPEGALAVEEASSVLKISDDERRS...   \n",
       "571281  MIVIEPTILYEAAALYTRYTANGGDSNDVVALLVGPIVQTLSIPSS...   \n",
       "\n",
       "                                                EC number  \n",
       "0                                        3.2.2.-; 3.2.2.6  \n",
       "1                                                3.2.1.39  \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...  \n",
       "3                                                3.1.1.74  \n",
       "4                                                3.2.1.91  \n",
       "...                                                   ...  \n",
       "571277                                                NaN  \n",
       "571278                                                NaN  \n",
       "571279                                                NaN  \n",
       "571280                                                NaN  \n",
       "571281                                                NaN  \n",
       "\n",
       "[571282 rows x 5 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "swissprot = pd.read_csv('raw_data/uniprotkb_AND_reviewed_true_2024_05_13.tsv', sep='\\t')\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58497313",
   "metadata": {},
   "source": [
    "Drop sequences that are fragments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ff073063",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>269</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>796</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>3423</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>248</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>471</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571277</th>\n",
       "      <td>V9HVX0</td>\n",
       "      <td>YPAA_ECOLI</td>\n",
       "      <td>61</td>\n",
       "      <td>MTIAERLRQEGHQIGWQEGKLEGLHEQAIKIALRMLEQGFDRDQVL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571278</th>\n",
       "      <td>W5X2N3</td>\n",
       "      <td>AP_FRAAN</td>\n",
       "      <td>193</td>\n",
       "      <td>MGWVWKDDDEQGGHVNPSAADISPRLDGDRCSTRKVVRTQCKTEEV...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571279</th>\n",
       "      <td>W6JGV7</td>\n",
       "      <td>REP_CPBDV</td>\n",
       "      <td>484</td>\n",
       "      <td>MSAPLSIEQDDLLTDDLKSWLSDIDFSNDNEEAIEMEPSDIEMSSP...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571280</th>\n",
       "      <td>W6JHZ8</td>\n",
       "      <td>VP1_CPBDV</td>\n",
       "      <td>210</td>\n",
       "      <td>MNVKGASDKAQLAMQAEWEEVLAPEGALAVEEASSVLKISDDERRS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571281</th>\n",
       "      <td>W6JQK2</td>\n",
       "      <td>YBW7_SCHPO</td>\n",
       "      <td>268</td>\n",
       "      <td>MIVIEPTILYEAAALYTRYTANGGDSNDVVALLVGPIVQTLSIPSS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>561990 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9     269   \n",
       "1       A0A023I7E1   ENG1_RHIMI     796   \n",
       "2       A0A024B7W1   POLG_ZIKVF    3423   \n",
       "3       A0A024SC78  CUTI1_HYPJR     248   \n",
       "4       A0A024SH76   GUX2_HYPJR     471   \n",
       "...            ...          ...     ...   \n",
       "571277      V9HVX0   YPAA_ECOLI      61   \n",
       "571278      W5X2N3     AP_FRAAN     193   \n",
       "571279      W6JGV7    REP_CPBDV     484   \n",
       "571280      W6JHZ8    VP1_CPBDV     210   \n",
       "571281      W6JQK2   YBW7_SCHPO     268   \n",
       "\n",
       "                                                 Sequence  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...   \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   \n",
       "...                                                   ...   \n",
       "571277  MTIAERLRQEGHQIGWQEGKLEGLHEQAIKIALRMLEQGFDRDQVL...   \n",
       "571278  MGWVWKDDDEQGGHVNPSAADISPRLDGDRCSTRKVVRTQCKTEEV...   \n",
       "571279  MSAPLSIEQDDLLTDDLKSWLSDIDFSNDNEEAIEMEPSDIEMSSP...   \n",
       "571280  MNVKGASDKAQLAMQAEWEEVLAPEGALAVEEASSVLKISDDERRS...   \n",
       "571281  MIVIEPTILYEAAALYTRYTANGGDSNDVVALLVGPIVQTLSIPSS...   \n",
       "\n",
       "                                                EC number  \n",
       "0                                        3.2.2.-; 3.2.2.6  \n",
       "1                                                3.2.1.39  \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...  \n",
       "3                                                3.1.1.74  \n",
       "4                                                3.2.1.91  \n",
       "...                                                   ...  \n",
       "571277                                                NaN  \n",
       "571278                                                NaN  \n",
       "571279                                                NaN  \n",
       "571280                                                NaN  \n",
       "571281                                                NaN  \n",
       "\n",
       "[561990 rows x 5 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fragments = pd.read_csv('raw_data/fragments.tsv', sep='\\t')['Entry'].values\n",
    "swissprot = swissprot[~swissprot['Entry'].isin(fragments)]\n",
    "swissprot\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee69eeba-0d87-4f0e-92f8-8d950cc616f3",
   "metadata": {},
   "source": [
    "## Explode out EC number so each EC has it's own row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2c501ce1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1763802/2647129001.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  swissprot['EC All'] = swissprot['EC number'].values\n",
      "/tmp/ipykernel_1763802/2647129001.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  swissprot['EC number'] = [ec.split(';') for ec in swissprot['EC number']]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18221\n",
      "5756\n",
      "230886\n"
     ]
    }
   ],
   "source": [
    "swissprot = swissprot[swissprot['EC number'].notna()]\n",
    "# Drop rows that don't have an ec number\n",
    "swissprot['EC All'] = swissprot['EC number'].values\n",
    "# Now expand out the ones we have left\n",
    "swissprot['EC number'] = [ec.split(';') for ec in swissprot['EC number']]\n",
    "# Then check how many have mulitple\n",
    "print(sum([1 if len(ec) > 1 else 0 for ec in swissprot['EC number'].values]))\n",
    "swissprot = swissprot.explode('EC number')\n",
    "# Clean the EC numbers\n",
    "swissprot['EC number'] = [ec.replace(' ', '') for ec in swissprot['EC number']]\n",
    "swissprot\n",
    "print(swissprot['EC number'].nunique())\n",
    "print(swissprot['Sequence'].nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4b81d7a",
   "metadata": {},
   "source": [
    "## Filtering steps \n",
    "\n",
    "* Filter to a certain sequence length\n",
    "* Drop duplicates\n",
    "* Remove EC numbers with dash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e42c854f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Partially filled EC numbers:  237556\n",
      "5399\n",
      "188331\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>EC All</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>269</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>796</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "      <td>3.2.1.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>248</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>3.1.1.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>471</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>3.2.1.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A044RE18</td>\n",
       "      <td>BLI_ONCVO</td>\n",
       "      <td>693</td>\n",
       "      <td>MYWQLVRILVLFDCLQKILAIEHDSICIADVDDACPEPSHTVMRLR...</td>\n",
       "      <td>3.4.21.75</td>\n",
       "      <td>3.4.21.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555831</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>240</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>4.1.1.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564731</th>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>Y3032_BACAN</td>\n",
       "      <td>584</td>\n",
       "      <td>MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>3.5.4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564754</th>\n",
       "      <td>Q6L032</td>\n",
       "      <td>Y1085_PICTO</td>\n",
       "      <td>573</td>\n",
       "      <td>MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>3.5.4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568711</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>172</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>3.6.1.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569733</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>124</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>3.1.1.29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200654 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9     269   \n",
       "1       A0A023I7E1   ENG1_RHIMI     796   \n",
       "3       A0A024SC78  CUTI1_HYPJR     248   \n",
       "4       A0A024SH76   GUX2_HYPJR     471   \n",
       "6       A0A044RE18    BLI_ONCVO     693   \n",
       "...            ...          ...     ...   \n",
       "555831      Q05115   AMDA_BORBO     240   \n",
       "564731      Q6HX62  Y3032_BACAN     584   \n",
       "564754      Q6L032  Y1085_PICTO     573   \n",
       "568711      Q94MV8   VG56_BPLZ5     172   \n",
       "569733      Q9J5H2   V029_FOWPN     124   \n",
       "\n",
       "                                                 Sequence  EC number  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.6   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...   3.2.1.39   \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   3.1.1.74   \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   3.2.1.91   \n",
       "6       MYWQLVRILVLFDCLQKILAIEHDSICIADVDDACPEPSHTVMRLR...  3.4.21.75   \n",
       "...                                                   ...        ...   \n",
       "555831  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...   4.1.1.76   \n",
       "564731  MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...    3.5.4.2   \n",
       "564754  MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...    3.5.4.2   \n",
       "568711  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...   3.6.1.12   \n",
       "569733  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...   3.1.1.29   \n",
       "\n",
       "                  EC All  \n",
       "0       3.2.2.-; 3.2.2.6  \n",
       "1               3.2.1.39  \n",
       "3               3.1.1.74  \n",
       "4               3.2.1.91  \n",
       "6              3.4.21.75  \n",
       "...                  ...  \n",
       "555831          4.1.1.76  \n",
       "564731           3.5.4.2  \n",
       "564754           3.5.4.2  \n",
       "568711          3.6.1.12  \n",
       "569733          3.1.1.29  \n",
       "\n",
       "[200654 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "swissprot = swissprot[swissprot['Length'] > 100]\n",
    "swissprot = swissprot[swissprot['Length'] < 1024] #used to be 1024\n",
    "swissprot = swissprot[~swissprot['EC number'].str.contains('-')]\n",
    "print('Partially filled EC numbers: ', len(swissprot))\n",
    "swissprot = swissprot.drop_duplicates(subset=['Sequence', 'EC number'])\n",
    "print(swissprot['EC number'].nunique())\n",
    "print(swissprot['Sequence'].nunique())\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8d17ad2",
   "metadata": {},
   "source": [
    "## 2. Reaction to EC processing\n",
    "\n",
    "EC and reaction data were downloaded from multiple sources, the first being ECReact.\n",
    "\n",
    "1. ECReact: https://www.nature.com/articles/s41467-022-28536-w, with the data downloaded from https://github.com/rxn4chemistry/biocatalysis-model specifically, https://github.com/rxn4chemistry/biocatalysis-model/blob/main/data/ecreact-1.0.csv\n",
    "2. EnzymeMap: https://chemrxiv.org/engage/chemrxiv/article-details/643e9f4034f4ff2ea4f4f84e https://zenodo.org/records/8254726 and the data https://github.com/hesther/enzymemap, using the data file https://github.com/hesther/enzymemap/blob/main/data/processed_reactions.csv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b6d71562",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>1.2.1.8</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C=C(C)CCOP(=O)([O-])OP(=O)([O-])[O-].CC(C)=CCO...</td>\n",
       "      <td>2.5.1.85</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O...</td>\n",
       "      <td>1.4.1.4</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>1.1.1.3</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cc1cn([C@H]2C[C@H](O)[C@@H](COP(=O)(O)OP(=O)(O...</td>\n",
       "      <td>2.7.7.9</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62217</th>\n",
       "      <td>CCCCC/C=C\\C/C=C\\C/C=C\\C/C=C\\C/C=C\\CCC(=O)O[C@@...</td>\n",
       "      <td>2.3.1.76</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62218</th>\n",
       "      <td>CCCCCCCCCCCC/C=C\\CCCCCCCC(=O)OC[C@H](COP(=O)(O...</td>\n",
       "      <td>2.7.8.41</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62219</th>\n",
       "      <td>CCCCCCCCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCCCCCCC...</td>\n",
       "      <td>2.7.8.41</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62220</th>\n",
       "      <td>CCCCCC/C=C\\CCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCC...</td>\n",
       "      <td>2.7.8.41</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62221</th>\n",
       "      <td>CCCCCCCC/C=C\\CCCCCCCC(=O)O[C@H](COC(=O)CCCCCCC...</td>\n",
       "      <td>4.1.1.65</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62222 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Reaction EC number  \\\n",
       "0      NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   1.2.1.8   \n",
       "1      C=C(C)CCOP(=O)([O-])OP(=O)([O-])[O-].CC(C)=CCO...  2.5.1.85   \n",
       "2      N.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O...   1.4.1.4   \n",
       "3      NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...   1.1.1.3   \n",
       "4      Cc1cn([C@H]2C[C@H](O)[C@@H](COP(=O)(O)OP(=O)(O...   2.7.7.9   \n",
       "...                                                  ...       ...   \n",
       "62217  CCCCC/C=C\\C/C=C\\C/C=C\\C/C=C\\C/C=C\\CCC(=O)O[C@@...  2.3.1.76   \n",
       "62218  CCCCCCCCCCCC/C=C\\CCCCCCCC(=O)OC[C@H](COP(=O)(O...  2.7.8.41   \n",
       "62219  CCCCCCCCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCCCCCCC...  2.7.8.41   \n",
       "62220  CCCCCC/C=C\\CCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCC...  2.7.8.41   \n",
       "62221  CCCCCCCC/C=C\\CCCCCCCC(=O)O[C@H](COC(=O)CCCCCCC...  4.1.1.65   \n",
       "\n",
       "                         Source  \n",
       "0        brenda_reaction_smiles  \n",
       "1        brenda_reaction_smiles  \n",
       "2        brenda_reaction_smiles  \n",
       "3        brenda_reaction_smiles  \n",
       "4        brenda_reaction_smiles  \n",
       "...                         ...  \n",
       "62217  pathbank_reaction_smiles  \n",
       "62218  pathbank_reaction_smiles  \n",
       "62219  pathbank_reaction_smiles  \n",
       "62220  pathbank_reaction_smiles  \n",
       "62221  pathbank_reaction_smiles  \n",
       "\n",
       "[62222 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ECreact = pd.read_csv('raw_data/ECReact.csv')\n",
    "ECreact['rxn_smiles'] = ECreact['rxn_smiles'].str.split('|').str[0] + '>>' + ECreact['rxn_smiles'].str.split('>>').str[1]\n",
    "ECreact.rename(columns={'rxn_smiles': 'Reaction', 'ec': 'EC number', 'source':'Source'}, inplace=True)\n",
    "ECreact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "21ac8cce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "349458\n",
      "4552\n",
      "47640\n",
      "47800\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...</td>\n",
       "      <td>[CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349453</th>\n",
       "      <td>CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + dimethylallyl diphosphate = dimethylally...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349454</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl diphosphate = geranyl-triphospho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349455</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl triphosphate = geranyl-tetraphos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349456</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl diphosphate = adenosine 5'-i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349457</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl triphosphate = isopentenyl-t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62896 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Reaction  \\\n",
       "0       CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1       CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...   \n",
       "25      CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...   \n",
       "26      CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...   \n",
       "27      NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                   ...   \n",
       "349453  CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349454  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...   \n",
       "349455  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...   \n",
       "349456  C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349457  C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...   \n",
       "\n",
       "                                          Mapped Reaction EC number  \\\n",
       "0       [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1       [CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...   1.1.1.1   \n",
       "25      [CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...   1.1.1.1   \n",
       "26      [CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...   1.1.1.1   \n",
       "27      [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                   ...       ...   \n",
       "349453  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...   6.5.1.3   \n",
       "349454  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349455  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349456  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "349457  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "\n",
       "                                            Reaction Text  \n",
       "0          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "25          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "26          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "27      2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...  \n",
       "...                                                   ...  \n",
       "349453  ATP + dimethylallyl diphosphate = dimethylally...  \n",
       "349454  ATP + geranyl diphosphate = geranyl-triphospho...  \n",
       "349455  ATP + geranyl triphosphate = geranyl-tetraphos...  \n",
       "349456  ATP + isopentenyl diphosphate = adenosine 5'-i...  \n",
       "349457  ATP + isopentenyl triphosphate = isopentenyl-t...  \n",
       "\n",
       "[62896 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enzymemap = pd.read_csv('raw_data/EnzymeMap.csv')\n",
    "print(len(enzymemap))\n",
    "enzymemap.rename(columns={'ec_num': 'EC number', 'unmapped': 'Reaction', 'mapped': 'Mapped Reaction', 'orig_rxn_text':'Reaction Text'}, inplace=True)\n",
    "enzymemap = enzymemap[['Reaction', 'Mapped Reaction', 'EC number', 'Reaction Text']]\n",
    "enzymemap.drop_duplicates(subset=['Reaction', 'EC number'], inplace=True)\n",
    "print(enzymemap['EC number'].nunique())\n",
    "print(enzymemap['Reaction'].nunique())\n",
    "print(enzymemap['Mapped Reaction'].nunique())\n",
    "enzymemap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a24320d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...</td>\n",
       "      <td>[CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349453</th>\n",
       "      <td>CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + dimethylallyl diphosphate = dimethylally...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349454</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl diphosphate = geranyl-triphospho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349455</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl triphosphate = geranyl-tetraphos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349456</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl diphosphate = adenosine 5'-i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349457</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl triphosphate = isopentenyl-t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62896 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Reaction  \\\n",
       "0       CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1       CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...   \n",
       "25      CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...   \n",
       "26      CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...   \n",
       "27      NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                   ...   \n",
       "349453  CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349454  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...   \n",
       "349455  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...   \n",
       "349456  C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349457  C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...   \n",
       "\n",
       "                                          Mapped Reaction EC number  \\\n",
       "0       [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1       [CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...   1.1.1.1   \n",
       "25      [CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...   1.1.1.1   \n",
       "26      [CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...   1.1.1.1   \n",
       "27      [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                   ...       ...   \n",
       "349453  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...   6.5.1.3   \n",
       "349454  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349455  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349456  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "349457  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "\n",
       "                                            Reaction Text  \n",
       "0          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "25          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "26          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "27      2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...  \n",
       "...                                                   ...  \n",
       "349453  ATP + dimethylallyl diphosphate = dimethylally...  \n",
       "349454  ATP + geranyl diphosphate = geranyl-triphospho...  \n",
       "349455  ATP + geranyl triphosphate = geranyl-tetraphos...  \n",
       "349456  ATP + isopentenyl diphosphate = adenosine 5'-i...  \n",
       "349457  ATP + isopentenyl triphosphate = isopentenyl-t...  \n",
       "\n",
       "[62896 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enzymemap.dropna(subset=['Mapped Reaction']) #just double check that all of them have mapped reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b244d31b",
   "metadata": {},
   "source": [
    "### Append EC numbers that are covered by ECreact but not by EnzymeMap, to EnzymeMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "94ce2b25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2353"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#find the EC numbers covered by ECreact but not by enzymemap\n",
    "ECreact = ECreact[~ECreact['EC number'].str.contains('-')]\n",
    "not_covered = ECreact[~ECreact['EC number'].isin(enzymemap['EC number'].unique())]\n",
    "not_covered['EC number'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "89e5d4dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6905\n",
      "56144\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1172</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1173</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1176</th>\n",
       "      <td>NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15228</th>\n",
       "      <td>Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.j</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21396</th>\n",
       "      <td>C[C@@]1(O)OC[C@H](O)C1(O)O.Nc1ncnc2c1ncn2[C@@H...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.l</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34012</th>\n",
       "      <td>*CS(=O)(=O)[O-].Nc1ncnc2c1ncn2[C@@H]1O[C@H](CO...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.o</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15541</th>\n",
       "      <td>Cc1ncc(C[n+]2csc(CCO)c2C)c(N)n1.Nc1ncnc2c1ncn2...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.p</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33659</th>\n",
       "      <td>Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.q</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72490 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Reaction  \\\n",
       "0      CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1172   CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...   \n",
       "1173   CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...   \n",
       "1176   NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...   \n",
       "1177   NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                  ...   \n",
       "15228  Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...   \n",
       "21396  C[C@@]1(O)OC[C@H](O)C1(O)O.Nc1ncnc2c1ncn2[C@@H...   \n",
       "34012  *CS(=O)(=O)[O-].Nc1ncnc2c1ncn2[C@@H]1O[C@H](CO...   \n",
       "15541  Cc1ncc(C[n+]2csc(CCO)c2C)c(N)n1.Nc1ncnc2c1ncn2...   \n",
       "33659  Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...   \n",
       "\n",
       "                                         Mapped Reaction EC number  \\\n",
       "0      [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1172   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1173   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1176   [H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...   1.1.1.1   \n",
       "1177   [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                  ...       ...   \n",
       "15228                                                NaN   7.6.2.j   \n",
       "21396                                                NaN   7.6.2.l   \n",
       "34012                                                NaN   7.6.2.o   \n",
       "15541                                                NaN   7.6.2.p   \n",
       "33659                                                NaN   7.6.2.q   \n",
       "\n",
       "                                           Reaction Text  \n",
       "0         acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1172   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1173   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1176   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "1177   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "...                                                  ...  \n",
       "15228                                                NaN  \n",
       "21396                                                NaN  \n",
       "34012                                                NaN  \n",
       "15541                                                NaN  \n",
       "33659                                                NaN  \n",
       "\n",
       "[72490 rows x 4 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reaction2EC = pd.concat((enzymemap, not_covered[['Reaction', 'EC number']])).sort_values(by='EC number')\n",
    "print(reaction2EC['EC number'].nunique())\n",
    "print(reaction2EC['Reaction'].nunique())\n",
    "reaction2EC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac489eea",
   "metadata": {},
   "source": [
    "## 3. Filter protein2EC and reaction2EC to intersection of their EC numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "83eb77eb-dcb7-423d-ad4d-b0246c207fd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>EC All</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A0A061I403</td>\n",
       "      <td>FICD_CRIGR</td>\n",
       "      <td>455</td>\n",
       "      <td>MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...</td>\n",
       "      <td>2.7.7.108</td>\n",
       "      <td>2.7.7.108; 3.1.4.-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A0A067XR63</td>\n",
       "      <td>XTH7_DIOKA</td>\n",
       "      <td>268</td>\n",
       "      <td>MNAEGGNLHREFEITWGDGRARIHNNGGLLTLSLDRASGSGFRSKN...</td>\n",
       "      <td>2.4.1.207</td>\n",
       "      <td>2.4.1.207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A0A067XRK9</td>\n",
       "      <td>XTH6_DIOKA</td>\n",
       "      <td>299</td>\n",
       "      <td>MASSLTLPMAMAFTLLALSFASAMGGSMNSSRFDELFQPSWAFDHF...</td>\n",
       "      <td>2.4.1.207</td>\n",
       "      <td>2.4.1.207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A0A067YMX8</td>\n",
       "      <td>XTH8_DIOKA</td>\n",
       "      <td>288</td>\n",
       "      <td>MAASPYSIFAVQLLLLASWMLSSSSSNFNQDFNIAWGGGRARILNN...</td>\n",
       "      <td>2.4.1.207</td>\n",
       "      <td>2.4.1.207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>A0A0C5Q4Y6</td>\n",
       "      <td>C76H2_ROSOF</td>\n",
       "      <td>493</td>\n",
       "      <td>MDSFPLLAALFFILAATWFISFRRPRNLPPGPFPYPIVGNMLQLGT...</td>\n",
       "      <td>1.14.14.175</td>\n",
       "      <td>1.14.14.-; 1.14.14.175; 1.14.14.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534459</th>\n",
       "      <td>Q9ZNA5</td>\n",
       "      <td>SAHH_ROSDO</td>\n",
       "      <td>462</td>\n",
       "      <td>MTKDFIVKDIALAEFGRKELDIAETEMPGLMALRAEYGDSKPLAGA...</td>\n",
       "      <td>3.13.2.1</td>\n",
       "      <td>3.13.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534465</th>\n",
       "      <td>Q9ZNJ9</td>\n",
       "      <td>MSRB_HATHI</td>\n",
       "      <td>159</td>\n",
       "      <td>MNHKKEKEYKKLDTEKLKENLTELQYNVTQRNATEKPFLNKYDKHF...</td>\n",
       "      <td>1.8.4.12</td>\n",
       "      <td>1.8.4.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534723</th>\n",
       "      <td>V5TD18</td>\n",
       "      <td>AMBI3_FISAU</td>\n",
       "      <td>273</td>\n",
       "      <td>MIVSTSVEQSAQFSVKSLTPFGALLEATEDHSDIQQLSIEQLCQLT...</td>\n",
       "      <td>1.14.20.11</td>\n",
       "      <td>1.14.20.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534724</th>\n",
       "      <td>V5TES5</td>\n",
       "      <td>AMBI2_FISAU</td>\n",
       "      <td>330</td>\n",
       "      <td>MTQIINITQSKVISEQILRHVFRHRRLISDTEPCVHQPCSLCLAPH...</td>\n",
       "      <td>4.1.99.25</td>\n",
       "      <td>4.1.99.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534725</th>\n",
       "      <td>V5TF65</td>\n",
       "      <td>AMBI1_FISAU</td>\n",
       "      <td>319</td>\n",
       "      <td>MISEKILRHIFQYRRLLSDTEPCAKEPCSICLAPHLPKIQSFIENN...</td>\n",
       "      <td>4.1.99.25</td>\n",
       "      <td>4.1.99.25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14659 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "11      A0A061I403   FICD_CRIGR     455   \n",
       "15      A0A067XR63   XTH7_DIOKA     268   \n",
       "16      A0A067XRK9   XTH6_DIOKA     299   \n",
       "17      A0A067YMX8   XTH8_DIOKA     288   \n",
       "93      A0A0C5Q4Y6  C76H2_ROSOF     493   \n",
       "...            ...          ...     ...   \n",
       "534459      Q9ZNA5   SAHH_ROSDO     462   \n",
       "534465      Q9ZNJ9   MSRB_HATHI     159   \n",
       "534723      V5TD18  AMBI3_FISAU     273   \n",
       "534724      V5TES5  AMBI2_FISAU     330   \n",
       "534725      V5TF65  AMBI1_FISAU     319   \n",
       "\n",
       "                                                 Sequence    EC number  \\\n",
       "11      MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...    2.7.7.108   \n",
       "15      MNAEGGNLHREFEITWGDGRARIHNNGGLLTLSLDRASGSGFRSKN...    2.4.1.207   \n",
       "16      MASSLTLPMAMAFTLLALSFASAMGGSMNSSRFDELFQPSWAFDHF...    2.4.1.207   \n",
       "17      MAASPYSIFAVQLLLLASWMLSSSSSNFNQDFNIAWGGGRARILNN...    2.4.1.207   \n",
       "93      MDSFPLLAALFFILAATWFISFRRPRNLPPGPFPYPIVGNMLQLGT...  1.14.14.175   \n",
       "...                                                   ...          ...   \n",
       "534459  MTKDFIVKDIALAEFGRKELDIAETEMPGLMALRAEYGDSKPLAGA...     3.13.2.1   \n",
       "534465  MNHKKEKEYKKLDTEKLKENLTELQYNVTQRNATEKPFLNKYDKHF...     1.8.4.12   \n",
       "534723  MIVSTSVEQSAQFSVKSLTPFGALLEATEDHSDIQQLSIEQLCQLT...   1.14.20.11   \n",
       "534724  MTQIINITQSKVISEQILRHVFRHRRLISDTEPCVHQPCSLCLAPH...    4.1.99.25   \n",
       "534725  MISEKILRHIFQYRRLLSDTEPCAKEPCSICLAPHLPKIQSFIENN...    4.1.99.25   \n",
       "\n",
       "                                    EC All  \n",
       "11                      2.7.7.108; 3.1.4.-  \n",
       "15                               2.4.1.207  \n",
       "16                               2.4.1.207  \n",
       "17                               2.4.1.207  \n",
       "93      1.14.14.-; 1.14.14.175; 1.14.14.60  \n",
       "...                                    ...  \n",
       "534459                            3.13.2.1  \n",
       "534465                            1.8.4.12  \n",
       "534723                          1.14.20.11  \n",
       "534724                           4.1.99.25  \n",
       "534725                           4.1.99.25  \n",
       "\n",
       "[14659 rows x 6 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ns_swissprot = swissprot[~swissprot['EC number'].isin(reaction2EC['EC number'].unique())]\n",
    "ns_swissprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8f2dc451-4b77-4da7-99dd-3d69bbbd964d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "439"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ns_swissprot['EC number'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "624fb47e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "174744\n",
      "4960\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>EC All</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>269</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>796</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "      <td>3.2.1.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>248</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>3.1.1.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>471</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>3.2.1.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A044RE18</td>\n",
       "      <td>BLI_ONCVO</td>\n",
       "      <td>693</td>\n",
       "      <td>MYWQLVRILVLFDCLQKILAIEHDSICIADVDDACPEPSHTVMRLR...</td>\n",
       "      <td>3.4.21.75</td>\n",
       "      <td>3.4.21.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555831</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>240</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>4.1.1.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564731</th>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>Y3032_BACAN</td>\n",
       "      <td>584</td>\n",
       "      <td>MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>3.5.4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564754</th>\n",
       "      <td>Q6L032</td>\n",
       "      <td>Y1085_PICTO</td>\n",
       "      <td>573</td>\n",
       "      <td>MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>3.5.4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568711</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>172</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>3.6.1.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569733</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>124</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>3.1.1.29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>185995 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9     269   \n",
       "1       A0A023I7E1   ENG1_RHIMI     796   \n",
       "3       A0A024SC78  CUTI1_HYPJR     248   \n",
       "4       A0A024SH76   GUX2_HYPJR     471   \n",
       "6       A0A044RE18    BLI_ONCVO     693   \n",
       "...            ...          ...     ...   \n",
       "555831      Q05115   AMDA_BORBO     240   \n",
       "564731      Q6HX62  Y3032_BACAN     584   \n",
       "564754      Q6L032  Y1085_PICTO     573   \n",
       "568711      Q94MV8   VG56_BPLZ5     172   \n",
       "569733      Q9J5H2   V029_FOWPN     124   \n",
       "\n",
       "                                                 Sequence  EC number  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.6   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...   3.2.1.39   \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   3.1.1.74   \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   3.2.1.91   \n",
       "6       MYWQLVRILVLFDCLQKILAIEHDSICIADVDDACPEPSHTVMRLR...  3.4.21.75   \n",
       "...                                                   ...        ...   \n",
       "555831  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...   4.1.1.76   \n",
       "564731  MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...    3.5.4.2   \n",
       "564754  MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...    3.5.4.2   \n",
       "568711  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...   3.6.1.12   \n",
       "569733  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...   3.1.1.29   \n",
       "\n",
       "                  EC All  \n",
       "0       3.2.2.-; 3.2.2.6  \n",
       "1               3.2.1.39  \n",
       "3               3.1.1.74  \n",
       "4               3.2.1.91  \n",
       "6              3.4.21.75  \n",
       "...                  ...  \n",
       "555831          4.1.1.76  \n",
       "564731           3.5.4.2  \n",
       "564754           3.5.4.2  \n",
       "568711          3.6.1.12  \n",
       "569733          3.1.1.29  \n",
       "\n",
       "[185995 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "swissprot = swissprot[swissprot['EC number'].isin(reaction2EC['EC number'].unique())]\n",
    "print(swissprot['Sequence'].nunique())\n",
    "print(swissprot['EC number'].nunique())\n",
    "swissprot # 149788"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b15e8b1d-d035-49a9-a524-d79453b5609e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4960"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(swissprot['EC number'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "443b6cdb-fe81-40ba-86a6-a52fe0103c29",
   "metadata": {},
   "source": [
    "## Save unique proteins\n",
    "\n",
    "#save the indices corresponding to unique proteins, for downstream use\n",
    "unique_protein_indices = swissprot.reset_index().drop_duplicates(subset=['Sequence']).index\n",
    "with open('unique_protein_indices.txt', 'w') as f:\n",
    "    for index in unique_protein_indices:\n",
    "        f.write(str(index) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "407b09c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4960\n",
      "48392\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1172</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1173</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1176</th>\n",
       "      <td>NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34226</th>\n",
       "      <td>CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@@H]8[C@H...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34223</th>\n",
       "      <td>CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@H]([C@H]...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15471</th>\n",
       "      <td>C[N+]1(C)CCC[C@H]1C(=O)[O-].Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)(O)OP(=O)(O...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8665</th>\n",
       "      <td>*[N+](*)(*)*.Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61766 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Reaction  \\\n",
       "0      CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1172   CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...   \n",
       "1173   CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...   \n",
       "1176   NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...   \n",
       "1177   NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                  ...   \n",
       "34226  CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@@H]8[C@H...   \n",
       "34223  CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@H]([C@H]...   \n",
       "15471  C[N+]1(C)CCC[C@H]1C(=O)[O-].Nc1ncnc2c1ncn2[C@@...   \n",
       "644    Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)(O)OP(=O)(O...   \n",
       "8665   *[N+](*)(*)*.Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=...   \n",
       "\n",
       "                                         Mapped Reaction EC number  \\\n",
       "0      [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1172   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1173   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1176   [H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...   1.1.1.1   \n",
       "1177   [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                  ...       ...   \n",
       "34226                                                NaN   7.6.2.8   \n",
       "34223                                                NaN   7.6.2.8   \n",
       "15471                                                NaN   7.6.2.9   \n",
       "644                                                  NaN   7.6.2.9   \n",
       "8665                                                 NaN   7.6.2.9   \n",
       "\n",
       "                                           Reaction Text  \n",
       "0         acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1172   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1173   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1176   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "1177   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "...                                                  ...  \n",
       "34226                                                NaN  \n",
       "34223                                                NaN  \n",
       "15471                                                NaN  \n",
       "644                                                  NaN  \n",
       "8665                                                 NaN  \n",
       "\n",
       "[61766 rows x 4 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reaction2EC = reaction2EC[reaction2EC['EC number'].isin(swissprot['EC number'].unique())]\n",
    "print(reaction2EC['EC number'].nunique())\n",
    "print(reaction2EC['Reaction'].nunique())\n",
    "reaction2EC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5721bdf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "reaction2EC.to_csv('../processed_data/reaction2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5250f2c8",
   "metadata": {},
   "source": [
    "## 4. Perform protein clustering using mmseqs\n",
    "\n",
    "Use mmseqs to identify clusters, https://github.com/soedinglab/MMseqs2\n",
    "\n",
    "Optional parameters:\n",
    "\n",
    "```\n",
    " --seed-sub-mat TWIN              Substitution matrix file for k-mer generation [aa:VTML80.out,nucl:nucleotide.out]\n",
    " -s FLOAT                         Sensitivity: 1.0 faster; 4.0 fast; 7.5 sensitive [4.000]\n",
    " -k INT                           k-mer length (0: automatically set to optimum) [0]\n",
    " --target-search-mode INT         target search mode (0: regular k-mer, 1: similar k-mer) [0]\n",
    " --k-score TWIN                   k-mer threshold for generating similar k-mer lists [seq:2147483647,prof:2147483647]\n",
    " --alph-size TWIN                 Alphabet size (range 2-21) [aa:21,nucl:5]\n",
    " --max-seqs INT                   Maximum results per query sequence allowed to pass the prefilter (affects sensitivity) [20]\n",
    " --split INT                      Split input into N equally distributed chunks. 0: set the best split automatically [0]\n",
    " --split-mode INT                 0: split target db; 1: split query db; 2: auto, depending on main memory [2]\n",
    " --split-memory-limit BYTE        Set max memory per split. E.g. 800B, 5K, 10M, 1G. Default (0) to all available system memory [0]\n",
    " --comp-bias-corr INT             Correct for locally biased amino acid composition (range 0-1) [1]\n",
    " --comp-bias-corr-scale FLOAT     Correct for locally biased amino acid composition (range 0-1) [1.000]\n",
    " --diag-score BOOL                Use ungapped diagonal scoring during prefilter [1]\n",
    " --exact-kmer-matching INT        Extract only exact k-mers for matching (range 0-1) [0]\n",
    " --mask INT                       Mask sequences in k-mer stage: 0: w/o low complexity masking, 1: with low complexity masking [1]\n",
    " --mask-prob FLOAT                Mask sequences is probablity is above threshold [0.900]\n",
    " --mask-lower-case INT            Lowercase letters will be excluded from k-mer search 0: include region, 1: exclude region [0]\n",
    " --min-ungapped-score INT         Accept only matches with ungapped alignment score above threshold [15]\n",
    " --add-self-matches BOOL          Artificially add entries of queries with themselves (for clustering) [0]\n",
    " --spaced-kmer-mode INT           0: use consecutive positions in k-mers; 1: use spaced k-mers [1]\n",
    " --spaced-kmer-pattern STR        User-specified spaced k-mer pattern []\n",
    " --local-tmp STR                  Path where some of the temporary files will be created []\n",
    "align:\n",
    " -c FLOAT                         List matches above this fraction of aligned (covered) residues (see --cov-mode) [0.800]\n",
    " --cov-mode INT                   0: coverage of query and target\n",
    "                                  1: coverage of target\n",
    "                                  2: coverage of query\n",
    "                                  3: target seq. length has to be at least x% of query length\n",
    "                                  4: query seq. length has to be at least x% of target length\n",
    "                                  5: short seq. needs to be at least x% of the other seq. length [0]\n",
    " -a BOOL                          Add backtrace string (convert to alignments with mmseqs convertalis module) [0]\n",
    " --alignment-mode INT             How to compute the alignment:\n",
    "                                  0: automatic\n",
    "                                  1: only score and end_pos\n",
    "                                  2: also start_pos and cov\n",
    "                                  3: also seq.id\n",
    "                                  4: only ungapped alignment [3]\n",
    " --alignment-output-mode INT      How to compute the alignment:\n",
    "                                  0: automatic\n",
    "                                  1: only score and end_pos\n",
    "                                  2: also start_pos and cov\n",
    "                                  3: also seq.id\n",
    "                                  4: only ungapped alignment\n",
    "                                  5: score only (output) cluster format [0]\n",
    " --wrapped-scoring BOOL           Double the (nucleotide) query sequence during the scoring process to allow wrapped diagonal scoring around end and start [0]\n",
    " -e DOUBLE                        List matches below this E-value (range 0.0-inf) [1.000E-03]\n",
    " --min-seq-id FLOAT               List matches above this sequence identity (for clustering) (range 0.0-1.0) [0.000]\n",
    " --min-aln-len INT                Minimum alignment length (range 0-INT_MAX) [0]\n",
    " --seq-id-mode INT                0: alignment length 1: shorter, 2: longer sequence [0]\n",
    " --alt-ali INT                    Show up to this many alternative alignments [0]\n",
    " --max-rejected INT               Maximum rejected alignments before alignment calculation for a query is stopped [2147483647]\n",
    " --max-accept INT                 Maximum accepted alignments before alignment calculation for a query is stopped [2147483647]\n",
    " --score-bias FLOAT               Score bias when computing SW alignment (in bits) [0.000]\n",
    " --realign BOOL                   Compute more conservative, shorter alignments (scores and E-values not changed) [0]\n",
    " --realign-score-bias FLOAT       Additional bias when computing realignment [-0.200]\n",
    " --realign-max-seqs INT           Maximum number of results to return in realignment [2147483647]\n",
    " --corr-score-weight FLOAT        Weight of backtrace correlation score that is added to the alignment score [0.000]\n",
    " --gap-open TWIN                  Gap open cost [aa:11,nucl:5]\n",
    " --gap-extend TWIN                Gap extension cost [aa:1,nucl:2]\n",
    " --zdrop INT                      Maximal allowed difference between score values before alignment is truncated  (nucleotide alignment only) [40]\n",
    "clust:\n",
    " --cluster-mode INT               0: Set-Cover (greedy)\n",
    "                                  1: Connected component (BLASTclust)\n",
    "                                  2,3: Greedy clustering by sequence length (CDHIT) [0]\n",
    " --max-iterations INT             Maximum depth of breadth first search in connected component clustering [1000]\n",
    " --similarity-type INT            Type of score used for clustering. 1: alignment score 2: sequence identity [2]\n",
    " --single-step-clustering BOOL    Switch from cascaded to simple clustering workflow [0]\n",
    " --cluster-steps INT              Cascaded clustering steps from 1 to -s [3]\n",
    " --cluster-reassign BOOL          Cascaded clustering can cluster sequence that do not fulfill the clustering criteria.\n",
    "                                  Cluster reassignment corrects these errors [0]\n",
    "kmermatcher:\n",
    " --weights STR                    Weights used for cluster priorization []\n",
    " --cluster-weight-threshold FLOAT Weight threshold used for cluster priorization [0.900]\n",
    " --kmer-per-seq INT               k-mers per sequence [21]\n",
    " --kmer-per-seq-scale TWIN        Scale k-mer per sequence based on sequence length as kmer-per-seq val + scale x seqlen [aa:0.000,nucl:0.200]\n",
    " --adjust-kmer-len BOOL           Adjust k-mer length based on specificity (only for nucleotides) [0]\n",
    " --hash-shift INT                 Shift k-mer hash initialization [67]\n",
    " --include-only-extendable BOOL   Include only extendable [0]\n",
    " --ignore-multi-kmer BOOL         Skip k-mers occurring multiple times (>=2) [0]\n",
    "profile:\n",
    " --pca                            Pseudo count admixture strength []\n",
    " --pcb                            Pseudo counts: Neff at half of maximum admixture (range 0.0-inf) []\n",
    "misc:\n",
    " --taxon-list STR                 Taxonomy ID, possibly multiple values separated by ',' []\n",
    " --rescore-mode INT               Rescore diagonals with:\n",
    "                                  0: Hamming distance\n",
    "                                  1: local alignment (score only)\n",
    "                                  2: local alignment\n",
    "                                  3: global alignment\n",
    "                                  4: longest alignment fulfilling window quality criterion [0]\n",
    " --dbtype INT                     Database type 0: auto, 1: amino acid 2: nucleotides [0]\n",
    " --shuffle BOOL                   Shuffle input database [1]\n",
    " --createdb-mode INT              Createdb mode 0: copy data, 1: soft link data and write new index (works only with single line fasta/q) [1]\n",
    " --id-offset INT                  Numeric ids in index file are offset by this value [0]\n",
    "common:\n",
    " --sub-mat TWIN                   Substitution matrix file [aa:blosum62.out,nucl:nucleotide.out]\n",
    " --max-seq-len INT                Maximum sequence length [65535]\n",
    " --db-load-mode INT               Database preload mode 0: auto, 1: fread, 2: mmap, 3: mmap+touch [0]\n",
    " --threads INT                    Number of CPU-cores used (all by default) [384]\n",
    " --compressed INT                 Write compressed output [0]\n",
    " -v INT                           Verbosity level: 0: quiet, 1: +errors, 2: +warnings, 3: +info [3]\n",
    " --remove-tmp-files BOOL          Delete temporary files [1]\n",
    " --force-reuse BOOL               Reuse tmp filse in tmp/latest folder ignoring parameters and version changes [0]\n",
    " --mpi-runner STR                 Use MPI on compute cluster with this MPI command (e.g. \"mpirun -np 42\") []\n",
    "expert:\n",
    " --filter-hits BOOL               Filter hits by seq.id. and coverage [0]\n",
    " --sort-results INT               Sort results: 0: no sorting, 1: sort by E-value (Alignment) or seq.id. (Hamming) [0]\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "19acec52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate a fasta file as input to mmseqs\n",
    "with open('clustering/swissprot.fasta', 'w') as f:\n",
    "    for index, row in swissprot.iterrows():\n",
    "        f.write(f'>{row[\"Entry\"]}\\n{row[\"Sequence\"]}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6d844f90-6f1a-40a7-b85b-1e490621c753",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes30 clustering/mmseqs_tmp --min-seq-id 0.3 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.3\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/4759399735664770999/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[169782] 0s 33ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 33ms\n",
      "Create directory clustering/mmseqs_tmp/4759399735664770999/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/4759399735664770999/input clustering/mmseqs_tmp/4759399735664770999/clu clustering/mmseqs_tmp/4759399735664770999/clu_tmp --max-seqs 20 -c 0.8 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.3 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 5.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 3\n",
      "linclust clustering/mmseqs_tmp/4759399735664770999/input clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/clu_redundancy clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/4759399735664770999/input clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/linclust/7067493748046126929/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/4759399735664770999/input clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/linclust/7067493748046126929/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 169865 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=========================================>                       ] 64.53% 109.62K eta 0s                                        ] 9.90% 16.82K eta 0s         ] 20.65% 35.09K eta 0s                         ] 41.38% 70.29K eta 0s                         ] 53.03% 90.08K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 57.53K 0s 223ms    \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 18ms=================>] 99.28% 57.11K eta 0s        \n",
      "Time for processing: 0h 0m 0s 482ms\n",
      "align clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/linclust/7067493748046126929/input_step_redundancy clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/linclust/7067493748046126929/input_step_redundancy clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/linclust/7067493748046126929/pref_rescore2 clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/linclust/7067493748046126929/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 57527 type: Aminoacid\n",
      "Target database size: 57527 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 57.53K 0s 318ms    K eta 0s       \n",
      "Time for merging to aln: 0h 0m 0s 21ms\n",
      "108151 alignments calculated\n",
      "104479 sequence pairs passed the thresholds (0.966047 of overall calculated)\n",
      "1.816173 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 558ms\n",
      "clust clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/linclust/7067493748046126929/input_step_redundancy clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/linclust/7067493748046126929/aln clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/linclust/7067493748046126929/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 57.53K 0s 156ms           ] 54.78% 31.51K eta 0s         ] 84.68% 48.72K eta 0s       \n",
      "Sort entries\n",
      "Find missing connections\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 34.37K 0s 190ms    \n",
      "[================================================================>] 99.82% 34.31K eta 0s       \n",
      "1.740811 k-mers per position\n",
      "301 DB matches per sequence\n",
      "0 overflows\n",
      "9 sequences passed prefiltering per query sequence\n",
      "8 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 16ms\n",
      "Time for processing: 0h 0m 1s 454ms\n",
      "align clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step_redundancy clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step_redundancy clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/pref_step0 clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 34367 type: Aminoacid\n",
      "Target database size: 34367 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 34.37K 0s 493ms    \n",
      "Time for merging to aln_step0: 0h 0m 0s 17ms\n",
      "298804 alignments calculated\n",
      "257111 sequence pairs passed the thresholds (0.860467 of overall calculated)\n",
      "7.481334 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 684ms\n",
      "clust clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step_redundancy clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/aln_step0 clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 34.37K 0s 121ms                                                 ] 1.92% 661 eta 0s       =====>                            ] 56.04% 19.26K eta 0s       \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 32048 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 34.37K 0s 197ms    ======================>                           ] 57.98% 19.92K eta 0s       K eta 0s       \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 34.37K 0s 5ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 420ms\n",
      "Total time: 0h 0m 0s 462ms\n",
      "\n",
      "Size of the sequence database: 34367\n",
      "Size of the alignment database: 34367\n",
      "Number of clusters: 13307\n",
      "\n",
      "Writing results 0h 0m 0s 3ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 475ms\n",
      "createsubdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/clu_step0 clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step_redundancy clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "prefilter clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step1 clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step1 clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 3 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 13307 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 13307 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[===============>                                                 ] 22.14% 2.95K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 9685\n",
      "Target db start 1 to 9685\n",
      "[=================================================================] 100.00% 9.69K 0s 199ms                         ] 64.03% 6.20K eta 0s       \n",
      "[================================================================>] 99.78% 9.66K eta 0s       \n",
      "181.126744 k-mers per position\n",
      "3400 DB matches per sequence\n",
      "0 overflows\n",
      "17 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 12ms\n",
      "Time for processing: 0h 0m 1s 71ms\n",
      "align clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step2 clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step2 clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/pref_step2 clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 9685 type: Aminoacid\n",
      "Target database size: 9685 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 9.69K 0s 191ms    \n",
      "Time for merging to aln_step2: 0h 0m 0s 13ms\n",
      "74590 alignments calculated\n",
      "11194 sequence pairs passed the thresholds (0.150074 of overall calculated)\n",
      "1.155808 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 428ms\n",
      "clust clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step2 clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/aln_step2 clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/clu_step2 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 9.69K 0s 32ms     \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 227 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 9.69K 0s 28ms     \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 9.69K 0s 0ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 128ms\n",
      "Total time: 0h 0m 0s 164ms\n",
      "\n",
      "Size of the sequence database: 9685\n",
      "Size of the alignment database: 9685\n",
      "Number of clusters: 9012\n",
      "\n",
      "Writing results 0h 0m 0s 2ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 172ms\n",
      "mergeclusters clustering/mmseqs_tmp/4759399735664770999/input clustering/mmseqs_tmp/4759399735664770999/clu clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/clu_redundancy clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/clu_step0 clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/clu_step1 clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/clu_step2 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 34.37K 0s 111ms    ==================>                               ] 52.06% 17.89K eta 0s       K eta 0s       \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 13.31K 0s 194ms    ==========>                                            ] 31.72% 4.22K eta 0s       \n",
      "Clustering step 3\n",
      "[=================================================================] 100.00% 9.69K 0s 261ms    >] 99.36% 9.62K eta 0s       \n",
      "Clustering step 4\n",
      "[=================================================================] 100.00% 9.01K 0s 304ms    \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 169.87K 0s 502ms    \n",
      "Time for merging to clu: 0h 0m 0s 18ms\n",
      "Time for processing: 0h 0m 0s 570ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/pref_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/aln_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/clu_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/pref_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/aln_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/clu_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step1_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4759399735664770999/clu_tmp/4997855282980051500/input_step2_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering/mmseqs_tmp/4759399735664770999/input clustering/mmseqs_tmp/4759399735664770999/input clustering/mmseqs_tmp/4759399735664770999/clu clustering/mmseqs_tmp/4759399735664770999/cluster.tsv --threads 384 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 13ms\n",
      "Time for processing: 0h 0m 0s 201ms\n",
      "result2repseq clustering/mmseqs_tmp/4759399735664770999/input clustering/mmseqs_tmp/4759399735664770999/clu clustering/mmseqs_tmp/4759399735664770999/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 9.01K 0s 53ms     ==============================>] 98.75% 8.90K eta 0s       \n",
      "Time for merging to clu_rep: 0h 0m 0s 19ms\n",
      "Time for processing: 0h 0m 0s 242ms\n",
      "result2flat clustering/mmseqs_tmp/4759399735664770999/input clustering/mmseqs_tmp/4759399735664770999/input clustering/mmseqs_tmp/4759399735664770999/clu_rep clustering/mmseqs_tmp/4759399735664770999/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 33ms\n",
      "createseqfiledb clustering/mmseqs_tmp/4759399735664770999/input clustering/mmseqs_tmp/4759399735664770999/clu clustering/mmseqs_tmp/4759399735664770999/clu_seqs --threads 384 -v 3 \n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster clustering/swissprot.fasta clustering/clusterRes30 clustering/mmseqs_tmp --min-seq-id 0.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5e6cdd85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes50 clustering/mmseqs_tmp --min-seq-id 0.5 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.5\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/2979914437703504827/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[169782] 0s 34ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 34ms\n",
      "Create directory clustering/mmseqs_tmp/2979914437703504827/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/2979914437703504827/input clustering/mmseqs_tmp/2979914437703504827/clu clustering/mmseqs_tmp/2979914437703504827/clu_tmp --max-seqs 20 -c 0.8 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.5 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 3.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 3\n",
      "linclust clustering/mmseqs_tmp/2979914437703504827/input clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/clu_redundancy clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/2979914437703504827/input clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/linclust/13347807278733767210/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/2979914437703504827/input clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/linclust/13347807278733767210/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 169865 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=========================================>                       ] 63.74% 108.28K eta 0s                                  ] 8.01% 13.61K eta 0s                                                         ] 19.44% 33.02K eta 0s                         ] 30.10% 51.14K eta 0s         ] 41.76% 70.94K eta 0s       K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 57.53K 0s 209ms                      ] 32.29% 18.58K eta 0s       =======>                          ] 58.81% 33.83K eta 0s       =======================================>          ] 83.57% 48.07K eta 0s       \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 18ms=================>] 99.83% 57.43K eta 0s       \n",
      "Time for processing: 0h 0m 0s 480ms\n",
      "align clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/linclust/13347807278733767210/input_step_redundancy clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/linclust/13347807278733767210/input_step_redundancy clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/linclust/13347807278733767210/pref_rescore2 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/linclust/13347807278733767210/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 57527 type: Aminoacid\n",
      "Target database size: 57527 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 57.53K 0s 357ms      ] 77.37% 44.51K eta 0s       ==========================================>       ] 88.55% 50.94K eta 0s       \n",
      "Time for merging to aln: 0h 0m 0s 26ms\n",
      "104016 alignments calculated\n",
      "83035 sequence pairs passed the thresholds (0.798291 of overall calculated)\n",
      "1.443409 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 631ms\n",
      "clust clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/linclust/13347807278733767210/input_step_redundancy clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/linclust/13347807278733767210/aln clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/linclust/13347807278733767210/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[============>                                                    ] 18.62% 10.71K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 42.35K 0s 351ms    \n",
      "Index statistics\n",
      "Entries:          6434080\n",
      "DB size:          525 MB\n",
      "Avg k-mer size:   0.100532\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t174\n",
      "    AYPWGT\t119\n",
      "    GRGHCS\t109\n",
      "    SPGGGW\t107\n",
      "    HGGDFP\t90\n",
      "    ITVGGP\t88\n",
      "    FREPGR\t87\n",
      "    PVGVMT\t87\n",
      "    LGGFVY\t82\n",
      "    IGGINY\t69\n",
      "Time for index table init: 0h 0m 1s 125ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 42345\n",
      "Target db start 1 to 42345\n",
      "[=========================================>                       ] 63.95% 27.08K eta 0s       K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "1.680336 k-mers per position\n",
      "366 DB matches per sequence\n",
      "0 overflows\n",
      "11 sequences passed prefiltering per query sequence\n",
      "12 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 18ms\n",
      "Time for processing: 0h 0m 1s 880ms\n",
      "align clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step_redundancy clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step_redundancy clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/pref_step0 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 42345 type: Aminoacid\n",
      "Target database size: 42345 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 42.35K 0s 652ms    \n",
      "Time for merging to aln_step0: 0h 0m 0s 18ms\n",
      "455941 alignments calculated\n",
      "129716 sequence pairs passed the thresholds (0.284502 of overall calculated)\n",
      "3.063313 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 841ms\n",
      "clust clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step_redundancy clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/aln_step0 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 42.35K 0s 143ms    =>                                                            ] 7.33% 3.10K eta 0s       K eta 0s       \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 26.53K 0s 316ms     \n",
      "[================================================================>] 99.59% 26.42K eta 0s       \n",
      "7.439837 k-mers per position\n",
      "549 DB matches per sequence\n",
      "0 overflows\n",
      "10 sequences passed prefiltering per query sequence\n",
      "9 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 15ms\n",
      "Time for processing: 0h 0m 1s 590ms\n",
      "align clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step1 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step1 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/pref_step1 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 26532 type: Aminoacid\n",
      "Target database size: 26532 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 26.53K 0s 431ms    \n",
      "Time for merging to aln_step1: 0h 0m 0s 15ms\n",
      "239959 alignments calculated\n",
      "27878 sequence pairs passed the thresholds (0.116178 of overall calculated)\n",
      "1.050731 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 619ms\n",
      "clust clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step1 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/aln_step1 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/clu_step1 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 26.53K 0s 98ms        ] 23.42% 6.22K eta 0s       K eta 0s       \n",
      "Sort entries\n",
      "Find missing connections\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "createsubdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/clu_step1 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step1 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "prefilter clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step2 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step2 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 3 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 25802 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 25802 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 25.80K 0s 169ms    K eta 0s       \n",
      "Index table: Masked residues: 41892\n",
      "Index table: fill\n",
      "[==============>                                                  ] 21.98% 5.67K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for processing: 0h 0m 1s 669ms\n",
      "align clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step2 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step2 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/pref_step2 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 25802 type: Aminoacid\n",
      "Target database size: 25802 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 25.80K 0s 457ms    \n",
      "Time for merging to aln_step2: 0h 0m 0s 15ms\n",
      "285562 alignments calculated\n",
      "25924 sequence pairs passed the thresholds (0.090782 of overall calculated)\n",
      "1.004728 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 648ms\n",
      "clust clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step2 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/aln_step2 clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/clu_step2 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 25.80K 0s 74ms                              ] 29.99% 7.74K eta 0s       ==========================>       ] 87.80% 22.66K eta 0s       \n",
      "Sort entries\n",
      "Find missing connections\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 42.35K 0s 139ms    \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 26.53K 0s 270ms    K eta 0s       \n",
      "Clustering step 3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for merging to clu: 0h 0m 0s 22ms\n",
      "Time for processing: 0h 0m 0s 668ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/pref_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/aln_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/clu_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/pref_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/aln_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/clu_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step1_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_tmp/17624418231649833369/input_step2_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering/mmseqs_tmp/2979914437703504827/input clustering/mmseqs_tmp/2979914437703504827/input clustering/mmseqs_tmp/2979914437703504827/clu clustering/mmseqs_tmp/2979914437703504827/cluster.tsv --threads 384 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 16ms\n",
      "Time for processing: 0h 0m 0s 311ms\n",
      "result2repseq clustering/mmseqs_tmp/2979914437703504827/input clustering/mmseqs_tmp/2979914437703504827/clu clustering/mmseqs_tmp/2979914437703504827/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 25.74K 0s 94ms       ] 78.25% 20.14K eta 0s       \n",
      "Time for merging to clu_rep: 0h 0m 0s 28ms\n",
      "Time for processing: 0h 0m 0s 314ms\n",
      "result2flat clustering/mmseqs_tmp/2979914437703504827/input clustering/mmseqs_tmp/2979914437703504827/input clustering/mmseqs_tmp/2979914437703504827/clu_rep clustering/mmseqs_tmp/2979914437703504827/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 51ms\n",
      "createseqfiledb clustering/mmseqs_tmp/2979914437703504827/input clustering/mmseqs_tmp/2979914437703504827/clu clustering/mmseqs_tmp/2979914437703504827/clu_seqs --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 25.74K 0s 160ms    K eta 0s       K eta 0s       \n",
      "Time for merging to clu_seqs: 0h 0m 0s 14ms\n",
      "Time for processing: 0h 0m 0s 405ms\n",
      "result2flat clustering/mmseqs_tmp/2979914437703504827/input clustering/mmseqs_tmp/2979914437703504827/input clustering/mmseqs_tmp/2979914437703504827/clu_seqs clustering/mmseqs_tmp/2979914437703504827/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 171ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering/mmseqs_tmp/2979914437703504827/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster clustering/swissprot.fasta clustering/clusterRes50 clustering/mmseqs_tmp --min-seq-id 0.5 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c87d03a8-8931-46fe-93ea-adfd02f557b9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes70 clustering/mmseqs_tmp --min-seq-id 0.7 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.7\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/1200429139742238655/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[169782] 0s 33ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 34ms\n",
      "Create directory clustering/mmseqs_tmp/1200429139742238655/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/1200429139742238655/input clustering/mmseqs_tmp/1200429139742238655/clu clustering/mmseqs_tmp/1200429139742238655/clu_tmp --max-seqs 20 -c 0.8 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.7 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 1.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 1\n",
      "linclust clustering/mmseqs_tmp/1200429139742238655/input clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/clu_redundancy clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.7 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/1200429139742238655/input clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/linclust/14664834640486547546/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.7 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/1200429139742238655/input clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/linclust/14664834640486547546/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.7 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 169865 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=========================================>                       ] 63.16% 107.28K eta 0s                                                                   ] 6.30% 10.69K eta 0s                                         ] 17.75% 30.16K eta 0s         ] 27.41% 46.56K eta 0s                                         ] 40.30% 68.46K eta 0s                         ] 52.34% 88.90K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 76.07K 0s 242ms    K eta 0s       \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 21ms=================>] 99.05% 75.34K eta 0s       K eta 0s       \n",
      "Time for processing: 0h 0m 0s 490ms\n",
      "align clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/linclust/14664834640486547546/input_step_redundancy clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/linclust/14664834640486547546/input_step_redundancy clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/linclust/14664834640486547546/pref_rescore2 clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/linclust/14664834640486547546/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.7 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 76068 type: Aminoacid\n",
      "Target database size: 76068 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 76.07K 0s 342ms                      ] 59.50% 45.26K eta 0s       ====================>             ] 79.69% 60.62K eta 0s       \n",
      "Time for merging to aln: 0h 0m 0s 26ms\n",
      "151444 alignments calculated\n",
      "86976 sequence pairs passed the thresholds (0.574311 of overall calculated)\n",
      "1.143398 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 562ms\n",
      "clust clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/linclust/14664834640486547546/input_step_redundancy clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/linclust/14664834640486547546/aln clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/linclust/14664834640486547546/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[============>                                                    ] 19.56% 14.88K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[===============================================>                 ] 73.81% 49.91K eta 0s       ====>                             ] 55.05% 37.22K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[============================================>                    ] 68.38% 46.24K eta 0s                         ] 21.39% 14.46K eta 0s         ] 37.65% 25.46K eta 0s       K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 67.62K 0s 201ms    \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 67.62K 0s 2ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 496ms\n",
      "Total time: 0h 0m 0s 539ms\n",
      "\n",
      "Size of the sequence database: 67622\n",
      "Size of the alignment database: 67622\n",
      "Number of clusters: 61736\n",
      "\n",
      "Writing results 0h 0m 0s 6ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 559ms\n",
      "mergeclusters clustering/mmseqs_tmp/1200429139742238655/input clustering/mmseqs_tmp/1200429139742238655/clu clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/clu_redundancy clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/clu_step0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 67.62K 0s 177ms    K eta 0s       ===========>      ] 89.29% 60.38K eta 0s       \n",
      "Clustering step 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 169.87K 0s 548ms    \n",
      "Time for merging to clu: 0h 0m 0s 29ms\n",
      "Time for processing: 0h 0m 0s 632ms\n",
      "rmdb clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/1200429139742238655/clu_tmp/1687116692744085081/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering/mmseqs_tmp/1200429139742238655/input clustering/mmseqs_tmp/1200429139742238655/input clustering/mmseqs_tmp/1200429139742238655/clu clustering/mmseqs_tmp/1200429139742238655/cluster.tsv --threads 384 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 21ms\n",
      "Time for processing: 0h 0m 0s 215ms\n",
      "result2repseq clustering/mmseqs_tmp/1200429139742238655/input clustering/mmseqs_tmp/1200429139742238655/clu clustering/mmseqs_tmp/1200429139742238655/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 61.74K 0s 155ms                      ] 61.05% 37.69K eta 0s       \n",
      "Time for merging to clu_rep: 0h 0m 0s 40ms\n",
      "Time for processing: 0h 0m 0s 420ms\n",
      "result2flat clustering/mmseqs_tmp/1200429139742238655/input clustering/mmseqs_tmp/1200429139742238655/input clustering/mmseqs_tmp/1200429139742238655/clu_rep clustering/mmseqs_tmp/1200429139742238655/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 89ms\n",
      "createseqfiledb clustering/mmseqs_tmp/1200429139742238655/input clustering/mmseqs_tmp/1200429139742238655/clu clustering/mmseqs_tmp/1200429139742238655/clu_seqs --threads 384 -v 3 \n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster clustering/swissprot.fasta clustering/clusterRes70 clustering/mmseqs_tmp --min-seq-id 0.7 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5665a723-85b5-458d-bc4a-9bbae4731d61",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes90 clustering/mmseqs_tmp --min-seq-id 0.9 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.9\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/17867687915490524099/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[169782] 0s 33ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 34ms\n",
      "Create directory clustering/mmseqs_tmp/17867687915490524099/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/17867687915490524099/input clustering/mmseqs_tmp/17867687915490524099/clu clustering/mmseqs_tmp/17867687915490524099/clu_tmp --max-seqs 20 -c 0.8 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.9 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 1.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 1\n",
      "linclust clustering/mmseqs_tmp/17867687915490524099/input clustering/mmseqs_tmp/17867687915490524099/clu_tmp/14639916083295034738/clu_redundancy clustering/mmseqs_tmp/17867687915490524099/clu_tmp/14639916083295034738/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/17867687915490524099/input clustering/mmseqs_tmp/17867687915490524099/clu_tmp/14639916083295034738/linclust/5117899077310921039/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.9 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/17867687915490524099/input clustering/mmseqs_tmp/17867687915490524099/clu_tmp/14639916083295034738/linclust/5117899077310921039/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.9 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 169865 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=========================================>                       ] 64.37% 109.34K eta 0s        0s       K eta 0s       =====>                                            ] 31.69% 53.82K eta 0s       K eta 0s       "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 109.75K 0s 355ms     eta 0s       K eta 0s                         ] 70.54% 77.42K eta 0s         ] 86.86% 95.32K eta 0s       \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 27ms=================>] 99.97% 109.71K eta 0s       \n",
      "Time for processing: 0h 0m 0s 602ms\n",
      "align clustering/mmseqs_tmp/17867687915490524099/clu_tmp/14639916083295034738/linclust/5117899077310921039/input_step_redundancy clustering/mmseqs_tmp/17867687915490524099/clu_tmp/14639916083295034738/linclust/5117899077310921039/input_step_redundancy clustering/mmseqs_tmp/17867687915490524099/clu_tmp/14639916083295034738/linclust/5117899077310921039/pref_rescore2 clustering/mmseqs_tmp/17867687915490524099/clu_tmp/14639916083295034738/linclust/5117899077310921039/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 109746 type: Aminoacid\n",
      "Target database size: 109746 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================>                                               ] 27.16% 29.81K eta 0s       \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 108.04K 0s 565ms    \n",
      "Index table: Masked residues: 106856\n",
      "Index table: fill\n",
      "[==============================>                                  ] 47.68% 51.52K eta 0s       1s       \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 108.04K 1s 896ms    \n",
      "Time for merging to aln_step0: 0h 0m 0s 29ms\n",
      "1753210 alignments calculated\n",
      "111245 sequence pairs passed the thresholds (0.063452 of overall calculated)\n",
      "1.029684 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 139ms\n",
      "clust clustering/mmseqs_tmp/17867687915490524099/clu_tmp/14639916083295034738/input_step_redundancy clustering/mmseqs_tmp/17867687915490524099/clu_tmp/14639916083295034738/aln_step0 clustering/mmseqs_tmp/17867687915490524099/clu_tmp/14639916083295034738/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 108.04K 0s 307ms     ] 16.47% 17.80K eta 0s         ] 27.59% 29.80K eta 0s                         ] 53.17% 57.44K eta 0s       \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 1 new connections.\n",
      "Reconstruct initial order\n",
      "[=======================================>                         ] 61.08% 65.99K eta 0s       \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[=================================================================] 100.00% 106.73K 0s 394ms    \n",
      "Time for merging to clu_seqs: 0h 0m 0s 30ms\n",
      "Time for processing: 0h 0m 0s 621ms\n",
      "result2flat clustering/mmseqs_tmp/17867687915490524099/input clustering/mmseqs_tmp/17867687915490524099/input clustering/mmseqs_tmp/17867687915490524099/clu_seqs clustering/mmseqs_tmp/17867687915490524099/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 211ms\n",
      "rmdb clustering/mmseqs_tmp/17867687915490524099/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/17867687915490524099/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/17867687915490524099/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 11ms\n",
      "rmdb clustering/mmseqs_tmp/17867687915490524099/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/17867687915490524099/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster clustering/swissprot.fasta clustering/clusterRes90 clustering/mmseqs_tmp --min-seq-id 0.9 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "02322b64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25739\n",
      "9012\n",
      "61736\n",
      "106727\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "      <th>EC All</th>\n",
       "      <th>clusterRes50</th>\n",
       "      <th>clusterRes30</th>\n",
       "      <th>clusterRes70</th>\n",
       "      <th>clusterRes90</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A1AY86</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>A8QPD8</td>\n",
       "      <td>A8QPD8</td>\n",
       "      <td>A8QPD8</td>\n",
       "      <td>A0A024SC78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>B2ABX7</td>\n",
       "      <td>B2AE04</td>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>A0A024SH76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>Q6K9A2</td>\n",
       "      <td>Q5FB34</td>\n",
       "      <td>Q9S9N9</td>\n",
       "      <td>A0A059TC02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A067XGX8</td>\n",
       "      <td>AROG2_PETHY</td>\n",
       "      <td>MALTATATTRGGSALPNSCLQTPKFQSLQKPTFISSFPTNKKTKPR...</td>\n",
       "      <td>2.5.1.54</td>\n",
       "      <td>512</td>\n",
       "      <td>2.5.1.54</td>\n",
       "      <td>P80574</td>\n",
       "      <td>P80576</td>\n",
       "      <td>Q75W16</td>\n",
       "      <td>A0A067XGX8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169860</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169861</th>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>Y3032_BACAN</td>\n",
       "      <td>MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>584</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>O34909</td>\n",
       "      <td>Q81BX6</td>\n",
       "      <td>Q6HX62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169862</th>\n",
       "      <td>Q6L032</td>\n",
       "      <td>Y1085_PICTO</td>\n",
       "      <td>MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>573</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>Q6L032</td>\n",
       "      <td>O34909</td>\n",
       "      <td>Q6L032</td>\n",
       "      <td>Q6L032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169863</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>P39262</td>\n",
       "      <td>P39262</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169864</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q9J5H2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>169865 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "1       A0A024SC78  CUTI1_HYPJR   \n",
       "2       A0A024SH76   GUX2_HYPJR   \n",
       "3       A0A059TC02   CCR1_PETHY   \n",
       "4       A0A067XGX8  AROG2_PETHY   \n",
       "...            ...          ...   \n",
       "169860      Q05115   AMDA_BORBO   \n",
       "169861      Q6HX62  Y3032_BACAN   \n",
       "169862      Q6L032  Y1085_PICTO   \n",
       "169863      Q94MV8   VG56_BPLZ5   \n",
       "169864      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence EC number  Length  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.6     269   \n",
       "1       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...  3.1.1.74     248   \n",
       "2       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...  3.2.1.91     471   \n",
       "3       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...  1.2.1.44     333   \n",
       "4       MALTATATTRGGSALPNSCLQTPKFQSLQKPTFISSFPTNKKTKPR...  2.5.1.54     512   \n",
       "...                                                   ...       ...     ...   \n",
       "169860  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...  4.1.1.76     240   \n",
       "169861  MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...   3.5.4.2     584   \n",
       "169862  MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...   3.5.4.2     573   \n",
       "169863  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...  3.6.1.12     172   \n",
       "169864  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...  3.1.1.29     124   \n",
       "\n",
       "                  EC All clusterRes50 clusterRes30 clusterRes70 clusterRes90  \n",
       "0       3.2.2.-; 3.2.2.6   A0A009IHW8       A1AY86   A0A009IHW8   A0A009IHW8  \n",
       "1               3.1.1.74       A8QPD8       A8QPD8       A8QPD8   A0A024SC78  \n",
       "2               3.2.1.91       B2ABX7       B2AE04   A0A024SH76   A0A024SH76  \n",
       "3               1.2.1.44       Q6K9A2       Q5FB34       Q9S9N9   A0A059TC02  \n",
       "4               2.5.1.54       P80574       P80576       Q75W16   A0A067XGX8  \n",
       "...                  ...          ...          ...          ...          ...  \n",
       "169860          4.1.1.76       Q05115       Q05115       Q05115       Q05115  \n",
       "169861           3.5.4.2       Q6HX62       O34909       Q81BX6       Q6HX62  \n",
       "169862           3.5.4.2       Q6L032       O34909       Q6L032       Q6L032  \n",
       "169863          3.6.1.12       P39262       P39262       Q94MV8       Q94MV8  \n",
       "169864          3.1.1.29       Q9J5H2       Q9J5H2       Q9J5H2       Q9J5H2  \n",
       "\n",
       "[169865 rows x 10 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkIAAAGdCAYAAAD+JxxnAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuNSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/xnp5ZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAtT0lEQVR4nO3de3BUdZ7//1fuF6STAEuajAEzowsi4SKRGLysLjHRzcwuyqJoRjOYAcVkhpAtUVAjoBiFUgFFWGZWmaoBRWpXRhGR/oYVVNoAEZSLoLMyiz+cTpyBpLlIp0k+vz92coaWi2HsGDqf56MqVZzzefc575N3F3lVd58kyhhjBAAAYKHozm4AAACgsxCEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWiu3sBs5nra2t+vLLL9W9e3dFRUV1djsAAKAdjDE6fPiwMjIyFB199td8CEJn8eWXXyozM7Oz2wAAAH+DL774QhdeeOFZawhCZ9G9e3dJ//eNdLlcYTtuMBjUunXrVFBQoLi4uLAdF+HFnCIDc4oMzCkydJU5+f1+ZWZmOj/Hz4YgdBZtb4e5XK6wB6Hk5GS5XK6IfqJ1dcwpMjCnyMCcIkNXm1N7PtbCh6UBAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQ6kR7c67QJwMu7ew2AACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFjrnIPQxo0b9ZOf/EQZGRmKiorSqlWrQtaNMaqqqlKfPn2UlJSk/Px8ffbZZyE1Bw8eVHFxsVwul1JTU1VaWqojR46E1Hz88ce65pprlJiYqMzMTM2ZM+eUXlauXKkBAwYoMTFR2dnZWrNmzTn3AgAA7HXOQejo0aMaMmSIFi5ceNr1OXPmaMGCBVq8eLFqa2vVrVs3FRYW6vjx405NcXGxdu3aJY/Ho9WrV2vjxo2aOHGis+73+1VQUKB+/fqprq5Oc+fO1YwZM7RkyRKnZtOmTbr99ttVWlqqbdu2afTo0Ro9erR27tx5Tr0AAACLme9Aknnttdec7dbWVuN2u83cuXOdfY2NjSYhIcG8/PLLxhhjdu/ebSSZLVu2ODVvvfWWiYqKMgcOHDDGGPPCCy+YtLQ0EwgEnJoHHnjA9O/f39m+9dZbTVFRUUg/ubm55p577ml3L9+mqanJSDJNTU3tqm+v5uZms2rVKrNj8BCzu/+AsB4b4dM2p+bm5s5uBWfBnCIDc4oMXWVO5/LzO6yfEdq3b598Pp/y8/OdfSkpKcrNzZXX65Ukeb1epaamKicnx6nJz89XdHS0amtrnZprr71W8fHxTk1hYaH27t2rQ4cOOTUnn6etpu087ekFAADYLTacB/P5fJKk9PT0kP3p6enOms/nU+/evUObiI1Vjx49QmqysrJOOUbbWlpamnw+37ee59t6+aZAIKBAIOBs+/1+SVIwGFQwGDzbpZ+TtmO1JiSEbOP80jYX5nN+Y06RgTlFhq4yp3PpP6xBKNJVV1dr5syZp+xft26dkpOTw36+fQ9NlyT9zzc+5I3zi8fj6ewW0A7MKTIwp8gQ6XM6duxYu2vDGoTcbrckqb6+Xn369HH219fXa+jQoU5NQ0NDyONOnDihgwcPOo93u92qr68PqWnb/raak9e/rZdvmjZtmiorK51tv9+vzMxMFRQUyOVyffs3oJ2CwaA8Ho+yZj+h6EBA/bduCduxET5tc7rhhhsUFxfX2e3gDJhTZGBOkaGrzKntHZ32CGsQysrKktvtVk1NjRM2/H6/amtrNWnSJElSXl6eGhsbVVdXp+HDh0uS1q9fr9bWVuXm5jo1Dz30kILBoDMIj8ej/v37Ky0tzampqalRRUWFc36Px6O8vLx29/JNCQkJSvjL21Uni4uL65AnRHQgoJhAIKKfbDboqPkjvJhTZGBOkSHS53QuvZ/zh6WPHDmi7du3a/v27ZL+70PJ27dv1/79+xUVFaWKigo9/vjjev3117Vjxw7dddddysjI0OjRoyVJl156qW688UZNmDBBmzdv1vvvv6/y8nKNGzdOGRkZkqQ77rhD8fHxKi0t1a5du7RixQrNnz8/5NWayZMna+3atXr66ae1Z88ezZgxQ1u3blV5ebkktasXAABgt3N+RWjr1q26/vrrne22cFJSUqKlS5dq6tSpOnr0qCZOnKjGxkZdffXVWrt2rRITE53HLFu2TOXl5Ro1apSio6M1ZswYLViwwFlPSUnRunXrVFZWpuHDh6tXr16qqqoK+V1DI0eO1PLly/Xwww9r+vTpuuSSS7Rq1SoNGjTIqWlPLwAAwF7nHISuu+46GWPOuB4VFaVZs2Zp1qxZZ6zp0aOHli9fftbzDB48WO++++5Za8aOHauxY8d+p14AAIC9+FtjAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaYQ9CLS0teuSRR5SVlaWkpCT96Ec/0mOPPSZjjFNjjFFVVZX69OmjpKQk5efn67PPPgs5zsGDB1VcXCyXy6XU1FSVlpbqyJEjITUff/yxrrnmGiUmJiozM1Nz5sw5pZ+VK1dqwIABSkxMVHZ2ttasWRPuSwYAABEq7EHoqaee0qJFi/T888/rk08+0VNPPaU5c+boueeec2rmzJmjBQsWaPHixaqtrVW3bt1UWFio48ePOzXFxcXatWuXPB6PVq9erY0bN2rixInOut/vV0FBgfr166e6ujrNnTtXM2bM0JIlS5yaTZs26fbbb1dpaam2bdum0aNHa/To0dq5c2e4LxsAAEQiE2ZFRUXm7rvvDtl3yy23mOLiYmOMMa2trcbtdpu5c+c6642NjSYhIcG8/PLLxhhjdu/ebSSZLVu2ODVvvfWWiYqKMgcOHDDGGPPCCy+YtLQ0EwgEnJoHHnjA9O/f39m+9dZbTVFRUUgvubm55p577mnXtTQ1NRlJpqmpqV317dXc3GxWrVpldgweYnb3HxDWYyN82ubU3Nzc2a3gLJhTZGBOkaGrzOlcfn7HhjtYjRw5UkuWLNGnn36qv//7v9dHH32k9957T88884wkad++ffL5fMrPz3cek5KSotzcXHm9Xo0bN05er1epqanKyclxavLz8xUdHa3a2lrdfPPN8nq9uvbaaxUfH+/UFBYW6qmnntKhQ4eUlpYmr9erysrKkP4KCwu1atWq0/YeCAQUCAScbb/fL0kKBoMKBoPf+XvTpu1YrQkJIds4v7TNhfmc35hTZGBOkaGrzOlc+g97EHrwwQfl9/s1YMAAxcTEqKWlRbNnz1ZxcbEkyefzSZLS09NDHpeenu6s+Xw+9e7dO7TR2Fj16NEjpCYrK+uUY7StpaWlyefznfU831RdXa2ZM2eesn/dunVKTk5u1/Wfi30PTZck/Q+fWzqveTyezm4B7cCcIgNzigyRPqdjx461uzbsQejVV1/VsmXLtHz5cl122WXavn27KioqlJGRoZKSknCfLqymTZsW8gqS3+9XZmamCgoK5HK5wnaeYDAoj8ejrNlPKDoQUP+tW8J2bIRP25xuuOEGxcXFdXY7OAPmFBmYU2ToKnNqe0enPcIehO6//349+OCDGjdunCQpOztb//u//6vq6mqVlJTI7XZLkurr69WnTx/ncfX19Ro6dKgkye12q6GhIeS4J06c0MGDB53Hu91u1dfXh9S0bX9bTdv6NyUkJCjhL29XnSwuLq5DnhDRgYBiAoGIfrLZoKPmj/BiTpGBOUWGSJ/TufQe9rvGjh07pujo0MPGxMSotbVVkpSVlSW3262amhpn3e/3q7a2Vnl5eZKkvLw8NTY2qq6uzqlZv369WltblZub69Rs3Lgx5H1Aj8ej/v37Ky0tzak5+TxtNW3nAQAAdgt7EPrJT36i2bNn680339Qf/vAHvfbaa3rmmWd08803S5KioqJUUVGhxx9/XK+//rp27Nihu+66SxkZGRo9erQk6dJLL9WNN96oCRMmaPPmzXr//fdVXl6ucePGKSMjQ5J0xx13KD4+XqWlpdq1a5dWrFih+fPnh7y1NXnyZK1du1ZPP/209uzZoxkzZmjr1q0qLy8P92UDAIAIFPa3xp577jk98sgjuu+++9TQ0KCMjAzdc889qqqqcmqmTp2qo0ePauLEiWpsbNTVV1+ttWvXKjEx0alZtmyZysvLNWrUKEVHR2vMmDFasGCBs56SkqJ169aprKxMw4cPV69evVRVVRXyu4ZGjhyp5cuX6+GHH9b06dN1ySWXaNWqVRo0aFC4LxsAAESgsAeh7t27a968eZo3b94Za6KiojRr1izNmjXrjDU9evTQ8uXLz3quwYMH69133z1rzdixYzV27Niz1gAAADvxt8YAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLU6JAgdOHBAP/3pT9WzZ08lJSUpOztbW7duddaNMaqqqlKfPn2UlJSk/Px8ffbZZyHHOHjwoIqLi+VyuZSamqrS0lIdOXIkpObjjz/WNddco8TERGVmZmrOnDmn9LJy5UoNGDBAiYmJys7O1po1azrikgEAQAQKexA6dOiQrrrqKsXFxemtt97S7t279fTTTystLc2pmTNnjhYsWKDFixertrZW3bp1U2FhoY4fP+7UFBcXa9euXfJ4PFq9erU2btyoiRMnOut+v18FBQXq16+f6urqNHfuXM2YMUNLlixxajZt2qTbb79dpaWl2rZtm0aPHq3Ro0dr586d4b5sAAAQiUyYPfDAA+bqq68+43pra6txu91m7ty5zr7GxkaTkJBgXn75ZWOMMbt37zaSzJYtW5yat956y0RFRZkDBw4YY4x54YUXTFpamgkEAiHn7t+/v7N96623mqKiopDz5+bmmnvuuadd19LU1GQkmaampnbVt1dzc7NZtWqV2TF4iNndf0BYj43waZtTc3NzZ7eCs2BOkYE5RYauMqdz+fkdG+5g9frrr6uwsFBjx47Vhg0b9IMf/ED33XefJkyYIEnat2+ffD6f8vPzncekpKQoNzdXXq9X48aNk9frVWpqqnJycpya/Px8RUdHq7a2VjfffLO8Xq+uvfZaxcfHOzWFhYV66qmndOjQIaWlpcnr9aqysjKkv8LCQq1ateq0vQcCAQUCAWfb7/dLkoLBoILB4Hf+3rRpO1ZrQkLINs4vbXNhPuc35hQZmFNk6CpzOpf+wx6EPv/8cy1atEiVlZWaPn26tmzZol/+8peKj49XSUmJfD6fJCk9PT3kcenp6c6az+dT7969QxuNjVWPHj1CarKysk45RttaWlqafD7fWc/zTdXV1Zo5c+Yp+9etW6fk5OT2fgvabd9D0yVJ/8Pnls5rHo+ns1tAOzCnyMCcIkOkz+nYsWPtrg17EGptbVVOTo6eeOIJSdKwYcO0c+dOLV68WCUlJeE+XVhNmzYt5BUkv9+vzMxMFRQUyOVyhe08wWBQHo9HWbOfUHQgoP5bt4Tt2AiftjndcMMNiouL6+x2cAbMKTIwp8jQVebU9o5Oe4Q9CPXp00cDBw4M2XfppZfqP//zPyVJbrdbklRfX68+ffo4NfX19Ro6dKhT09DQEHKMEydO6ODBg87j3W636uvrQ2ratr+tpm39mxISEpTwl7erThYXF9chT4joQEAxgUBEP9ls0FHzR3gxp8jAnCJDpM/pXHoP+11jV111lfbu3Ruy79NPP1W/fv0kSVlZWXK73aqpqXHW/X6/amtrlZeXJ0nKy8tTY2Oj6urqnJr169ertbVVubm5Ts3GjRtD3gf0eDzq37+/c4daXl5eyHnaatrOAwAA7Bb2IDRlyhR98MEHeuKJJ/T73/9ey5cv15IlS1RWViZJioqKUkVFhR5//HG9/vrr2rFjh+666y5lZGRo9OjRkv7vFaQbb7xREyZM0ObNm/X++++rvLxc48aNU0ZGhiTpjjvuUHx8vEpLS7Vr1y6tWLFC8+fPD3lra/LkyVq7dq2efvpp7dmzRzNmzNDWrVtVXl4e7ssGAAARKOxvjV1xxRV67bXXNG3aNM2aNUtZWVmaN2+eiouLnZqpU6fq6NGjmjhxohobG3X11Vdr7dq1SkxMdGqWLVum8vJyjRo1StHR0RozZowWLFjgrKekpGjdunUqKyvT8OHD1atXL1VVVYX8rqGRI0dq+fLlevjhhzV9+nRdcsklWrVqlQYNGhTuywYAABEo7EFIkn784x/rxz/+8RnXo6KiNGvWLM2aNeuMNT169NDy5cvPep7Bgwfr3XffPWvN2LFjNXbs2LM3DAAArMTfGgMAANYiCJ0Hsn+T3dktAABgJYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1OjwIPfnkk4qKilJFRYWz7/jx4yorK1PPnj11wQUXaMyYMaqvrw953P79+1VUVKTk5GT17t1b999/v06cOBFS88477+jyyy9XQkKCLr74Yi1duvSU8y9cuFAXXXSREhMTlZubq82bN3fEZQIAgAjUoUFoy5Yt+vd//3cNHjw4ZP+UKVP0xhtvaOXKldqwYYO+/PJL3XLLLc56S0uLioqK1NzcrE2bNuk3v/mNli5dqqqqKqdm3759Kioq0vXXX6/t27eroqJCP//5z/X22287NStWrFBlZaUeffRRffjhhxoyZIgKCwvV0NDQkZcNAAAiRIcFoSNHjqi4uFi/+tWvlJaW5uxvamrSf/zHf+iZZ57RP/7jP2r48OF66aWXtGnTJn3wwQeSpHXr1mn37t367W9/q6FDh+qmm27SY489poULF6q5uVmStHjxYmVlZenpp5/WpZdeqvLycv3rv/6rnn32WedczzzzjCZMmKDx48dr4MCBWrx4sZKTk/Xiiy921GUDAIAIEttRBy4rK1NRUZHy8/P1+OOPO/vr6uoUDAaVn5/v7BswYID69u0rr9erK6+8Ul6vV9nZ2UpPT3dqCgsLNWnSJO3atUvDhg2T1+sNOUZbTdtbcM3Nzaqrq9O0adOc9ejoaOXn58vr9Z6250AgoEAg4Gz7/X5JUjAYVDAY/Nu/Gd/QdqzWhARJUoJiw3p8hEfbTJjN+Y05RQbmFBm6ypzOpf8OCUKvvPKKPvzwQ23ZsuWUNZ/Pp/j4eKWmpobsT09Pl8/nc2pODkFt621rZ6vx+/36+uuvdejQIbW0tJy2Zs+ePaftu7q6WjNnzjxl/7p165ScnHyWK/7b7HtouiTpEUlr1qwJ+/ERHh6Pp7NbQDswp8jAnCJDpM/p2LFj7a4NexD64osvNHnyZHk8HiUmJob78B1q2rRpqqysdLb9fr8yMzNVUFAgl8sVtvMEg0F5PB5lzX5C0YGAflYZK+8dp3+VCp2nbU433HCD4uLiOrsdnAFzigzMKTJ0lTm1vaPTHmEPQnV1dWpoaNDll1/u7GtpadHGjRv1/PPP6+2331Zzc7MaGxtDXhWqr6+X2+2WJLnd7lPu7mq7q+zkmm/eaVZfXy+Xy6WkpCTFxMQoJibmtDVtx/imhIQEJfzl7aqTxcXFdcgTIjoQUEwgoIBaIvoJ19V11PwRXswpMjCnyBDpczqX3sP+YelRo0Zpx44d2r59u/OVk5Oj4uJi599xcXGqqalxHrN3717t379feXl5kqS8vDzt2LEj5O4uj8cjl8ulgQMHOjUnH6Otpu0Y8fHxGj58eEhNa2urampqnBoAAGC3sL8i1L17dw0aNChkX7du3dSzZ09nf2lpqSorK9WjRw+5XC794he/UF5enq688kpJUkFBgQYOHKg777xTc+bMkc/n08MPP6yysjLnFZt7771Xzz//vKZOnaq7775b69ev16uvvqo333zTOW9lZaVKSkqUk5OjESNGaN68eTp69KjGjx8f7ssGAAARqMPuGjubZ599VtHR0RozZowCgYAKCwv1wgsvOOsxMTFavXq1Jk2apLy8PHXr1k0lJSWaNWuWU5OVlaU333xTU6ZM0fz583XhhRfq17/+tQoLC52a2267TV999ZWqqqrk8/k0dOhQrV279pQPUAMAADt9L0HonXfeCdlOTEzUwoULtXDhwjM+pl+/ft96J9V1112nbdu2nbWmvLxc5eXl7e4VAADYg781BgAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEzhMXPfhmZ7cAAIB1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKwV9iBUXV2tK664Qt27d1fv3r01evRo7d27N6Tm+PHjKisrU8+ePXXBBRdozJgxqq+vD6nZv3+/ioqKlJycrN69e+v+++/XiRMnQmreeecdXX755UpISNDFF1+spUuXntLPwoULddFFFykxMVG5ubnavHlzuC8ZAABEqLAHoQ0bNqisrEwffPCBPB6PgsGgCgoKdPToUadmypQpeuONN7Ry5Upt2LBBX375pW655RZnvaWlRUVFRWpubtamTZv0m9/8RkuXLlVVVZVTs2/fPhUVFen666/X9u3bVVFRoZ///Od6++23nZoVK1aosrJSjz76qD788EMNGTJEhYWFamhoCPdlAwCASGQ6WENDg5FkNmzYYIwxprGx0cTFxZmVK1c6NZ988omRZLxerzHGmDVr1pjo6Gjj8/mcmkWLFhmXy2UCgYAxxpipU6eayy67LORct912myksLHS2R4wYYcrKypztlpYWk5GRYaqrq9vVe1NTk5FkmpqazvGqz665udmsWrXK7Bg8xOzuP8AMWjrI9HtgdVjPge+ubU7Nzc2d3QrOgjlFBuYUGbrKnM7l53dsRwetpqYmSVKPHj0kSXV1dQoGg8rPz3dqBgwYoL59+8rr9erKK6+U1+tVdna20tPTnZrCwkJNmjRJu3bt0rBhw+T1ekOO0VZTUVEhSWpublZdXZ2mTZvmrEdHRys/P19er/e0vQYCAQUCAWfb7/dLkoLBoILB4Hf4LoRqO1ZrQoIkKUGxao4xYT0Hvru2eTCX8xtzigzMKTJ0lTmdS/8dGoRaW1tVUVGhq666SoMGDZIk+Xw+xcfHKzU1NaQ2PT1dPp/PqTk5BLWtt62drcbv9+vrr7/WoUOH1NLSctqaPXv2nLbf6upqzZw585T969atU3Jycjuvuv32PTRdkvSIJI1o0Zo1a8J+Dnx3Ho+ns1tAOzCnyMCcIkOkz+nYsWPtru3QIFRWVqadO3fqvffe68jThM20adNUWVnpbPv9fmVmZqqgoEAulyts5wkGg/J4PMqa/YSiAwH9rDJWhz+doZ0zCsN2Dnx3bXO64YYbFBcX19nt4AyYU2RgTpGhq8yp7R2d9uiwIFReXq7Vq1dr48aNuvDCC539brdbzc3NamxsDHlVqL6+Xm6326n55t1dbXeVnVzzzTvN6uvr5XK5lJSUpJiYGMXExJy2pu0Y35SQkKCEv7xddbK4uLgOeUJEBwKKCQQUUIsCLVER/aTryjpq/ggv5hQZmFNkiPQ5nUvvYb9rzBij8vJyvfbaa1q/fr2ysrJC1ocPH664uDjV1NQ4+/bu3av9+/crLy9PkpSXl6cdO3aE3N3l8Xjkcrk0cOBAp+bkY7TVtB0jPj5ew4cPD6lpbW1VTU2NUwMAAOwW9leEysrKtHz5cv3ud79T9+7dnc/0pKSkKCkpSSkpKSotLVVlZaV69Oghl8ulX/ziF8rLy9OVV14pSSooKNDAgQN15513as6cOfL5fHr44YdVVlbmvGJz77336vnnn9fUqVN19913a/369Xr11Vf15pt//SvulZWVKikpUU5OjkaMGKF58+bp6NGjGj9+fLgvGwAARKCwB6FFixZJkq677rqQ/S+99JJ+9rOfSZKeffZZRUdHa8yYMQoEAiosLNQLL7zg1MbExGj16tWaNGmS8vLy1K1bN5WUlGjWrFlOTVZWlt58801NmTJF8+fP14UXXqhf//rXKiz86+dsbrvtNn311VeqqqqSz+fT0KFDtXbt2lM+QA0AAOwU9iBkjPnWmsTERC1cuFALFy48Y02/fv2+9S6q6667Ttu2bTtrTXl5ucrLy7+1JwAAYB/+1hgAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIInSf+kHhHZ7cAAIB1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEzhPZWX2V/Zvszm4DAACrEIQAAIC1CEIAAMBaBKHzwKvVJ/Rq9YnObgMAAOtYEYQWLlyoiy66SImJicrNzdXmzZs7u6UzuujBN50vAADQsbp8EFqxYoUqKyv16KOP6sMPP9SQIUNUWFiohoaGzm4NAAB0si4fhJ555hlNmDBB48eP18CBA7V48WIlJyfrxRdf7OzWTqv7pQ86X5qR0tntAADQpcV2dgMdqbm5WXV1dZo2bZqzLzo6Wvn5+fJ6vafUBwIBBQIBZ7upqUmSdPDgQQWDwbD1FQwGdezYMfljYxXd0uLs/9Xcv9ZMKo/VMPcPpcXDJEn/b///J0nKbV4YcqzaaaPC1hdCtc3pz3/+s+Li4jq7HZwBc4oMzCkydJU5HT58WJJkjPnW2i4dhP70pz+ppaVF6enpIfvT09O1Z8+eU+qrq6s1c+bMU/ZnZWV1WI9nNCl0s5fzr9tC9z/9fTQDAEDkOXz4sFJSzv7uSpcOQudq2rRpqqysdLZbW1t18OBB9ezZU1FRUWE7j9/vV2Zmpr744gu5XK6wHRfhxZwiA3OKDMwpMnSVORljdPjwYWVkZHxrbZcOQr169VJMTIzq6+tD9tfX18vtdp9Sn5CQoISEhJB9qampHdafy+WK6CeaLZhTZGBOkYE5RYauMKdveyWoTZf+sHR8fLyGDx+umpoaZ19ra6tqamqUl5fXiZ0BAIDzQZd+RUiSKisrVVJSopycHI0YMULz5s3T0aNHNX78+M5uDQAAdLIuH4Ruu+02ffXVV6qqqpLP59PQoUO1du3aUz5A/X1KSEjQo48+esrbcDi/MKfIwJwiA3OKDDbOKcq0594yAACALqhLf0YIAADgbAhCAADAWgQhAABgLYIQAACwFkGoEyxcuFAXXXSREhMTlZubq82bN3d2S9aorq7WFVdcoe7du6t3794aPXq09u7dG1Jz/PhxlZWVqWfPnrrgggs0ZsyYU34p5/79+1VUVKTk5GT17t1b999/v06cOPF9XopVnnzySUVFRamiosLZx5zODwcOHNBPf/pT9ezZU0lJScrOztbWrVuddWOMqqqq1KdPHyUlJSk/P1+fffZZyDEOHjyo4uJiuVwupaamqrS0VEeOHPm+L6XLamlp0SOPPKKsrCwlJSXpRz/6kR577LGQv8Nl9ZwMvlevvPKKiY+PNy+++KLZtWuXmTBhgklNTTX19fWd3ZoVCgsLzUsvvWR27txptm/fbv7pn/7J9O3b1xw5csSpuffee01mZqapqakxW7duNVdeeaUZOXKks37ixAkzaNAgk5+fb7Zt22bWrFljevXqZaZNm9YZl9Tlbd682Vx00UVm8ODBZvLkyc5+5tT5Dh48aPr162d+9rOfmdraWvP555+bt99+2/z+9793ap588kmTkpJiVq1aZT766CPzz//8zyYrK8t8/fXXTs2NN95ohgwZYj744APz7rvvmosvvtjcfvvtnXFJXdLs2bNNz549zerVq82+ffvMypUrzQUXXGDmz5/v1Ng8J4LQ92zEiBGmrKzM2W5paTEZGRmmurq6E7uyV0NDg5FkNmzYYIwxprGx0cTFxZmVK1c6NZ988omRZLxerzHGmDVr1pjo6Gjj8/mcmkWLFhmXy2UCgcD3ewFd3OHDh80ll1xiPB6P+Yd/+AcnCDGn88MDDzxgrr766jOut7a2GrfbbebOnevsa2xsNAkJCebll182xhize/duI8ls2bLFqXnrrbdMVFSUOXDgQMc1b5GioiJz9913h+y75ZZbTHFxsTGGOfHW2PeoublZdXV1ys/Pd/ZFR0crPz9fXq+3EzuzV1NTkySpR48ekqS6ujoFg8GQGQ0YMEB9+/Z1ZuT1epWdnR3ySzkLCwvl9/u1a9eu77H7rq+srExFRUUh85CY0/ni9ddfV05OjsaOHavevXtr2LBh+tWvfuWs79u3Tz6fL2ROKSkpys3NDZlTamqqcnJynJr8/HxFR0ertrb2+7uYLmzkyJGqqanRp59+Kkn66KOP9N577+mmm26SxJy6/G+WPp/86U9/UktLyym/1To9PV179uzppK7s1draqoqKCl111VUaNGiQJMnn8yk+Pv6UP7abnp4un8/n1Jxuhm1rCI9XXnlFH374obZs2XLKGnM6P3z++edatGiRKisrNX36dG3ZskW//OUvFR8fr5KSEuf7fLo5nDyn3r17h6zHxsaqR48ezClMHnzwQfn9fg0YMEAxMTFqaWnR7NmzVVxcLEnWz4kgBGuVlZVp586deu+99zq7FXzDF198ocmTJ8vj8SgxMbGz28EZtLa2KicnR0888YQkadiwYdq5c6cWL16skpKSTu4ObV599VUtW7ZMy5cv12WXXabt27eroqJCGRkZzEncNfa96tWrl2JiYk65s6W+vl5ut7uTurJTeXm5Vq9erf/+7//WhRde6Ox3u91qbm5WY2NjSP3JM3K73aedYdsavru6ujo1NDTo8ssvV2xsrGJjY7VhwwYtWLBAsbGxSk9PZ07ngT59+mjgwIEh+y699FLt379f0l+/z2f7P8/tdquhoSFk/cSJEzp48CBzCpP7779fDz74oMaNG6fs7GzdeeedmjJliqqrqyUxJ4LQ9yg+Pl7Dhw9XTU2Ns6+1tVU1NTXKy8vrxM7sYYxReXm5XnvtNa1fv15ZWVkh68OHD1dcXFzIjPbu3av9+/c7M8rLy9OOHTtC/lPweDxyuVyn/FDA32bUqFHasWOHtm/f7nzl5OSouLjY+Tdz6nxXXXXVKb9+4tNPP1W/fv0kSVlZWXK73SFz8vv9qq2tDZlTY2Oj6urqnJr169ertbVVubm538NVdH3Hjh1TdHToj/uYmBi1trZKYk7cNfY9e+WVV0xCQoJZunSp2b17t5k4caJJTU0NubMFHWfSpEkmJSXFvPPOO+aPf/yj83Xs2DGn5t577zV9+/Y169evN1u3bjV5eXkmLy/PWW+7LbugoMBs377drF271vzd3/0dt2V3sJPvGjOGOZ0PNm/ebGJjY83s2bPNZ599ZpYtW2aSk5PNb3/7W6fmySefNKmpqeZ3v/ud+fjjj82//Mu/nPa27GHDhpna2lrz3nvvmUsuuaRL3JZ9vigpKTE/+MEPnNvn/+u//sv06tXLTJ061amxeU4EoU7w3HPPmb59+5r4+HgzYsQI88EHH3R2S9aQdNqvl156yan5+uuvzX333WfS0tJMcnKyufnmm80f//jHkOP84Q9/MDfddJNJSkoyvXr1Mv/2b/9mgsHg93w1dvlmEGJO54c33njDDBo0yCQkJJgBAwaYJUuWhKy3traaRx55xKSnp5uEhAQzatQos3fv3pCaP//5z+b22283F1xwgXG5XGb8+PHm8OHD3+dldGl+v99MnjzZ9O3b1yQmJpof/vCH5qGHHgr5NRI2zynKmJN+tSQAAIBF+IwQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANb6/wHFBDE8luJljAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def process_clustering(name):\n",
    "    clustering = pd.read_csv('clustering/{}_cluster.tsv'.format(name), delimiter='\\t', header=None)\n",
    "    #rename heading as cluster reference and id\n",
    "    clustering.columns = [name, 'Entry']\n",
    "    clustering.drop_duplicates(subset='Entry', keep='first', inplace=True)\n",
    "    print(clustering[name].nunique())\n",
    "    clustering[name].value_counts().hist(bins=100)\n",
    "    return clustering\n",
    "\n",
    "swissprot = swissprot[['Entry', 'Entry Name', 'Sequence', 'EC number', 'Length', 'EC All']]\n",
    "\n",
    "cluster50 = process_clustering('clusterRes50')\n",
    "cluster30 = process_clustering('clusterRes30')\n",
    "cluster70 = process_clustering('clusterRes70')\n",
    "cluster90 = process_clustering('clusterRes90')\n",
    "\n",
    "swissprot = pd.merge(swissprot, cluster50, on='Entry', how='left')\n",
    "swissprot = pd.merge(swissprot, cluster30, on='Entry', how='left')\n",
    "swissprot = pd.merge(swissprot, cluster70, on='Entry', how='left')\n",
    "swissprot = pd.merge(swissprot, cluster90, on='Entry', how='left')\n",
    "\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "63ca5f5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1763802/4218798898.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  swissprot['EC3'] = swissprot['EC number'].str.split('.').str[:3].str.join('.')\n",
      "/tmp/ipykernel_1763802/4218798898.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  swissprot['EC2'] = swissprot['EC number'].str.split('.').str[:2].str.join('.')\n",
      "/tmp/ipykernel_1763802/4218798898.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  swissprot['EC1'] = swissprot['EC number'].str.split('.').str[:1].str.join('.')\n"
     ]
    }
   ],
   "source": [
    "swissprot['EC3'] = swissprot['EC number'].str.split('.').str[:3].str.join('.')\n",
    "swissprot['EC2'] = swissprot['EC number'].str.split('.').str[:2].str.join('.')\n",
    "swissprot['EC1'] = swissprot['EC number'].str.split('.').str[:1].str.join('.')\n",
    "swissprot.to_csv('../processed_data/protein2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "756c5423",
   "metadata": {},
   "source": [
    "# Finally process textual description of EC numbers\n",
    "This is optionally used in CREEP\n",
    "\n",
    "`raw_data/ECtoGO_raw.txt` was downloaded from http://current.geneontology.org/ontology/external2go/ec2go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b5f94ed4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EC number</th>\n",
       "      <th>Text</th>\n",
       "      <th>Text Incomplete</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1658</th>\n",
       "      <td>1.1.1.100</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2147</th>\n",
       "      <td>1.1.1.101</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2119</th>\n",
       "      <td>1.1.1.102</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>7.6.2.5</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1727</th>\n",
       "      <td>7.6.2.6</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4695</th>\n",
       "      <td>7.6.2.7</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>934</th>\n",
       "      <td>7.6.2.8</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3176</th>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4960 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      EC number                                               Text  \\\n",
       "327     1.1.1.1  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "2573   1.1.1.10  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "1658  1.1.1.100  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "2147  1.1.1.101  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "2119  1.1.1.102  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "...         ...                                                ...   \n",
       "83      7.6.2.5  primary active transmembrane transporter; ABC-...   \n",
       "1727    7.6.2.6  primary active transmembrane transporter; ABC-...   \n",
       "4695    7.6.2.7  primary active transmembrane transporter; ABC-...   \n",
       "934     7.6.2.8  primary active transmembrane transporter; ABC-...   \n",
       "3176    7.6.2.9  primary active transmembrane transporter; ABC-...   \n",
       "\n",
       "      Text Incomplete  \n",
       "327             False  \n",
       "2573            False  \n",
       "1658            False  \n",
       "2147            False  \n",
       "2119            False  \n",
       "...               ...  \n",
       "83              False  \n",
       "1727            False  \n",
       "4695            False  \n",
       "934             False  \n",
       "3176            False  \n",
       "\n",
       "[4960 rows x 3 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read ECtoGO.txt line by line\n",
    "ECtoGO = open(\"raw_data/ECtoGO_raw.txt\", \"r\")\n",
    "ECtoGO_lines = ECtoGO.readlines()\n",
    "ECtoGO.close()\n",
    "\n",
    "#skip the first two lines\n",
    "ECtoGO_lines = ECtoGO_lines[2:]\n",
    "EC2desc = {}\n",
    "for line in ECtoGO_lines:\n",
    "    line = line.strip().split(\">\")\n",
    "    EC = line[0][3:-1]\n",
    "    desc = line[1].split(\";\")[0][4:-1]\n",
    "    EC2desc[EC] = desc\n",
    "\n",
    "def get_EC_desc(EC):\n",
    "    num_dashes = EC.count('-')\n",
    "\n",
    "    description_missing = False\n",
    "    EC1 = '.'.join(EC.split('.')[:1])\n",
    "    if EC1 + '.-.-.-' in EC2desc and num_dashes < 3:\n",
    "        desc1 = EC2desc[EC1 + '.-.-.-']\n",
    "    else:\n",
    "         desc1 = ''\n",
    "    EC2 = '.'.join(EC.split('.')[:2]) \n",
    "    if EC2 + '.-.-' in EC2desc and num_dashes < 2:\n",
    "        desc2 = EC2desc[EC2 + '.-.-']\n",
    "    else:\n",
    "        desc2 = ''\n",
    "    EC3 = '.'.join(EC.split('.')[:3])\n",
    "    if EC3 + '.-' in EC2desc and num_dashes < 1:\n",
    "        desc3 = EC2desc[EC3 + '.-']\n",
    "    else:\n",
    "        desc3 = ''\n",
    "        \n",
    "    if EC in EC2desc:\n",
    "        desc4 = EC2desc[EC]\n",
    "    else:\n",
    "        description_missing = True\n",
    "        desc4 = ''\n",
    "\n",
    "  \n",
    "    description = desc1 + '; ' + desc2 + '; ' + desc3 + '; ' + desc4\n",
    "    description = description.replace(' ; ', ' ')\n",
    "    description = description.replace(' ; ', ' ')\n",
    "    description = description.replace(' ; ', ' ')\n",
    "    description = description.replace(' activity', '')\n",
    "    #if string starts with ;, replace with space\n",
    "    if description[0] == ';':\n",
    "        description = description[2:]\n",
    "    if description[-2:] == '; ':\n",
    "        description = description[:-2]\n",
    "    return description\n",
    "\n",
    "#subset to the EC numbers in swissprot\n",
    "EC2desc_df = pd.DataFrame(swissprot['EC number'].unique(), columns=['EC number'])\n",
    "EC2desc_df['Text'] = EC2desc_df['EC number'].apply(get_EC_desc)\n",
    "EC2desc_df['Text Incomplete'] = ~EC2desc_df['EC number'].isin(EC2desc.keys())\n",
    "EC2desc_df = EC2desc_df.sort_values(by='EC number')\n",
    "EC2desc_df = EC2desc_df[EC2desc_df['EC number'].isin(swissprot['EC number'].unique())]\n",
    "EC2desc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bda4ac06",
   "metadata": {},
   "outputs": [],
   "source": [
    "EC2desc_df.to_csv('../processed_data/text2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c71ca7a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save ECs to a txt as the order of the cluster centers for downstream tasks\n",
    "with open('../processed_data/EC_list.txt', 'w') as f:\n",
    "    for ec in EC2desc_df['EC number']:\n",
    "        f.write(ec + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aed0f440",
   "metadata": {},
   "source": [
    "raw_data/price.tsv was downloaded from the CLEAN repo."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
