{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "06e5ee54-4659-4353-b6ad-de92dd9ab319",
   "metadata": {},
   "source": [
    "# Data download and processing from Swissprot\n",
    "\n",
    "Downloaded the data on 13th of May 2024. 571,282 results filtering for reviewed \"Swiss-Prot\". (https://www.uniprot.org/uniprotkb?query=*&facets=reviewed%3Atrue)\n",
    "\n",
    "Selected download TSV and the columns, Seqeunce (under Sequences tab), EC number (under Function).\n",
    "\n",
    "Unzipped the downloaded file.\n",
    "\n",
    "## 1. Protein to EC processing\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4f961afc-aedd-41b9-8635-4c6eabd017ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>269</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>796</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>3423</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>248</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>471</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571277</th>\n",
       "      <td>V9HVX0</td>\n",
       "      <td>YPAA_ECOLI</td>\n",
       "      <td>61</td>\n",
       "      <td>MTIAERLRQEGHQIGWQEGKLEGLHEQAIKIALRMLEQGFDRDQVL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571278</th>\n",
       "      <td>W5X2N3</td>\n",
       "      <td>AP_FRAAN</td>\n",
       "      <td>193</td>\n",
       "      <td>MGWVWKDDDEQGGHVNPSAADISPRLDGDRCSTRKVVRTQCKTEEV...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571279</th>\n",
       "      <td>W6JGV7</td>\n",
       "      <td>REP_CPBDV</td>\n",
       "      <td>484</td>\n",
       "      <td>MSAPLSIEQDDLLTDDLKSWLSDIDFSNDNEEAIEMEPSDIEMSSP...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571280</th>\n",
       "      <td>W6JHZ8</td>\n",
       "      <td>VP1_CPBDV</td>\n",
       "      <td>210</td>\n",
       "      <td>MNVKGASDKAQLAMQAEWEEVLAPEGALAVEEASSVLKISDDERRS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571281</th>\n",
       "      <td>W6JQK2</td>\n",
       "      <td>YBW7_SCHPO</td>\n",
       "      <td>268</td>\n",
       "      <td>MIVIEPTILYEAAALYTRYTANGGDSNDVVALLVGPIVQTLSIPSS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>571282 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9     269   \n",
       "1       A0A023I7E1   ENG1_RHIMI     796   \n",
       "2       A0A024B7W1   POLG_ZIKVF    3423   \n",
       "3       A0A024SC78  CUTI1_HYPJR     248   \n",
       "4       A0A024SH76   GUX2_HYPJR     471   \n",
       "...            ...          ...     ...   \n",
       "571277      V9HVX0   YPAA_ECOLI      61   \n",
       "571278      W5X2N3     AP_FRAAN     193   \n",
       "571279      W6JGV7    REP_CPBDV     484   \n",
       "571280      W6JHZ8    VP1_CPBDV     210   \n",
       "571281      W6JQK2   YBW7_SCHPO     268   \n",
       "\n",
       "                                                 Sequence  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...   \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   \n",
       "...                                                   ...   \n",
       "571277  MTIAERLRQEGHQIGWQEGKLEGLHEQAIKIALRMLEQGFDRDQVL...   \n",
       "571278  MGWVWKDDDEQGGHVNPSAADISPRLDGDRCSTRKVVRTQCKTEEV...   \n",
       "571279  MSAPLSIEQDDLLTDDLKSWLSDIDFSNDNEEAIEMEPSDIEMSSP...   \n",
       "571280  MNVKGASDKAQLAMQAEWEEVLAPEGALAVEEASSVLKISDDERRS...   \n",
       "571281  MIVIEPTILYEAAALYTRYTANGGDSNDVVALLVGPIVQTLSIPSS...   \n",
       "\n",
       "                                                EC number  \n",
       "0                                        3.2.2.-; 3.2.2.6  \n",
       "1                                                3.2.1.39  \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...  \n",
       "3                                                3.1.1.74  \n",
       "4                                                3.2.1.91  \n",
       "...                                                   ...  \n",
       "571277                                                NaN  \n",
       "571278                                                NaN  \n",
       "571279                                                NaN  \n",
       "571280                                                NaN  \n",
       "571281                                                NaN  \n",
       "\n",
       "[571282 rows x 5 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "swissprot = pd.read_csv('raw_data/uniprotkb_AND_reviewed_true_2024_05_13.tsv', sep='\\t')\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58497313",
   "metadata": {},
   "source": [
    "Drop sequences that are fragments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ff073063",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>269</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>796</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024B7W1</td>\n",
       "      <td>POLG_ZIKVF</td>\n",
       "      <td>3423</td>\n",
       "      <td>MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...</td>\n",
       "      <td>2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>248</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>471</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571277</th>\n",
       "      <td>V9HVX0</td>\n",
       "      <td>YPAA_ECOLI</td>\n",
       "      <td>61</td>\n",
       "      <td>MTIAERLRQEGHQIGWQEGKLEGLHEQAIKIALRMLEQGFDRDQVL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571278</th>\n",
       "      <td>W5X2N3</td>\n",
       "      <td>AP_FRAAN</td>\n",
       "      <td>193</td>\n",
       "      <td>MGWVWKDDDEQGGHVNPSAADISPRLDGDRCSTRKVVRTQCKTEEV...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571279</th>\n",
       "      <td>W6JGV7</td>\n",
       "      <td>REP_CPBDV</td>\n",
       "      <td>484</td>\n",
       "      <td>MSAPLSIEQDDLLTDDLKSWLSDIDFSNDNEEAIEMEPSDIEMSSP...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571280</th>\n",
       "      <td>W6JHZ8</td>\n",
       "      <td>VP1_CPBDV</td>\n",
       "      <td>210</td>\n",
       "      <td>MNVKGASDKAQLAMQAEWEEVLAPEGALAVEEASSVLKISDDERRS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571281</th>\n",
       "      <td>W6JQK2</td>\n",
       "      <td>YBW7_SCHPO</td>\n",
       "      <td>268</td>\n",
       "      <td>MIVIEPTILYEAAALYTRYTANGGDSNDVVALLVGPIVQTLSIPSS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>561990 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9     269   \n",
       "1       A0A023I7E1   ENG1_RHIMI     796   \n",
       "2       A0A024B7W1   POLG_ZIKVF    3423   \n",
       "3       A0A024SC78  CUTI1_HYPJR     248   \n",
       "4       A0A024SH76   GUX2_HYPJR     471   \n",
       "...            ...          ...     ...   \n",
       "571277      V9HVX0   YPAA_ECOLI      61   \n",
       "571278      W5X2N3     AP_FRAAN     193   \n",
       "571279      W6JGV7    REP_CPBDV     484   \n",
       "571280      W6JHZ8    VP1_CPBDV     210   \n",
       "571281      W6JQK2   YBW7_SCHPO     268   \n",
       "\n",
       "                                                 Sequence  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...   \n",
       "2       MKNPKKKSGGFRIVNMLKRGVARVSPFGGLKRLPAGLLLGHGPIRM...   \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   \n",
       "...                                                   ...   \n",
       "571277  MTIAERLRQEGHQIGWQEGKLEGLHEQAIKIALRMLEQGFDRDQVL...   \n",
       "571278  MGWVWKDDDEQGGHVNPSAADISPRLDGDRCSTRKVVRTQCKTEEV...   \n",
       "571279  MSAPLSIEQDDLLTDDLKSWLSDIDFSNDNEEAIEMEPSDIEMSSP...   \n",
       "571280  MNVKGASDKAQLAMQAEWEEVLAPEGALAVEEASSVLKISDDERRS...   \n",
       "571281  MIVIEPTILYEAAALYTRYTANGGDSNDVVALLVGPIVQTLSIPSS...   \n",
       "\n",
       "                                                EC number  \n",
       "0                                        3.2.2.-; 3.2.2.6  \n",
       "1                                                3.2.1.39  \n",
       "2       2.1.1.56; 2.1.1.57; 2.7.7.48; 3.4.21.91; 3.6.1...  \n",
       "3                                                3.1.1.74  \n",
       "4                                                3.2.1.91  \n",
       "...                                                   ...  \n",
       "571277                                                NaN  \n",
       "571278                                                NaN  \n",
       "571279                                                NaN  \n",
       "571280                                                NaN  \n",
       "571281                                                NaN  \n",
       "\n",
       "[561990 rows x 5 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fragments = pd.read_csv('raw_data/fragments.tsv', sep='\\t')['Entry'].values\n",
    "swissprot = swissprot[~swissprot['Entry'].isin(fragments)]\n",
    "swissprot\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee69eeba-0d87-4f0e-92f8-8d950cc616f3",
   "metadata": {},
   "source": [
    "## Explode out EC number so each EC has it's own row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2c501ce1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "5756\n",
      "230886\n"
     ]
    }
   ],
   "source": [
    "swissprot = swissprot[swissprot['EC number'].notna()]\n",
    "# Drop rows that don't have an ec number\n",
    "swissprot['EC All'] = swissprot['EC number'].values\n",
    "# Now expand out the ones we have left\n",
    "swissprot['EC number'] = [ec.split(';') for ec in swissprot['EC number']]\n",
    "# Then check how many have mulitple\n",
    "print(sum([1 if len(ec) > 1 else 0 for ec in swissprot['EC number'].values]))\n",
    "swissprot = swissprot.explode('EC number')\n",
    "# Clean the EC numbers\n",
    "swissprot['EC number'] = [ec.replace(' ', '') for ec in swissprot['EC number']]\n",
    "swissprot\n",
    "print(swissprot['EC number'].nunique())\n",
    "print(swissprot['Sequence'].nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4b81d7a",
   "metadata": {},
   "source": [
    "## Filtering steps \n",
    "\n",
    "* Filter to a certain sequence length\n",
    "* Drop duplicates\n",
    "* Remove EC numbers with dash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e42c854f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Partially filled EC numbers:  237556\n",
      "5399\n",
      "188331\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>EC All</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>269</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>3.2.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>796</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "      <td>3.2.1.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>248</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>3.1.1.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>471</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>3.2.1.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A044RE18</td>\n",
       "      <td>BLI_ONCVO</td>\n",
       "      <td>693</td>\n",
       "      <td>MYWQLVRILVLFDCLQKILAIEHDSICIADVDDACPEPSHTVMRLR...</td>\n",
       "      <td>3.4.21.75</td>\n",
       "      <td>3.4.21.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555831</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>240</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>4.1.1.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564731</th>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>Y3032_BACAN</td>\n",
       "      <td>584</td>\n",
       "      <td>MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>3.5.4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564754</th>\n",
       "      <td>Q6L032</td>\n",
       "      <td>Y1085_PICTO</td>\n",
       "      <td>573</td>\n",
       "      <td>MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>3.5.4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568711</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>172</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>3.6.1.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569733</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>124</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>3.1.1.29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200654 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9     269   \n",
       "1       A0A023I7E1   ENG1_RHIMI     796   \n",
       "3       A0A024SC78  CUTI1_HYPJR     248   \n",
       "4       A0A024SH76   GUX2_HYPJR     471   \n",
       "6       A0A044RE18    BLI_ONCVO     693   \n",
       "...            ...          ...     ...   \n",
       "555831      Q05115   AMDA_BORBO     240   \n",
       "564731      Q6HX62  Y3032_BACAN     584   \n",
       "564754      Q6L032  Y1085_PICTO     573   \n",
       "568711      Q94MV8   VG56_BPLZ5     172   \n",
       "569733      Q9J5H2   V029_FOWPN     124   \n",
       "\n",
       "                                                 Sequence  EC number  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.6   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...   3.2.1.39   \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   3.1.1.74   \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   3.2.1.91   \n",
       "6       MYWQLVRILVLFDCLQKILAIEHDSICIADVDDACPEPSHTVMRLR...  3.4.21.75   \n",
       "...                                                   ...        ...   \n",
       "555831  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...   4.1.1.76   \n",
       "564731  MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...    3.5.4.2   \n",
       "564754  MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...    3.5.4.2   \n",
       "568711  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...   3.6.1.12   \n",
       "569733  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...   3.1.1.29   \n",
       "\n",
       "           EC All  \n",
       "0         3.2.2.6  \n",
       "1        3.2.1.39  \n",
       "3        3.1.1.74  \n",
       "4        3.2.1.91  \n",
       "6       3.4.21.75  \n",
       "...           ...  \n",
       "555831   4.1.1.76  \n",
       "564731    3.5.4.2  \n",
       "564754    3.5.4.2  \n",
       "568711   3.6.1.12  \n",
       "569733   3.1.1.29  \n",
       "\n",
       "[200654 rows x 6 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "swissprot = swissprot[swissprot['Length'] > 100]\n",
    "swissprot = swissprot[swissprot['Length'] < 1024] #used to be 1024\n",
    "swissprot = swissprot[~swissprot['EC number'].str.contains('-')]\n",
    "print('Partially filled EC numbers: ', len(swissprot))\n",
    "swissprot = swissprot.drop_duplicates(subset=['Sequence', 'EC number'])\n",
    "print(swissprot['EC number'].nunique())\n",
    "print(swissprot['Sequence'].nunique())\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8d17ad2",
   "metadata": {},
   "source": [
    "## 2. Reaction to EC processing\n",
    "\n",
    "EC and reaction data were downloaded from multiple sources, the first being ECReact.\n",
    "\n",
    "1. ECReact: https://www.nature.com/articles/s41467-022-28536-w, with the data downloaded from https://github.com/rxn4chemistry/biocatalysis-model specifically, https://github.com/rxn4chemistry/biocatalysis-model/blob/main/data/ecreact-1.0.csv\n",
    "2. EnzymeMap: https://chemrxiv.org/engage/chemrxiv/article-details/643e9f4034f4ff2ea4f4f84e https://zenodo.org/records/8254726 and the data https://github.com/hesther/enzymemap, using the data file https://github.com/hesther/enzymemap/blob/main/data/processed_reactions.csv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b6d71562",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>1.2.1.8</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C=C(C)CCOP(=O)([O-])OP(=O)([O-])[O-].CC(C)=CCO...</td>\n",
       "      <td>2.5.1.85</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O...</td>\n",
       "      <td>1.4.1.4</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>1.1.1.3</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cc1cn([C@H]2C[C@H](O)[C@@H](COP(=O)(O)OP(=O)(O...</td>\n",
       "      <td>2.7.7.9</td>\n",
       "      <td>brenda_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62217</th>\n",
       "      <td>CCCCC/C=C\\C/C=C\\C/C=C\\C/C=C\\C/C=C\\CCC(=O)O[C@@...</td>\n",
       "      <td>2.3.1.76</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62218</th>\n",
       "      <td>CCCCCCCCCCCC/C=C\\CCCCCCCC(=O)OC[C@H](COP(=O)(O...</td>\n",
       "      <td>2.7.8.41</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62219</th>\n",
       "      <td>CCCCCCCCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCCCCCCC...</td>\n",
       "      <td>2.7.8.41</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62220</th>\n",
       "      <td>CCCCCC/C=C\\CCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCC...</td>\n",
       "      <td>2.7.8.41</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62221</th>\n",
       "      <td>CCCCCCCC/C=C\\CCCCCCCC(=O)O[C@H](COC(=O)CCCCCCC...</td>\n",
       "      <td>4.1.1.65</td>\n",
       "      <td>pathbank_reaction_smiles</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62222 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Reaction EC number  \\\n",
       "0      NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   1.2.1.8   \n",
       "1      C=C(C)CCOP(=O)([O-])OP(=O)([O-])[O-].CC(C)=CCO...  2.5.1.85   \n",
       "2      N.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O...   1.4.1.4   \n",
       "3      NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...   1.1.1.3   \n",
       "4      Cc1cn([C@H]2C[C@H](O)[C@@H](COP(=O)(O)OP(=O)(O...   2.7.7.9   \n",
       "...                                                  ...       ...   \n",
       "62217  CCCCC/C=C\\C/C=C\\C/C=C\\C/C=C\\C/C=C\\CCC(=O)O[C@@...  2.3.1.76   \n",
       "62218  CCCCCCCCCCCC/C=C\\CCCCCCCC(=O)OC[C@H](COP(=O)(O...  2.7.8.41   \n",
       "62219  CCCCCCCCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCCCCCCC...  2.7.8.41   \n",
       "62220  CCCCCC/C=C\\CCCCCCCCCCCC(=O)O[C@H](COC(=O)CCCCC...  2.7.8.41   \n",
       "62221  CCCCCCCC/C=C\\CCCCCCCC(=O)O[C@H](COC(=O)CCCCCCC...  4.1.1.65   \n",
       "\n",
       "                         Source  \n",
       "0        brenda_reaction_smiles  \n",
       "1        brenda_reaction_smiles  \n",
       "2        brenda_reaction_smiles  \n",
       "3        brenda_reaction_smiles  \n",
       "4        brenda_reaction_smiles  \n",
       "...                         ...  \n",
       "62217  pathbank_reaction_smiles  \n",
       "62218  pathbank_reaction_smiles  \n",
       "62219  pathbank_reaction_smiles  \n",
       "62220  pathbank_reaction_smiles  \n",
       "62221  pathbank_reaction_smiles  \n",
       "\n",
       "[62222 rows x 3 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ECreact = pd.read_csv('raw_data/ECReact.csv')\n",
    "ECreact['rxn_smiles'] = ECreact['rxn_smiles'].str.split('|').str[0] + '>>' + ECreact['rxn_smiles'].str.split('>>').str[1]\n",
    "ECreact.rename(columns={'rxn_smiles': 'Reaction', 'ec': 'EC number', 'source':'Source'}, inplace=True)\n",
    "ECreact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "21ac8cce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "349458\n",
      "4552\n",
      "47640\n",
      "47800\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...</td>\n",
       "      <td>[CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349453</th>\n",
       "      <td>CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + dimethylallyl diphosphate = dimethylally...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349454</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl diphosphate = geranyl-triphospho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349455</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl triphosphate = geranyl-tetraphos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349456</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl diphosphate = adenosine 5'-i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349457</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl triphosphate = isopentenyl-t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62896 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Reaction  \\\n",
       "0       CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1       CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...   \n",
       "25      CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...   \n",
       "26      CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...   \n",
       "27      NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                   ...   \n",
       "349453  CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349454  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...   \n",
       "349455  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...   \n",
       "349456  C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349457  C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...   \n",
       "\n",
       "                                          Mapped Reaction EC number  \\\n",
       "0       [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1       [CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...   1.1.1.1   \n",
       "25      [CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...   1.1.1.1   \n",
       "26      [CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...   1.1.1.1   \n",
       "27      [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                   ...       ...   \n",
       "349453  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...   6.5.1.3   \n",
       "349454  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349455  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349456  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "349457  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "\n",
       "                                            Reaction Text  \n",
       "0          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "25          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "26          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "27      2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...  \n",
       "...                                                   ...  \n",
       "349453  ATP + dimethylallyl diphosphate = dimethylally...  \n",
       "349454  ATP + geranyl diphosphate = geranyl-triphospho...  \n",
       "349455  ATP + geranyl triphosphate = geranyl-tetraphos...  \n",
       "349456  ATP + isopentenyl diphosphate = adenosine 5'-i...  \n",
       "349457  ATP + isopentenyl triphosphate = isopentenyl-t...  \n",
       "\n",
       "[62896 rows x 4 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enzymemap = pd.read_csv('raw_data/EnzymeMap.csv')\n",
    "print(len(enzymemap))\n",
    "enzymemap.rename(columns={'ec_num': 'EC number', 'unmapped': 'Reaction', 'mapped': 'Mapped Reaction', 'orig_rxn_text':'Reaction Text'}, inplace=True)\n",
    "enzymemap = enzymemap[['Reaction', 'Mapped Reaction', 'EC number', 'Reaction Text']]\n",
    "enzymemap.drop_duplicates(subset=['Reaction', 'EC number'], inplace=True)\n",
    "print(enzymemap['EC number'].nunique())\n",
    "print(enzymemap['Reaction'].nunique())\n",
    "print(enzymemap['Mapped Reaction'].nunique())\n",
    "enzymemap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a24320d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...</td>\n",
       "      <td>[CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...</td>\n",
       "      <td>[CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>methylglyoxal + NADH + H+ = acetol + NAD+ {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349453</th>\n",
       "      <td>CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + dimethylallyl diphosphate = dimethylally...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349454</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl diphosphate = geranyl-triphospho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349455</th>\n",
       "      <td>CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + geranyl triphosphate = geranyl-tetraphos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349456</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl diphosphate = adenosine 5'-i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349457</th>\n",
       "      <td>C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...</td>\n",
       "      <td>[CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...</td>\n",
       "      <td>6.5.1.3</td>\n",
       "      <td>ATP + isopentenyl triphosphate = isopentenyl-t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62896 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Reaction  \\\n",
       "0       CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1       CCO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP...   \n",
       "25      CC(=O)C=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)...   \n",
       "26      CC(=O)CO.NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)...   \n",
       "27      NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                   ...   \n",
       "349453  CC(C)=CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349454  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)O.Nc1ncnc2...   \n",
       "349455  CC(C)=CCC/C(C)=C/COP(=O)(O)OP(=O)(O)OP(=O)(O)O...   \n",
       "349456  C=C(C)CCOP(=O)(O)OP(=O)(O)O.Nc1ncnc2c1ncn2[C@@...   \n",
       "349457  C=C(C)CCOP(=O)(O)OP(=O)(O)OP(=O)(O)O.Nc1ncnc2c...   \n",
       "\n",
       "                                          Mapped Reaction EC number  \\\n",
       "0       [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1       [CH3:1][CH2:2][OH:3].[NH2:4][C:5](=[O:6])[c:7]...   1.1.1.1   \n",
       "25      [CH3:1][C:2](=[O:3])[CH:4]=[O:5].[H+].[NH2:6][...   1.1.1.1   \n",
       "26      [CH3:1][C:2](=[O:3])[CH2:4][OH:5].[NH2:6][C:7]...   1.1.1.1   \n",
       "27      [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                   ...       ...   \n",
       "349453  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][O:6][P:7](...   6.5.1.3   \n",
       "349454  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349455  [CH3:1][C:2]([CH3:3])=[CH:4][CH2:5][CH2:6]/[C:...   6.5.1.3   \n",
       "349456  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "349457  [CH2:1]=[C:2]([CH3:3])[CH2:4][CH2:5][O:6][P:7]...   6.5.1.3   \n",
       "\n",
       "                                            Reaction Text  \n",
       "0          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1          acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "25          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "26          methylglyoxal + NADH + H+ = acetol + NAD+ {r}  \n",
       "27      2-dehydro-3-deoxy-D-gluconate + NADH + H+ = 4-...  \n",
       "...                                                   ...  \n",
       "349453  ATP + dimethylallyl diphosphate = dimethylally...  \n",
       "349454  ATP + geranyl diphosphate = geranyl-triphospho...  \n",
       "349455  ATP + geranyl triphosphate = geranyl-tetraphos...  \n",
       "349456  ATP + isopentenyl diphosphate = adenosine 5'-i...  \n",
       "349457  ATP + isopentenyl triphosphate = isopentenyl-t...  \n",
       "\n",
       "[62896 rows x 4 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enzymemap.dropna(subset=['Mapped Reaction']) #just double check that all of them have mapped reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b244d31b",
   "metadata": {},
   "source": [
    "### Append EC numbers that are covered by ECreact but not by EnzymeMap, to EnzymeMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "94ce2b25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2353"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#find the EC numbers covered by ECreact but not by enzymemap\n",
    "ECreact = ECreact[~ECreact['EC number'].str.contains('-')]\n",
    "not_covered = ECreact[~ECreact['EC number'].isin(enzymemap['EC number'].unique())]\n",
    "not_covered['EC number'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "89e5d4dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6905\n",
      "56144\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1172</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1173</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1176</th>\n",
       "      <td>NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15228</th>\n",
       "      <td>Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.j</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21396</th>\n",
       "      <td>C[C@@]1(O)OC[C@H](O)C1(O)O.Nc1ncnc2c1ncn2[C@@H...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.l</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34012</th>\n",
       "      <td>*CS(=O)(=O)[O-].Nc1ncnc2c1ncn2[C@@H]1O[C@H](CO...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.o</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15541</th>\n",
       "      <td>Cc1ncc(C[n+]2csc(CCO)c2C)c(N)n1.Nc1ncnc2c1ncn2...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.p</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33659</th>\n",
       "      <td>Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.q</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72490 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Reaction  \\\n",
       "0      CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1172   CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...   \n",
       "1173   CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...   \n",
       "1176   NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...   \n",
       "1177   NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                  ...   \n",
       "15228  Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...   \n",
       "21396  C[C@@]1(O)OC[C@H](O)C1(O)O.Nc1ncnc2c1ncn2[C@@H...   \n",
       "34012  *CS(=O)(=O)[O-].Nc1ncnc2c1ncn2[C@@H]1O[C@H](CO...   \n",
       "15541  Cc1ncc(C[n+]2csc(CCO)c2C)c(N)n1.Nc1ncnc2c1ncn2...   \n",
       "33659  Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)([O-])OP(=O...   \n",
       "\n",
       "                                         Mapped Reaction EC number  \\\n",
       "0      [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1172   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1173   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1176   [H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...   1.1.1.1   \n",
       "1177   [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                  ...       ...   \n",
       "15228                                                NaN   7.6.2.j   \n",
       "21396                                                NaN   7.6.2.l   \n",
       "34012                                                NaN   7.6.2.o   \n",
       "15541                                                NaN   7.6.2.p   \n",
       "33659                                                NaN   7.6.2.q   \n",
       "\n",
       "                                           Reaction Text  \n",
       "0         acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1172   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1173   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1176   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "1177   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "...                                                  ...  \n",
       "15228                                                NaN  \n",
       "21396                                                NaN  \n",
       "34012                                                NaN  \n",
       "15541                                                NaN  \n",
       "33659                                                NaN  \n",
       "\n",
       "[72490 rows x 4 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reaction2EC = pd.concat((enzymemap, not_covered[['Reaction', 'EC number']])).sort_values(by='EC number')\n",
    "print(reaction2EC['EC number'].nunique())\n",
    "print(reaction2EC['Reaction'].nunique())\n",
    "reaction2EC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac489eea",
   "metadata": {},
   "source": [
    "## 3. Filter protein2EC and reaction2EC to intersection of their EC numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "83eb77eb-dcb7-423d-ad4d-b0246c207fd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>EC All</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A0A061I403</td>\n",
       "      <td>FICD_CRIGR</td>\n",
       "      <td>455</td>\n",
       "      <td>MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...</td>\n",
       "      <td>2.7.7.108</td>\n",
       "      <td>2.7.7.108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A0A067XR63</td>\n",
       "      <td>XTH7_DIOKA</td>\n",
       "      <td>268</td>\n",
       "      <td>MNAEGGNLHREFEITWGDGRARIHNNGGLLTLSLDRASGSGFRSKN...</td>\n",
       "      <td>2.4.1.207</td>\n",
       "      <td>2.4.1.207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A0A067XRK9</td>\n",
       "      <td>XTH6_DIOKA</td>\n",
       "      <td>299</td>\n",
       "      <td>MASSLTLPMAMAFTLLALSFASAMGGSMNSSRFDELFQPSWAFDHF...</td>\n",
       "      <td>2.4.1.207</td>\n",
       "      <td>2.4.1.207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A0A067YMX8</td>\n",
       "      <td>XTH8_DIOKA</td>\n",
       "      <td>288</td>\n",
       "      <td>MAASPYSIFAVQLLLLASWMLSSSSSNFNQDFNIAWGGGRARILNN...</td>\n",
       "      <td>2.4.1.207</td>\n",
       "      <td>2.4.1.207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>A0A0C5Q4Y6</td>\n",
       "      <td>C76H2_ROSOF</td>\n",
       "      <td>493</td>\n",
       "      <td>MDSFPLLAALFFILAATWFISFRRPRNLPPGPFPYPIVGNMLQLGT...</td>\n",
       "      <td>1.14.14.175</td>\n",
       "      <td>1.14.14.175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534459</th>\n",
       "      <td>Q9ZNA5</td>\n",
       "      <td>SAHH_ROSDO</td>\n",
       "      <td>462</td>\n",
       "      <td>MTKDFIVKDIALAEFGRKELDIAETEMPGLMALRAEYGDSKPLAGA...</td>\n",
       "      <td>3.13.2.1</td>\n",
       "      <td>3.13.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534465</th>\n",
       "      <td>Q9ZNJ9</td>\n",
       "      <td>MSRB_HATHI</td>\n",
       "      <td>159</td>\n",
       "      <td>MNHKKEKEYKKLDTEKLKENLTELQYNVTQRNATEKPFLNKYDKHF...</td>\n",
       "      <td>1.8.4.12</td>\n",
       "      <td>1.8.4.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534723</th>\n",
       "      <td>V5TD18</td>\n",
       "      <td>AMBI3_FISAU</td>\n",
       "      <td>273</td>\n",
       "      <td>MIVSTSVEQSAQFSVKSLTPFGALLEATEDHSDIQQLSIEQLCQLT...</td>\n",
       "      <td>1.14.20.11</td>\n",
       "      <td>1.14.20.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534724</th>\n",
       "      <td>V5TES5</td>\n",
       "      <td>AMBI2_FISAU</td>\n",
       "      <td>330</td>\n",
       "      <td>MTQIINITQSKVISEQILRHVFRHRRLISDTEPCVHQPCSLCLAPH...</td>\n",
       "      <td>4.1.99.25</td>\n",
       "      <td>4.1.99.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534725</th>\n",
       "      <td>V5TF65</td>\n",
       "      <td>AMBI1_FISAU</td>\n",
       "      <td>319</td>\n",
       "      <td>MISEKILRHIFQYRRLLSDTEPCAKEPCSICLAPHLPKIQSFIENN...</td>\n",
       "      <td>4.1.99.25</td>\n",
       "      <td>4.1.99.25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14659 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "11      A0A061I403   FICD_CRIGR     455   \n",
       "15      A0A067XR63   XTH7_DIOKA     268   \n",
       "16      A0A067XRK9   XTH6_DIOKA     299   \n",
       "17      A0A067YMX8   XTH8_DIOKA     288   \n",
       "93      A0A0C5Q4Y6  C76H2_ROSOF     493   \n",
       "...            ...          ...     ...   \n",
       "534459      Q9ZNA5   SAHH_ROSDO     462   \n",
       "534465      Q9ZNJ9   MSRB_HATHI     159   \n",
       "534723      V5TD18  AMBI3_FISAU     273   \n",
       "534724      V5TES5  AMBI2_FISAU     330   \n",
       "534725      V5TF65  AMBI1_FISAU     319   \n",
       "\n",
       "                                                 Sequence    EC number  \\\n",
       "11      MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...    2.7.7.108   \n",
       "15      MNAEGGNLHREFEITWGDGRARIHNNGGLLTLSLDRASGSGFRSKN...    2.4.1.207   \n",
       "16      MASSLTLPMAMAFTLLALSFASAMGGSMNSSRFDELFQPSWAFDHF...    2.4.1.207   \n",
       "17      MAASPYSIFAVQLLLLASWMLSSSSSNFNQDFNIAWGGGRARILNN...    2.4.1.207   \n",
       "93      MDSFPLLAALFFILAATWFISFRRPRNLPPGPFPYPIVGNMLQLGT...  1.14.14.175   \n",
       "...                                                   ...          ...   \n",
       "534459  MTKDFIVKDIALAEFGRKELDIAETEMPGLMALRAEYGDSKPLAGA...     3.13.2.1   \n",
       "534465  MNHKKEKEYKKLDTEKLKENLTELQYNVTQRNATEKPFLNKYDKHF...     1.8.4.12   \n",
       "534723  MIVSTSVEQSAQFSVKSLTPFGALLEATEDHSDIQQLSIEQLCQLT...   1.14.20.11   \n",
       "534724  MTQIINITQSKVISEQILRHVFRHRRLISDTEPCVHQPCSLCLAPH...    4.1.99.25   \n",
       "534725  MISEKILRHIFQYRRLLSDTEPCAKEPCSICLAPHLPKIQSFIENN...    4.1.99.25   \n",
       "\n",
       "             EC All  \n",
       "11        2.7.7.108  \n",
       "15        2.4.1.207  \n",
       "16        2.4.1.207  \n",
       "17        2.4.1.207  \n",
       "93      1.14.14.175  \n",
       "...             ...  \n",
       "534459     3.13.2.1  \n",
       "534465     1.8.4.12  \n",
       "534723   1.14.20.11  \n",
       "534724    4.1.99.25  \n",
       "534725    4.1.99.25  \n",
       "\n",
       "[14659 rows x 6 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ns_swissprot = swissprot[~swissprot['EC number'].isin(reaction2EC['EC number'].unique())]\n",
    "ns_swissprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8f2dc451-4b77-4da7-99dd-3d69bbbd964d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "439"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ns_swissprot['EC number'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "624fb47e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "174744\n",
      "4960\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>EC All</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>269</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>3.2.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>796</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "      <td>3.2.1.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>248</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>3.1.1.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>471</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>3.2.1.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A044RE18</td>\n",
       "      <td>BLI_ONCVO</td>\n",
       "      <td>693</td>\n",
       "      <td>MYWQLVRILVLFDCLQKILAIEHDSICIADVDDACPEPSHTVMRLR...</td>\n",
       "      <td>3.4.21.75</td>\n",
       "      <td>3.4.21.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555831</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>240</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>4.1.1.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564731</th>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>Y3032_BACAN</td>\n",
       "      <td>584</td>\n",
       "      <td>MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>3.5.4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564754</th>\n",
       "      <td>Q6L032</td>\n",
       "      <td>Y1085_PICTO</td>\n",
       "      <td>573</td>\n",
       "      <td>MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>3.5.4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568711</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>172</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>3.6.1.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569733</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>124</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>3.1.1.29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>185995 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  Length  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9     269   \n",
       "1       A0A023I7E1   ENG1_RHIMI     796   \n",
       "3       A0A024SC78  CUTI1_HYPJR     248   \n",
       "4       A0A024SH76   GUX2_HYPJR     471   \n",
       "6       A0A044RE18    BLI_ONCVO     693   \n",
       "...            ...          ...     ...   \n",
       "555831      Q05115   AMDA_BORBO     240   \n",
       "564731      Q6HX62  Y3032_BACAN     584   \n",
       "564754      Q6L032  Y1085_PICTO     573   \n",
       "568711      Q94MV8   VG56_BPLZ5     172   \n",
       "569733      Q9J5H2   V029_FOWPN     124   \n",
       "\n",
       "                                                 Sequence  EC number  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.6   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...   3.2.1.39   \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   3.1.1.74   \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   3.2.1.91   \n",
       "6       MYWQLVRILVLFDCLQKILAIEHDSICIADVDDACPEPSHTVMRLR...  3.4.21.75   \n",
       "...                                                   ...        ...   \n",
       "555831  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...   4.1.1.76   \n",
       "564731  MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...    3.5.4.2   \n",
       "564754  MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...    3.5.4.2   \n",
       "568711  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...   3.6.1.12   \n",
       "569733  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...   3.1.1.29   \n",
       "\n",
       "           EC All  \n",
       "0         3.2.2.6  \n",
       "1        3.2.1.39  \n",
       "3        3.1.1.74  \n",
       "4        3.2.1.91  \n",
       "6       3.4.21.75  \n",
       "...           ...  \n",
       "555831   4.1.1.76  \n",
       "564731    3.5.4.2  \n",
       "564754    3.5.4.2  \n",
       "568711   3.6.1.12  \n",
       "569733   3.1.1.29  \n",
       "\n",
       "[185995 rows x 6 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "swissprot = swissprot[swissprot['EC number'].isin(reaction2EC['EC number'].unique())]\n",
    "print(swissprot['Sequence'].nunique())\n",
    "print(swissprot['EC number'].nunique())\n",
    "swissprot # 149788"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b15e8b1d-d035-49a9-a524-d79453b5609e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4960"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(swissprot['EC number'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "443b6cdb-fe81-40ba-86a6-a52fe0103c29",
   "metadata": {},
   "source": [
    "## Save unique proteins\n",
    "\n",
    "#save the indices corresponding to unique proteins, for downstream use\n",
    "unique_protein_indices = swissprot.reset_index().drop_duplicates(subset=['Sequence']).index\n",
    "with open('unique_protein_indices.txt', 'w') as f:\n",
    "    for index in unique_protein_indices:\n",
    "        f.write(str(index) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "407b09c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4960\n",
      "48392\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Mapped Reaction</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Reaction Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...</td>\n",
       "      <td>[CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>acetaldehyde + NADH + H+ = ethanol + NAD+  {r}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1172</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1173</th>\n",
       "      <td>CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...</td>\n",
       "      <td>[CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1176</th>\n",
       "      <td>NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...</td>\n",
       "      <td>[H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...</td>\n",
       "      <td>[NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...</td>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34226</th>\n",
       "      <td>CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@@H]8[C@H...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34223</th>\n",
       "      <td>CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@H]([C@H]...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15471</th>\n",
       "      <td>C[N+]1(C)CCC[C@H]1C(=O)[O-].Nc1ncnc2c1ncn2[C@@...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)(O)OP(=O)(O...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8665</th>\n",
       "      <td>*[N+](*)(*)*.Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61766 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Reaction  \\\n",
       "0      CC=O.NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O...   \n",
       "1172   CC(C)(C)OC(=O)N1CCC(=O)C1.NC(=O)C1=CN([C@@H]2O...   \n",
       "1173   CC(C)(C)OC(=O)N1CC[C@H](O)C1.NC(=O)c1ccc[n+]([...   \n",
       "1176   NC(=O)C1=CN([C@@H]2O[C@H](COP(=O)(O)OP(=O)(O)O...   \n",
       "1177   NC(=O)c1ccc[n+]([C@@H]2O[C@H](COP(=O)(O)OP(=O)...   \n",
       "...                                                  ...   \n",
       "34226  CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@@H]8[C@H...   \n",
       "34223  CC1=C2[N+]3=C(C=C4[N+]5=C(C(C)=C6N7[C@H]([C@H]...   \n",
       "15471  C[N+]1(C)CCC[C@H]1C(=O)[O-].Nc1ncnc2c1ncn2[C@@...   \n",
       "644    Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)(O)OP(=O)(O...   \n",
       "8665   *[N+](*)(*)*.Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=...   \n",
       "\n",
       "                                         Mapped Reaction EC number  \\\n",
       "0      [CH3:1][CH:2]=[O:3].[H+].[NH2:4][C:5](=[O:6])[...   1.1.1.1   \n",
       "1172   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1173   [CH3:1][C:2]([CH3:3])([CH3:4])[O:5][C:6](=[O:7...   1.1.1.1   \n",
       "1176   [H+].[NH2:1][C:2](=[O:3])[C:4]1=[CH:5][N:6]([C...   1.1.1.1   \n",
       "1177   [NH2:1][C:2](=[O:3])[c:4]1[cH:5][cH:6][cH:7][n...   1.1.1.1   \n",
       "...                                                  ...       ...   \n",
       "34226                                                NaN   7.6.2.8   \n",
       "34223                                                NaN   7.6.2.8   \n",
       "15471                                                NaN   7.6.2.9   \n",
       "644                                                  NaN   7.6.2.9   \n",
       "8665                                                 NaN   7.6.2.9   \n",
       "\n",
       "                                           Reaction Text  \n",
       "0         acetaldehyde + NADH + H+ = ethanol + NAD+  {r}  \n",
       "1172   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1173   N-tert-butoxycarbonyl-3-pyrrolidinone + NADH +...  \n",
       "1176   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "1177   p-nitrobenzyloxycarbonyl-3-pyrrolidinone + NAD...  \n",
       "...                                                  ...  \n",
       "34226                                                NaN  \n",
       "34223                                                NaN  \n",
       "15471                                                NaN  \n",
       "644                                                  NaN  \n",
       "8665                                                 NaN  \n",
       "\n",
       "[61766 rows x 4 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reaction2EC = reaction2EC[reaction2EC['EC number'].isin(swissprot['EC number'].unique())]\n",
    "print(reaction2EC['EC number'].nunique())\n",
    "print(reaction2EC['Reaction'].nunique())\n",
    "reaction2EC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5721bdf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "reaction2EC.to_csv('../processed_data/reaction2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5250f2c8",
   "metadata": {},
   "source": [
    "## 4. Perform protein clustering using mmseqs\n",
    "\n",
    "Use mmseqs to identify clusters, https://github.com/soedinglab/MMseqs2\n",
    "\n",
    "From the documentation:\n",
    "```\n",
    "With --cov-mode 0 only sequences are clustered that have a sequence length overlap greater than X% of the longer of the two sequences. This coverage mode should be used to cluster full length protein sequences. The multi domain structure of proteins will be most likely preserved when using a coverage > 80% (-c 0.8). Default --cluster-mode is the greedy set cover.\n",
    "```\n",
    "\n",
    "\n",
    "Optional parameters:\n",
    "\n",
    "```\n",
    " --seed-sub-mat TWIN              Substitution matrix file for k-mer generation [aa:VTML80.out,nucl:nucleotide.out]\n",
    " -s FLOAT                         Sensitivity: 1.0 faster; 4.0 fast; 7.5 sensitive [4.000]\n",
    " -k INT                           k-mer length (0: automatically set to optimum) [0]\n",
    " --target-search-mode INT         target search mode (0: regular k-mer, 1: similar k-mer) [0]\n",
    " --k-score TWIN                   k-mer threshold for generating similar k-mer lists [seq:2147483647,prof:2147483647]\n",
    " --alph-size TWIN                 Alphabet size (range 2-21) [aa:21,nucl:5]\n",
    " --max-seqs INT                   Maximum results per query sequence allowed to pass the prefilter (affects sensitivity) [20]\n",
    " --split INT                      Split input into N equally distributed chunks. 0: set the best split automatically [0]\n",
    " --split-mode INT                 0: split target db; 1: split query db; 2: auto, depending on main memory [2]\n",
    " --split-memory-limit BYTE        Set max memory per split. E.g. 800B, 5K, 10M, 1G. Default (0) to all available system memory [0]\n",
    " --comp-bias-corr INT             Correct for locally biased amino acid composition (range 0-1) [1]\n",
    " --comp-bias-corr-scale FLOAT     Correct for locally biased amino acid composition (range 0-1) [1.000]\n",
    " --diag-score BOOL                Use ungapped diagonal scoring during prefilter [1]\n",
    " --exact-kmer-matching INT        Extract only exact k-mers for matching (range 0-1) [0]\n",
    " --mask INT                       Mask sequences in k-mer stage: 0: w/o low complexity masking, 1: with low complexity masking [1]\n",
    " --mask-prob FLOAT                Mask sequences is probablity is above threshold [0.900]\n",
    " --mask-lower-case INT            Lowercase letters will be excluded from k-mer search 0: include region, 1: exclude region [0]\n",
    " --min-ungapped-score INT         Accept only matches with ungapped alignment score above threshold [15]\n",
    " --add-self-matches BOOL          Artificially add entries of queries with themselves (for clustering) [0]\n",
    " --spaced-kmer-mode INT           0: use consecutive positions in k-mers; 1: use spaced k-mers [1]\n",
    " --spaced-kmer-pattern STR        User-specified spaced k-mer pattern []\n",
    " --local-tmp STR                  Path where some of the temporary files will be created []\n",
    "align:\n",
    " -c FLOAT                         List matches above this fraction of aligned (covered) residues (see --cov-mode) [0.800]\n",
    " --cov-mode INT                   0: coverage of query and target\n",
    "                                  1: coverage of target\n",
    "                                  2: coverage of query\n",
    "                                  3: target seq. length has to be at least x% of query length\n",
    "                                  4: query seq. length has to be at least x% of target length\n",
    "                                  5: short seq. needs to be at least x% of the other seq. length [0]\n",
    " -a BOOL                          Add backtrace string (convert to alignments with mmseqs convertalis module) [0]\n",
    " --alignment-mode INT             How to compute the alignment:\n",
    "                                  0: automatic\n",
    "                                  1: only score and end_pos\n",
    "                                  2: also start_pos and cov\n",
    "                                  3: also seq.id\n",
    "                                  4: only ungapped alignment [3]\n",
    " --alignment-output-mode INT      How to compute the alignment:\n",
    "                                  0: automatic\n",
    "                                  1: only score and end_pos\n",
    "                                  2: also start_pos and cov\n",
    "                                  3: also seq.id\n",
    "                                  4: only ungapped alignment\n",
    "                                  5: score only (output) cluster format [0]\n",
    " --wrapped-scoring BOOL           Double the (nucleotide) query sequence during the scoring process to allow wrapped diagonal scoring around end and start [0]\n",
    " -e DOUBLE                        List matches below this E-value (range 0.0-inf) [1.000E-03]\n",
    " --min-seq-id FLOAT               List matches above this sequence identity (for clustering) (range 0.0-1.0) [0.000]\n",
    " --min-aln-len INT                Minimum alignment length (range 0-INT_MAX) [0]\n",
    " --seq-id-mode INT                0: alignment length 1: shorter, 2: longer sequence [0]\n",
    " --alt-ali INT                    Show up to this many alternative alignments [0]\n",
    " --max-rejected INT               Maximum rejected alignments before alignment calculation for a query is stopped [2147483647]\n",
    " --max-accept INT                 Maximum accepted alignments before alignment calculation for a query is stopped [2147483647]\n",
    " --score-bias FLOAT               Score bias when computing SW alignment (in bits) [0.000]\n",
    " --realign BOOL                   Compute more conservative, shorter alignments (scores and E-values not changed) [0]\n",
    " --realign-score-bias FLOAT       Additional bias when computing realignment [-0.200]\n",
    " --realign-max-seqs INT           Maximum number of results to return in realignment [2147483647]\n",
    " --corr-score-weight FLOAT        Weight of backtrace correlation score that is added to the alignment score [0.000]\n",
    " --gap-open TWIN                  Gap open cost [aa:11,nucl:5]\n",
    " --gap-extend TWIN                Gap extension cost [aa:1,nucl:2]\n",
    " --zdrop INT                      Maximal allowed difference between score values before alignment is truncated  (nucleotide alignment only) [40]\n",
    "clust:\n",
    " --cluster-mode INT               0: Set-Cover (greedy)\n",
    "                                  1: Connected component (BLASTclust)\n",
    "                                  2,3: Greedy clustering by sequence length (CDHIT) [0]\n",
    " --max-iterations INT             Maximum depth of breadth first search in connected component clustering [1000]\n",
    " --similarity-type INT            Type of score used for clustering. 1: alignment score 2: sequence identity [2]\n",
    " --single-step-clustering BOOL    Switch from cascaded to simple clustering workflow [0]\n",
    " --cluster-steps INT              Cascaded clustering steps from 1 to -s [3]\n",
    " --cluster-reassign BOOL          Cascaded clustering can cluster sequence that do not fulfill the clustering criteria.\n",
    "                                  Cluster reassignment corrects these errors [0]\n",
    "kmermatcher:\n",
    " --weights STR                    Weights used for cluster priorization []\n",
    " --cluster-weight-threshold FLOAT Weight threshold used for cluster priorization [0.900]\n",
    " --kmer-per-seq INT               k-mers per sequence [21]\n",
    " --kmer-per-seq-scale TWIN        Scale k-mer per sequence based on sequence length as kmer-per-seq val + scale x seqlen [aa:0.000,nucl:0.200]\n",
    " --adjust-kmer-len BOOL           Adjust k-mer length based on specificity (only for nucleotides) [0]\n",
    " --hash-shift INT                 Shift k-mer hash initialization [67]\n",
    " --include-only-extendable BOOL   Include only extendable [0]\n",
    " --ignore-multi-kmer BOOL         Skip k-mers occurring multiple times (>=2) [0]\n",
    "profile:\n",
    " --pca                            Pseudo count admixture strength []\n",
    " --pcb                            Pseudo counts: Neff at half of maximum admixture (range 0.0-inf) []\n",
    "misc:\n",
    " --taxon-list STR                 Taxonomy ID, possibly multiple values separated by ',' []\n",
    " --rescore-mode INT               Rescore diagonals with:\n",
    "                                  0: Hamming distance\n",
    "                                  1: local alignment (score only)\n",
    "                                  2: local alignment\n",
    "                                  3: global alignment\n",
    "                                  4: longest alignment fulfilling window quality criterion [0]\n",
    " --dbtype INT                     Database type 0: auto, 1: amino acid 2: nucleotides [0]\n",
    " --shuffle BOOL                   Shuffle input database [1]\n",
    " --createdb-mode INT              Createdb mode 0: copy data, 1: soft link data and write new index (works only with single line fasta/q) [1]\n",
    " --id-offset INT                  Numeric ids in index file are offset by this value [0]\n",
    "common:\n",
    " --sub-mat TWIN                   Substitution matrix file [aa:blosum62.out,nucl:nucleotide.out]\n",
    " --max-seq-len INT                Maximum sequence length [65535]\n",
    " --db-load-mode INT               Database preload mode 0: auto, 1: fread, 2: mmap, 3: mmap+touch [0]\n",
    " --threads INT                    Number of CPU-cores used (all by default) [384]\n",
    " --compressed INT                 Write compressed output [0]\n",
    " -v INT                           Verbosity level: 0: quiet, 1: +errors, 2: +warnings, 3: +info [3]\n",
    " --remove-tmp-files BOOL          Delete temporary files [1]\n",
    " --force-reuse BOOL               Reuse tmp filse in tmp/latest folder ignoring parameters and version changes [0]\n",
    " --mpi-runner STR                 Use MPI on compute cluster with this MPI command (e.g. \"mpirun -np 42\") []\n",
    "expert:\n",
    " --filter-hits BOOL               Filter hits by seq.id. and coverage [0]\n",
    " --sort-results INT               Sort results: 0: no sorting, 1: sort by E-value (Alignment) or seq.id. (Hamming) [0]\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "19acec52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate a fasta file as input to mmseqs\n",
    "with open('clustering/swissprot.fasta', 'w') as f:\n",
    "    for index, row in swissprot.iterrows():\n",
    "        f.write(f'>{row[\"Entry\"]}\\n{row[\"Sequence\"]}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6d844f90-6f1a-40a7-b85b-1e490621c753",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes30 clustering/mmseqs_tmp --min-seq-id 0.3 --cov-mode 0 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.3\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/12942347142416062949/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Shuffle database cannot be combined with --createdb-mode 0\n",
      "We recompute with --shuffle 0\n",
      "Converting sequences\n",
      "[==================\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 47ms\n",
      "Create directory clustering/mmseqs_tmp/12942347142416062949/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu clustering/mmseqs_tmp/12942347142416062949/clu_tmp --max-seqs 20 -c 0.8 --cov-mode 0 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.3 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 5.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 3\n",
      "linclust clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 185995 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 185.99K 0s 80ms\n",
      "Sort kmer 0h 0m 0s 49ms\n",
      "Sort by rep. sequence 0h 0m 0s 37ms\n",
      "Time for fill: 0h 0m 0s 37ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 340ms\n",
      "rescorediagonal clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[===============================================================] 185.99K 0s 125ms\n",
      "==Time for merging to pref_rescore1: 0h 0m 0s 60ms\n",
      "Time for processing: 0h 0m 0s 469ms\n",
      "clust clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref_rescore1 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pre_clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 185.99K 0s 14ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 369984 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 185.99K 0s 10ms\n",
      "Add missing connections\n",
      "[=================================================================] 185.99K 0s 11ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 261ms\n",
      "Total time: 0h 0m 0s 313ms\n",
      "\n",
      "Size of the sequence database: 185995\n",
      "Size of the alignment database: 185995\n",
      "Number of clusters: 64569\n",
      "\n",
      "Writing results 0h 0m 0s 9ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 358ms\n",
      "createsubdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/order_redundancy clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "createsubdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/order_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 23ms\n",
      "filterdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref_filter1 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref_filter2 --filter-file clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/order_redundancy --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 64.57K 0s 55ms\n",
      "Time for merging to pref_filter2: 0h 0m 0s 33ms\n",
      "Time for processing: 0h 0m 0s 340ms\n",
      "rescorediagonal clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/input_step_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/input_step_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref_filter2 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 64.57K 0s 72ms\n",
      "Time for merging to pref_rescore2: 0h 0m 0s 29ms\n",
      "Time for processing: 0h 0m 0s 392ms\n",
      "align clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/input_step_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/input_step_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref_rescore2 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 64569 type: Aminoacid\n",
      "Target database size: 64569 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 64.57K 0s 691ms\n",
      "Time for merging to aln: 0h 0m 0s 36ms\n",
      "123904 alignments calculated\n",
      "119706 sequence pairs passed the thresholds (0.966119 of overall calculated)\n",
      "1.853924 hits per query sequence\n",
      "Time for processing: 0h 0m 1s 476ms\n",
      "clust clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/input_step_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/aln clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 64.57K 0s 12ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 55137 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 64.57K 0s 7ms\n",
      "Add missing connections\n",
      "[=================================================================] 64.57K 0s 3ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 131ms\n",
      "Total time: 0h 0m 0s 168ms\n",
      "\n",
      "Size of the sequence database: 64569\n",
      "Size of the alignment database: 64569\n",
      "Number of clusters: 38011\n",
      "\n",
      "Writing results 0h 0m 0s 6ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 189ms\n",
      "mergeclusters clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pre_clust clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/clust --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 64.57K 0s 33ms\n",
      "Clustering step 2\n",
      "[=================================================================] 38.01K 0s 80ms\n",
      "Write merged clustering\n",
      "[=================================================================] 185.99K 0s 279ms\n",
      "Time for merging to clu_redundancy: 0h 0m 0s 28ms\n",
      "Time for processing: 0h 0m 0s 353ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 10ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/linclust/12818586189046854761/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_redundancy clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 15ms\n",
      "prefilter clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 38011 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 38011 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 38.01K 0s 69ms\n",
      "Index table: Masked residues: 99857\n",
      "Index table: fill\n",
      "[=================================================================] 38.01K 0s 81ms\n",
      "Index statistics\n",
      "Entries:          6731339\n",
      "DB size:          526 MB\n",
      "Avg k-mer size:   0.105177\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t249\n",
      "    GPGGTL\t145\n",
      "    LGGFVY\t135\n",
      "    HRLPLL\t79\n",
      "    RNAGRQ\t79\n",
      "    AYPWGT\t70\n",
      "    GRGHCS\t67\n",
      "    IHDKNV\t67\n",
      "    SPGGGW\t61\n",
      "    SSMQNP\t53\n",
      "Time for index table init: 0h 0m 0s 606ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 38011\n",
      "Target db start 1 to 38011\n",
      "[=================================================================] 38.01K 0s 135ms\n",
      "\n",
      "1.760973 k-mers per position\n",
      "378 DB matches per sequence\n",
      "0 overflows\n",
      "10 sequences passed prefiltering per query sequence\n",
      "9 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 29ms\n",
      "Time for processing: 0h 0m 1s 236ms\n",
      "align clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/pref_step0 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 38011 type: Aminoacid\n",
      "Target database size: 38011 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 38.01K 0s 792ms\n",
      "Time for merging to aln_step0: 0h 0m 0s 30ms\n",
      "338652 alignments calculated\n",
      "288260 sequence pairs passed the thresholds (0.851198 of overall calculated)\n",
      "7.583594 hits per query sequence\n",
      "Time for processing: 0h 0m 1s 23ms\n",
      "clust clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/aln_step0 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 38.01K 0s 5ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 37803 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 38.01K 0s 6ms\n",
      "Add missing connections\n",
      "[=================================================================] 38.01K 0s 6ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 105ms\n",
      "Total time: 0h 0m 0s 149ms\n",
      "\n",
      "Size of the sequence database: 38011\n",
      "Size of the alignment database: 38011\n",
      "Number of clusters: 14480\n",
      "\n",
      "Writing results 0h 0m 0s 4ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 164ms\n",
      "createsubdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_step0 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "prefilter clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step1 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step1 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 3 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 14480 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 14480 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 14.48K 0s 39ms\n",
      "Index table: Masked residues: 57853\n",
      "Index table: fill\n",
      "[=================================================================] 14.48K 0s 39ms\n",
      "Index statistics\n",
      "Entries:          5266618\n",
      "DB size:          518 MB\n",
      "Avg k-mer size:   0.082291\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t146\n",
      "    RDKELL\t78\n",
      "    LGGFVY\t63\n",
      "    IHDKNV\t42\n",
      "    GPGGTL\t36\n",
      "    IVRLEN\t35\n",
      "    DISGLY\t30\n",
      "    DVSGLY\t27\n",
      "    HRISIL\t23\n",
      "    SGTALP\t22\n",
      "Time for index table init: 0h 0m 0s 526ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 14480\n",
      "Target db start 1 to 14480\n",
      "[=================================================================] 14.48K 0s 140ms\n",
      "\n",
      "22.852811 k-mers per position\n",
      "1000 DB matches per sequence\n",
      "0 overflows\n",
      "8 sequences passed prefiltering per query sequence\n",
      "7 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 25ms\n",
      "Time for processing: 0h 0m 1s 239ms\n",
      "align clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step1 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step1 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/pref_step1 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 14480 type: Aminoacid\n",
      "Target database size: 14480 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 14.48K 0s 368ms\n",
      "Time for merging to aln_step1: 0h 0m 0s 25ms\n",
      "67672 alignments calculated\n",
      "29248 sequence pairs passed the thresholds (0.432202 of overall calculated)\n",
      "2.019890 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 576ms\n",
      "clust clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step1 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/aln_step1 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_step1 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "==[===============================================================] 14.48K 0s 11ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 1926 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 14.48K 0s 6ms\n",
      "Add missing connections\n",
      "[=================================================================] 14.48K 0s 0ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 99ms\n",
      "Total time: 0h 0m 0s 139ms\n",
      "\n",
      "Size of the sequence database: 14480\n",
      "Size of the alignment database: 14480\n",
      "Number of clusters: 10731\n",
      "\n",
      "Writing results 0h 0m 0s 3ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 151ms\n",
      "createsubdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_step1 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step1 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "prefilter clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step2 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step2 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 5 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 10731 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 10731 type: Aminoacid\n",
      "Index table k-mer threshold: 118 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 10.73K 0s 42ms\n",
      "Index table: Masked residues: 54096\n",
      "Index table: fill\n",
      "[=================================================================] 10.73K 0s 51ms\n",
      "Index statistics\n",
      "Entries:          4392711\n",
      "DB size:          513 MB\n",
      "Avg k-mer size:   0.068636\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t137\n",
      "    RDKELL\t76\n",
      "    LGGFVY\t59\n",
      "    IHDKNV\t40\n",
      "    IVRLEN\t32\n",
      "    DISGLY\t28\n",
      "    DVSGLW\t26\n",
      "    DVSGLY\t25\n",
      "    GPGGTL\t24\n",
      "    SGTALP\t22\n",
      "Time for index table init: 0h 0m 0s 537ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 118\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 10731\n",
      "Target db start 1 to 10731\n",
      "[=================================================================] 10.73K 0s 272ms\n",
      "\n",
      "180.516467 k-mers per position\n",
      "5198 DB matches per sequence\n",
      "0 overflows\n",
      "18 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 24ms\n",
      "Time for processing: 0h 0m 1s 304ms\n",
      "align clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step2 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step2 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/pref_step2 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 10731 type: Aminoacid\n",
      "Target database size: 10731 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 10.73K 0s 327ms\n",
      "Time for merging to aln_step2: 0h 0m 0s 27ms\n",
      "74756 alignments calculated\n",
      "12257 sequence pairs passed the thresholds (0.163960 of overall calculated)\n",
      "1.142205 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 532ms\n",
      "clust clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step2 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/aln_step2 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_step2 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 10.73K 0s 5ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 228 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 10.73K 0s 5ms\n",
      "Add missing connections\n",
      "[=================================================================] 10.73K 0s 0ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 71ms\n",
      "Total time: 0h 0m 0s 100ms\n",
      "\n",
      "Size of the sequence database: 10731\n",
      "Size of the alignment database: 10731\n",
      "Number of clusters: 10057\n",
      "\n",
      "Writing results 0h 0m 0s 3ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 111ms\n",
      "mergeclusters clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_redundancy clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_step0 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_step1 clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_step2 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 38.01K 0s 31ms\n",
      "Clustering step 2\n",
      "[=================================================================] 14.48K 0s 80ms\n",
      "Clustering step 3\n",
      "[=================================================================] 10.73K 0s 123ms\n",
      "Clustering step 4\n",
      "[=================================================================] 10.06K 0s 161ms\n",
      "Write merged clustering\n",
      "[=================================================================] 185.99K 0s 334ms\n",
      "Time for merging to clu: 0h 0m 0s 22ms\n",
      "Time for processing: 0h 0m 0s 407ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/pref_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/aln_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/pref_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/aln_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/clu_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step1_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_tmp/18226833203064600488/input_step2_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu clustering/mmseqs_tmp/12942347142416062949/cluster.tsv --threads 384 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 22ms\n",
      "Time for processing: 0h 0m 0s 277ms\n",
      "result2repseq clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu clustering/mmseqs_tmp/12942347142416062949/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 10.06K 0s 16ms\n",
      "Time for merging to clu_rep: 0h 0m 0s 27ms\n",
      "Time for processing: 0h 0m 0s 291ms\n",
      "result2flat clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu_rep clustering/mmseqs_tmp/12942347142416062949/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 55ms\n",
      "createseqfiledb clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu clustering/mmseqs_tmp/12942347142416062949/clu_seqs --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 10.06K 0s 50ms\n",
      "Time for merging to clu_seqs: 0h 0m 0s 15ms\n",
      "Time for processing: 0h 0m 0s 364ms\n",
      "result2flat clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/input clustering/mmseqs_tmp/12942347142416062949/clu_seqs clustering/mmseqs_tmp/12942347142416062949/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 263ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 24ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering/mmseqs_tmp/12942347142416062949/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes50 clustering/mmseqs_tmp --min-seq-id 0.5 --cov-mode 0 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.5\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "cluster clustering/mmseqs_tmp/11162861844454796777/input clustering/mmseqs_tmp/11162861844454796777/clu clustering/mmseqs_tmp/11162861844454796777/clu_tmp --max-seqs 20 -c 0.8 --cov-mode 0 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.5 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 3.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 3\n",
      "prefilter clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step_redundancy clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step_redundancy clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 46908 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 46908 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 46.91K 0s 80ms\n",
      "Index table: Masked residues: 111813\n",
      "Index table: fill\n",
      "[=================================================================] 46.91K 0s 84ms\n",
      "Index statistics\n",
      "Entries:          8240418\n",
      "DB size:          535 MB\n",
      "Avg k-mer size:   0.128757\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t272\n",
      "    GPGGTL\t176\n",
      "    LGGFVY\t164\n",
      "    AYPWGT\t119\n",
      "    GRGHCS\t113\n",
      "    SPGGGW\t108\n",
      "    IGGFVY\t101\n",
      "    NQRRRQ\t92\n",
      "    DNLGGR\t92\n",
      "    PVGVMT\t92\n",
      "Time for index table init: 0h 0m 0s 613ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 46908\n",
      "Target db start 1 to 46908\n",
      "[=================================================================] 46.91K 0s 154ms\n",
      "\n",
      "1.705059 k-mers per position\n",
      "471 DB matches per sequence\n",
      "0 overflows\n",
      "12 sequences passed prefiltering per query sequence\n",
      "13 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 31ms\n",
      "Time for processing: 0h 0m 1s 252ms\n",
      "align clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step_redundancy clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step_redundancy clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/pref_step0 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 46908 type: Aminoacid\n",
      "Target database size: 46908 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 46.91K 1s 10ms\n",
      "Time for merging to aln_step0: 0h 0m 0s 31ms\n",
      "512409 alignments calculated\n",
      "145605 sequence pairs passed the thresholds (0.284158 of overall calculated)\n",
      "3.104055 hits per query sequence\n",
      "Time for processing: 0h 0m 1s 289ms\n",
      "clust clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step_redundancy clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/aln_step0 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 46.91K 0s 3ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 6139 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 46.91K 0s 7ms\n",
      "Add missing connections\n",
      "[=================================================================] 46.91K 0s 3ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 102ms\n",
      "Total time: 0h 0m 0s 135ms\n",
      "\n",
      "Size of the sequence database: 46908\n",
      "Size of the alignment database: 46908\n",
      "Number of clusters: 29481\n",
      "\n",
      "Writing results 0h 0m 0s 5ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 152ms\n",
      "createsubdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/clu_step0 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step_redundancy clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "prefilter clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step1 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step1 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 2 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 29481 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 29481 type: Aminoacid\n",
      "Index table k-mer threshold: 145 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 29.48K 0s 55ms\n",
      "Index table: Masked residues: 89997\n",
      "Index table: fill\n",
      "[=================================================================] 29.48K 0s 85ms\n",
      "Index statistics\n",
      "Entries:          8173553\n",
      "DB size:          535 MB\n",
      "Avg k-mer size:   0.127712\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t224\n",
      "    LGGFVY\t112\n",
      "    GPGGTL\t89\n",
      "    IGGFVY\t82\n",
      "    IHDKNV\t66\n",
      "    FPHNAQ\t54\n",
      "    GPGGSL\t51\n",
      "    SGTAIP\t51\n",
      "    GRGHCS\t50\n",
      "    SPGGGW\t49\n",
      "Time for index table init: 0h 0m 0s 591ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 145\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 29481\n",
      "Target db start 1 to 29481\n",
      "[=================================================================] 29.48K 0s 156ms\n",
      "\n",
      "7.472793 k-mers per position\n",
      "728 DB matches per sequence\n",
      "0 overflows\n",
      "11 sequences passed prefiltering per query sequence\n",
      "11 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 28ms\n",
      "Time for processing: 0h 0m 1s 303ms\n",
      "align clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step1 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step1 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/pref_step1 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 29481 type: Aminoacid\n",
      "Target database size: 29481 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 29.48K 0s 688ms\n",
      "Time for merging to aln_step1: 0h 0m 0s 27ms\n",
      "269864 alignments calculated\n",
      "30930 sequence pairs passed the thresholds (0.114613 of overall calculated)\n",
      "1.049150 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 951ms\n",
      "clust clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step1 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/aln_step1 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/clu_step1 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "=[================================================================] 29.48K 0s 5ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 307 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 29.48K 0s 6ms\n",
      "Add missing connections\n",
      "[=================================================================] 29.48K 0s 0ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 90ms\n",
      "Total time: 0h 0m 0s 122ms\n",
      "\n",
      "Size of the sequence database: 29481\n",
      "Size of the alignment database: 29481\n",
      "Number of clusters: 28692\n",
      "\n",
      "Writing results 0h 0m 0s 5ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 137ms\n",
      "createsubdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/clu_step1 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step1 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "prefilter clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step2 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step2 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 3 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 28692 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 28692 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 28.69K 0s 54ms\n",
      "Index table: Masked residues: 88804\n",
      "Index table: fill\n",
      "[=================================================================] 28.69K 0s 105ms\n",
      "Index statistics\n",
      "Entries:          10702093\n",
      "DB size:          549 MB\n",
      "Avg k-mer size:   0.167220\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t223\n",
      "    RDKELL\t117\n",
      "    LGGFVY\t112\n",
      "    GQQVAR\t68\n",
      "    IHDKNV\t66\n",
      "    PGSSTI\t54\n",
      "    FPHNAQ\t54\n",
      "    SGTAIP\t51\n",
      "    GRGHCS\t50\n",
      "    SGTALP\t49\n",
      "Time for index table init: 0h 0m 0s 599ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 28692\n",
      "Target db start 1 to 28692\n",
      "[=================================================================] 28.69K 0s 187ms\n",
      "\n",
      "22.464352 k-mers per position\n",
      "1793 DB matches per sequence\n",
      "0 overflows\n",
      "15 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 28ms\n",
      "Time for processing: 0h 0m 1s 286ms\n",
      "align clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step2 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step2 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/pref_step2 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 28692 type: Aminoacid\n",
      "Target database size: 28692 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 28.69K 0s 733ms\n",
      "Time for merging to aln_step2: 0h 0m 0s 28ms\n",
      "317695 alignments calculated\n",
      "28808 sequence pairs passed the thresholds (0.090678 of overall calculated)\n",
      "1.004043 hits per query sequence\n",
      "Time for processing: 0h 0m 1s 26ms\n",
      "clust clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step2 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/aln_step2 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/clu_step2 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "===[==============================================================] 28.69K 0s 5ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 22 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 28.69K 0s 6ms\n",
      "Add missing connections\n",
      "[=================================================================] 28.69K 0s 0ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 90ms\n",
      "Total time: 0h 0m 0s 120ms\n",
      "\n",
      "Size of the sequence database: 28692\n",
      "Size of the alignment database: 28692\n",
      "Number of clusters: 28632\n",
      "\n",
      "Writing results 0h 0m 0s 5ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 135ms\n",
      "mergeclusters clustering/mmseqs_tmp/11162861844454796777/input clustering/mmseqs_tmp/11162861844454796777/clu clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/clu_redundancy clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/clu_step0 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/clu_step1 clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/clu_step2 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 46.91K 0s 32ms\n",
      "Clustering step 2\n",
      "[=================================================================] 29.48K 0s 80ms\n",
      "Clustering step 3\n",
      "[=================================================================] 28.69K 0s 123ms\n",
      "Clustering step 4\n",
      "[=================================================================] 28.63K 0s 166ms\n",
      "Write merged clustering\n",
      "[=================================================================] 185.99K 0s 396ms\n",
      "Time for merging to clu: 0h 0m 0s 28ms\n",
      "Time for processing: 0h 0m 0s 472ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/pref_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/aln_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/clu_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/pref_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/aln_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/clu_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step1_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_tmp/15086314160690248260/input_step2_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering/mmseqs_tmp/11162861844454796777/input clustering/mmseqs_tmp/11162861844454796777/input clustering/mmseqs_tmp/11162861844454796777/clu clustering/mmseqs_tmp/11162861844454796777/cluster.tsv --threads 384 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 29ms\n",
      "Time for processing: 0h 0m 0s 308ms\n",
      "result2repseq clustering/mmseqs_tmp/11162861844454796777/input clustering/mmseqs_tmp/11162861844454796777/clu clustering/mmseqs_tmp/11162861844454796777/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 28.63K 0s 21ms\n",
      "Time for merging to clu_rep: 0h 0m 0s 52ms\n",
      "Time for processing: 0h 0m 0s 326ms\n",
      "result2flat clustering/mmseqs_tmp/11162861844454796777/input clustering/mmseqs_tmp/11162861844454796777/input clustering/mmseqs_tmp/11162861844454796777/clu_rep clustering/mmseqs_tmp/11162861844454796777/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 85ms\n",
      "createseqfiledb clustering/mmseqs_tmp/11162861844454796777/input clustering/mmseqs_tmp/11162861844454796777/clu clustering/mmseqs_tmp/11162861844454796777/clu_seqs --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 28.63K 0s 65ms\n",
      "Time for merging to clu_seqs: 0h 0m 0s 26ms\n",
      "Time for processing: 0h 0m 0s 364ms\n",
      "result2flat clustering/mmseqs_tmp/11162861844454796777/input clustering/mmseqs_tmp/11162861844454796777/input clustering/mmseqs_tmp/11162861844454796777/clu_seqs clustering/mmseqs_tmp/11162861844454796777/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 278ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 26ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb clustering/mmseqs_tmp/11162861844454796777/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes70 clustering/mmseqs_tmp --min-seq-id 0.7 --cov-mode 0 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.7\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/9383376546493530605/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Shuffle database cannot be combined with --createdb-mode 0\n",
      "We recompute with --shuffle 0\n",
      "Converting sequences\n",
      "[==================\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 47ms\n",
      "Create directory clustering/mmseqs_tmp/9383376546493530605/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu clustering/mmseqs_tmp/9383376546493530605/clu_tmp --max-seqs 20 -c 0.8 --cov-mode 0 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.7 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 1.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 1\n",
      "linclust clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/clu_redundancy clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.7 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.7 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.7 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 185995 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 185.99K 0s 68ms\n",
      "Sort kmer 0h 0m 0s 52ms\n",
      "Sort by rep. sequence 0h 0m 0s 39ms\n",
      "Time for fill: 0h 0m 0s 37ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 334ms\n",
      "rescorediagonal clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.7 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 185.99K 0s 128ms\n",
      "Time for merging to pref_rescore1: 0h 0m 0s 55ms\n",
      "Time for processing: 0h 0m 0s 433ms\n",
      "clust clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref_rescore1 clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pre_clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 185.99K 0s 14ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 238202 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 185.99K 0s 6ms\n",
      "Add missing connections\n",
      "[=================================================================] 185.99K 0s 8ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 250ms\n",
      "Total time: 0h 0m 0s 298ms\n",
      "\n",
      "Size of the sequence database: 185995\n",
      "Size of the alignment database: 185995\n",
      "Number of clusters: 84571\n",
      "\n",
      "Writing results 0h 0m 0s 11ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 344ms\n",
      "createsubdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/order_redundancy clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 19ms\n",
      "createsubdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/order_redundancy clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 24ms\n",
      "filterdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref_filter1 clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref_filter2 --filter-file clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/order_redundancy --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 84.57K 0s 56ms\n",
      "Time for merging to pref_filter2: 0h 0m 0s 38ms\n",
      "Time for processing: 0h 0m 0s 285ms\n",
      "rescorediagonal clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/input_step_redundancy clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/input_step_redundancy clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref_filter2 clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.7 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 84.57K 0s 169ms\n",
      "Time for merging to pref_rescore2: 0h 0m 0s 34ms\n",
      "Time for processing: 0h 0m 0s 508ms\n",
      "align clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/input_step_redundancy clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/input_step_redundancy clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref_rescore2 clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.7 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 84571 type: Aminoacid\n",
      "Target database size: 84571 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 84.57K 0s 429ms\n",
      "Time for merging to aln: 0h 0m 0s 40ms\n",
      "166591 alignments calculated\n",
      "97965 sequence pairs passed the thresholds (0.588057 of overall calculated)\n",
      "1.158376 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 678ms\n",
      "clust clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/input_step_redundancy clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/aln clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 84.57K 0s 7ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 13394 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 84.57K 0s 8ms\n",
      "Add missing connections\n",
      "[=================================================================] 84.57K 0s 2ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 145ms\n",
      "Total time: 0h 0m 0s 180ms\n",
      "\n",
      "Size of the sequence database: 84571\n",
      "Size of the alignment database: 84571\n",
      "Number of clusters: 74336\n",
      "\n",
      "Writing results 0h 0m 0s 8ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 207ms\n",
      "mergeclusters clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/clu_redundancy clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pre_clust clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/clust --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 84.57K 0s 32ms\n",
      "Clustering step 2\n",
      "[=================================================================] 74.34K 0s 80ms\n",
      "Write merged clustering\n",
      "[=================================================================] 185.99K 0s 268ms\n",
      "Time for merging to clu_redundancy: 0h 0m 0s 35ms\n",
      "Time for processing: 0h 0m 0s 346ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/linclust/911987756442916214/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/clu_redundancy clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 19ms\n",
      "prefilter clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/input_step_redundancy clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/input_step_redundancy clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 74336 type: Aminoacid\n",
      "Estimated memory consumption: 2G\n",
      "Target database size: 74336 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 74.34K 0s 92ms\n",
      "Index table: Masked residues: 141991\n",
      "Index table: fill\n",
      "[=================================================================] 74.34K 0s 95ms\n",
      "Index statistics\n",
      "Entries:          12897000\n",
      "DB size:          562 MB\n",
      "Avg k-mer size:   0.201516\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t414\n",
      "    SRRWPI\t289\n",
      "    RNAGRQ\t261\n",
      "    GRGHCS\t245\n",
      "    RRDQGR\t229\n",
      "    SPGGGW\t223\n",
      "    LGGFVY\t222\n",
      "    MLNSDY\t204\n",
      "    RHSRNQ\t198\n",
      "    DTGPLT\t193\n",
      "Time for index table init: 0h 0m 0s 639ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 74336\n",
      "Target db start 1 to 74336\n",
      "[=================================================================] 74.34K 0s 179ms\n",
      "\n",
      "1.654127 k-mers per position\n",
      "884 DB matches per sequence\n",
      "0 overflows\n",
      "15 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 37ms\n",
      "Time for processing: 0h 0m 1s 366ms\n",
      "align clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/input_step_redundancy clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/input_step_redundancy clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/pref_step0 clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.7 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 74336 type: Aminoacid\n",
      "Target database size: 74336 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 74.34K 1s 690ms\n",
      "Time for merging to aln_step0: 0h 0m 0s 36ms\n",
      "1050675 alignments calculated\n",
      "94161 sequence pairs passed the thresholds (0.089620 of overall calculated)\n",
      "1.266694 hits per query sequence\n",
      "Time for processing: 0h 0m 1s 933ms\n",
      "clust clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/input_step_redundancy clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/aln_step0 clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 74.34K 0s 10ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 141 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 74.34K 0s 8ms\n",
      "Add missing connections\n",
      "[=================================================================] 74.34K 0s 1ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 133ms\n",
      "Total time: 0h 0m 0s 168ms\n",
      "\n",
      "Size of the sequence database: 74336\n",
      "Size of the alignment database: 74336\n",
      "Number of clusters: 67754\n",
      "\n",
      "Writing results 0h 0m 0s 7ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 192ms\n",
      "mergeclusters clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/clu_redundancy clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/clu_step0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 74.34K 0s 34ms\n",
      "Clustering step 2\n",
      "[=================================================================] 67.75K 0s 89ms\n",
      "Write merged clustering\n",
      "[=================================================================] 185.99K 0s 277ms\n",
      "Time for merging to clu: 0h 0m 0s 36ms\n",
      "Time for processing: 0h 0m 0s 362ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_tmp/10688742340621065901/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu clustering/mmseqs_tmp/9383376546493530605/cluster.tsv --threads 384 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 37ms\n",
      "Time for processing: 0h 0m 0s 274ms\n",
      "result2repseq clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu clustering/mmseqs_tmp/9383376546493530605/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 67.75K 0s 22ms\n",
      "Time for merging to clu_rep: 0h 0m 0s 77ms\n",
      "Time for processing: 0h 0m 0s 356ms\n",
      "result2flat clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu_rep clustering/mmseqs_tmp/9383376546493530605/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 141ms\n",
      "createseqfiledb clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu clustering/mmseqs_tmp/9383376546493530605/clu_seqs --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 67.75K 0s 64ms\n",
      "Time for merging to clu_seqs: 0h 0m 0s 37ms\n",
      "Time for processing: 0h 0m 0s 361ms\n",
      "result2flat clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/input clustering/mmseqs_tmp/9383376546493530605/clu_seqs clustering/mmseqs_tmp/9383376546493530605/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 303ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 26ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/9383376546493530605/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes90 clustering/mmseqs_tmp --min-seq-id 0.9 --cov-mode 0 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.9\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/7603891248532264433/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Shuffle database cannot be combined with --createdb-mode 0\n",
      "We recompute with --shuffle 0\n",
      "Converting sequences\n",
      "[==================\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 47ms\n",
      "Create directory clustering/mmseqs_tmp/7603891248532264433/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu clustering/mmseqs_tmp/7603891248532264433/clu_tmp --max-seqs 20 -c 0.8 --cov-mode 0 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.9 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 1.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 1\n",
      "linclust clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/clu_redundancy clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.9 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.9 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 185995 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 185.99K 0s 76ms\n",
      "Sort kmer 0h 0m 0s 45ms\n",
      "Sort by rep. sequence 0h 0m 0s 60ms\n",
      "Time for fill: 0h 0m 0s 29ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 348ms\n",
      "rescorediagonal clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 185.99K 0s 93ms\n",
      "Time for merging to pref_rescore1: 0h 0m 0s 58ms\n",
      "Time for processing: 0h 0m 0s 395ms\n",
      "clust clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref_rescore1 clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pre_clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 185.99K 0s 14ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 104829 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 185.99K 0s 10ms\n",
      "Add missing connections\n",
      "[=================================================================] 185.99K 0s 4ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 247ms\n",
      "Total time: 0h 0m 0s 294ms\n",
      "\n",
      "Size of the sequence database: 185995\n",
      "Size of the alignment database: 185995\n",
      "Number of clusters: 120358\n",
      "\n",
      "Writing results 0h 0m 0s 13ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 344ms\n",
      "createsubdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/order_redundancy clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 23ms\n",
      "createsubdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/order_redundancy clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 28ms\n",
      "filterdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref_filter1 clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref_filter2 --filter-file clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/order_redundancy --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 120.36K 0s 28ms\n",
      "Time for merging to pref_filter2: 0h 0m 0s 46ms\n",
      "Time for processing: 0h 0m 0s 332ms\n",
      "rescorediagonal clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/input_step_redundancy clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/input_step_redundancy clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref_filter2 clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 120.36K 0s 227ms\n",
      "Time for merging to pref_rescore2: 0h 0m 0s 42ms\n",
      "Time for processing: 0h 0m 0s 491ms\n",
      "align clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/input_step_redundancy clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/input_step_redundancy clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref_rescore2 clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 120358 type: Aminoacid\n",
      "Target database size: 120358 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 120.36K 0s 387ms\n",
      "Time for merging to aln: 0h 0m 0s 48ms\n",
      "217793 alignments calculated\n",
      "122691 sequence pairs passed the thresholds (0.563338 of overall calculated)\n",
      "1.019384 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 664ms\n",
      "clust clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/input_step_redundancy clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/aln clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 120.36K 0s 8ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 2333 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 120.36K 0s 9ms\n",
      "Add missing connections\n",
      "[=================================================================] 120.36K 0s 1ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 183ms\n",
      "Total time: 0h 0m 0s 231ms\n",
      "\n",
      "Size of the sequence database: 120358\n",
      "Size of the alignment database: 120358\n",
      "Number of clusters: 118140\n",
      "\n",
      "Writing results 0h 0m 0s 12ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 267ms\n",
      "mergeclusters clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/clu_redundancy clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pre_clust clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/clust --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 120.36K 0s 33ms\n",
      "Clustering step 2\n",
      "[=================================================================] 118.14K 0s 86ms\n",
      "Write merged clustering\n",
      "[=================================================================] 185.99K 0s 288ms\n",
      "Time for merging to clu_redundancy: 0h 0m 0s 43ms\n",
      "Time for processing: 0h 0m 0s 373ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/linclust/16766760123448016298/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/clu_redundancy clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 23ms\n",
      "prefilter clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/input_step_redundancy clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/input_step_redundancy clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 118140 type: Aminoacid\n",
      "Estimated memory consumption: 3G\n",
      "Target database size: 118140 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 118.14K 0s 110ms\n",
      "Index table: Masked residues: 187832\n",
      "Index table: fill\n",
      "[=================================================================] 118.14K 0s 126ms\n",
      "Index statistics\n",
      "Entries:          20422118\n",
      "DB size:          605 MB\n",
      "Avg k-mer size:   0.319096\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t705\n",
      "    RRDQGR\t403\n",
      "    GRGHCS\t399\n",
      "    RNAGRQ\t386\n",
      "    DERPIS\t380\n",
      "    SPGGGW\t362\n",
      "    FYHRER\t355\n",
      "    VKPFRR\t355\n",
      "    PVGVMT\t338\n",
      "    YAGPRY\t332\n",
      "Time for index table init: 0h 0m 0s 674ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 118140\n",
      "Target db start 1 to 118140\n",
      "[=================================================================] 118.14K 0s 186ms\n",
      "\n",
      "1.644596 k-mers per position\n",
      "1848 DB matches per sequence\n",
      "0 overflows\n",
      "17 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 48ms\n",
      "Time for processing: 0h 0m 1s 560ms\n",
      "align clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/input_step_redundancy clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/input_step_redundancy clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/pref_step0 clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 118140 type: Aminoacid\n",
      "Target database size: 118140 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 118.14K 2s 792ms\n",
      "Time for merging to aln_step0: 0h 0m 0s 45ms\n",
      "1913766 alignments calculated\n",
      "121997 sequence pairs passed the thresholds (0.063747 of overall calculated)\n",
      "1.032648 hits per query sequence\n",
      "Time for processing: 0h 0m 3s 59ms\n",
      "clust clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/input_step_redundancy clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/aln_step0 clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "=[================================================================] 118.14K 0s 12ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 1 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 118.14K 0s 9ms\n",
      "Add missing connections\n",
      "[=================================================================] 118.14K 0s 1ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 184ms\n",
      "Total time: 0h 0m 0s 222ms\n",
      "\n",
      "Size of the sequence database: 118140\n",
      "Size of the alignment database: 118140\n",
      "Number of clusters: 116606\n",
      "\n",
      "Writing results 0h 0m 0s 11ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 258ms\n",
      "mergeclusters clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/clu_redundancy clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/clu_step0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 118.14K 0s 35ms\n",
      "Clustering step 2\n",
      "[=================================================================] 116.61K 0s 93ms\n",
      "Write merged clustering\n",
      "[=================================================================] 185.99K 0s 260ms\n",
      "Time for merging to clu: 0h 0m 0s 45ms\n",
      "Time for processing: 0h 0m 0s 357ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 12ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_tmp/4604907506740475307/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu clustering/mmseqs_tmp/7603891248532264433/cluster.tsv --threads 384 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 48ms\n",
      "Time for processing: 0h 0m 0s 277ms\n",
      "result2repseq clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu clustering/mmseqs_tmp/7603891248532264433/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 116.61K 0s 28ms\n",
      "Time for merging to clu_rep: 0h 0m 0s 112ms\n",
      "Time for processing: 0h 0m 0s 413ms\n",
      "result2flat clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu_rep clustering/mmseqs_tmp/7603891248532264433/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 213ms\n",
      "createseqfiledb clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu clustering/mmseqs_tmp/7603891248532264433/clu_seqs --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 116.61K 0s 51ms\n",
      "Time for merging to clu_seqs: 0h 0m 0s 49ms\n",
      "Time for processing: 0h 0m 0s 384ms\n",
      "result2flat clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/input clustering/mmseqs_tmp/7603891248532264433/clu_seqs clustering/mmseqs_tmp/7603891248532264433/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 329ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 21ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "rmdb clustering/mmseqs_tmp/7603891248532264433/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "for cov in [0.3, 0.5, 0.7, 0.9]:\n",
    "    os.system(f'mmseqs easy-cluster clustering/swissprot.fasta clustering/clusterRes{str(int(cov*100))} clustering/mmseqs_tmp --min-seq-id {cov} --cov-mode 0 ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "02322b64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10057\n",
      "28632\n",
      "67754\n",
      "116606\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "      <th>EC All</th>\n",
       "      <th>clusterRes30</th>\n",
       "      <th>clusterRes50</th>\n",
       "      <th>clusterRes70</th>\n",
       "      <th>clusterRes90</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>A1AY86</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>ENG1_RHIMI</td>\n",
       "      <td>MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...</td>\n",
       "      <td>3.2.1.39</td>\n",
       "      <td>796</td>\n",
       "      <td>3.2.1.39</td>\n",
       "      <td>D4AZ24</td>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>A0A023I7E1</td>\n",
       "      <td>A0A023I7E1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>A8QPD8</td>\n",
       "      <td>A8QPD8</td>\n",
       "      <td>A8QPD8</td>\n",
       "      <td>A0A024SC78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>B2AE04</td>\n",
       "      <td>A1CCN4</td>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>A0A024SH76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A044RE18</td>\n",
       "      <td>BLI_ONCVO</td>\n",
       "      <td>MYWQLVRILVLFDCLQKILAIEHDSICIADVDDACPEPSHTVMRLR...</td>\n",
       "      <td>3.4.21.75</td>\n",
       "      <td>693</td>\n",
       "      <td>3.4.21.75</td>\n",
       "      <td>Q9VBC7</td>\n",
       "      <td>A0A044RE18</td>\n",
       "      <td>A0A044RE18</td>\n",
       "      <td>A0A044RE18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185990</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185991</th>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>Y3032_BACAN</td>\n",
       "      <td>MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>584</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>Q9KF49</td>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>Q6HX62</td>\n",
       "      <td>Q6HX62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185992</th>\n",
       "      <td>Q6L032</td>\n",
       "      <td>Y1085_PICTO</td>\n",
       "      <td>MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>573</td>\n",
       "      <td>3.5.4.2</td>\n",
       "      <td>Q9KF49</td>\n",
       "      <td>Q6L032</td>\n",
       "      <td>Q6L032</td>\n",
       "      <td>Q6L032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185993</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>P39262</td>\n",
       "      <td>P39262</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185994</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q9J5H2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>185995 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "1       A0A023I7E1   ENG1_RHIMI   \n",
       "2       A0A024SC78  CUTI1_HYPJR   \n",
       "3       A0A024SH76   GUX2_HYPJR   \n",
       "4       A0A044RE18    BLI_ONCVO   \n",
       "...            ...          ...   \n",
       "185990      Q05115   AMDA_BORBO   \n",
       "185991      Q6HX62  Y3032_BACAN   \n",
       "185992      Q6L032  Y1085_PICTO   \n",
       "185993      Q94MV8   VG56_BPLZ5   \n",
       "185994      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence  EC number  Length  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.6     269   \n",
       "1       MRFQVIVAAATITMITSYIPGVASQSTSDGDDLFVPVSNFDPKSIF...   3.2.1.39     796   \n",
       "2       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   3.1.1.74     248   \n",
       "3       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   3.2.1.91     471   \n",
       "4       MYWQLVRILVLFDCLQKILAIEHDSICIADVDDACPEPSHTVMRLR...  3.4.21.75     693   \n",
       "...                                                   ...        ...     ...   \n",
       "185990  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...   4.1.1.76     240   \n",
       "185991  MGQNQFRWSNEQLREHVEIIDGTRSPHKLLKNATYLNSYIREWMQA...    3.5.4.2     584   \n",
       "185992  MLLKNIKISNDYNIFMIIASRKPSLKDIYKIIKVSKFDEPADLIIE...    3.5.4.2     573   \n",
       "185993  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...   3.6.1.12     172   \n",
       "185994  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...   3.1.1.29     124   \n",
       "\n",
       "           EC All clusterRes30 clusterRes50 clusterRes70 clusterRes90  \n",
       "0         3.2.2.6       A1AY86   A0A009IHW8   A0A009IHW8   A0A009IHW8  \n",
       "1        3.2.1.39       D4AZ24   A0A023I7E1   A0A023I7E1   A0A023I7E1  \n",
       "2        3.1.1.74       A8QPD8       A8QPD8       A8QPD8   A0A024SC78  \n",
       "3        3.2.1.91       B2AE04       A1CCN4   A0A024SH76   A0A024SH76  \n",
       "4       3.4.21.75       Q9VBC7   A0A044RE18   A0A044RE18   A0A044RE18  \n",
       "...           ...          ...          ...          ...          ...  \n",
       "185990   4.1.1.76       Q05115       Q05115       Q05115       Q05115  \n",
       "185991    3.5.4.2       Q9KF49       Q6HX62       Q6HX62       Q6HX62  \n",
       "185992    3.5.4.2       Q9KF49       Q6L032       Q6L032       Q6L032  \n",
       "185993   3.6.1.12       P39262       P39262       Q94MV8       Q94MV8  \n",
       "185994   3.1.1.29       Q9J5H2       Q9J5H2       Q9J5H2       Q9J5H2  \n",
       "\n",
       "[185995 rows x 10 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkIAAAGdCAYAAAD+JxxnAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuNSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/xnp5ZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAs5ElEQVR4nO3de3BUZZ7/8U8Skk6CdBJgkyZjwOzocpGrRGLwsrrERDczu1EWRTOawQx4SRwgW6IwGgHFKJQKKMIys8pUDShSuzKKiPQvrCDSBoigXASdkllcnE6cgaS5SNIkz++PqZyhCUKY6RCa5/2qosp+nm+f8z35psyn+nQnUcYYIwAAAAtFd3YDAAAAnYUgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwVpfObuBC1tLSom+++UbdunVTVFRUZ7cDAADawRijw4cPKz09XdHRZ37NhyB0Bt98840yMjI6uw0AAPBX+Prrr3XppZeesYYgdAbdunWT9OcvpNvtDttxg8Gg1q5dq7y8PMXGxobtuOgYzCuyMK/IwrwiRyTNKhAIKCMjw/k5fiYEoTNovR3mdrvDHoQSExPldrsv+G8mMK9Iw7wiC/OKHJE4q/a8rYU3SwMAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyDUifZmXd3ZLQAAYDWCEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrnXMQ2rBhg3784x8rPT1dUVFRWrlyZci+MUYVFRXq1auXEhISlJubqy+//DKk5uDBgyoqKpLb7VZycrJKSkp05MiRkJrPPvtM119/veLj45WRkaHZs2e36WXFihXq16+f4uPjNWjQIK1evfqcewEAAPY65yB09OhRDRkyRAsWLDjt/uzZszV//nwtWrRI1dXV6tq1q/Lz83X8+HGnpqioSLt27ZLX69WqVau0YcMGTZgwwdkPBALKy8tTnz59VFNTozlz5mj69OlavHixU7Np0ybdddddKikp0bZt21RYWKjCwkLt3LnznHoBAAAWM38DSeatt95yHre0tBiPx2PmzJnjrNXX1xuXy2Vef/11Y4wxu3fvNpLMli1bnJr33nvPREVFmQMHDhhjjHnllVdMSkqKaWxsdGoeffRR07dvX+fxHXfcYQoKCkL6yc7ONvfff3+7ezmbhoYGI8k0NDS0q769mpqazMqVK82OwUPCelx0jNZ5NTU1dXYraAfmFVmYV+SIpFmdy8/vLuEMVfv27ZPf71dubq6zlpSUpOzsbPl8Po0dO1Y+n0/JycnKyspyanJzcxUdHa3q6mrddttt8vl8uuGGGxQXF+fU5Ofn67nnntOhQ4eUkpIin8+n8vLykPPn5+c7t+ra08upGhsb1djY6DwOBAKSpGAwqGAw+Ld9cU7SeqwWlyusx0XHaJ0Rs4oMzCuyMK/IEUmzOpcewxqE/H6/JCktLS1kPS0tzdnz+/1KTU0NbaJLF3Xv3j2kJjMzs80xWvdSUlLk9/vPep6z9XKqyspKzZgxo8362rVrlZiY+D1X/dfb94tp2nfK+5pw4fJ6vZ3dAs4B84oszCtyRMKsjh071u7asAahSDd16tSQV5kCgYAyMjKUl5cnt9sdtvMEg0F5vV5lznpG/T/aGLbjomO0zuvmm29WbGxsZ7eDs2BekYV5RY5ImlXrHZ32CGsQ8ng8kqTa2lr16tXLWa+trdXQoUOdmrq6upDnnThxQgcPHnSe7/F4VFtbG1LT+vhsNSfvn62XU7lcLrlcrjbrsbGxHTL06MbGC/6bCX/RUd8H6BjMK7Iwr8gRCbM6l/7C+nuEMjMz5fF4VFVV5awFAgFVV1crJydHkpSTk6P6+nrV1NQ4NevWrVNLS4uys7Odmg0bNoTc4/N6verbt69SUlKcmpPP01rTep729AIAAOx2zkHoyJEj2r59u7Zv3y7pz29K3r59u/bv36+oqChNmjRJTz/9tN5++23t2LFD9957r9LT01VYWChJ6t+/v2655RaNHz9emzdv1kcffaSysjKNHTtW6enpkqS7775bcXFxKikp0a5du7R8+XLNmzcv5LbVxIkTtWbNGj3//PPas2ePpk+frq1bt6qsrEyS2tULAACw2znfGtu6datuuukm53FrOCkuLtaSJUs0ZcoUHT16VBMmTFB9fb2uu+46rVmzRvHx8c5zli5dqrKyMo0aNUrR0dEaPXq05s+f7+wnJSVp7dq1Ki0t1fDhw9WzZ09VVFSE/K6hkSNHatmyZXr88cc1bdo0XXHFFVq5cqUGDhzo1LSnFwAAYK9zDkI33nijjDHfux8VFaWZM2dq5syZ31vTvXt3LVu27IznGTx4sD788MMz1owZM0Zjxoz5m3oBAAD24m+NAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWCnsQam5u1hNPPKHMzEwlJCTohz/8oZ566ikZY5waY4wqKirUq1cvJSQkKDc3V19++WXIcQ4ePKiioiK53W4lJyerpKRER44cCan57LPPdP311ys+Pl4ZGRmaPXt2m35WrFihfv36KT4+XoMGDdLq1avDfckAACBChT0IPffcc1q4cKFefvllff7553ruuec0e/ZsvfTSS07N7NmzNX/+fC1atEjV1dXq2rWr8vPzdfz4caemqKhIu3btktfr1apVq7RhwwZNmDDB2Q8EAsrLy1OfPn1UU1OjOXPmaPr06Vq8eLFTs2nTJt11110qKSnRtm3bVFhYqMLCQu3cuTPclw0AACKRCbOCggJz3333hazdfvvtpqioyBhjTEtLi/F4PGbOnDnOfn19vXG5XOb11183xhize/duI8ls2bLFqXnvvfdMVFSUOXDggDHGmFdeecWkpKSYxsZGp+bRRx81ffv2dR7fcccdpqCgIKSX7Oxsc//997frWhoaGowk09DQ0K769mpqajIrV640OwYPCetx0TFa59XU1NTZraAdmFdkYV6RI5JmdS4/v7uEO1iNHDlSixcv1hdffKF/+Id/0KeffqqNGzfqhRdekCTt27dPfr9fubm5znOSkpKUnZ0tn8+nsWPHyufzKTk5WVlZWU5Nbm6uoqOjVV1drdtuu00+n0833HCD4uLinJr8/Hw999xzOnTokFJSUuTz+VReXh7SX35+vlauXHna3hsbG9XY2Og8DgQCkqRgMKhgMPg3f21atR6rxeUK63HRMVpnxKwiA/OKLMwrckTSrM6lx7AHoccee0yBQED9+vVTTEyMmpubNWvWLBUVFUmS/H6/JCktLS3keWlpac6e3+9XampqaKNduqh79+4hNZmZmW2O0bqXkpIiv99/xvOcqrKyUjNmzGizvnbtWiUmJrbr+s/Fvl9M0z7esxQxvF5vZ7eAc8C8IgvzihyRMKtjx461uzbsQejNN9/U0qVLtWzZMl155ZXavn27Jk2apPT0dBUXF4f7dGE1derUkFeQAoGAMjIylJeXJ7fbHbbzBINBeb1eZc56Rv0/2hi246JjtM7r5ptvVmxsbGe3g7NgXpGFeUWOSJpV6x2d9gh7EHrkkUf02GOPaezYsZKkQYMG6X//939VWVmp4uJieTweSVJtba169erlPK+2tlZDhw6VJHk8HtXV1YUc98SJEzp48KDzfI/Ho9ra2pCa1sdnq2ndP5XL5ZLL5WqzHhsb2yFDj25svOC/mfAXHfV9gI7BvCIL84ockTCrc+kv7J8aO3bsmKKjQw8bExOjlpYWSVJmZqY8Ho+qqqqc/UAgoOrqauXk5EiScnJyVF9fr5qaGqdm3bp1amlpUXZ2tlOzYcOGkPuAXq9Xffv2VUpKilNz8nlaa1rPAwAA7Bb2IPTjH/9Ys2bN0rvvvqvf//73euutt/TCCy/otttukyRFRUVp0qRJevrpp/X2229rx44duvfee5Wenq7CwkJJUv/+/XXLLbdo/Pjx2rx5sz766COVlZVp7NixSk9PlyTdfffdiouLU0lJiXbt2qXly5dr3rx5Ibe2Jk6cqDVr1uj555/Xnj17NH36dG3dulVlZWXhvmwAABCBwn5r7KWXXtITTzyhhx56SHV1dUpPT9f999+viooKp2bKlCk6evSoJkyYoPr6el133XVas2aN4uPjnZqlS5eqrKxMo0aNUnR0tEaPHq358+c7+0lJSVq7dq1KS0s1fPhw9ezZUxUVFSG/a2jkyJFatmyZHn/8cU2bNk1XXHGFVq5cqYEDB4b7sgEAQAQKexDq1q2b5s6dq7lz535vTVRUlGbOnKmZM2d+b0337t21bNmyM55r8ODB+vDDD89YM2bMGI0ZM+aMNQAAwE78rTEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFodEoQOHDign/zkJ+rRo4cSEhI0aNAgbd261dk3xqiiokK9evVSQkKCcnNz9eWXX4Yc4+DBgyoqKpLb7VZycrJKSkp05MiRkJrPPvtM119/veLj45WRkaHZs2e36WXFihXq16+f4uPjNWjQIK1evbojLhkAAESgsAehQ4cO6dprr1VsbKzee+897d69W88//7xSUlKcmtmzZ2v+/PlatGiRqqur1bVrV+Xn5+v48eNOTVFRkXbt2iWv16tVq1Zpw4YNmjBhgrMfCASUl5enPn36qKamRnPmzNH06dO1ePFip2bTpk266667VFJSom3btqmwsFCFhYXauXNnuC8bAABEIhNmjz76qLnuuuu+d7+lpcV4PB4zZ84cZ62+vt64XC7z+uuvG2OM2b17t5FktmzZ4tS89957Jioqyhw4cMAYY8wrr7xiUlJSTGNjY8i5+/bt6zy+4447TEFBQcj5s7Ozzf3339+ua2loaDCSTENDQ7vq26upqcmsXLnS7Bg8JKzHRcdonVdTU1Nnt4J2YF6RhXlFjkia1bn8/O4S7mD19ttvKz8/X2PGjNH69ev1gx/8QA899JDGjx8vSdq3b5/8fr9yc3Od5yQlJSk7O1s+n09jx46Vz+dTcnKysrKynJrc3FxFR0erurpat912m3w+n2644QbFxcU5Nfn5+Xruued06NAhpaSkyOfzqby8PKS//Px8rVy58rS9NzY2qrGx0XkcCAQkScFgUMFg8G/+2rRqPVaLyxXW46JjtM6IWUUG5hVZmFfkiKRZnUuPYQ9CX331lRYuXKjy8nJNmzZNW7Zs0c9//nPFxcWpuLhYfr9fkpSWlhbyvLS0NGfP7/crNTU1tNEuXdS9e/eQmszMzDbHaN1LSUmR3+8/43lOVVlZqRkzZrRZX7t2rRITE9v7JWi3fb+Ypn28ZylieL3ezm4B54B5RRbmFTkiYVbHjh1rd23Yg1BLS4uysrL0zDPPSJKGDRumnTt3atGiRSouLg736cJq6tSpIa8gBQIBZWRkKC8vT263O2znCQaD8nq9ypz1jPp/tDFsx0XHaJ3XzTffrNjY2M5uB2fBvCIL84ockTSr1js67RH2INSrVy8NGDAgZK1///76r//6L0mSx+ORJNXW1qpXr15OTW1trYYOHerU1NXVhRzjxIkTOnjwoPN8j8ej2trakJrWx2erad0/lcvlksvlarMeGxvbIUOPbmy84L+Z8Bcd9X2AjsG8IgvzihyRMKtz6S/snxq79tprtXfv3pC1L774Qn369JEkZWZmyuPxqKqqytkPBAKqrq5WTk6OJCknJ0f19fWqqalxatatW6eWlhZlZ2c7NRs2bAi5D+j1etW3b1/nE2o5OTkh52mtaT0PAACwW9iD0OTJk/Xxxx/rmWee0e9+9zstW7ZMixcvVmlpqSQpKipKkyZN0tNPP623335bO3bs0L333qv09HQVFhZK+vMrSLfccovGjx+vzZs366OPPlJZWZnGjh2r9PR0SdLdd9+tuLg4lZSUaNeuXVq+fLnmzZsXcmtr4sSJWrNmjZ5//nnt2bNH06dP19atW1VWVhbuywYAABEo7LfGrr76ar311luaOnWqZs6cqczMTM2dO1dFRUVOzZQpU3T06FFNmDBB9fX1uu6667RmzRrFx8c7NUuXLlVZWZlGjRql6OhojR49WvPnz3f2k5KStHbtWpWWlmr48OHq2bOnKioqQn7X0MiRI7Vs2TI9/vjjmjZtmq644gqtXLlSAwcODPdlAwCACBT2ICRJP/rRj/SjH/3oe/ejoqI0c+ZMzZw583trunfvrmXLlp3xPIMHD9aHH354xpoxY8ZozJgxZ24YAABYib81BgAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYq8OD0LPPPquoqChNmjTJWTt+/LhKS0vVo0cPXXLJJRo9erRqa2tDnrd//34VFBQoMTFRqampeuSRR3TixImQmg8++EBXXXWVXC6XLr/8ci1ZsqTN+RcsWKDLLrtM8fHxys7O1ubNmzviMgEAQATq0CC0ZcsW/cd//IcGDx4csj558mS98847WrFihdavX69vvvlGt99+u7Pf3NysgoICNTU1adOmTfr1r3+tJUuWqKKiwqnZt2+fCgoKdNNNN2n79u2aNGmSfvazn+n99993apYvX67y8nI9+eST+uSTTzRkyBDl5+errq6uIy8bAABEiA4LQkeOHFFRUZF++ctfKiUlxVlvaGjQf/7nf+qFF17QP/3TP2n48OF67bXXtGnTJn388ceSpLVr12r37t36zW9+o6FDh+rWW2/VU089pQULFqipqUmStGjRImVmZur5559X//79VVZWpn/7t3/Tiy++6JzrhRde0Pjx4zVu3DgNGDBAixYtUmJiol599dWOumwAABBBunTUgUtLS1VQUKDc3Fw9/fTTznpNTY2CwaByc3OdtX79+ql3797y+Xy65ppr5PP5NGjQIKWlpTk1+fn5evDBB7Vr1y4NGzZMPp8v5BitNa234JqamlRTU6OpU6c6+9HR0crNzZXP5zttz42NjWpsbHQeBwIBSVIwGFQwGPzrvxinaD1Wi8sV1uOiY7TOiFlFBuYVWZhX5IikWZ1Ljx0ShN544w198skn2rJlS5s9v9+vuLg4JScnh6ynpaXJ7/c7NSeHoNb91r0z1QQCAX333Xc6dOiQmpubT1uzZ8+e0/ZdWVmpGTNmtFlfu3atEhMTz3DFf519v5imfatXh/246Bher7ezW8A5YF6RhXlFjkiY1bFjx9pdG/Yg9PXXX2vixInyer2Kj48P9+E71NSpU1VeXu48DgQCysjIUF5entxud9jOEwwG5fV6lTnrGfX/aGPYjouO0Tqvm2++WbGxsZ3dDs6CeUUW5hU5ImlWrXd02iPsQaimpkZ1dXW66qqrnLXm5mZt2LBBL7/8st5//301NTWpvr4+5FWh2tpaeTweSZLH42nz6a7WT5WdXHPqJ81qa2vldruVkJCgmJgYxcTEnLam9RincrlccrlcbdZjY2M7ZOjRjY0X/DcT/qKjvg/QMZhXZGFekSMSZnUu/YX9zdKjRo3Sjh07tH37dudfVlaWioqKnP+OjY1VVVWV85y9e/dq//79ysnJkSTl5ORox44dIZ/u8nq9crvdGjBggFNz8jFaa1qPERcXp+HDh4fUtLS0qKqqyqkBAAB2C/srQt26ddPAgQND1rp27aoePXo46yUlJSovL1f37t3ldrv18MMPKycnR9dcc40kKS8vTwMGDNA999yj2bNny+/36/HHH1dpaanzis0DDzygl19+WVOmTNF9992ndevW6c0339S7777rnLe8vFzFxcXKysrSiBEjNHfuXB09elTjxo0L92UDAIAI1GGfGjuTF198UdHR0Ro9erQaGxuVn5+vV155xdmPiYnRqlWr9OCDDyonJ0ddu3ZVcXGxZs6c6dRkZmbq3Xff1eTJkzVv3jxdeuml+tWvfqX8/Hyn5s4779S3336riooK+f1+DR06VGvWrGnzBmoAAGCn8xKEPvjgg5DH8fHxWrBggRYsWPC9z+nTp49Wn+UTVTfeeKO2bdt2xpqysjKVlZW1u1cAAGAP/tYYAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsFaXzm7AetOTTvrvhs7rAwAAC/GKEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrhT0IVVZW6uqrr1a3bt2UmpqqwsJC7d27N6Tm+PHjKi0tVY8ePXTJJZdo9OjRqq2tDanZv3+/CgoKlJiYqNTUVD3yyCM6ceJESM0HH3ygq666Si6XS5dffrmWLFnSpp8FCxbosssuU3x8vLKzs7V58+ZwXzIAAIhQYQ9C69evV2lpqT7++GN5vV4Fg0Hl5eXp6NGjTs3kyZP1zjvvaMWKFVq/fr2++eYb3X777c5+c3OzCgoK1NTUpE2bNunXv/61lixZooqKCqdm3759Kigo0E033aTt27dr0qRJ+tnPfqb333/fqVm+fLnKy8v15JNP6pNPPtGQIUOUn5+vurq6cF82AACIRKaD1dXVGUlm/fr1xhhj6uvrTWxsrFmxYoVT8/nnnxtJxufzGWOMWb16tYmOjjZ+v9+pWbhwoXG73aaxsdEYY8yUKVPMlVdeGXKuO++80+Tn5zuPR4wYYUpLS53Hzc3NJj093VRWVrar94aGBiPJNDQ0nONVn1lTU5NZuXKl2TF4iDFPuv/yDxek1nk1NTV1ditoB+YVWZhX5IikWZ3Lz+8Of49QQ0ODJKl79+6SpJqaGgWDQeXm5jo1/fr1U+/eveXz+SRJPp9PgwYNUlpamlOTn5+vQCCgXbt2OTUnH6O1pvUYTU1NqqmpCamJjo5Wbm6uUwMAAOzWpSMP3tLSokmTJunaa6/VwIEDJUl+v19xcXFKTk4OqU1LS5Pf73dqTg5Brfute2eqCQQC+u6773To0CE1NzeftmbPnj2n7bexsVGNjY3O40AgIEkKBoMKBoPnculn1HqsFpdLwej4kzfCdg6ET+u8wvk9gI7DvCIL84ockTSrc+mxQ4NQaWmpdu7cqY0bN3bkacKmsrJSM2bMaLO+du1aJSYmhv18+34xTftOXli9OuznQPh4vd7ObgHngHlFFuYVOSJhVseOHWt3bYcFobKyMq1atUobNmzQpZde6qx7PB41NTWpvr4+5FWh2tpaeTwep+bUT3e1fqrs5JpTP2lWW1srt9uthIQExcTEKCYm5rQ1rcc41dSpU1VeXu48DgQCysjIUF5entxu9zl+Bb5fMBiU1+tV5qxn1P9Hvz+pgf8L2zkQPq3zuvnmmxUbG9vZ7eAsmFdkYV6RI5Jm1XpHpz3CHoSMMXr44Yf11ltv6YMPPlBmZmbI/vDhwxUbG6uqqiqNHj1akrR3717t379fOTk5kqScnBzNmjVLdXV1Sk1NlfTnBOp2uzVgwACnZvUpr6B4vV7nGHFxcRo+fLiqqqpUWFgo6c+36qqqqlRWVnba3l0ul1wuV5v12NjYDhl6dGOjYluOn3yisJ8D4dNR3wfoGMwrsjCvyBEJszqX/sIehEpLS7Vs2TL99re/Vbdu3Zz39CQlJSkhIUFJSUkqKSlReXm5unfvLrfbrYcfflg5OTm65pprJEl5eXkaMGCA7rnnHs2ePVt+v1+PP/64SktLnaDywAMP6OWXX9aUKVN03333ad26dXrzzTf17rvvOr2Ul5eruLhYWVlZGjFihObOnaujR49q3Lhx4b5sAAAQgcIehBYuXChJuvHGG0PWX3vtNf30pz+VJL344ouKjo7W6NGj1djYqPz8fL3yyitObUxMjFatWqUHH3xQOTk56tq1q4qLizVz5kynJjMzU++++64mT56sefPm6dJLL9WvfvUr5efnOzV33nmnvv32W1VUVMjv92vo0KFas2ZNmzdQAwAAO3XIrbGziY+P14IFC7RgwYLvrenTp0+bW1+nuvHGG7Vt27Yz1pSVlX3vrTAAAGA3/tYYAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQAAIC1CEIAAMBaBCEAAGAtghAAALAWQQgAAFiLIAQAAKxFEAIAANYiCHWyz99I7+wWAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQuAIMye3d2CwAAWIkgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWgQhAABgLYIQAACwFkEIAABYiyAEAACsRRACAADWIggBAABrEYQuAG9WnujsFgAAsBJBCAAAWKtLZzeAk0xPOs1aw/nvAwAAS1jxitCCBQt02WWXKT4+XtnZ2dq8eXNnt9TGoMzeGpTZu7PbAADAKhd9EFq+fLnKy8v15JNP6pNPPtGQIUOUn5+vurq6zm4NAAB0sos+CL3wwgsaP368xo0bpwEDBmjRokVKTEzUq6++2tmtAQCATnZRv0eoqalJNTU1mjp1qrMWHR2t3Nxc+Xy+NvWNjY1qbGx0Hjc0/Pn9OQcPHlQwGAxbX8FgUMeOHVOgSxdFNzdLkn455897w8r+PrR40TD9v/3/9/0H+/c9YesLp9c6rz/96U+KjY3t7HZwFswrsjCvyBFJszp8+LAkyRhz1tqLOgj98Y9/VHNzs9LS0kLW09LStGdP2wBRWVmpGTNmtFnPzMzssB7beLDtUs8z1VeecRcAAGsdPnxYSUmn+SDSSS7qIHSupk6dqvLycudxS0uLDh48qB49eigqKips5wkEAsrIyNDXX38tt9sdtuOiYzCvyMK8IgvzihyRNCtjjA4fPqz09PSz1l7UQahnz56KiYlRbW1tyHptba08Hk+bepfLJZfLFbKWnJzcYf253e4L/psJf8G8IgvziizMK3JEyqzO9kpQq4v6zdJxcXEaPny4qqqqnLWWlhZVVVUpJyenEzsDAAAXgov6FSFJKi8vV3FxsbKysjRixAjNnTtXR48e1bhx4zq7NQAA0Mku+iB055136ttvv1VFRYX8fr+GDh2qNWvWtHkD9fnkcrn05JNPtrkNhwsT84oszCuyMK/IcbHOKsq057NlAAAAF6GL+j1CAAAAZ0IQAgAA1iIIAQAAaxGEAACAtQhCnWDBggW67LLLFB8fr+zsbG3evLmzW7JOZWWlrr76anXr1k2pqakqLCzU3r17Q2qOHz+u0tJS9ejRQ5dccolGjx7d5pdz7t+/XwUFBUpMTFRqaqoeeeQRnThx4nxeinWeffZZRUVFadKkSc4as7qwHDhwQD/5yU/Uo0cPJSQkaNCgQdq6dauzb4xRRUWFevXqpYSEBOXm5urLL78MOcbBgwdVVFQkt9ut5ORklZSU6MiRI+f7Ui56zc3NeuKJJ5SZmamEhAT98Ic/1FNPPRXyN7ou+nkZnFdvvPGGiYuLM6+++qrZtWuXGT9+vElOTja1tbWd3ZpV8vPzzWuvvWZ27txptm/fbv75n//Z9O7d2xw5csSpeeCBB0xGRoapqqoyW7duNddcc40ZOXKks3/ixAkzcOBAk5uba7Zt22ZWr15tevbsaaZOndoZl2SFzZs3m8suu8wMHjzYTJw40VlnVheOgwcPmj59+pif/vSnprq62nz11Vfm/fffN7/73e+cmmeffdYkJSWZlStXmk8//dT8y7/8i8nMzDTfffedU3PLLbeYIUOGmI8//th8+OGH5vLLLzd33XVXZ1zSRW3WrFmmR48eZtWqVWbfvn1mxYoV5pJLLjHz5s1zai72eRGEzrMRI0aY0tJS53Fzc7NJT083lZWVndgV6urqjCSzfv16Y4wx9fX1JjY21qxYscKp+fzzz40k4/P5jDHGrF692kRHRxu/3+/ULFy40LjdbtPY2Hh+L8AChw8fNldccYXxer3mH//xH50gxKwuLI8++qi57rrrvne/paXFeDweM2fOHGetvr7euFwu8/rrrxtjjNm9e7eRZLZs2eLUvPfeeyYqKsocOHCg45q3UEFBgbnvvvtC1m6//XZTVFRkjLFjXtwaO4+amppUU1Oj3NxcZy06Olq5ubny+Xyd2BkaGhokSd27d5ck1dTUKBgMhsyqX79+6t27tzMrn8+nQYMGhfxyzvz8fAUCAe3ates8dm+H0tJSFRQUhMxEYlYXmrfffltZWVkaM2aMUlNTNWzYMP3yl7909vft2ye/3x8yr6SkJGVnZ4fMKzk5WVlZWU5Nbm6uoqOjVV1dff4uxgIjR45UVVWVvvjiC0nSp59+qo0bN+rWW2+VZMe8LvrfLH0h+eMf/6jm5uY2v9U6LS1Ne/bs6aSu0NLSokmTJunaa6/VwIEDJUl+v19xcXFt/uhuWlqa/H6/U3O6WbbuIXzeeOMNffLJJ9qyZUubPWZ1Yfnqq6+0cOFClZeXa9q0adqyZYt+/vOfKy4uTsXFxc7X+3TzOHleqampIftdunRR9+7dmVeYPfbYYwoEAurXr59iYmLU3NysWbNmqaioSJKsmBdBCNYrLS3Vzp07tXHjxs5uBafx9ddfa+LEifJ6vYqPj+/sdnAWLS0tysrK0jPPPCNJGjZsmHbu3KlFixapuLi4k7vDqd58800tXbpUy5Yt05VXXqnt27dr0qRJSk9Pt2Ze3Bo7j3r27KmYmJg2n2apra2Vx+PppK7sVlZWplWrVul//ud/dOmllzrrHo9HTU1Nqq+vD6k/eVYej+e0s2zdQ3jU1NSorq5OV111lbp06aIuXbpo/fr1mj9/vrp06aK0tDRmdQHp1auXBgwYELLWv39/7d+/X9Jfvt5n+v+gx+NRXV1dyP6JEyd08OBB5hVmjzzyiB577DGNHTtWgwYN0j333KPJkyersrJSkh3zIgidR3FxcRo+fLiqqqqctZaWFlVVVSknJ6cTO7OPMUZlZWV66623tG7dOmVmZobsDx8+XLGxsSGz2rt3r/bv3+/MKicnRzt27Aj5H4DX65Xb7W7zgwB/vVGjRmnHjh3avn278y8rK0tFRUXOfzOrC8e1117b5ldRfPHFF+rTp48kKTMzUx6PJ2RegUBA1dXVIfOqr69XTU2NU7Nu3Tq1tLQoOzv7PFyFPY4dO6bo6NAoEBMTo5aWFkmWzKuz361tmzfeeMO4XC6zZMkSs3v3bjNhwgSTnJwc8mkWdLwHH3zQJCUlmQ8++MD84Q9/cP4dO3bMqXnggQdM7969zbp168zWrVtNTk6OycnJcfZbP5Kdl5dntm/fbtasWWP+7u/+jo9knwcnf2rMGGZ1Idm8ebPp0qWLmTVrlvnyyy/N0qVLTWJiovnNb37j1Dz77LMmOTnZ/Pa3vzWfffaZ+dd//dfTfhx72LBhprq62mzcuNFcccUVEfNx7EhSXFxsfvCDHzgfn//v//5v07NnTzNlyhSn5mKfF0GoE7z00kumd+/eJi4uzowYMcJ8/PHHnd2SdSSd9t9rr73m1Hz33XfmoYceMikpKSYxMdHcdttt5g9/+EPIcX7/+9+bW2+91SQkJJiePXuaf//3fzfBYPA8X419Tg1CzOrC8s4775iBAwcal8tl+vXrZxYvXhyy39LSYp544gmTlpZmXC6XGTVqlNm7d29IzZ/+9Cdz1113mUsuucS43W4zbtw4c/jw4fN5GVYIBAJm4sSJpnfv3iY+Pt78/d//vfnFL34R8mslLvZ5RRlz0q+PBAAAsAjvEQIAANYiCAEAAGsRhAAAgLUIQgAAwFoEIQAAYC2CEAAAsBZBCAAAWIsgBAAArEUQAgAA1iIIAQAAaxGEAACAtQhCAADAWv8f14kQYzaoY5gAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def process_clustering(name):\n",
    "    clustering = pd.read_csv('clustering/{}_cluster.tsv'.format(name), delimiter='\\t', header=None)\n",
    "    #rename heading as cluster reference and id\n",
    "    clustering.columns = [name, 'Entry']\n",
    "    clustering.drop_duplicates(subset='Entry', keep='first', inplace=True)\n",
    "    print(clustering[name].nunique())\n",
    "    clustering[name].value_counts().hist(bins=100)\n",
    "    return clustering\n",
    "\n",
    "swissprot = swissprot[['Entry', 'Entry Name', 'Sequence', 'EC number', 'Length', 'EC All']]\n",
    "\n",
    "cluster30 = process_clustering('clusterRes30')\n",
    "cluster50 = process_clustering('clusterRes50')\n",
    "cluster70 = process_clustering('clusterRes70')\n",
    "cluster90 = process_clustering('clusterRes90')\n",
    "\n",
    "swissprot = pd.merge(swissprot, cluster30, on='Entry', how='left')\n",
    "swissprot = pd.merge(swissprot, cluster50, on='Entry', how='left')\n",
    "swissprot = pd.merge(swissprot, cluster70, on='Entry', how='left')\n",
    "swissprot = pd.merge(swissprot, cluster90, on='Entry', how='left')\n",
    "\n",
    "# 25739\n",
    "# 9012\n",
    "# 61736\n",
    "# 106727\n",
    "swissprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "63ca5f5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "swissprot['EC3'] = swissprot['EC number'].str.split('.').str[:3].str.join('.')\n",
    "swissprot['EC2'] = swissprot['EC number'].str.split('.').str[:2].str.join('.')\n",
    "swissprot['EC1'] = swissprot['EC number'].str.split('.').str[:1].str.join('.')\n",
    "swissprot.to_csv('../processed_data/protein2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "756c5423",
   "metadata": {},
   "source": [
    "# Finally process textual description of EC numbers\n",
    "This is optionally used in CREEP\n",
    "\n",
    "`raw_data/ECtoGO_raw.txt` was downloaded from http://current.geneontology.org/ontology/external2go/ec2go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b5f94ed4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EC number</th>\n",
       "      <th>Text</th>\n",
       "      <th>Text Incomplete</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1658</th>\n",
       "      <td>1.1.1.100</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2147</th>\n",
       "      <td>1.1.1.101</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2119</th>\n",
       "      <td>1.1.1.102</td>\n",
       "      <td>oxidoreductase; oxidoreductase, acting on CH-O...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>7.6.2.5</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1727</th>\n",
       "      <td>7.6.2.6</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4695</th>\n",
       "      <td>7.6.2.7</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>934</th>\n",
       "      <td>7.6.2.8</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3176</th>\n",
       "      <td>7.6.2.9</td>\n",
       "      <td>primary active transmembrane transporter; ABC-...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4960 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      EC number                                               Text  \\\n",
       "327     1.1.1.1  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "2573   1.1.1.10  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "1658  1.1.1.100  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "2147  1.1.1.101  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "2119  1.1.1.102  oxidoreductase; oxidoreductase, acting on CH-O...   \n",
       "...         ...                                                ...   \n",
       "83      7.6.2.5  primary active transmembrane transporter; ABC-...   \n",
       "1727    7.6.2.6  primary active transmembrane transporter; ABC-...   \n",
       "4695    7.6.2.7  primary active transmembrane transporter; ABC-...   \n",
       "934     7.6.2.8  primary active transmembrane transporter; ABC-...   \n",
       "3176    7.6.2.9  primary active transmembrane transporter; ABC-...   \n",
       "\n",
       "      Text Incomplete  \n",
       "327             False  \n",
       "2573            False  \n",
       "1658            False  \n",
       "2147            False  \n",
       "2119            False  \n",
       "...               ...  \n",
       "83              False  \n",
       "1727            False  \n",
       "4695            False  \n",
       "934             False  \n",
       "3176            False  \n",
       "\n",
       "[4960 rows x 3 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read ECtoGO.txt line by line\n",
    "ECtoGO = open(\"raw_data/ECtoGO_raw.txt\", \"r\")\n",
    "ECtoGO_lines = ECtoGO.readlines()\n",
    "ECtoGO.close()\n",
    "\n",
    "#skip the first two lines\n",
    "ECtoGO_lines = ECtoGO_lines[2:]\n",
    "EC2desc = {}\n",
    "for line in ECtoGO_lines:\n",
    "    line = line.strip().split(\">\")\n",
    "    EC = line[0][3:-1]\n",
    "    desc = line[1].split(\";\")[0][4:-1]\n",
    "    EC2desc[EC] = desc\n",
    "\n",
    "def get_EC_desc(EC):\n",
    "    num_dashes = EC.count('-')\n",
    "\n",
    "    description_missing = False\n",
    "    EC1 = '.'.join(EC.split('.')[:1])\n",
    "    if EC1 + '.-.-.-' in EC2desc and num_dashes < 3:\n",
    "        desc1 = EC2desc[EC1 + '.-.-.-']\n",
    "    else:\n",
    "         desc1 = ''\n",
    "    EC2 = '.'.join(EC.split('.')[:2]) \n",
    "    if EC2 + '.-.-' in EC2desc and num_dashes < 2:\n",
    "        desc2 = EC2desc[EC2 + '.-.-']\n",
    "    else:\n",
    "        desc2 = ''\n",
    "    EC3 = '.'.join(EC.split('.')[:3])\n",
    "    if EC3 + '.-' in EC2desc and num_dashes < 1:\n",
    "        desc3 = EC2desc[EC3 + '.-']\n",
    "    else:\n",
    "        desc3 = ''\n",
    "        \n",
    "    if EC in EC2desc:\n",
    "        desc4 = EC2desc[EC]\n",
    "    else:\n",
    "        description_missing = True\n",
    "        desc4 = ''\n",
    "\n",
    "  \n",
    "    description = desc1 + '; ' + desc2 + '; ' + desc3 + '; ' + desc4\n",
    "    description = description.replace(' ; ', ' ')\n",
    "    description = description.replace(' ; ', ' ')\n",
    "    description = description.replace(' ; ', ' ')\n",
    "    description = description.replace(' activity', '')\n",
    "    #if string starts with ;, replace with space\n",
    "    if description[0] == ';':\n",
    "        description = description[2:]\n",
    "    if description[-2:] == '; ':\n",
    "        description = description[:-2]\n",
    "    return description\n",
    "\n",
    "#subset to the EC numbers in swissprot\n",
    "EC2desc_df = pd.DataFrame(swissprot['EC number'].unique(), columns=['EC number'])\n",
    "EC2desc_df['Text'] = EC2desc_df['EC number'].apply(get_EC_desc)\n",
    "EC2desc_df['Text Incomplete'] = ~EC2desc_df['EC number'].isin(EC2desc.keys())\n",
    "EC2desc_df = EC2desc_df.sort_values(by='EC number')\n",
    "EC2desc_df = EC2desc_df[EC2desc_df['EC number'].isin(swissprot['EC number'].unique())]\n",
    "EC2desc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "bda4ac06",
   "metadata": {},
   "outputs": [],
   "source": [
    "EC2desc_df.to_csv('../processed_data/text2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c71ca7a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save ECs to a txt as the order of the cluster centers for downstream tasks\n",
    "with open('../processed_data/EC_list.txt', 'w') as f:\n",
    "    for ec in EC2desc_df['EC number']:\n",
    "        f.write(ec + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aed0f440",
   "metadata": {},
   "source": [
    "raw_data/price.tsv was downloaded from the CLEAN repo."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
