{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "06e5ee54-4659-4353-b6ad-de92dd9ab319",
   "metadata": {},
   "source": [
    "# Data download and processing from Uniprot\n",
    "\n",
    "Downloaded the data on 01 May 2024. 254,929 results filtering for \"Catalytic activity\" and Reviewed \"Swiss-Prot\".  \n",
    "\n",
    "Selected download TSV and the columns, Seqeunce (under Sequences tab), EC number (under Function).\n",
    "\n",
    "Unzipped the downloaded file.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4f961afc-aedd-41b9-8635-4c6eabd017ab",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'data/uniprot/uniprotkb_AND_reviewed_true_AND_protein_2024_05_01.tsv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpd\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mdata/uniprot/uniprotkb_AND_reviewed_true_AND_protein_2024_05_01.tsv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m df\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/parsers/readers.py:912\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m    899\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    900\u001b[0m     dialect,\n\u001b[1;32m    901\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    908\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m    909\u001b[0m )\n\u001b[1;32m    910\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 912\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/parsers/readers.py:577\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    574\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    576\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 577\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    579\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    580\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/parsers/readers.py:1407\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1404\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1406\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1407\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/parsers/readers.py:1661\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1659\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1660\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1661\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1662\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1663\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1664\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1665\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1666\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1667\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1668\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1669\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1670\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1671\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1672\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/common.py:859\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    854\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    855\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    856\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    857\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    858\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 859\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    860\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    861\u001b[0m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    862\u001b[0m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    863\u001b[0m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    864\u001b[0m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    865\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    866\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    867\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    868\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/uniprot/uniprotkb_AND_reviewed_true_AND_protein_2024_05_01.tsv'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv('data/uniprot/uniprotkb_AND_reviewed_true_AND_protein_2024_05_01.tsv', sep='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "160aa2c9-277e-427f-8e52-58d77299d79e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A0A061I403</td>\n",
       "      <td>FICD_CRIGR</td>\n",
       "      <td>MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...</td>\n",
       "      <td>2.7.7.108; 3.1.4.-</td>\n",
       "      <td>455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254921</th>\n",
       "      <td>P83855</td>\n",
       "      <td>ACLP_MOUSE</td>\n",
       "      <td>GVYMEIGRCRXEAXRRRKEAV</td>\n",
       "      <td>4.6.1.1</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254922</th>\n",
       "      <td>P85957</td>\n",
       "      <td>PER_PSEMZ</td>\n",
       "      <td>AMAGTATVQGQGTR</td>\n",
       "      <td>1.11.1.7</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254923</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254927</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254928</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>202388 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "3       A0A024SC78  CUTI1_HYPJR   \n",
       "4       A0A024SH76   GUX2_HYPJR   \n",
       "6       A0A059TC02   CCR1_PETHY   \n",
       "8       A0A061I403   FICD_CRIGR   \n",
       "...            ...          ...   \n",
       "254921      P83855   ACLP_MOUSE   \n",
       "254922      P85957    PER_PSEMZ   \n",
       "254923      Q05115   AMDA_BORBO   \n",
       "254927      Q94MV8   VG56_BPLZ5   \n",
       "254928      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence           EC number  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.-; 3.2.2.6   \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...            3.1.1.74   \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...            3.2.1.91   \n",
       "6       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...            1.2.1.44   \n",
       "8       MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...  2.7.7.108; 3.1.4.-   \n",
       "...                                                   ...                 ...   \n",
       "254921                              GVYMEIGRCRXEAXRRRKEAV             4.6.1.1   \n",
       "254922                                     AMAGTATVQGQGTR            1.11.1.7   \n",
       "254923  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...            4.1.1.76   \n",
       "254927  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...            3.6.1.12   \n",
       "254928  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...            3.1.1.29   \n",
       "\n",
       "        Length  \n",
       "0          269  \n",
       "3          248  \n",
       "4          471  \n",
       "6          333  \n",
       "8          455  \n",
       "...        ...  \n",
       "254921      21  \n",
       "254922      14  \n",
       "254923     240  \n",
       "254927     172  \n",
       "254928     124  \n",
       "\n",
       "[202388 rows x 5 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter to only include sequneces < 512\n",
    "df = df[df['Length'] < 512]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee69eeba-0d87-4f0e-92f8-8d950cc616f3",
   "metadata": {},
   "source": [
    "# Explode out EC number so each EC has it's own row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "724377c7-9690-4cb2-ad79-e081cc7f9fb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "199220\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/gq/6ljhmvm1713fykdjqbl188pm0000gn/T/ipykernel_34484/1265130750.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['EC number'] = [ec.split(';') for ec in df['EC number']]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254921</th>\n",
       "      <td>P83855</td>\n",
       "      <td>ACLP_MOUSE</td>\n",
       "      <td>GVYMEIGRCRXEAXRRRKEAV</td>\n",
       "      <td>4.6.1.1</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254922</th>\n",
       "      <td>P85957</td>\n",
       "      <td>PER_PSEMZ</td>\n",
       "      <td>AMAGTATVQGQGTR</td>\n",
       "      <td>1.11.1.7</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254923</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254927</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254928</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>213091 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "3       A0A024SC78  CUTI1_HYPJR   \n",
       "4       A0A024SH76   GUX2_HYPJR   \n",
       "6       A0A059TC02   CCR1_PETHY   \n",
       "...            ...          ...   \n",
       "254921      P83855   ACLP_MOUSE   \n",
       "254922      P85957    PER_PSEMZ   \n",
       "254923      Q05115   AMDA_BORBO   \n",
       "254927      Q94MV8   VG56_BPLZ5   \n",
       "254928      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence EC number  Length  \n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.-     269  \n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.6     269  \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...  3.1.1.74     248  \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...  3.2.1.91     471  \n",
       "6       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...  1.2.1.44     333  \n",
       "...                                                   ...       ...     ...  \n",
       "254921                              GVYMEIGRCRXEAXRRRKEAV   4.6.1.1      21  \n",
       "254922                                     AMAGTATVQGQGTR  1.11.1.7      14  \n",
       "254923  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...  4.1.1.76     240  \n",
       "254927  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...  3.6.1.12     172  \n",
       "254928  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...  3.1.1.29     124  \n",
       "\n",
       "[213091 rows x 5 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop rows that don't have an ec number\n",
    "df = df[df['EC number'].notna()]\n",
    "print(len(df))\n",
    "# Now expand out the ones we have left\n",
    "df['EC number'] = [ec.split(';') for ec in df['EC number']]\n",
    "df = df.explode('EC number')\n",
    "# Clean the EC numbers\n",
    "df['EC number'] = [ec.replace(' ', '') for ec in df['EC number']]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a1e43b52-f7c4-444d-a064-baaf45c77f76",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('data/uniprot/processed.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4b81d7a",
   "metadata": {},
   "source": [
    "### Added a bit here to get rid of the short ones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e42c854f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('data/uniprot/processed.csv')\n",
    "df = df[df['Length'] >= 100 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "aad366c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213078</th>\n",
       "      <td>P36352</td>\n",
       "      <td>POLR_PHMV</td>\n",
       "      <td>VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213079</th>\n",
       "      <td>P39262</td>\n",
       "      <td>VG56_BPT4</td>\n",
       "      <td>MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213088</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213089</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213090</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>208984 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "1       A0A009IHW8  ABTIR_ACIB9   \n",
       "2       A0A024SC78  CUTI1_HYPJR   \n",
       "3       A0A024SH76   GUX2_HYPJR   \n",
       "4       A0A059TC02   CCR1_PETHY   \n",
       "...            ...          ...   \n",
       "213078      P36352    POLR_PHMV   \n",
       "213079      P39262    VG56_BPT4   \n",
       "213088      Q05115   AMDA_BORBO   \n",
       "213089      Q94MV8   VG56_BPLZ5   \n",
       "213090      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence EC number  Length  \n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.-     269  \n",
       "1       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.6     269  \n",
       "2       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...  3.1.1.74     248  \n",
       "3       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...  3.2.1.91     471  \n",
       "4       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...  1.2.1.44     333  \n",
       "...                                                   ...       ...     ...  \n",
       "213078  VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...  2.7.7.48     178  \n",
       "213079  MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...  3.6.1.12     171  \n",
       "213088  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...  4.1.1.76     240  \n",
       "213089  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...  3.6.1.12     172  \n",
       "213090  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...  3.1.1.29     124  \n",
       "\n",
       "[208984 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5250f2c8",
   "metadata": {},
   "source": [
    "### Perform clustering using mmseqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "19acec52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate a fast file as input\n",
    "with open('data/swissprot/swissprot.fasta', 'w') as f:\n",
    "    for index, row in df.iterrows():\n",
    "        f.write(f'>{row[\"Entry\"]}\\n{row[\"Sequence\"]}\\n')\n",
    "\n",
    "# !mmseqs createdb examples/swissprot.fasta swissprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5e6cdd85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Create directory data/swissprot/mmseqs_tmp\n",
      "easy-cluster data/swissprot/swissprot.fasta data/swissprot/clusterRes50 data/swissprot/mmseqs_tmp --min-seq-id 0.5 -c 0.8 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.5\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb data/swissprot/swissprot.fasta data/swissprot/mmseqs_tmp/8532145161567998619/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[208970] 0s 65ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 66ms\n",
      "Create directory data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp\n",
      "cluster data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp --max-seqs 20 -c 0.8 --cov-mode 1 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.5 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 3.000000\n",
      "Set cluster mode GREEDY MEM\n",
      "Set cluster iterations to 3\n",
      "linclust data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 208984 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 100.00% 208.98K 1s 874ms    \n",
      "Sort kmer 0h 0m 0s 329ms\n",
      "Sort by rep. sequence 0h 0m 0s 274ms\n",
      "Time for fill: 0h 0m 0s 70ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 3s 477ms\n",
      "rescorediagonal data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 208.98K 0s 423ms    \n",
      "Time for merging to pref_rescore1: 0h 0m 0s 59ms============>     ] 91.01% 190.20K eta 0s       \n",
      "Time for processing: 0h 0m 1s 866ms\n",
      "clust data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref_rescore1 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pre_clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 969ms\n",
      "\n",
      "Size of the sequence database: 208984\n",
      "Size of the alignment database: 208984\n",
      "Number of clusters: 58426\n",
      "\n",
      "Writing results 0h 0m 0s 32ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 1s 54ms\n",
      "createsubdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/order_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 23ms\n",
      "createsubdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/order_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 32ms\n",
      "filterdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref_filter1 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref_filter2 --filter-file data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/order_redundancy --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 100.00% 58.43K 0s 839ms    \n",
      "Time for merging to pref_filter2: 0h 0m 0s 25ms\n",
      "Time for processing: 0h 0m 1s 414ms\n",
      "rescorediagonal data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/input_step_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/input_step_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref_filter2 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 58.43K 0s 142ms    \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 28ms                  ] 52.16% 30.48K eta 0s       \n",
      "Time for processing: 0h 0m 1s 386ms\n",
      "align data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/input_step_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/input_step_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref_rescore2 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 58426 type: Aminoacid\n",
      "Target database size: 58426 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 58.43K 2s 484ms    \n",
      "Time for merging to aln: 0h 0m 0s 33ms\n",
      "119492 alignments calculated\n",
      "90483 sequence pairs passed the thresholds (0.757231 of overall calculated)\n",
      "1.548677 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 907ms\n",
      "clust data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/input_step_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/aln data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 585ms\n",
      "\n",
      "Size of the sequence database: 58426\n",
      "Size of the alignment database: 58426\n",
      "Number of clusters: 41795\n",
      "\n",
      "Writing results 0h 0m 0s 19ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 622ms\n",
      "mergeclusters data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pre_clust data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/clust --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 58.43K 0s 718ms    \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 41.79K 0s 926ms      ] 68.07% 28.45K eta 0s       \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 208.98K 1s 412ms    K eta 2s       \n",
      "Time for merging to clu_redundancy: 0h 0m 0s 24ms\n",
      "Time for processing: 0h 0m 1s 532ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/linclust/14965783017100437856/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "prefilter data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 41795 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 41795 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 41.79K 1s 342ms    \n",
      "Index table: Masked residues: 47641\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 41.79K 0s 453ms    \n",
      "Index statistics\n",
      "Entries:          5815575\n",
      "DB size:          521 MB\n",
      "Avg k-mer size:   0.090868\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t163\n",
      "    AYPWGT\t121\n",
      "    GRGHCS\t97\n",
      "    GPGGTQ\t91\n",
      "    SPGGGW\t88\n",
      "    HGGDFP\t79\n",
      "    ITVGGP\t76\n",
      "    PHEEQS\t74\n",
      "    DPAGIQ\t69\n",
      "    PARFLR\t66\n",
      "Time for index table init: 0h 0m 3s 549ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 41795\n",
      "Target db start 1 to 41795\n",
      "[=================================================================] 100.00% 41.79K 1s 432ms    \n",
      "\n",
      "1.708625 k-mers per position\n",
      "315 DB matches per sequence\n",
      "0 overflows\n",
      "11 sequences passed prefiltering per query sequence\n",
      "10 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 22ms\n",
      "Time for processing: 0h 0m 7s 489ms\n",
      "align data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/pref_step0 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 41795 type: Aminoacid\n",
      "Target database size: 41795 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 41.79K 4s 282ms    \n",
      "Time for merging to aln_step0: 0h 0m 0s 22ms\n",
      "452487 alignments calculated\n",
      "118024 sequence pairs passed the thresholds (0.260834 of overall calculated)\n",
      "2.823879 hits per query sequence\n",
      "Time for processing: 0h 0m 4s 750ms\n",
      "clust data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/aln_step0 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_step0 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 911ms\n",
      "\n",
      "Size of the sequence database: 41795\n",
      "Size of the alignment database: 41795\n",
      "Number of clusters: 28575\n",
      "\n",
      "Writing results 0h 0m 0s 5ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 939ms\n",
      "createsubdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_step0 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 10ms\n",
      "prefilter data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step1 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step1 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 2 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 28575 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 28575 type: Aminoacid\n",
      "Index table k-mer threshold: 145 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 28.58K 1s 294ms    \n",
      "Index table: Masked residues: 39213\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 28.58K 0s 267ms    K eta 0s       \n",
      "Index statistics\n",
      "Entries:          6233577\n",
      "DB size:          523 MB\n",
      "Avg k-mer size:   0.097400\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t93\n",
      "    GQQVAR\t81\n",
      "    AYPWGT\t54\n",
      "    GSELFN\t53\n",
      "    PTGLGH\t50\n",
      "    FNLPNP\t50\n",
      "    FPHNAQ\t49\n",
      "    FREPGR\t49\n",
      "    GRGHCS\t48\n",
      "    TGNKTT\t44\n",
      "Time for index table init: 0h 0m 3s 139ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 145\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 28575\n",
      "Target db start 1 to 28575\n",
      "[=================================================================] 100.00% 28.58K 1s 371ms    \n",
      "\n",
      "7.382307 k-mers per position\n",
      "503 DB matches per sequence\n",
      "0 overflows\n",
      "11 sequences passed prefiltering per query sequence\n",
      "10 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 21ms\n",
      "Time for processing: 0h 0m 8s 242ms\n",
      "align data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step1 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step1 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/pref_step1 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 28575 type: Aminoacid\n",
      "Target database size: 28575 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 28.58K 3s 510ms    \n",
      "Time for merging to aln_step1: 0h 0m 0s 18ms\n",
      "296767 alignments calculated\n",
      "30799 sequence pairs passed the thresholds (0.103782 of overall calculated)\n",
      "1.077830 hits per query sequence\n",
      "Time for processing: 0h 0m 4s 27ms\n",
      "clust data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step1 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/aln_step1 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_step1 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 622ms\n",
      "\n",
      "Size of the sequence database: 28575\n",
      "Size of the alignment database: 28575\n",
      "Number of clusters: 27520\n",
      "\n",
      "Writing results 0h 0m 0s 4ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 646ms\n",
      "createsubdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_step1 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step1 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "prefilter data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step2 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step2 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 3 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 27520 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 27520 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 27.52K 1s 103ms    K eta 0s       \n",
      "Index table: Masked residues: 38119\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 27.52K 0s 347ms     36.85% 10.14K eta 0s       \n",
      "Index statistics\n",
      "Entries:          8139435\n",
      "DB size:          534 MB\n",
      "Avg k-mer size:   0.127179\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t90\n",
      "    GQQVAR\t73\n",
      "    GGGHPA\t53\n",
      "    STYLGG\t53\n",
      "    GSELFN\t53\n",
      "    GGSEVS\t51\n",
      "    AYPWGT\t49\n",
      "    RDKELL\t45\n",
      "    TGNKTT\t44\n",
      "    SPGGGW\t44\n",
      "Time for index table init: 0h 0m 3s 7ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 27520\n",
      "Target db start 1 to 27520\n",
      "[=================================================================] 100.00% 27.52K 1s 664ms    \n",
      "\n",
      "22.239691 k-mers per position\n",
      "1175 DB matches per sequence\n",
      "0 overflows\n",
      "14 sequences passed prefiltering per query sequence\n",
      "17 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 20ms\n",
      "Time for processing: 0h 0m 7s 988ms\n",
      "align data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step2 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step2 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/pref_step2 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 27520 type: Aminoacid\n",
      "Target database size: 27520 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 27.52K 3s 819ms    \n",
      "Time for merging to aln_step2: 0h 0m 0s 21ms\n",
      "357859 alignments calculated\n",
      "27696 sequence pairs passed the thresholds (0.077394 of overall calculated)\n",
      "1.006395 hits per query sequence\n",
      "Time for processing: 0h 0m 4s 238ms\n",
      "clust data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step2 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/aln_step2 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_step2 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 667ms\n",
      "\n",
      "Size of the sequence database: 27520\n",
      "Size of the alignment database: 27520\n",
      "Number of clusters: 27415\n",
      "\n",
      "Writing results 0h 0m 0s 11ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 690ms\n",
      "mergeclusters data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_redundancy data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_step0 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_step1 data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_step2 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 41.79K 0s 809ms    \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 28.58K 1s 32ms     \n",
      "Clustering step 3\n",
      "[=================================================================] 100.00% 27.52K 1s 229ms                                                         ] 4.31% 1.19K eta 26s       \n",
      "Clustering step 4\n",
      "[=================================================================] 100.00% 27.42K 1s 406ms    \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 208.98K 1s 853ms     \n",
      "Time for merging to clu: 0h 0m 0s 20ms\n",
      "Time for processing: 0h 0m 1s 951ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/pref_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/aln_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/pref_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/aln_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/clu_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step1_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_tmp/1428806242270519819/input_step2_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu data/swissprot/mmseqs_tmp/8532145161567998619/cluster.tsv --threads 384 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 33ms\n",
      "Time for processing: 0h 0m 0s 692ms\n",
      "result2repseq data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu data/swissprot/mmseqs_tmp/8532145161567998619/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 27.42K 0s 209ms    \n",
      "Time for merging to clu_rep: 0h 0m 0s 30ms\n",
      "Time for processing: 0h 0m 0s 647ms\n",
      "result2flat data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu_rep data/swissprot/mmseqs_tmp/8532145161567998619/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 90ms\n",
      "createseqfiledb data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu data/swissprot/mmseqs_tmp/8532145161567998619/clu_seqs --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 27.42K 1s 39ms     K eta 0s       \n",
      "Time for merging to clu_seqs: 0h 0m 0s 22ms\n",
      "Time for processing: 0h 0m 1s 558ms\n",
      "result2flat data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/input data/swissprot/mmseqs_tmp/8532145161567998619/clu_seqs data/swissprot/mmseqs_tmp/8532145161567998619/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 306ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb data/swissprot/mmseqs_tmp/8532145161567998619/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster data/swissprot/swissprot.fasta data/swissprot/clusterRes50 data/swissprot/mmseqs_tmp --min-seq-id 0.5 -c 0.8 --cov-mode 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "286e6959",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster data/swissprot/swissprot.fasta data/swissprot/clusterRes30 data/swissprot/mmseqs_tmp --min-seq-id 0.3 -c 0.8 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.3\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb data/swissprot/swissprot.fasta data/swissprot/mmseqs_tmp/14689927414933344151/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[208970] 0s 64ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 65ms\n",
      "Create directory data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp\n",
      "cluster data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp --max-seqs 20 -c 0.8 --cov-mode 1 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.3 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 5.000000\n",
      "Set cluster mode GREEDY MEM\n",
      "Set cluster iterations to 3\n",
      "linclust data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 208984 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 100.00% 208.98K 1s 734ms    .74% 106.05K eta 1s       \n",
      "Sort kmer 0h 0m 0s 356ms\n",
      "Sort by rep. sequence 0h 0m 0s 271ms\n",
      "Time for fill: 0h 0m 0s 55ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 3s 367ms\n",
      "rescorediagonal data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 208.98K 0s 401ms    \n",
      "Time for merging to pref_rescore1: 0h 0m 0s 73ms==============>   ] 95.32% 199.21K eta 0s       \n",
      "Time for processing: 0h 0m 1s 629ms\n",
      "clust data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref_rescore1 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pre_clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 922ms\n",
      "\n",
      "Size of the sequence database: 208984\n",
      "Size of the alignment database: 208984\n",
      "Number of clusters: 58426\n",
      "\n",
      "Writing results 0h 0m 0s 16ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 983ms\n",
      "createsubdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/order_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 24ms\n",
      "createsubdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/order_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 35ms\n",
      "filterdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref_filter1 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref_filter2 --filter-file data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/order_redundancy --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 100.00% 58.43K 0s 882ms    \n",
      "Time for merging to pref_filter2: 0h 0m 0s 23ms\n",
      "Time for processing: 0h 0m 1s 433ms\n",
      "rescorediagonal data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/input_step_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/input_step_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref_filter2 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 58.43K 0s 151ms    \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 24ms                  ] 8.99% 5.25K eta 6s         \n",
      "Time for processing: 0h 0m 1s 468ms\n",
      "align data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/input_step_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/input_step_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref_rescore2 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 58426 type: Aminoacid\n",
      "Target database size: 58426 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 58.43K 2s 703ms    \n",
      "Time for merging to aln: 0h 0m 0s 30ms\n",
      "125279 alignments calculated\n",
      "120511 sequence pairs passed the thresholds (0.961941 of overall calculated)\n",
      "2.062626 hits per query sequence\n",
      "Time for processing: 0h 0m 3s 144ms\n",
      "clust data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/input_step_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/aln data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 909ms\n",
      "\n",
      "Size of the sequence database: 58426\n",
      "Size of the alignment database: 58426\n",
      "Number of clusters: 32714\n",
      "\n",
      "Writing results 0h 0m 0s 16ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 946ms\n",
      "mergeclusters data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pre_clust data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/clust --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 58.43K 0s 434ms    eta 0s       \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 32.71K 0s 680ms    ======>                                               ] 27.04% 8.85K eta 1s       \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 208.98K 1s 203ms                                      ] 48.40% 101.16K eta 1s       \n",
      "Time for merging to clu_redundancy: 0h 0m 0s 32ms\n",
      "Time for processing: 0h 0m 1s 315ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/linclust/1981412250453875402/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 21ms\n",
      "prefilter data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 32714 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 32714 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 32.71K 1s 196ms    \n",
      "Index table: Masked residues: 40035\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 32.71K 0s 440ms      ] 60.25% 19.71K eta 0s       \n",
      "Index statistics\n",
      "Entries:          4615880\n",
      "DB size:          514 MB\n",
      "Avg k-mer size:   0.072123\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t122\n",
      "    AYPWGT\t74\n",
      "    IHDKNI\t68\n",
      "    GPGGTQ\t63\n",
      "    FREPGR\t50\n",
      "    TFPHHL\t48\n",
      "    RKYYLP\t48\n",
      "    GRGHCS\t48\n",
      "    EARPGP\t43\n",
      "    PVGVMT\t43\n",
      "Time for index table init: 0h 0m 3s 186ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 32714\n",
      "Target db start 1 to 32714\n",
      "[=================================================================] 100.00% 32.71K 1s 242ms    \n",
      "\n",
      "1.779371 k-mers per position\n",
      "253 DB matches per sequence\n",
      "0 overflows\n",
      "8 sequences passed prefiltering per query sequence\n",
      "6 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 19ms\n",
      "Time for processing: 0h 0m 8s 143ms\n",
      "align data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/pref_step0 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 32714 type: Aminoacid\n",
      "Target database size: 32714 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 32.71K 4s 294ms    \n",
      "Time for merging to aln_step0: 0h 0m 0s 25ms\n",
      "273591 alignments calculated\n",
      "235333 sequence pairs passed the thresholds (0.860164 of overall calculated)\n",
      "7.193648 hits per query sequence\n",
      "Time for processing: 0h 0m 4s 684ms\n",
      "clust data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/aln_step0 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_step0 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 750ms\n",
      "\n",
      "Size of the sequence database: 32714\n",
      "Size of the alignment database: 32714\n",
      "Number of clusters: 14440\n",
      "\n",
      "Writing results 0h 0m 0s 5ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 776ms\n",
      "createsubdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_step0 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "prefilter data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step1 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step1 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 3 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 14440 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 14440 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 14.44K 0s 985ms    \n",
      "Index table: Masked residues: 24995\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 14.44K 0s 238ms    \n",
      "Index statistics\n",
      "Entries:          4192495\n",
      "DB size:          512 MB\n",
      "Avg k-mer size:   0.065508\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t43\n",
      "    GPGGTL\t37\n",
      "    GLGIGK\t24\n",
      "    RDKELL\t23\n",
      "    AVGGGL\t21\n",
      "    VVGGGL\t21\n",
      "    GQQVAR\t20\n",
      "    HRLPIL\t15\n",
      "    VHDKNL\t15\n",
      "    RLFGVL\t15\n",
      "Time for index table init: 0h 0m 3s 115ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 14440\n",
      "Target db start 1 to 14440\n",
      "[=================================================================] 100.00% 14.44K 1s 31ms     \n",
      "\n",
      "22.510530 k-mers per position\n",
      "672 DB matches per sequence\n",
      "0 overflows\n",
      "7 sequences passed prefiltering per query sequence\n",
      "6 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 15ms\n",
      "Time for processing: 0h 0m 8s 359ms\n",
      "align data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step1 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step1 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/pref_step1 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 14440 type: Aminoacid\n",
      "Target database size: 14440 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 14.44K 2s 295ms    \n",
      "Time for merging to aln_step1: 0h 0m 0s 22ms\n",
      "91185 alignments calculated\n",
      "39907 sequence pairs passed the thresholds (0.437649 of overall calculated)\n",
      "2.763643 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 734ms\n",
      "clust data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step1 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/aln_step1 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_step1 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 859ms\n",
      "\n",
      "Size of the sequence database: 14440\n",
      "Size of the alignment database: 14440\n",
      "Number of clusters: 9653\n",
      "\n",
      "Writing results 0h 0m 0s 11ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 882ms\n",
      "createsubdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_step1 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step1 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "prefilter data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step2 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step2 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 5 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 9653 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 9653 type: Aminoacid\n",
      "Index table k-mer threshold: 118 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 9.65K 1s 66ms     \n",
      "Index table: Masked residues: 21786\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 9.65K 0s 171ms    \n",
      "Index statistics\n",
      "Entries:          3077606\n",
      "DB size:          505 MB\n",
      "Avg k-mer size:   0.048088\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t42\n",
      "    VHDKNI\t22\n",
      "    RDKELL\t22\n",
      "    GPGGTL\t18\n",
      "    AVGGGL\t14\n",
      "    HRLPIL\t14\n",
      "    VHDKNL\t14\n",
      "    QCGGGP\t13\n",
      "    RVGLSR\t13\n",
      "    LGGFVY\t13\n",
      "Time for index table init: 0h 0m 2s 942ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 118\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 9653\n",
      "Target db start 1 to 9653\n",
      "[=================================================================] 100.00% 9.65K 1s 591ms    \n",
      "\n",
      "178.818277 k-mers per position\n",
      "2885 DB matches per sequence\n",
      "0 overflows\n",
      "17 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 15ms\n",
      "Time for processing: 0h 0m 7s 499ms\n",
      "align data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step2 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step2 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/pref_step2 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 9653 type: Aminoacid\n",
      "Target database size: 9653 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 9.65K 2s 389ms    \n",
      "Time for merging to aln_step2: 0h 0m 0s 23ms\n",
      "121652 alignments calculated\n",
      "12985 sequence pairs passed the thresholds (0.106739 of overall calculated)\n",
      "1.345178 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 799ms\n",
      "clust data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step2 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/aln_step2 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_step2 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 793ms\n",
      "\n",
      "Size of the sequence database: 9653\n",
      "Size of the alignment database: 9653\n",
      "Number of clusters: 8444\n",
      "\n",
      "Writing results 0h 0m 0s 6ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 810ms\n",
      "mergeclusters data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_redundancy data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_step0 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_step1 data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_step2 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 32.71K 0s 700ms      ] 72.67% 23.77K eta 0s       \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 14.44K 0s 931ms                                       ] 12.01% 1.74K eta 6s       \n",
      "Clustering step 3\n",
      "[=================================================================] 100.00% 9.65K 1s 117ms     \n",
      "Clustering step 4\n",
      "[=================================================================] 100.00% 8.44K 1s 326ms    \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 208.98K 1s 804ms     ====================>                                           ] 32.39% 67.68K eta 3s       \n",
      "Time for merging to clu: 0h 0m 0s 14ms\n",
      "Time for processing: 0h 0m 1s 912ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/pref_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/aln_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/pref_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/aln_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/clu_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step1_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_tmp/13807509212710083353/input_step2_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu data/swissprot/mmseqs_tmp/14689927414933344151/cluster.tsv --threads 384 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 23ms\n",
      "Time for processing: 0h 0m 0s 761ms\n",
      "result2repseq data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu data/swissprot/mmseqs_tmp/14689927414933344151/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 8.44K 0s 195ms    \n",
      "Time for merging to clu_rep: 0h 0m 0s 19ms\n",
      "Time for processing: 0h 0m 0s 649ms\n",
      "result2flat data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu_rep data/swissprot/mmseqs_tmp/14689927414933344151/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 58ms\n",
      "createseqfiledb data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu data/swissprot/mmseqs_tmp/14689927414933344151/clu_seqs --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 8.44K 1s 35ms     \n",
      "Time for merging to clu_seqs: 0h 0m 0s 64ms\n",
      "Time for processing: 0h 0m 1s 611ms\n",
      "result2flat data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/input data/swissprot/mmseqs_tmp/14689927414933344151/clu_seqs data/swissprot/mmseqs_tmp/14689927414933344151/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 287ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb data/swissprot/mmseqs_tmp/14689927414933344151/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster data/swissprot/swissprot.fasta data/swissprot/clusterRes30 data/swissprot/mmseqs_tmp --min-seq-id 0.3 -c 0.8 --cov-mode 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "02322b64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27415\n",
      "8444\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "      <th>clusterRes50</th>\n",
       "      <th>clusterRes30</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-</td>\n",
       "      <td>269</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A1AY86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A1AY86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>A0A024SC78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "      <td>G4MM92</td>\n",
       "      <td>G4MM92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "      <td>Q9S9N9</td>\n",
       "      <td>P14721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208979</th>\n",
       "      <td>P36352</td>\n",
       "      <td>POLR_PHMV</td>\n",
       "      <td>VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>178</td>\n",
       "      <td>P36352</td>\n",
       "      <td>P36352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208980</th>\n",
       "      <td>P39262</td>\n",
       "      <td>VG56_BPT4</td>\n",
       "      <td>MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>171</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208981</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208982</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208983</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q54ZD0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>208984 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "1       A0A009IHW8  ABTIR_ACIB9   \n",
       "2       A0A024SC78  CUTI1_HYPJR   \n",
       "3       A0A024SH76   GUX2_HYPJR   \n",
       "4       A0A059TC02   CCR1_PETHY   \n",
       "...            ...          ...   \n",
       "208979      P36352    POLR_PHMV   \n",
       "208980      P39262    VG56_BPT4   \n",
       "208981      Q05115   AMDA_BORBO   \n",
       "208982      Q94MV8   VG56_BPLZ5   \n",
       "208983      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence EC number  Length  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.-     269   \n",
       "1       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.6     269   \n",
       "2       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...  3.1.1.74     248   \n",
       "3       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...  3.2.1.91     471   \n",
       "4       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...  1.2.1.44     333   \n",
       "...                                                   ...       ...     ...   \n",
       "208979  VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...  2.7.7.48     178   \n",
       "208980  MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...  3.6.1.12     171   \n",
       "208981  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...  4.1.1.76     240   \n",
       "208982  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...  3.6.1.12     172   \n",
       "208983  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...  3.1.1.29     124   \n",
       "\n",
       "       clusterRes50 clusterRes30  \n",
       "0        A0A009IHW8       A1AY86  \n",
       "1        A0A009IHW8       A1AY86  \n",
       "2        A0A024SC78   A0A024SC78  \n",
       "3            G4MM92       G4MM92  \n",
       "4            Q9S9N9       P14721  \n",
       "...             ...          ...  \n",
       "208979       P36352       P36352  \n",
       "208980       Q94MV8       Q94MV8  \n",
       "208981       Q05115       Q05115  \n",
       "208982       Q94MV8       Q94MV8  \n",
       "208983       Q9J5H2       Q54ZD0  \n",
       "\n",
       "[208984 rows x 7 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkAAAAGdCAYAAAD60sxaAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuNSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/xnp5ZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAskElEQVR4nO3de3BUZZ7G8ScJ6U4idMLFJGQImFGHi1wNQ2gvLA4hDaYcUcoSZN2giCub7BgyixDFCLJuHBxQVDTrOoBbA17YUhyBDWmDgAwNSCRCUBgvuDgjHRwQwkWSJjn7x1SONAlIoNMZfL+fqlRx3vfX57zn1wSeOt2nO8KyLEsAAAAGiWzrBQAAAIQbAQgAABiHAAQAAIxDAAIAAMYhAAEAAOMQgAAAgHEIQAAAwDgEIAAAYJx2bb2AttTQ0KCvv/5aHTp0UERERFsvBwAAnAfLsnT06FGlpKQoMvLCruUYHYC+/vprpaamtvUyAADABfjqq6/UrVu3C3qs0QGoQ4cOkv7WQJfLFbL9BgIBlZWVKSsrS9HR0SHbL5pHv8OPnocfPQ8v+h1+Lel5TU2NUlNT7f/HL0SLAlBxcbHefPNN7d69W7Gxsbruuuv0m9/8Rj179rRrhg8frvXr1wc97p//+Z9VUlJib+/bt09TpkzRe++9p/bt2ysnJ0fFxcVq1+775axbt04FBQXatWuXUlNTNXPmTE2cODFovwsXLtRTTz0lv9+vAQMG6LnnntOQIUPO+3waX/ZyuVwhD0BxcXFyuVz84oQB/Q4/eh5+9Dy86Hf4XUjPL+btKy164Wz9+vXKzc3V5s2b5fV6FQgElJWVpePHjwfVTZ48Wfv377d/5s6da8/V19crOztbdXV12rRpk1555RUtWbJERUVFds3evXuVnZ2tm266SZWVlcrPz9d9992nNWvW2DWvv/66CgoK9Nhjj+nDDz/UgAED5PF4dODAgQvtBQAAMESLrgCVlpYGbS9ZskSJiYmqqKjQsGHD7PG4uDglJyc3u4+ysjJ9/PHHevfdd5WUlKSBAwdqzpw5mj59umbNmiWHw6GSkhKlpaVp3rx5kqTevXtr48aNevrpp+XxeCRJ8+fP1+TJk3XPPfdIkkpKSrRq1SotWrRIM2bMaMlpAQAAw1zUe4COHDkiSerUqVPQ+NKlS/X73/9eycnJuuWWW/Too48qLi5OkuTz+dSvXz8lJSXZ9R6PR1OmTNGuXbs0aNAg+Xw+ZWZmBu3T4/EoPz9fklRXV6eKigoVFhba85GRkcrMzJTP5zvremtra1VbW2tv19TUSPrbZbdAIHABHWhe475CuU+cHf0OP3oefvQ8vOh3+LWk56F4Xi44ADU0NCg/P1/XX3+9+vbta4/fdddd6tGjh1JSUrRjxw5Nnz5de/bs0ZtvvilJ8vv9QeFHkr3t9/vPWVNTU6PvvvtO3377rerr65ut2b1791nXXFxcrNmzZzcZLysrswNaKHm93pDvE2dHv8OPnocfPQ8v+h1+59PzEydOXPRxLjgA5ebmqqqqShs3bgwav//+++0/9+vXT127dtWIESP0+eef68orr7zwlYZAYWGhCgoK7O3Gd5FnZWWF/E3QXq9XI0eO5M1zYUC/w4+ehx89Dy/6HX4t6XnjKzgX44ICUF5enlauXKkNGzb84P33GRkZkqTPPvtMV155pZKTk7V169agmurqakmy3zeUnJxsj51e43K5FBsbq6ioKEVFRTVbc7b3HkmS0+mU0+lsMh4dHd0qf8Fba79oHv0OP3oefvQ8vOh3+J1Pz0PxnLToLjDLspSXl6e33npLa9euVVpa2g8+prKyUpLUtWtXSZLb7dbOnTuD7tbyer1yuVzq06ePXVNeXh60H6/XK7fbLUlyOBxKT08PqmloaFB5ebldAwAAcDYtugKUm5urZcuW6e2331aHDh3s9+zEx8crNjZWn3/+uZYtW6abb75ZnTt31o4dOzR16lQNGzZM/fv3lyRlZWWpT58+uvvuuzV37lz5/X7NnDlTubm59tWZBx54QM8//7weeugh3XvvvVq7dq3eeOMNrVq1yl5LQUGBcnJyNHjwYA0ZMkTPPPOMjh8/bt8VBgAAcDYtCkAvvviipL992OHpFi9erIkTJ8rhcOjdd9+1w0hqaqrGjh2rmTNn2rVRUVFauXKlpkyZIrfbrcsuu0w5OTl6/PHH7Zq0tDStWrVKU6dO1YIFC9StWze9/PLL9i3wknTnnXfqm2++UVFRkfx+vwYOHKjS0tImb4wGAAA4U4sCkGVZ55xPTU1t8inQzenRo4dWr159zprhw4dr+/bt56zJy8tTXl7eDx4PAADgdBf2FaoAAACXMAIQAAAwDgEIAAAYhwAEAACMc1HfBYZz6ztrjWrrI+ztL5/MbsPVAACARlwBAgAAxiEAAQAA4xCAAACAcQhAAADAOAQgAABgHAIQAAAwDgEIAAAYhwAEAACMQwACAADGIQABAADjEIAAAIBxCEAAAMA4BCAAAGAcAhAAADAOAQgAABiHAAQAAIxDAAIAAMYhAAEAAOMQgAAAgHEIQAAAwDgEIAAAYBwCEAAAMA4BCAAAGIcABAAAjEMAAgAAxiEAAQAA4xCAAACAcQhAAADAOAQgAABgHAIQAAAwDgEIAAAYhwAEAACMQwACAADGIQABAADjEIAAAIBxCEAAAMA4BCAAAGAcAhAAADAOAQgAABiHAAQAAIxDAAIAAMYhAAEAAOMQgAAAgHEIQAAAwDgEIAAAYBwCEAAAMA4BCAAAGIcABAAAjEMAAgAAxiEAAQAA4xCAAACAcQhAAADAOAQgAABgHAIQAAAwDgEIAAAYhwAEAACMQwACAADGaVEAKi4u1s9//nN16NBBiYmJGjNmjPbs2RNUc/LkSeXm5qpz585q3769xo4dq+rq6qCaffv2KTs7W3FxcUpMTNS0adN06tSpoJp169bp2muvldPp1FVXXaUlS5Y0Wc/ChQt1xRVXKCYmRhkZGdq6dWtLTgcAABiqRQFo/fr1ys3N1ebNm+X1ehUIBJSVlaXjx4/bNVOnTtU777yj5cuXa/369fr66691++232/P19fXKzs5WXV2dNm3apFdeeUVLlixRUVGRXbN3715lZ2frpptuUmVlpfLz83XfffdpzZo1ds3rr7+ugoICPfbYY/rwww81YMAAeTweHThw4GL6AQAADNCuJcWlpaVB20uWLFFiYqIqKio0bNgwHTlyRL/73e+0bNky/eIXv5AkLV68WL1799bmzZs1dOhQlZWV6eOPP9a7776rpKQkDRw4UHPmzNH06dM1a9YsORwOlZSUKC0tTfPmzZMk9e7dWxs3btTTTz8tj8cjSZo/f74mT56se+65R5JUUlKiVatWadGiRZoxY8ZFNwYAAPx4tSgAnenIkSOSpE6dOkmSKioqFAgElJmZadf06tVL3bt3l8/n09ChQ+Xz+dSvXz8lJSXZNR6PR1OmTNGuXbs0aNAg+Xy+oH001uTn50uS6urqVFFRocLCQns+MjJSmZmZ8vl8Z11vbW2tamtr7e2amhpJUiAQUCAQuMAuNNW4L2ek1ew4Qquxr/Q3fOh5+NHz8KLf4deSnofiebngANTQ0KD8/Hxdf/316tu3ryTJ7/fL4XAoISEhqDYpKUl+v9+uOT38NM43zp2rpqamRt99952+/fZb1dfXN1uze/fus665uLhYs2fPbjJeVlamuLi48zjrlpkzuCFoe/Xq1SE/Br7n9XrbegnGoefhR8/Di36H3/n0/MSJExd9nAsOQLm5uaqqqtLGjRsvehHhUlhYqIKCAnu7pqZGqampysrKksvlCtlxAoGAvF6vHt0WqdqGCHu8apYnZMfA9xr7PXLkSEVHR7f1coxAz8OPnocX/Q6/lvS88RWci3FBASgvL08rV67Uhg0b1K1bN3s8OTlZdXV1Onz4cNBVoOrqaiUnJ9s1Z96t1XiX2Ok1Z945Vl1dLZfLpdjYWEVFRSkqKqrZmsZ9NMfpdMrpdDYZj46ObpW/4LUNEaqt/z4A8UvUulrrecTZ0fPwo+fhRb/D73x6HornpEV3gVmWpby8PL311ltau3at0tLSgubT09MVHR2t8vJye2zPnj3at2+f3G63JMntdmvnzp1Bd2t5vV65XC716dPHrjl9H401jftwOBxKT08PqmloaFB5ebldAwAAcDYtugKUm5urZcuW6e2331aHDh3s9+zEx8crNjZW8fHxmjRpkgoKCtSpUye5XC7967/+q9xut4YOHSpJysrKUp8+fXT33Xdr7ty58vv9mjlzpnJzc+2rMw888ICef/55PfTQQ7r33nu1du1avfHGG1q1apW9loKCAuXk5Gjw4MEaMmSInnnmGR0/fty+KwwAAOBsWhSAXnzxRUnS8OHDg8YXL16siRMnSpKefvppRUZGauzYsaqtrZXH49ELL7xg10ZFRWnlypWaMmWK3G63LrvsMuXk5Ojxxx+3a9LS0rRq1SpNnTpVCxYsULdu3fTyyy/bt8BL0p133qlvvvlGRUVF8vv9GjhwoEpLS5u8MRoAAOBMLQpAlmX9YE1MTIwWLlyohQsXnrWmR48eP3hH1PDhw7V9+/Zz1uTl5SkvL+8H1wQAAHA6vgsMAAAYhwAEAACMQwACAADGIQABAADjEIAAAIBxCEAAAMA4BCAAAGAcAhAAADAOAQgAABiHAAQAAIxDAAIAAMYhAAEAAOMQgAAAgHEIQAAAwDgEIAAAYBwCEAAAMA4BCAAAGIcABAAAjEMAAgAAxiEAAQAA4xCAAACAcQhAAADAOAQgAABgHAIQAAAwDgEIAAAYhwAEAACMQwACAADGIQABAADjEIAAAIBxCEAAAMA4BCAAAGAcAhAAADAOAQgAABiHAAQAAIxDAAIAAMYhAAEAAOMQgAAAgHEIQAAAwDgEIAAAYBwCEAAAMA4BCAAAGIcABAAAjEMAAgAAxiEAAQAA4xCAAACAcQhAAADAOAQgAABgHAIQAAAwDgEIAAAYhwAEAACMQwACAADGIQABAADjEIAAAIBxCEAAAMA4BCAAAGAcAhAAADAOAQgAABiHAAQAAIxDAAIAAMYhAAEAAOMQgAAAgHFaHIA2bNigW265RSkpKYqIiNCKFSuC5idOnKiIiIign1GjRgXVHDp0SBMmTJDL5VJCQoImTZqkY8eOBdXs2LFDN954o2JiYpSamqq5c+c2Wcvy5cvVq1cvxcTEqF+/flq9enVLTwcAABioxQHo+PHjGjBggBYuXHjWmlGjRmn//v32z6uvvho0P2HCBO3atUter1crV67Uhg0bdP/999vzNTU1ysrKUo8ePVRRUaGnnnpKs2bN0ksvvWTXbNq0SePHj9ekSZO0fft2jRkzRmPGjFFVVVVLTwkAABimXUsfMHr0aI0ePfqcNU6nU8nJyc3OffLJJyotLdUHH3ygwYMHS5Kee+453Xzzzfrtb3+rlJQULV26VHV1dVq0aJEcDoeuueYaVVZWav78+XZQWrBggUaNGqVp06ZJkubMmSOv16vnn39eJSUlLT0tAABgkFZ5D9C6deuUmJionj17asqUKTp48KA95/P5lJCQYIcfScrMzFRkZKS2bNli1wwbNkwOh8Ou8Xg82rNnj7799lu7JjMzM+i4Ho9HPp+vNU4JAAD8iLT4CtAPGTVqlG6//XalpaXp888/18MPP6zRo0fL5/MpKipKfr9fiYmJwYto106dOnWS3++XJPn9fqWlpQXVJCUl2XMdO3aU3++3x06vadxHc2pra1VbW2tv19TUSJICgYACgcCFn/QZGvfljLSaHUdoNfaV/oYPPQ8/eh5e9Dv8WtLzUDwvIQ9A48aNs//cr18/9e/fX1deeaXWrVunESNGhPpwLVJcXKzZs2c3GS8rK1NcXFzIjzdncEPQNm/Sbl1er7etl2Aceh5+9Dy86Hf4nU/PT5w4cdHHCXkAOtNPf/pTdenSRZ999plGjBih5ORkHThwIKjm1KlTOnTokP2+oeTkZFVXVwfVNG7/UM3Z3nskSYWFhSooKLC3a2pqlJqaqqysLLlcrgs/yTMEAgF5vV49ui1StQ0R9njVLE/IjoHvNfZ75MiRio6ObuvlGIGehx89Dy/6HX4t6XnjKzgXo9UD0J///GcdPHhQXbt2lSS53W4dPnxYFRUVSk9PlyStXbtWDQ0NysjIsGseeeQRBQIBuwler1c9e/ZUx44d7Zry8nLl5+fbx/J6vXK73Wddi9PplNPpbDIeHR3dKn/BaxsiVFv/fQDil6h1tdbziLOj5+FHz8OLfoff+fQ8FM9Ji98EfezYMVVWVqqyslKStHfvXlVWVmrfvn06duyYpk2bps2bN+vLL79UeXm5br31Vl111VXyeP529aN3794aNWqUJk+erK1bt+qPf/yj8vLyNG7cOKWkpEiS7rrrLjkcDk2aNEm7du3S66+/rgULFgRdvXnwwQdVWlqqefPmaffu3Zo1a5a2bdumvLy8i24KAAD4cWtxANq2bZsGDRqkQYMGSZIKCgo0aNAgFRUVKSoqSjt27NAvf/lL/exnP9OkSZOUnp6u999/P+jKy9KlS9WrVy+NGDFCN998s2644Yagz/iJj49XWVmZ9u7dq/T0dP36179WUVFR0GcFXXfddVq2bJleeuklDRgwQP/zP/+jFStWqG/fvhfTDwAAYIAWvwQ2fPhwWZZ11vk1a9b84D46deqkZcuWnbOmf//+ev/9989Zc8cdd+iOO+74weMBAACcju8CAwAAxiEAAQAA4xCAAACAcQhAAADAOAQgAABgHAIQAAAwDgEIAAAYhwAEAACMQwACAADGIQABAADjEIAAAIBxCEAAAMA4BCAAAGAcAhAAADAOAQgAABiHAAQAAIxDAAIAAMYhAAEAAOMQgAAAgHEIQAAAwDgEIAAAYBwCEAAAMA4BCAAAGIcABAAAjEMAAgAAxiEAAQAA4xCAAACAcQhAAADAOAQgAABgHAIQAAAwDgEIAAAYhwAEAACMQwACAADGIQABAADjEIAAAIBxCEAAAMA4BCAAAGAcAhAAADAOAQgAABiHAAQAAIxDAAIAAMYhAAEAAOMQgAAAgHEIQAAAwDgEIAAAYBwCEAAAMA4BCAAAGIcABAAAjEMAAgAAxiEAAQAA4xCAAACAcQhAAADAOAQgAABgHAIQAAAwDgEIAAAYhwAEAACMQwACAADGIQABAADjEIAAAIBxCEAAAMA4BCAAAGAcAhAAADBOiwPQhg0bdMsttyglJUURERFasWJF0LxlWSoqKlLXrl0VGxurzMxMffrpp0E1hw4d0oQJE+RyuZSQkKBJkybp2LFjQTU7duzQjTfeqJiYGKWmpmru3LlN1rJ8+XL16tVLMTEx6tevn1avXt3S0wEAAAZqcQA6fvy4BgwYoIULFzY7P3fuXD377LMqKSnRli1bdNlll8nj8ejkyZN2zYQJE7Rr1y55vV6tXLlSGzZs0P3332/P19TUKCsrSz169FBFRYWeeuopzZo1Sy+99JJds2nTJo0fP16TJk3S9u3bNWbMGI0ZM0ZVVVUtPSUAAGCYdi19wOjRozV69Ohm5yzL0jPPPKOZM2fq1ltvlST993//t5KSkrRixQqNGzdOn3zyiUpLS/XBBx9o8ODBkqTnnntON998s377298qJSVFS5cuVV1dnRYtWiSHw6FrrrlGlZWVmj9/vh2UFixYoFGjRmnatGmSpDlz5sjr9er5559XSUnJBTUDAACYocUB6Fz27t0rv9+vzMxMeyw+Pl4ZGRny+XwaN26cfD6fEhIS7PAjSZmZmYqMjNSWLVt02223yefzadiwYXI4HHaNx+PRb37zG3377bfq2LGjfD6fCgoKgo7v8XiavCR3utraWtXW1trbNTU1kqRAIKBAIHCxp29r3Jcz0mp2HKHV2Ff6Gz70PPzoeXjR7/BrSc9D8byENAD5/X5JUlJSUtB4UlKSPef3+5WYmBi8iHbt1KlTp6CatLS0JvtonOvYsaP8fv85j9Oc4uJizZ49u8l4WVmZ4uLizucUW2TO4Iagbd6j1Lq8Xm9bL8E49Dz86Hl40e/wO5+enzhx4qKPE9IA9PeusLAw6KpRTU2NUlNTlZWVJZfLFbLjBAIBeb1ePbotUrUNEfZ41SxPyI6B7zX2e+TIkYqOjm7r5RiBnocfPQ8v+h1+Lel54ys4FyOkASg5OVmSVF1dra5du9rj1dXVGjhwoF1z4MCBoMedOnVKhw4dsh+fnJys6urqoJrG7R+qaZxvjtPplNPpbDIeHR3dKn/BaxsiVFv/fQDil6h1tdbziLOj5+FHz8OLfoff+fQ8FM9JSD8HKC0tTcnJySovL7fHampqtGXLFrndbkmS2+3W4cOHVVFRYdesXbtWDQ0NysjIsGs2bNgQ9Bqf1+tVz5491bFjR7vm9OM01jQeBwAA4GxaHICOHTumyspKVVZWSvrbG58rKyu1b98+RUREKD8/X//+7/+uP/zhD9q5c6f+6Z/+SSkpKRozZowkqXfv3ho1apQmT56srVu36o9//KPy8vI0btw4paSkSJLuuusuORwOTZo0Sbt27dLrr7+uBQsWBL189eCDD6q0tFTz5s3T7t27NWvWLG3btk15eXkX3xUAAPCj1uKXwLZt26abbrrJ3m4MJTk5OVqyZIkeeughHT9+XPfff78OHz6sG264QaWlpYqJibEfs3TpUuXl5WnEiBGKjIzU2LFj9eyzz9rz8fHxKisrU25urtLT09WlSxcVFRUFfVbQddddp2XLlmnmzJl6+OGHdfXVV2vFihXq27fvBTUCAACYo8UBaPjw4bIs66zzERERevzxx/X444+ftaZTp05atmzZOY/Tv39/vf/+++esueOOO3THHXece8EAAABn4LvAAACAcQhAAADAOAQgAABgHAIQAAAwDgEIAAAYhwAEAACMQwACAADGIQABAADjEIAAAIBxCEAAAMA4BCAAAGAcAhAAADAOAQgAABiHAAQAAIxDAAIAAMYhAAEAAOMQgAAAgHEIQAAAwDgEIAAAYBwCEAAAMA4BCAAAGIcABAAAjEMAAgAAxiEAAQAA4xCAAACAcQhAAADAOAQgAABgHAIQAAAwDgEIAAAYhwAEAACMQwACAADGIQABAADjEIAAAIBxCEAAAMA4BCAAAGAcAhAAADAOAQgAABiHAAQAAIxDAAIAAMYhAAEAAOMQgAAAgHEIQAAAwDgEIAAAYBwCEAAAMA4BCAAAGIcABAAAjNOurRfwY1blnKTohpOnjRxps7UAAIDvcQUIAAAYhwAEAACMQwACAADGIQABAADjEIAAAIBxCEAAAMA4BCAAAGAcAhAAADAOAQgAABiHAAQAAIxDAAIAAMYhAAEAAOMQgAAAgHEIQAAAwDgEIAAAYJyQB6BZs2YpIiIi6KdXr172/MmTJ5Wbm6vOnTurffv2Gjt2rKqrq4P2sW/fPmVnZysuLk6JiYmaNm2aTp06FVSzbt06XXvttXI6nbrqqqu0ZMmSUJ8KAAD4kWqVK0DXXHON9u/fb/9s3LjRnps6dareeecdLV++XOvXr9fXX3+t22+/3Z6vr69Xdna26urqtGnTJr3yyitasmSJioqK7Jq9e/cqOztbN910kyorK5Wfn6/77rtPa9asaY3TAQAAPzLtWmWn7dopOTm5yfiRI0f0u9/9TsuWLdMvfvELSdLixYvVu3dvbd68WUOHDlVZWZk+/vhjvfvuu0pKStLAgQM1Z84cTZ8+XbNmzZLD4VBJSYnS0tI0b948SVLv3r21ceNGPf300/J4PK1xSgAA4EekVQLQp59+qpSUFMXExMjtdqu4uFjdu3dXRUWFAoGAMjMz7dpevXqpe/fu8vl8Gjp0qHw+n/r166ekpCS7xuPxaMqUKdq1a5cGDRokn88XtI/Gmvz8/HOuq7a2VrW1tfZ2TU2NJCkQCCgQCITgzGXvT5ICkTFnToTsGPie3W/6Gzb0PPzoeXjR7/BrSc9D8byEPABlZGRoyZIl6tmzp/bv36/Zs2frxhtvVFVVlfx+vxwOhxISEoIek5SUJL/fL0ny+/1B4adxvnHuXDU1NTX67rvvFBsb2+zaiouLNXv27CbjZWVliouLu6DzPRdvv2eDB1avDvkx8D2v19vWSzAOPQ8/eh5e9Dv8zqfnJ06cuOjjhDwAjR492v5z//79lZGRoR49euiNN944azAJl8LCQhUUFNjbNTU1Sk1NVVZWllwuV8iOEwgE5PV6NXLnrxTdcPK0Bfw5ZMfA9+x+jxyp6Ojotl6OEeh5+NHz8KLf4deSnje+gnMxWuUlsNMlJCToZz/7mT777DONHDlSdXV1Onz4cNBVoOrqavs9Q8nJydq6dWvQPhrvEju95sw7x6qrq+Vyuc4ZspxOp5xOZ5Px6OjoVvkLHt1wMjgA8UvUqlrrecTZ0fPwo+fhRb/D73x6HornpNU/B+jYsWP6/PPP1bVrV6Wnpys6Olrl5eX2/J49e7Rv3z653W5Jktvt1s6dO3XgwAG7xuv1yuVyqU+fPnbN6ftorGncBwAAwLmEPAD927/9m9avX68vv/xSmzZt0m233aaoqCiNHz9e8fHxmjRpkgoKCvTee++poqJC99xzj9xut4YOHSpJysrKUp8+fXT33Xfro48+0po1azRz5kzl5ubaV28eeOABffHFF3rooYe0e/duvfDCC3rjjTc0derUUJ8OAAD4EQr5S2B//vOfNX78eB08eFCXX365brjhBm3evFmXX365JOnpp59WZGSkxo4dq9raWnk8Hr3wwgv246OiorRy5UpNmTJFbrdbl112mXJycvT444/bNWlpaVq1apWmTp2qBQsWqFu3bnr55Ze5BR4AAJyXkAeg11577ZzzMTExWrhwoRYuXHjWmh49emj1D9wxNXz4cG3fvv2C1ggAAMzGd4EBAADjEIAAAIBxCEAAAMA4BCAAAGAcAhAAADAOAQgAABiHAAQAAIxDAAIAAMYhAAEAAOMQgAAAgHEIQAAAwDgEIAAAYBwCEAAAMA4BCAAAGIcABAAAjEMAAgAAxiEAAQAA4xCAAACAcQhAAADAOAQgAABgHAIQAAAwDgEIAAAYhwAEAACMQwACAADGIQABAADjEIAAAIBxCEAAAMA4BCAAAGAcAhAAADAOAQgAABiHAAQAAIxDAAIAAMYhAAEAAOMQgAAAgHEIQAAAwDgEIAAAYBwCEAAAMA4BCAAAGIcABAAAjEMAAgAAxiEAAQAA4xCAAACAcQhAAADAOAQgAABgnHZtvQCTXDFjVdD2l09mt9FKAAAwG1eAAACAcQhAAADAOAQgAABgHAIQAAAwDgEIAAAYhwAEAACMQwACAADGIQABAADjEIAAAIBxCEAAAMA4fBVGGH0Zc9cZI0faZB0AAJiOK0AAAMA4BCAAAGAcXgJrQ2d+O7zEN8QDABAOXAECAADGIQABAADjXPIBaOHChbriiisUExOjjIwMbd26ta2XdN6+jLmryQ8AAGh9l/R7gF5//XUVFBSopKREGRkZeuaZZ+TxeLRnzx4lJia29fIuyJnvC+I9QQAAhN4lHYDmz5+vyZMn65577pEklZSUaNWqVVq0aJFmzJjRxqu7ME2uAs1qWnPFyWXBjyEkAQDQIpdsAKqrq1NFRYUKCwvtscjISGVmZsrn8zX7mNraWtXW1trbR4787YMIDx06pEAgELK1BQIBnThxQgfrHIpuaAjZfhtVRE4M2j74cNOajLqFQdtbHLk/XFM44qLX1hbsfh88qOjo6LZejhHoefjR8/Ci3+HXkp4fPXpUkmRZ1gUf75INQH/9619VX1+vpKSkoPGkpCTt3r272ccUFxdr9uzZTcbT0tJaZY1t686grS7nUzOv1RYDAEDIHT16VPHx8Rf02Es2AF2IwsJCFRQU2NsNDQ06dOiQOnfurIiIiJAdp6amRqmpqfrqq6/kcrlCtl80j36HHz0PP3oeXvQ7/FrSc8uydPToUaWkpFzw8S7ZANSlSxdFRUWpuro6aLy6ulrJycnNPsbpdMrpdAaNJSQktNYS5XK5+MUJI/odfvQ8/Oh5eNHv8Dvfnl/olZ9Gl+xt8A6HQ+np6SovL7fHGhoaVF5eLrfb3YYrAwAAf+8u2StAklRQUKCcnBwNHjxYQ4YM0TPPPKPjx4/bd4UBAAA055IOQHfeeae++eYbFRUVye/3a+DAgSotLW3yxuhwczqdeuyxx5q83IbWQb/Dj56HHz0PL/odfuHueYR1MfeQAQAAXIIu2fcAAQAAXCgCEAAAMA4BCAAAGIcABAAAjEMACrGFCxfqiiuuUExMjDIyMrR169a2XtIlqbi4WD//+c/VoUMHJSYmasyYMdqzZ09QzcmTJ5Wbm6vOnTurffv2Gjt2bJMPxty3b5+ys7MVFxenxMRETZs2TadOnQrnqVyynnzySUVERCg/P98eo+eh95e//EX/+I//qM6dOys2Nlb9+vXTtm3b7HnLslRUVKSuXbsqNjZWmZmZ+vTTT4P2cejQIU2YMEEul0sJCQmaNGmSjh07Fu5T+btXX1+vRx99VGlpaYqNjdWVV16pOXPmBH2fFP2+OBs2bNAtt9yilJQURUREaMWKFUHzoervjh07dOONNyomJkapqamaO3duyxdrIWRee+01y+FwWIsWLbJ27dplTZ482UpISLCqq6vbemmXHI/HYy1evNiqqqqyKisrrZtvvtnq3r27dezYMbvmgQcesFJTU63y8nJr27Zt1tChQ63rrrvOnj916pTVt29fKzMz09q+fbu1evVqq0uXLlZhYWFbnNIlZevWrdYVV1xh9e/f33rwwQftcXoeWocOHbJ69OhhTZw40dqyZYv1xRdfWGvWrLE+++wzu+bJJ5+04uPjrRUrVlgfffSR9ctf/tJKS0uzvvvuO7tm1KhR1oABA6zNmzdb77//vnXVVVdZ48ePb4tT+rv2xBNPWJ07d7ZWrlxp7d2711q+fLnVvn17a8GCBXYN/b44q1evth555BHrzTfftCRZb731VtB8KPp75MgRKykpyZowYYJVVVVlvfrqq1ZsbKz1n//5ny1aKwEohIYMGWLl5uba2/X19VZKSopVXFzchqv6cThw4IAlyVq/fr1lWZZ1+PBhKzo62lq+fLld88knn1iSLJ/PZ1nW334RIyMjLb/fb9e8+OKLlsvlsmpra8N7ApeQo0ePWldffbXl9Xqtf/iHf7ADED0PvenTp1s33HDDWecbGhqs5ORk66mnnrLHDh8+bDmdTuvVV1+1LMuyPv74Y0uS9cEHH9g1//u//2tFRERYf/nLX1pv8Zeg7Oxs69577w0au/32260JEyZYlkW/Q+3MABSq/r7wwgtWx44dg/5NmT59utWzZ88WrY+XwEKkrq5OFRUVyszMtMciIyOVmZkpn8/Xhiv7cThy5IgkqVOnTpKkiooKBQKBoH736tVL3bt3t/vt8/nUr1+/oA/G9Hg8qqmp0a5du8K4+ktLbm6usrOzg3or0fPW8Ic//EGDBw/WHXfcocTERA0aNEj/9V//Zc/v3btXfr8/qOfx8fHKyMgI6nlCQoIGDx5s12RmZioyMlJbtmwJ38lcAq677jqVl5frT3/6kyTpo48+0saNGzV69GhJ9Lu1haq/Pp9Pw4YNk8PhsGs8Ho/27Nmjb7/99rzXc0l/EvTfk7/+9a+qr69v8inUSUlJ2r17dxut6sehoaFB+fn5uv7669W3b19Jkt/vl8PhaPJltklJSfL7/XZNc89H4xyaeu211/Thhx/qgw8+aDJHz0Pviy++0IsvvqiCggI9/PDD+uCDD/SrX/1KDodDOTk5ds+a6+npPU9MTAyab9eunTp16kTPzzBjxgzV1NSoV69eioqKUn19vZ544glNmDBBkuh3KwtVf/1+v9LS0prso3GuY8eO57UeAhD+7uXm5qqqqkobN25s66X8qH311Vd68MEH5fV6FRMT09bLMUJDQ4MGDx6s//iP/5AkDRo0SFVVVSopKVFOTk4br+7H54033tDSpUu1bNkyXXPNNaqsrFR+fr5SUlLot4F4CSxEunTpoqioqCZ3xFRXVys5ObmNVnXpy8vL08qVK/Xee++pW7du9nhycrLq6up0+PDhoPrT+52cnNzs89E4h2AVFRU6cOCArr32WrVr107t2rXT+vXr9eyzz6pdu3ZKSkqi5yHWtWtX9enTJ2isd+/e2rdvn6Tve3auf1eSk5N14MCBoPlTp07p0KFD9PwM06ZN04wZMzRu3Dj169dPd999t6ZOnari4mJJ9Lu1haq/ofp3hgAUIg6HQ+np6SovL7fHGhoaVF5eLrfb3YYruzRZlqW8vDy99dZbWrt2bZPLnenp6YqOjg7q9549e7Rv3z673263Wzt37gz6ZfJ6vXK5XE3+04E0YsQI7dy5U5WVlfbP4MGDNWHCBPvP9Dy0rr/++iYf7/CnP/1JPXr0kCSlpaUpOTk5qOc1NTXasmVLUM8PHz6siooKu2bt2rVqaGhQRkZGGM7i0nHixAlFRgb/txcVFaWGhgZJ9Lu1haq/brdbGzZsUCAQsGu8Xq969ux53i9/SeI2+FB67bXXLKfTaS1ZssT6+OOPrfvvv99KSEgIuiMG52fKlClWfHy8tW7dOmv//v32z4kTJ+yaBx54wOrevbu1du1aa9u2bZbb7bbcbrc933hLdlZWllVZWWmVlpZal19+Obdkt8Dpd4FZFj0Pta1bt1rt2rWznnjiCevTTz+1li5dasXFxVm///3v7Zonn3zSSkhIsN5++21rx44d1q233trsbcODBg2ytmzZYm3cuNG6+uqruS27GTk5OdZPfvIT+zb4N9980+rSpYv10EMP2TX0++IcPXrU2r59u7V9+3ZLkjV//nxr+/bt1v/93/9ZlhWa/h4+fNhKSkqy7r77bquqqsp67bXXrLi4OG6Db2vPPfec1b17d8vhcFhDhgyxNm/e3NZLuiRJavZn8eLFds13331n/cu//IvVsWNHKy4uzrrtttus/fv3B+3nyy+/tEaPHm3FxsZaXbp0sX79619bgUAgzGdz6TozANHz0HvnnXesvn37Wk6n0+rVq5f10ksvBc03NDRYjz76qJWUlGQ5nU5rxIgR1p49e4JqDh48aI0fP95q37695XK5rHvuucc6evRoOE/jklBTU2M9+OCDVvfu3a2YmBjrpz/9qfXII48E3U5Nvy/Oe++91+y/3Tk5OZZlha6/H330kXXDDTdYTqfT+slPfmI9+eSTLV5rhGWd9hGYAAAABuA9QAAAwDgEIAAAYBwCEAAAMA4BCAAAGIcABAAAjEMAAgAAxiEAAQAA4xCAAACAcQhAAADAOAQgAABgHAIQAAAwDgEIAAAY5/8BSYcFP0j+FA8AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def process_clustering(name):\n",
    "    clustering = pd.read_csv('data/swissprot/{}_cluster.tsv'.format(name), delimiter='\\t', header=None)\n",
    "    #rename heading as cluster reference and id\n",
    "    clustering.columns = [name, 'Entry']\n",
    "    clustering.drop_duplicates(subset='Entry', keep='first', inplace=True)\n",
    "    print(clustering[name].nunique())\n",
    "    clustering[name].value_counts().hist(bins=100)\n",
    "    return clustering\n",
    "\n",
    "cluster50 = process_clustering('clusterRes50')\n",
    "cluster30 = process_clustering('clusterRes30')\n",
    "\n",
    "df = pd.merge(df, cluster50, on='Entry', how='left')\n",
    "df = pd.merge(df, cluster30, on='Entry', how='left')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "63ca5f5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('../processed_data/protein2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "756c5423",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
