{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "06e5ee54-4659-4353-b6ad-de92dd9ab319",
   "metadata": {},
   "source": [
    "# Data download and processing from Uniprot\n",
    "\n",
    "Downloaded the data on 01 May 2024. 254,929 results filtering for \"Catalytic activity\" and Reviewed \"Swiss-Prot\".  \n",
    "\n",
    "Selected download TSV and the columns, Seqeunce (under Sequences tab), EC number (under Function).\n",
    "\n",
    "Unzipped the downloaded file.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4f961afc-aedd-41b9-8635-4c6eabd017ab",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'data/uniprot/uniprotkb_AND_reviewed_true_AND_protein_2024_05_01.tsv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpd\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mdata/uniprot/uniprotkb_AND_reviewed_true_AND_protein_2024_05_01.tsv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m df\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/parsers/readers.py:912\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m    899\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    900\u001b[0m     dialect,\n\u001b[1;32m    901\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    908\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m    909\u001b[0m )\n\u001b[1;32m    910\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 912\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/parsers/readers.py:577\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    574\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    576\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 577\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    579\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    580\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/parsers/readers.py:1407\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1404\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1406\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1407\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/parsers/readers.py:1661\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1659\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1660\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1661\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1662\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1663\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1664\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1665\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1666\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1667\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1668\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1669\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1670\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1671\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1672\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/common.py:859\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    854\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    855\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    856\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    857\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    858\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 859\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    860\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    861\u001b[0m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    862\u001b[0m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    863\u001b[0m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    864\u001b[0m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    865\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    866\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    867\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    868\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/uniprot/uniprotkb_AND_reviewed_true_AND_protein_2024_05_01.tsv'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv('data/uniprot/uniprotkb_AND_reviewed_true_AND_protein_2024_05_01.tsv', sep='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "160aa2c9-277e-427f-8e52-58d77299d79e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A0A061I403</td>\n",
       "      <td>FICD_CRIGR</td>\n",
       "      <td>MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...</td>\n",
       "      <td>2.7.7.108; 3.1.4.-</td>\n",
       "      <td>455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254921</th>\n",
       "      <td>P83855</td>\n",
       "      <td>ACLP_MOUSE</td>\n",
       "      <td>GVYMEIGRCRXEAXRRRKEAV</td>\n",
       "      <td>4.6.1.1</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254922</th>\n",
       "      <td>P85957</td>\n",
       "      <td>PER_PSEMZ</td>\n",
       "      <td>AMAGTATVQGQGTR</td>\n",
       "      <td>1.11.1.7</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254923</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254927</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254928</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>202388 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "3       A0A024SC78  CUTI1_HYPJR   \n",
       "4       A0A024SH76   GUX2_HYPJR   \n",
       "6       A0A059TC02   CCR1_PETHY   \n",
       "8       A0A061I403   FICD_CRIGR   \n",
       "...            ...          ...   \n",
       "254921      P83855   ACLP_MOUSE   \n",
       "254922      P85957    PER_PSEMZ   \n",
       "254923      Q05115   AMDA_BORBO   \n",
       "254927      Q94MV8   VG56_BPLZ5   \n",
       "254928      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence           EC number  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.-; 3.2.2.6   \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...            3.1.1.74   \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...            3.2.1.91   \n",
       "6       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...            1.2.1.44   \n",
       "8       MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...  2.7.7.108; 3.1.4.-   \n",
       "...                                                   ...                 ...   \n",
       "254921                              GVYMEIGRCRXEAXRRRKEAV             4.6.1.1   \n",
       "254922                                     AMAGTATVQGQGTR            1.11.1.7   \n",
       "254923  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...            4.1.1.76   \n",
       "254927  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...            3.6.1.12   \n",
       "254928  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...            3.1.1.29   \n",
       "\n",
       "        Length  \n",
       "0          269  \n",
       "3          248  \n",
       "4          471  \n",
       "6          333  \n",
       "8          455  \n",
       "...        ...  \n",
       "254921      21  \n",
       "254922      14  \n",
       "254923     240  \n",
       "254927     172  \n",
       "254928     124  \n",
       "\n",
       "[202388 rows x 5 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter to only include sequneces < 512\n",
    "df = df[df['Length'] < 512]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee69eeba-0d87-4f0e-92f8-8d950cc616f3",
   "metadata": {},
   "source": [
    "# Explode out EC number so each EC has it's own row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "724377c7-9690-4cb2-ad79-e081cc7f9fb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "199220\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/gq/6ljhmvm1713fykdjqbl188pm0000gn/T/ipykernel_34484/1265130750.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['EC number'] = [ec.split(';') for ec in df['EC number']]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254921</th>\n",
       "      <td>P83855</td>\n",
       "      <td>ACLP_MOUSE</td>\n",
       "      <td>GVYMEIGRCRXEAXRRRKEAV</td>\n",
       "      <td>4.6.1.1</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254922</th>\n",
       "      <td>P85957</td>\n",
       "      <td>PER_PSEMZ</td>\n",
       "      <td>AMAGTATVQGQGTR</td>\n",
       "      <td>1.11.1.7</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254923</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254927</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254928</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>213091 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "3       A0A024SC78  CUTI1_HYPJR   \n",
       "4       A0A024SH76   GUX2_HYPJR   \n",
       "6       A0A059TC02   CCR1_PETHY   \n",
       "...            ...          ...   \n",
       "254921      P83855   ACLP_MOUSE   \n",
       "254922      P85957    PER_PSEMZ   \n",
       "254923      Q05115   AMDA_BORBO   \n",
       "254927      Q94MV8   VG56_BPLZ5   \n",
       "254928      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence EC number  Length  \n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.-     269  \n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.6     269  \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...  3.1.1.74     248  \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...  3.2.1.91     471  \n",
       "6       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...  1.2.1.44     333  \n",
       "...                                                   ...       ...     ...  \n",
       "254921                              GVYMEIGRCRXEAXRRRKEAV   4.6.1.1      21  \n",
       "254922                                     AMAGTATVQGQGTR  1.11.1.7      14  \n",
       "254923  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...  4.1.1.76     240  \n",
       "254927  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...  3.6.1.12     172  \n",
       "254928  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...  3.1.1.29     124  \n",
       "\n",
       "[213091 rows x 5 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop rows that don't have an ec number\n",
    "df = df[df['EC number'].notna()]\n",
    "print(len(df))\n",
    "# Now expand out the ones we have left\n",
    "df['EC number'] = [ec.split(';') for ec in df['EC number']]\n",
    "df = df.explode('EC number')\n",
    "# Clean the EC numbers\n",
    "df['EC number'] = [ec.replace(' ', '') for ec in df['EC number']]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a1e43b52-f7c4-444d-a064-baaf45c77f76",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('data/uniprot/processed.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4b81d7a",
   "metadata": {},
   "source": [
    "### Added a bit here to get rid of the short ones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e42c854f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('data/uniprot/processed.csv')\n",
    "df = df[df['Length'] >= 100 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "da480c10",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop_duplicates(subset=['Sequence', 'EC number'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "aad366c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "      <th>clusterRes50</th>\n",
       "      <th>clusterRes30</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-</td>\n",
       "      <td>269</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A1AY86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A1AY86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>A0A024SC78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "      <td>G4MM92</td>\n",
       "      <td>G4MM92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "      <td>Q9S9N9</td>\n",
       "      <td>P14721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208979</th>\n",
       "      <td>P36352</td>\n",
       "      <td>POLR_PHMV</td>\n",
       "      <td>VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>178</td>\n",
       "      <td>P36352</td>\n",
       "      <td>P36352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208980</th>\n",
       "      <td>P39262</td>\n",
       "      <td>VG56_BPT4</td>\n",
       "      <td>MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>171</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208981</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208982</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208983</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q54ZD0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>173694 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "1       A0A009IHW8  ABTIR_ACIB9   \n",
       "2       A0A024SC78  CUTI1_HYPJR   \n",
       "3       A0A024SH76   GUX2_HYPJR   \n",
       "4       A0A059TC02   CCR1_PETHY   \n",
       "...            ...          ...   \n",
       "208979      P36352    POLR_PHMV   \n",
       "208980      P39262    VG56_BPT4   \n",
       "208981      Q05115   AMDA_BORBO   \n",
       "208982      Q94MV8   VG56_BPLZ5   \n",
       "208983      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence EC number  Length  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.-     269   \n",
       "1       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.6     269   \n",
       "2       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...  3.1.1.74     248   \n",
       "3       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...  3.2.1.91     471   \n",
       "4       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...  1.2.1.44     333   \n",
       "...                                                   ...       ...     ...   \n",
       "208979  VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...  2.7.7.48     178   \n",
       "208980  MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...  3.6.1.12     171   \n",
       "208981  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...  4.1.1.76     240   \n",
       "208982  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...  3.6.1.12     172   \n",
       "208983  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...  3.1.1.29     124   \n",
       "\n",
       "       clusterRes50 clusterRes30  \n",
       "0        A0A009IHW8       A1AY86  \n",
       "1        A0A009IHW8       A1AY86  \n",
       "2        A0A024SC78   A0A024SC78  \n",
       "3            G4MM92       G4MM92  \n",
       "4            Q9S9N9       P14721  \n",
       "...             ...          ...  \n",
       "208979       P36352       P36352  \n",
       "208980       Q94MV8       Q94MV8  \n",
       "208981       Q05115       Q05115  \n",
       "208982       Q94MV8       Q94MV8  \n",
       "208983       Q9J5H2       Q54ZD0  \n",
       "\n",
       "[173694 rows x 7 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4607689a",
   "metadata": {},
   "source": [
    "### Comparison to the dataset used in CLEAN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9d4113b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P51451</td>\n",
       "      <td>2.7.10.2</td>\n",
       "      <td>MGLVSSKKPDKEKPIKEKDKGQWSPLKVSAQDKDAPPLPPLVVFNH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C7C422</td>\n",
       "      <td>3.5.2.6</td>\n",
       "      <td>MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q07457</td>\n",
       "      <td>2.3.2.27</td>\n",
       "      <td>MTAEPATKKIKLELSDPSEPLTQSDVIAFQKEALFRCINRRRVDFE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q07009</td>\n",
       "      <td>3.4.22.53</td>\n",
       "      <td>MAGIAMKLAKDREAAEGLGSHERAIKYLNQDYETLRNECLEAGALF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q9TTH8</td>\n",
       "      <td>3.4.22.54</td>\n",
       "      <td>MPTVISASVAPRTGAEPRSPGPIAQAAQGKGTEAGGGNPSGIYSAI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227357</th>\n",
       "      <td>Q97HD1</td>\n",
       "      <td>3.1.11.6</td>\n",
       "      <td>MPSKKESYESMIKELEKIVSSMENEELPLEEAMKNYEDGVKLCDKL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227358</th>\n",
       "      <td>A0A140CUL9</td>\n",
       "      <td>1.1.99.13</td>\n",
       "      <td>MAENNTYDAIVIGSGISGGWATKELTEKGLKTLMLERGRNIEHVKD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227359</th>\n",
       "      <td>W6JQJ6</td>\n",
       "      <td>1.13.12.2</td>\n",
       "      <td>MNKNNRHPADGKKPITIFGPDFPFAFDDWLEHPAGLGSIPAARHGE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227360</th>\n",
       "      <td>B3IVI7</td>\n",
       "      <td>3.5.1.30</td>\n",
       "      <td>MRIALYQGAPKPLDVPGNLQRLRHQAQLAAERGAQLLVCPEMFLTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227361</th>\n",
       "      <td>Q88QV2</td>\n",
       "      <td>3.5.1.30</td>\n",
       "      <td>MRIALYQGAPKPLDVPGNLQRLRHQAQLAADRGAQLLVCPEMFLSG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>191488 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry  EC number  \\\n",
       "0           P51451   2.7.10.2   \n",
       "1           C7C422    3.5.2.6   \n",
       "2           Q07457   2.3.2.27   \n",
       "3           Q07009  3.4.22.53   \n",
       "4           Q9TTH8  3.4.22.54   \n",
       "...            ...        ...   \n",
       "227357      Q97HD1   3.1.11.6   \n",
       "227358  A0A140CUL9  1.1.99.13   \n",
       "227359      W6JQJ6  1.13.12.2   \n",
       "227360      B3IVI7   3.5.1.30   \n",
       "227361      Q88QV2   3.5.1.30   \n",
       "\n",
       "                                                 Sequence  \n",
       "0       MGLVSSKKPDKEKPIKEKDKGQWSPLKVSAQDKDAPPLPPLVVFNH...  \n",
       "1       MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...  \n",
       "2       MTAEPATKKIKLELSDPSEPLTQSDVIAFQKEALFRCINRRRVDFE...  \n",
       "3       MAGIAMKLAKDREAAEGLGSHERAIKYLNQDYETLRNECLEAGALF...  \n",
       "4       MPTVISASVAPRTGAEPRSPGPIAQAAQGKGTEAGGGNPSGIYSAI...  \n",
       "...                                                   ...  \n",
       "227357  MPSKKESYESMIKELEKIVSSMENEELPLEEAMKNYEDGVKLCDKL...  \n",
       "227358  MAENNTYDAIVIGSGISGGWATKELTEKGLKTLMLERGRNIEHVKD...  \n",
       "227359  MNKNNRHPADGKKPITIFGPDFPFAFDDWLEHPAGLGSIPAARHGE...  \n",
       "227360  MRIALYQGAPKPLDVPGNLQRLRHQAQLAAERGAQLLVCPEMFLTG...  \n",
       "227361  MRIALYQGAPKPLDVPGNLQRLRHQAQLAADRGAQLLVCPEMFLSG...  \n",
       "\n",
       "[191488 rows x 3 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_df = pd.read_csv('/disk1/jyang4/repos/CLEAN/app/data/split100.csv', sep='\\t')\n",
    "clean_df = clean_df.drop_duplicates(subset=['Sequence', 'EC number'])\n",
    "clean_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3cdeac1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P51451</td>\n",
       "      <td>2.7.10.2</td>\n",
       "      <td>MGLVSSKKPDKEKPIKEKDKGQWSPLKVSAQDKDAPPLPPLVVFNH...</td>\n",
       "      <td>505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C7C422</td>\n",
       "      <td>3.5.2.6</td>\n",
       "      <td>MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...</td>\n",
       "      <td>270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P00917</td>\n",
       "      <td>4.2.1.1</td>\n",
       "      <td>MAHSDWGYDSPNGPZEWVKLYPIANGNNQSPIDIKTSETKHDTSLK...</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q9ULX7</td>\n",
       "      <td>4.2.1.1</td>\n",
       "      <td>MLFSALLLEVIWILAADGGQHWTYEGPHGQDHWPASYPECGNNAQS...</td>\n",
       "      <td>337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>P10506</td>\n",
       "      <td>2.7.12.2</td>\n",
       "      <td>MFKRRRNPKGLVLNPNASVKSSDNDHKEELINNQKSFESNVEAFME...</td>\n",
       "      <td>340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227341</th>\n",
       "      <td>Q9KQH9</td>\n",
       "      <td>2.3.1.179</td>\n",
       "      <td>MSKRRVVVTGMGMLSPVGNTVESSWKALLAGQSGIVNIEHFDTTNF...</td>\n",
       "      <td>414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227352</th>\n",
       "      <td>Q11WK3</td>\n",
       "      <td>6.3.5.2</td>\n",
       "      <td>MTEKILILDFGSQYTQLIARRVRELNVYCEIHPYNKAPQIDGTVKG...</td>\n",
       "      <td>509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227354</th>\n",
       "      <td>B1KNT2</td>\n",
       "      <td>4.2.1.59</td>\n",
       "      <td>MSEQLNTMDIKEIMSFLPHRYPFLLIDRVLDYTPGETLHAIKNVTI...</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227360</th>\n",
       "      <td>B3IVI7</td>\n",
       "      <td>3.5.1.30</td>\n",
       "      <td>MRIALYQGAPKPLDVPGNLQRLRHQAQLAAERGAQLLVCPEMFLTG...</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227361</th>\n",
       "      <td>Q88QV2</td>\n",
       "      <td>3.5.1.30</td>\n",
       "      <td>MRIALYQGAPKPLDVPGNLQRLRHQAQLAADRGAQLLVCPEMFLSG...</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>151801 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Entry  EC number                                           Sequence  \\\n",
       "0       P51451   2.7.10.2  MGLVSSKKPDKEKPIKEKDKGQWSPLKVSAQDKDAPPLPPLVVFNH...   \n",
       "1       C7C422    3.5.2.6  MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...   \n",
       "6       P00917    4.2.1.1  MAHSDWGYDSPNGPZEWVKLYPIANGNNQSPIDIKTSETKHDTSLK...   \n",
       "7       Q9ULX7    4.2.1.1  MLFSALLLEVIWILAADGGQHWTYEGPHGQDHWPASYPECGNNAQS...   \n",
       "11      P10506   2.7.12.2  MFKRRRNPKGLVLNPNASVKSSDNDHKEELINNQKSFESNVEAFME...   \n",
       "...        ...        ...                                                ...   \n",
       "227341  Q9KQH9  2.3.1.179  MSKRRVVVTGMGMLSPVGNTVESSWKALLAGQSGIVNIEHFDTTNF...   \n",
       "227352  Q11WK3    6.3.5.2  MTEKILILDFGSQYTQLIARRVRELNVYCEIHPYNKAPQIDGTVKG...   \n",
       "227354  B1KNT2   4.2.1.59  MSEQLNTMDIKEIMSFLPHRYPFLLIDRVLDYTPGETLHAIKNVTI...   \n",
       "227360  B3IVI7   3.5.1.30  MRIALYQGAPKPLDVPGNLQRLRHQAQLAAERGAQLLVCPEMFLTG...   \n",
       "227361  Q88QV2   3.5.1.30  MRIALYQGAPKPLDVPGNLQRLRHQAQLAADRGAQLLVCPEMFLSG...   \n",
       "\n",
       "        Length  \n",
       "0          505  \n",
       "1          270  \n",
       "6          261  \n",
       "7          337  \n",
       "11         340  \n",
       "...        ...  \n",
       "227341     414  \n",
       "227352     509  \n",
       "227354     152  \n",
       "227360     264  \n",
       "227361     264  \n",
       "\n",
       "[151801 rows x 4 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_df['Length'] = [len(seq) for seq in clean_df['Sequence']]\n",
    "clean_df = clean_df[clean_df['Length'] < 512]\n",
    "clean_df = clean_df[clean_df['Length'] >= 100]\n",
    "clean_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5250f2c8",
   "metadata": {},
   "source": [
    "### Perform clustering using mmseqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "19acec52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate a fasta file as input to mmseqs\n",
    "with open('clustering/swissprot.fasta', 'w') as f:\n",
    "    for index, row in df.iterrows():\n",
    "        f.write(f'>{row[\"Entry\"]}\\n{row[\"Sequence\"]}\\n')\n",
    "\n",
    "# !mmseqs createdb examples/swissprot.fasta swissprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5e6cdd85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes50 clustering/mmseqs_tmp --min-seq-id 0.5 -c 0.8 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.5\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/14239173291168504146/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[173620] 0s 47ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 48ms\n",
      "Create directory clustering/mmseqs_tmp/14239173291168504146/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu clustering/mmseqs_tmp/14239173291168504146/clu_tmp --max-seqs 20 -c 0.8 --cov-mode 1 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.5 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 3.000000\n",
      "Set cluster mode GREEDY MEM\n",
      "Set cluster iterations to 3\n",
      "linclust clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 173694 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 100.00% 173.69K 1s 808ms    \n",
      "Sort kmer 0h 0m 0s 311ms\n",
      "Sort by rep. sequence 0h 0m 0s 237ms\n",
      "Time for fill: 0h 0m 0s 47ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 3s 285ms\n",
      "rescorediagonal clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 173.69K 0s 343ms    \n",
      "Time for merging to pref_rescore1: 0h 0m 0s 62ms                  ] 38.59% 67.04K eta 0s       \n",
      "Time for processing: 0h 0m 1s 566ms\n",
      "clust clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref_rescore1 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pre_clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 872ms\n",
      "\n",
      "Size of the sequence database: 173694\n",
      "Size of the alignment database: 173694\n",
      "Number of clusters: 58426\n",
      "\n",
      "Writing results 0h 0m 0s 12ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 925ms\n",
      "createsubdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/order_redundancy clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 23ms\n",
      "createsubdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/order_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 28ms\n",
      "filterdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref_filter1 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref_filter2 --filter-file clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/order_redundancy --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 100.00% 58.43K 0s 882ms    \n",
      "Time for merging to pref_filter2: 0h 0m 0s 31ms\n",
      "Time for processing: 0h 0m 1s 417ms\n",
      "rescorediagonal clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/input_step_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/input_step_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref_filter2 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 58.43K 0s 147ms    \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 23ms                  ] 63.50% 37.10K eta 0s       \n",
      "Time for processing: 0h 0m 1s 439ms\n",
      "align clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/input_step_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/input_step_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref_rescore2 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 58426 type: Aminoacid\n",
      "Target database size: 58426 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 58.43K 2s 812ms    \n",
      "Time for merging to aln: 0h 0m 0s 26ms\n",
      "119492 alignments calculated\n",
      "90483 sequence pairs passed the thresholds (0.757231 of overall calculated)\n",
      "1.548677 hits per query sequence\n",
      "Time for processing: 0h 0m 3s 284ms\n",
      "clust clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/input_step_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/aln clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 778ms\n",
      "\n",
      "Size of the sequence database: 58426\n",
      "Size of the alignment database: 58426\n",
      "Number of clusters: 41795\n",
      "\n",
      "Writing results 0h 0m 0s 12ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 811ms\n",
      "mergeclusters clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pre_clust clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/clust --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 58.43K 0s 366ms                      ] 65.83% 38.47K eta 0s       \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 41.79K 0s 602ms              ] 4.45% 1.86K eta 10s       \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 173.69K 1s 129ms       ] 39.67% 68.90K eta 1s       \n",
      "Time for merging to clu_redundancy: 0h 0m 0s 22ms\n",
      "Time for processing: 0h 0m 1s 242ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/linclust/15699254948364400700/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_redundancy clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 19ms\n",
      "prefilter clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 41795 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 41795 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 41.79K 1s 277ms    \n",
      "Index table: Masked residues: 47641\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 41.79K 0s 235ms    .56% 11.10K eta 0s       \n",
      "Index statistics\n",
      "Entries:          5815575\n",
      "DB size:          521 MB\n",
      "Avg k-mer size:   0.090868\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t163\n",
      "    AYPWGT\t121\n",
      "    GRGHCS\t97\n",
      "    GPGGTQ\t91\n",
      "    SPGGGW\t88\n",
      "    HGGDFP\t79\n",
      "    ITVGGP\t76\n",
      "    PHEEQS\t74\n",
      "    DPAGIQ\t69\n",
      "    PARFLR\t66\n",
      "Time for index table init: 0h 0m 3s 210ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 41795\n",
      "Target db start 1 to 41795\n",
      "[=================================================================] 100.00% 41.79K 1s 369ms    \n",
      "\n",
      "1.708625 k-mers per position\n",
      "315 DB matches per sequence\n",
      "0 overflows\n",
      "11 sequences passed prefiltering per query sequence\n",
      "10 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 26ms\n",
      "Time for processing: 0h 0m 7s 562ms\n",
      "align clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/pref_step0 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 41795 type: Aminoacid\n",
      "Target database size: 41795 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 41.79K 4s 116ms    \n",
      "Time for merging to aln_step0: 0h 0m 0s 31ms\n",
      "452487 alignments calculated\n",
      "118024 sequence pairs passed the thresholds (0.260834 of overall calculated)\n",
      "2.823879 hits per query sequence\n",
      "Time for processing: 0h 0m 4s 579ms\n",
      "clust clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/aln_step0 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_step0 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 828ms\n",
      "\n",
      "Size of the sequence database: 41795\n",
      "Size of the alignment database: 41795\n",
      "Number of clusters: 28575\n",
      "\n",
      "Writing results 0h 0m 0s 12ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 857ms\n",
      "createsubdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_step0 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "prefilter clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step1 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step1 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 2 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 28575 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 28575 type: Aminoacid\n",
      "Index table k-mer threshold: 145 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 28.58K 1s 146ms    \n",
      "Index table: Masked residues: 39213\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 28.58K 0s 284ms    \n",
      "Index statistics\n",
      "Entries:          6233577\n",
      "DB size:          523 MB\n",
      "Avg k-mer size:   0.097400\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t93\n",
      "    GQQVAR\t81\n",
      "    AYPWGT\t54\n",
      "    GSELFN\t53\n",
      "    PTGLGH\t50\n",
      "    FNLPNP\t50\n",
      "    FPHNAQ\t49\n",
      "    FREPGR\t49\n",
      "    GRGHCS\t48\n",
      "    TGNKTT\t44\n",
      "Time for index table init: 0h 0m 2s 983ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 145\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 28575\n",
      "Target db start 1 to 28575\n",
      "[=================================================================] 100.00% 28.58K 1s 195ms    \n",
      "\n",
      "7.382307 k-mers per position\n",
      "503 DB matches per sequence\n",
      "0 overflows\n",
      "11 sequences passed prefiltering per query sequence\n",
      "10 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 21ms\n",
      "Time for processing: 0h 0m 7s 437ms\n",
      "align clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step1 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step1 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/pref_step1 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 28575 type: Aminoacid\n",
      "Target database size: 28575 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 28.58K 3s 686ms    \n",
      "Time for merging to aln_step1: 0h 0m 0s 29ms\n",
      "296767 alignments calculated\n",
      "30799 sequence pairs passed the thresholds (0.103782 of overall calculated)\n",
      "1.077830 hits per query sequence\n",
      "Time for processing: 0h 0m 4s 141ms\n",
      "clust clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step1 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/aln_step1 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_step1 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 729ms\n",
      "\n",
      "Size of the sequence database: 28575\n",
      "Size of the alignment database: 28575\n",
      "Number of clusters: 27520\n",
      "\n",
      "Writing results 0h 0m 0s 5ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 747ms\n",
      "createsubdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_step1 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step1 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "prefilter clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step2 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step2 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 3 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 27520 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 27520 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 27.52K 1s 45ms     \n",
      "Index table: Masked residues: 38119\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 27.52K 0s 644ms    \n",
      "Index statistics\n",
      "Entries:          8139435\n",
      "DB size:          534 MB\n",
      "Avg k-mer size:   0.127179\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t90\n",
      "    GQQVAR\t73\n",
      "    GGGHPA\t53\n",
      "    STYLGG\t53\n",
      "    GSELFN\t53\n",
      "    GGSEVS\t51\n",
      "    AYPWGT\t49\n",
      "    RDKELL\t45\n",
      "    TGNKTT\t44\n",
      "    SPGGGW\t44\n",
      "Time for index table init: 0h 0m 3s 250ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 27520\n",
      "Target db start 1 to 27520\n",
      "[=================================================================] 100.00% 27.52K 1s 855ms    \n",
      "\n",
      "22.239691 k-mers per position\n",
      "1175 DB matches per sequence\n",
      "0 overflows\n",
      "14 sequences passed prefiltering per query sequence\n",
      "17 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 18ms\n",
      "Time for processing: 0h 0m 8s 292ms\n",
      "align clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step2 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step2 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/pref_step2 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 27520 type: Aminoacid\n",
      "Target database size: 27520 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 27.52K 3s 510ms    \n",
      "Time for merging to aln_step2: 0h 0m 0s 34ms\n",
      "357859 alignments calculated\n",
      "27696 sequence pairs passed the thresholds (0.077394 of overall calculated)\n",
      "1.006395 hits per query sequence\n",
      "Time for processing: 0h 0m 3s 966ms\n",
      "clust clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step2 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/aln_step2 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_step2 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 685ms\n",
      "\n",
      "Size of the sequence database: 27520\n",
      "Size of the alignment database: 27520\n",
      "Number of clusters: 27415\n",
      "\n",
      "Writing results 0h 0m 0s 5ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 704ms\n",
      "mergeclusters clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_redundancy clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_step0 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_step1 clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_step2 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 41.79K 0s 454ms    K eta 0s       \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 28.58K 0s 689ms    >                                                     ] 18.22% 5.21K eta 2s       \n",
      "Clustering step 3\n",
      "[=================================================================] 100.00% 27.52K 0s 901ms    ta 1s       \n",
      "Clustering step 4\n",
      "[=================================================================] 100.00% 27.42K 1s 95ms     \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 173.69K 1s 538ms     \n",
      "Time for merging to clu: 0h 0m 0s 20ms\n",
      "Time for processing: 0h 0m 1s 642ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/pref_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/aln_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/pref_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/aln_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/clu_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step1_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_tmp/13391587180916045397/input_step2_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu clustering/mmseqs_tmp/14239173291168504146/cluster.tsv --threads 384 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 22ms\n",
      "Time for processing: 0h 0m 0s 802ms\n",
      "result2repseq clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu clustering/mmseqs_tmp/14239173291168504146/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 27.42K 0s 475ms    \n",
      "Time for merging to clu_rep: 0h 0m 0s 29ms\n",
      "Time for processing: 0h 0m 0s 920ms\n",
      "result2flat clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu_rep clustering/mmseqs_tmp/14239173291168504146/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 74ms\n",
      "createseqfiledb clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu clustering/mmseqs_tmp/14239173291168504146/clu_seqs --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 27.42K 0s 999ms    \n",
      "Time for merging to clu_seqs: 0h 0m 0s 16ms\n",
      "Time for processing: 0h 0m 1s 450ms\n",
      "result2flat clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/input clustering/mmseqs_tmp/14239173291168504146/clu_seqs clustering/mmseqs_tmp/14239173291168504146/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 238ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 12ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb clustering/mmseqs_tmp/14239173291168504146/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster clustering/swissprot.fasta clustering/clusterRes50 clustering/mmseqs_tmp --min-seq-id 0.5 -c 0.8 --cov-mode 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "286e6959",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes30 clustering/mmseqs_tmp --min-seq-id 0.3 -c 0.8 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.3\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/16018658589129770318/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[173620] 0s 48ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 49ms\n",
      "Create directory clustering/mmseqs_tmp/16018658589129770318/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu clustering/mmseqs_tmp/16018658589129770318/clu_tmp --max-seqs 20 -c 0.8 --cov-mode 1 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.3 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 5.000000\n",
      "Set cluster mode GREEDY MEM\n",
      "Set cluster iterations to 3\n",
      "linclust clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 173694 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 100.00% 173.69K 1s 925ms    \n",
      "Sort kmer 0h 0m 0s 290ms\n",
      "Sort by rep. sequence 0h 0m 0s 304ms\n",
      "Time for fill: 0h 0m 0s 46ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 3s 517ms\n",
      "rescorediagonal clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 173.69K 0s 346ms    \n",
      "Time for merging to pref_rescore1: 0h 0m 0s 54ms======>           ] 82.39% 143.10K eta 0s       \n",
      "Time for processing: 0h 0m 1s 538ms\n",
      "clust clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref_rescore1 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pre_clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 1s 20ms\n",
      "\n",
      "Size of the sequence database: 173694\n",
      "Size of the alignment database: 173694\n",
      "Number of clusters: 58426\n",
      "\n",
      "Writing results 0h 0m 0s 13ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 1s 74ms\n",
      "createsubdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/order_redundancy clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 22ms\n",
      "createsubdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/order_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 27ms\n",
      "filterdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref_filter1 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref_filter2 --filter-file clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/order_redundancy --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 100.00% 58.43K 0s 583ms    \n",
      "Time for merging to pref_filter2: 0h 0m 0s 40ms\n",
      "Time for processing: 0h 0m 1s 557ms\n",
      "rescorediagonal clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/input_step_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/input_step_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref_filter2 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 58.43K 0s 144ms    \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 27ms                  ] 36.73% 21.46K eta 1s       \n",
      "Time for processing: 0h 0m 1s 300ms\n",
      "align clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/input_step_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/input_step_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref_rescore2 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 58426 type: Aminoacid\n",
      "Target database size: 58426 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 58.43K 2s 764ms    \n",
      "Time for merging to aln: 0h 0m 0s 30ms\n",
      "125279 alignments calculated\n",
      "120511 sequence pairs passed the thresholds (0.961941 of overall calculated)\n",
      "2.062626 hits per query sequence\n",
      "Time for processing: 0h 0m 3s 217ms\n",
      "clust clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/input_step_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/aln clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 915ms\n",
      "\n",
      "Size of the sequence database: 58426\n",
      "Size of the alignment database: 58426\n",
      "Number of clusters: 32714\n",
      "\n",
      "Writing results 0h 0m 0s 19ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 953ms\n",
      "mergeclusters clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pre_clust clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/clust --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 58.43K 0s 439ms      ] 75.12% 43.89K eta 0s       \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 32.71K 0s 656ms    38% 11.25K eta 1s       \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 173.69K 1s 162ms     eta 0s       \n",
      "Time for merging to clu_redundancy: 0h 0m 0s 23ms\n",
      "Time for processing: 0h 0m 1s 293ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/linclust/15117454218751531870/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_redundancy clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 18ms\n",
      "prefilter clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 32714 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 32714 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 32.71K 1s 165ms    \n",
      "Index table: Masked residues: 40035\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 32.71K 0s 236ms    \n",
      "Index statistics\n",
      "Entries:          4615880\n",
      "DB size:          514 MB\n",
      "Avg k-mer size:   0.072123\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t122\n",
      "    AYPWGT\t74\n",
      "    IHDKNI\t68\n",
      "    GPGGTQ\t63\n",
      "    FREPGR\t50\n",
      "    TFPHHL\t48\n",
      "    RKYYLP\t48\n",
      "    GRGHCS\t48\n",
      "    EARPGP\t43\n",
      "    PVGVMT\t43\n",
      "Time for index table init: 0h 0m 2s 932ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 32714\n",
      "Target db start 1 to 32714\n",
      "[=================================================================] 100.00% 32.71K 1s 217ms    \n",
      "\n",
      "1.779371 k-mers per position\n",
      "253 DB matches per sequence\n",
      "0 overflows\n",
      "8 sequences passed prefiltering per query sequence\n",
      "6 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 18ms\n",
      "Time for processing: 0h 0m 7s 557ms\n",
      "align clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/pref_step0 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 32714 type: Aminoacid\n",
      "Target database size: 32714 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 32.71K 3s 787ms    \n",
      "Time for merging to aln_step0: 0h 0m 0s 22ms\n",
      "273591 alignments calculated\n",
      "235333 sequence pairs passed the thresholds (0.860164 of overall calculated)\n",
      "7.193648 hits per query sequence\n",
      "Time for processing: 0h 0m 4s 316ms\n",
      "clust clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/aln_step0 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_step0 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 966ms\n",
      "\n",
      "Size of the sequence database: 32714\n",
      "Size of the alignment database: 32714\n",
      "Number of clusters: 14440\n",
      "\n",
      "Writing results 0h 0m 0s 6ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 987ms\n",
      "createsubdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_step0 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "prefilter clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step1 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step1 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 3 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 14440 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 14440 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 14.44K 0s 863ms    \n",
      "Index table: Masked residues: 24995\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 14.44K 0s 174ms    0s       \n",
      "Index statistics\n",
      "Entries:          4192495\n",
      "DB size:          512 MB\n",
      "Avg k-mer size:   0.065508\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t43\n",
      "    GPGGTL\t37\n",
      "    GLGIGK\t24\n",
      "    RDKELL\t23\n",
      "    AVGGGL\t21\n",
      "    VVGGGL\t21\n",
      "    GQQVAR\t20\n",
      "    HRLPIL\t15\n",
      "    VHDKNL\t15\n",
      "    RLFGVL\t15\n",
      "Time for index table init: 0h 0m 2s 617ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 14440\n",
      "Target db start 1 to 14440\n",
      "[=================================================================] 100.00% 14.44K 1s 101ms    \n",
      "\n",
      "22.510530 k-mers per position\n",
      "672 DB matches per sequence\n",
      "0 overflows\n",
      "7 sequences passed prefiltering per query sequence\n",
      "6 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 18ms\n",
      "Time for processing: 0h 0m 6s 881ms\n",
      "align clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step1 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step1 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/pref_step1 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 14440 type: Aminoacid\n",
      "Target database size: 14440 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 14.44K 2s 496ms    \n",
      "Time for merging to aln_step1: 0h 0m 0s 21ms\n",
      "91185 alignments calculated\n",
      "39907 sequence pairs passed the thresholds (0.437649 of overall calculated)\n",
      "2.763643 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 923ms\n",
      "clust clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step1 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/aln_step1 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_step1 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 832ms\n",
      "\n",
      "Size of the sequence database: 14440\n",
      "Size of the alignment database: 14440\n",
      "Number of clusters: 9653\n",
      "\n",
      "Writing results 0h 0m 0s 4ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 845ms\n",
      "createsubdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_step1 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step1 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "prefilter clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step2 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step2 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 5 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 9653 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 9653 type: Aminoacid\n",
      "Index table k-mer threshold: 118 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 9.65K 1s 2ms      \n",
      "Index table: Masked residues: 21786\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 9.65K 0s 379ms    \n",
      "Index statistics\n",
      "Entries:          3077606\n",
      "DB size:          505 MB\n",
      "Avg k-mer size:   0.048088\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t42\n",
      "    VHDKNI\t22\n",
      "    RDKELL\t22\n",
      "    GPGGTL\t18\n",
      "    AVGGGL\t14\n",
      "    HRLPIL\t14\n",
      "    VHDKNL\t14\n",
      "    QCGGGP\t13\n",
      "    RVGLSR\t13\n",
      "    LGGFVY\t13\n",
      "Time for index table init: 0h 0m 3s 197ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 118\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 9653\n",
      "Target db start 1 to 9653\n",
      "[=================================================================] 100.00% 9.65K 1s 881ms    \n",
      "\n",
      "178.818277 k-mers per position\n",
      "2885 DB matches per sequence\n",
      "0 overflows\n",
      "17 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 15ms\n",
      "Time for processing: 0h 0m 8s 48ms\n",
      "align clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step2 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step2 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/pref_step2 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 9653 type: Aminoacid\n",
      "Target database size: 9653 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 9.65K 2s 322ms    \n",
      "Time for merging to aln_step2: 0h 0m 0s 14ms\n",
      "121652 alignments calculated\n",
      "12985 sequence pairs passed the thresholds (0.106739 of overall calculated)\n",
      "1.345178 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 738ms\n",
      "clust clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step2 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/aln_step2 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_step2 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 523ms\n",
      "\n",
      "Size of the sequence database: 9653\n",
      "Size of the alignment database: 9653\n",
      "Number of clusters: 8444\n",
      "\n",
      "Writing results 0h 0m 0s 3ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 538ms\n",
      "mergeclusters clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_redundancy clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_step0 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_step1 clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_step2 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 32.71K 0s 605ms    \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 14.44K 0s 807ms    =============>                                          ] 35.29% 5.10K eta 1s       \n",
      "Clustering step 3\n",
      "[=================================================================] 100.00% 9.65K 1s 2ms       \n",
      "Clustering step 4\n",
      "[=================================================================] 100.00% 8.44K 1s 169ms    \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 173.69K 1s 682ms          ] 51.74% 89.88K eta 1s       \n",
      "Time for merging to clu: 0h 0m 0s 15ms\n",
      "Time for processing: 0h 0m 1s 776ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/pref_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/aln_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/pref_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/aln_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/clu_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step1_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_tmp/7937821247301186169/input_step2_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu clustering/mmseqs_tmp/16018658589129770318/cluster.tsv --threads 384 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 27ms\n",
      "Time for processing: 0h 0m 0s 947ms\n",
      "result2repseq clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu clustering/mmseqs_tmp/16018658589129770318/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 8.44K 0s 163ms    \n",
      "Time for merging to clu_rep: 0h 0m 0s 26ms\n",
      "Time for processing: 0h 0m 0s 569ms\n",
      "result2flat clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu_rep clustering/mmseqs_tmp/16018658589129770318/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 45ms\n",
      "createseqfiledb clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu clustering/mmseqs_tmp/16018658589129770318/clu_seqs --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 8.44K 0s 885ms    \n",
      "Time for merging to clu_seqs: 0h 0m 0s 15ms\n",
      "Time for processing: 0h 0m 1s 375ms\n",
      "result2flat clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/input clustering/mmseqs_tmp/16018658589129770318/clu_seqs clustering/mmseqs_tmp/16018658589129770318/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 226ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 12ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/16018658589129770318/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster clustering/swissprot.fasta clustering/clusterRes30 clustering/mmseqs_tmp --min-seq-id 0.3 -c 0.8 --cov-mode 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "02322b64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27415\n",
      "8444\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "      <th>clusterRes50_x</th>\n",
       "      <th>clusterRes30_x</th>\n",
       "      <th>clusterRes50_y</th>\n",
       "      <th>clusterRes30_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-</td>\n",
       "      <td>269</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A1AY86</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A1AY86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A1AY86</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A1AY86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>A0A024SC78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "      <td>G4MM92</td>\n",
       "      <td>G4MM92</td>\n",
       "      <td>G4MM92</td>\n",
       "      <td>G4MM92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "      <td>Q9S9N9</td>\n",
       "      <td>P14721</td>\n",
       "      <td>Q9S9N9</td>\n",
       "      <td>P14721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173689</th>\n",
       "      <td>P36352</td>\n",
       "      <td>POLR_PHMV</td>\n",
       "      <td>VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>178</td>\n",
       "      <td>P36352</td>\n",
       "      <td>P36352</td>\n",
       "      <td>P36352</td>\n",
       "      <td>P36352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173690</th>\n",
       "      <td>P39262</td>\n",
       "      <td>VG56_BPT4</td>\n",
       "      <td>MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>171</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173691</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173692</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173693</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q54ZD0</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q54ZD0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>173694 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "1       A0A009IHW8  ABTIR_ACIB9   \n",
       "2       A0A024SC78  CUTI1_HYPJR   \n",
       "3       A0A024SH76   GUX2_HYPJR   \n",
       "4       A0A059TC02   CCR1_PETHY   \n",
       "...            ...          ...   \n",
       "173689      P36352    POLR_PHMV   \n",
       "173690      P39262    VG56_BPT4   \n",
       "173691      Q05115   AMDA_BORBO   \n",
       "173692      Q94MV8   VG56_BPLZ5   \n",
       "173693      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence EC number  Length  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.-     269   \n",
       "1       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.6     269   \n",
       "2       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...  3.1.1.74     248   \n",
       "3       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...  3.2.1.91     471   \n",
       "4       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...  1.2.1.44     333   \n",
       "...                                                   ...       ...     ...   \n",
       "173689  VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...  2.7.7.48     178   \n",
       "173690  MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...  3.6.1.12     171   \n",
       "173691  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...  4.1.1.76     240   \n",
       "173692  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...  3.6.1.12     172   \n",
       "173693  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...  3.1.1.29     124   \n",
       "\n",
       "       clusterRes50_x clusterRes30_x clusterRes50_y clusterRes30_y  \n",
       "0          A0A009IHW8         A1AY86     A0A009IHW8         A1AY86  \n",
       "1          A0A009IHW8         A1AY86     A0A009IHW8         A1AY86  \n",
       "2          A0A024SC78     A0A024SC78     A0A024SC78     A0A024SC78  \n",
       "3              G4MM92         G4MM92         G4MM92         G4MM92  \n",
       "4              Q9S9N9         P14721         Q9S9N9         P14721  \n",
       "...               ...            ...            ...            ...  \n",
       "173689         P36352         P36352         P36352         P36352  \n",
       "173690         Q94MV8         Q94MV8         Q94MV8         Q94MV8  \n",
       "173691         Q05115         Q05115         Q05115         Q05115  \n",
       "173692         Q94MV8         Q94MV8         Q94MV8         Q94MV8  \n",
       "173693         Q9J5H2         Q54ZD0         Q9J5H2         Q54ZD0  \n",
       "\n",
       "[173694 rows x 9 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjkAAAGfCAYAAAC9RsMDAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuNSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/xnp5ZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAucUlEQVR4nO3de1zUdb7H8Te34ZIBXgJkBaKbl7ymidNtLRFUTpvlo0eap8UyO3lgj0jHlNZMczu0tZWmFo9OqXseyZbuI63U1AlvWajJSoql3WxtNwfbTMdLAsL3/NGD3zqCF1wI+PZ6Ph7zWOf3+8xvvu8Zln3vzPyGAGOMEQAAgGUCm3sBAAAATYGSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsFNyQ4fz8fL3xxhvavXu3wsPDdd111+n3v/+9Onfu7MwMHDhQGzZs8Lvdf/zHf6igoMC5vm/fPo0fP17r1q1TmzZtlJmZqfz8fAUH/3M569evV25urnbt2qWEhARNnTpVY8aM8TvuvHnz9PTTT8vr9apXr16aM2eO+vfvf955ampq9M033+jiiy9WQEBAQx4KAADQTIwxOnLkiOLj4xUYeJbXa0wDpKenmwULFpiysjJTWlpqhg0bZhITE83Ro0edmV/+8pdm3LhxZv/+/c7l8OHDzv6TJ0+a7t27m9TUVLN9+3azcuVK06FDB5OXl+fMfPnllyYiIsLk5uaajz/+2MyZM8cEBQWZVatWOTOvvfaacblcZv78+WbXrl1m3LhxJjo62pSXl593nq+//tpI4sKFCxcuXLi0wsvXX3991v+dDzDmwv9A57fffquYmBht2LBBN910k6QfX8np3bu3Zs2aVe9t3nnnHf3bv/2bvvnmG8XGxkqSCgoKNHnyZH377bdyuVyaPHmyVqxYobKyMud2I0eO1KFDh7Rq1SpJUkpKiq699lrNnTtX0o+vyiQkJOg3v/mNpkyZcl7rP3z4sKKjo/X1118rMjLyQh+GOqqqqrRmzRqlpaUpJCSk0Y7bUtieTyKjDWzPJ5HRFrZnbIp8Pp9PCQkJOnTokKKios4416C3q053+PBhSVK7du38ti9atEivvvqq4uLidOutt+rRRx9VRESEJKm4uFg9evRwCo4kpaena/z48dq1a5f69Omj4uJipaam+h0zPT1dOTk5kqTKykqVlJQoLy/P2R8YGKjU1FQVFxefcb0VFRWqqKhwrh85ckSSFB4ervDw8At4BOoXHBysiIgIhYeHW/kDa3s+iYw2sD2fREZb2J6xKfJVVVVJ0jk/anLBJaempkY5OTm6/vrr1b17d2f73XffraSkJMXHx2vHjh2aPHmy9uzZozfeeEOS5PV6/QqOJOe61+s964zP59MPP/yg77//XtXV1fXO7N69+4xrzs/P14wZM+psX7NmjVPCGpPH42n0Y7YktueTyGgD2/NJZLSF7RkbM9/x48fPa+6CS05WVpbKysq0adMmv+0PPPCA8+8ePXqoY8eOGjRokL744gtdfvnlF3p3jSIvL0+5ubnO9dqXu9LS0hr97SqPx6PBgwdb2cptzyeR0Qa255PIaAvbMzZFPp/Pd15zF1RysrOztXz5cm3cuFGdOnU662xKSook6fPPP9fll1+uuLg4bd261W+mvLxckhQXF+f8Z+22U2ciIyMVHh6uoKAgBQUF1TtTe4z6hIaGKjQ0tM72kJCQJvnBaqrjthS255PIaAPb80lktIXtGRsz3/kep0Hfk2OMUXZ2tpYuXaq1a9cqOTn5nLcpLS2VJHXs2FGS5Ha7tXPnTh04cMCZ8Xg8ioyMVLdu3ZyZoqIiv+N4PB653W5JksvlUt++ff1mampqVFRU5MwAAICftwa9kpOVlaXCwkK9+eabuvjii53P0ERFRSk8PFxffPGFCgsLNWzYMLVv3147duzQxIkTddNNN6lnz56SpLS0NHXr1k333HOPnnrqKXm9Xk2dOlVZWVnOqywPPvig5s6dq4cfflj33Xef1q5dq8WLF2vFihXOWnJzc5WZmal+/fqpf//+mjVrlo4dO6Z77723sR4bAADQijWo5Lz44ouSfjxN/FQLFizQmDFj5HK59O677zqFIyEhQSNGjNDUqVOd2aCgIC1fvlzjx4+X2+3WRRddpMzMTD3++OPOTHJyslasWKGJEydq9uzZ6tSpk15++WWlp6c7M3fddZe+/fZbTZs2TV6vV71799aqVavqfBgZAAD8PDWo5JzrK3USEhLqfNtxfZKSkrRy5cqzzgwcOFDbt28/60x2drays7PPeX8AAODnh79dBQAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACw0gX/gU6cW/fpq1VR/c8/A//VkxnNuBoAAH5eeCUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKDSo5+fn5uvbaa3XxxRcrJiZGw4cP1549e/xmTpw4oaysLLVv315t2rTRiBEjVF5e7jezb98+ZWRkKCIiQjExMZo0aZJOnjzpN7N+/Xpdc801Cg0N1RVXXKGFCxfWWc+8efN06aWXKiwsTCkpKdq6dWtD4gAAAIs1qORs2LBBWVlZ2rx5szwej6qqqpSWlqZjx445MxMnTtTbb7+tJUuWaMOGDfrmm290xx13OPurq6uVkZGhyspKffDBB/rjH/+ohQsXatq0ac7M3r17lZGRoZtvvlmlpaXKycnR/fffr9WrVzszr7/+unJzc/XYY4/pL3/5i3r16qX09HQdOHDgX3k8AACAJYIbMrxq1Sq/6wsXLlRMTIxKSkp000036fDhw3rllVdUWFioW265RZK0YMECde3aVZs3b9aAAQO0Zs0affzxx3r33XcVGxur3r17a+bMmZo8ebKmT58ul8ulgoICJScn65lnnpEkde3aVZs2bdJzzz2n9PR0SdKzzz6rcePG6d5775UkFRQUaMWKFZo/f76mTJnyLz8wAACgdWtQyTnd4cOHJUnt2rWTJJWUlKiqqkqpqanOTJcuXZSYmKji4mINGDBAxcXF6tGjh2JjY52Z9PR0jR8/Xrt27VKfPn1UXFzsd4zamZycHElSZWWlSkpKlJeX5+wPDAxUamqqiouLz7jeiooKVVRUONd9Pp8kqaqqSlVVVRf4KNRVe6zQQFPv9tauNocteepDxtbP9nwSGW1he8amyHe+x7rgklNTU6OcnBxdf/316t69uyTJ6/XK5XIpOjrabzY2NlZer9eZObXg1O6v3Xe2GZ/Ppx9++EHff/+9qqur653ZvXv3Gdecn5+vGTNm1Nm+Zs0aRUREnEfqhpnZr8bv+sqVKxv9PpqTx+Np7iU0OTK2frbnk8hoC9szNma+48ePn9fcBZecrKwslZWVadOmTRd6iJ9cXl6ecnNznes+n08JCQlKS0tTZGRko91PVVWVPB6PHt0WqIqaAGd72fT0RruP5lSbb/DgwQoJCWnu5TQJMrZ+tueTyGgL2zM2Rb7ad2LO5YJKTnZ2tpYvX66NGzeqU6dOzva4uDhVVlbq0KFDfq/mlJeXKy4uzpk5/Syo2rOvTp05/Yys8vJyRUZGKjw8XEFBQQoKCqp3pvYY9QkNDVVoaGid7SEhIU3yg1VRE6CK6n+WHNt+eJvqcWtJyNj62Z5PIqMtbM/YmPnO9zgNOrvKGKPs7GwtXbpUa9euVXJyst/+vn37KiQkREVFRc62PXv2aN++fXK73ZIkt9utnTt3+p0F5fF4FBkZqW7dujkzpx6jdqb2GC6XS3379vWbqampUVFRkTMDAAB+3hr0Sk5WVpYKCwv15ptv6uKLL3Y+QxMVFaXw8HBFRUVp7Nixys3NVbt27RQZGanf/OY3crvdGjBggCQpLS1N3bp10z333KOnnnpKXq9XU6dOVVZWlvMqy4MPPqi5c+fq4Ycf1n333ae1a9dq8eLFWrFihbOW3NxcZWZmql+/furfv79mzZqlY8eOOWdbAQCAn7cGlZwXX3xRkjRw4EC/7QsWLNCYMWMkSc8995wCAwM1YsQIVVRUKD09XS+88IIzGxQUpOXLl2v8+PFyu9266KKLlJmZqccff9yZSU5O1ooVKzRx4kTNnj1bnTp10ssvv+ycPi5Jd911l7799ltNmzZNXq9XvXv31qpVq+p8GBkAAPw8NajkGGPOORMWFqZ58+Zp3rx5Z5xJSko655lGAwcO1Pbt2886k52drezs7HOuCQAA/Pzwt6sAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVmpwydm4caNuvfVWxcfHKyAgQMuWLfPbP2bMGAUEBPhdhgwZ4jdz8OBBjR49WpGRkYqOjtbYsWN19OhRv5kdO3boxhtvVFhYmBISEvTUU0/VWcuSJUvUpUsXhYWFqUePHlq5cmVD4wAAAEs1uOQcO3ZMvXr10rx58844M2TIEO3fv9+5/OlPf/LbP3r0aO3atUsej0fLly/Xxo0b9cADDzj7fT6f0tLSlJSUpJKSEj399NOaPn26XnrpJWfmgw8+0KhRozR27Fht375dw4cP1/Dhw1VWVtbQSAAAwELBDb3B0KFDNXTo0LPOhIaGKi4urt59n3zyiVatWqUPP/xQ/fr1kyTNmTNHw4YN0x/+8AfFx8dr0aJFqqys1Pz58+VyuXT11VertLRUzz77rFOGZs+erSFDhmjSpEmSpJkzZ8rj8Wju3LkqKCio974rKipUUVHhXPf5fJKkqqoqVVVVNeyBOIvaY4UGmnq3t3a1OWzJUx8ytn6255PIaAvbMzZFvvM9VoAxxpx77Aw3DgjQ0qVLNXz4cGfbmDFjtGzZMrlcLrVt21a33HKLfve736l9+/aSpPnz5+uhhx7S999/79zm5MmTCgsL05IlS3T77bfr17/+tXw+n99bYevWrdMtt9yigwcPqm3btkpMTFRubq5ycnKcmccee0zLli3TRx99VO96p0+frhkzZtTZXlhYqIiIiAt9GAAAwE/o+PHjuvvuu3X48GFFRkaeca7Br+Scy5AhQ3THHXcoOTlZX3zxhR555BENHTpUxcXFCgoKktfrVUxMjP8igoPVrl07eb1eSZLX61VycrLfTGxsrLOvbdu28nq9zrZTZ2qPUZ+8vDzl5uY6130+nxISEpSWlnbWB6mhqqqq5PF49Oi2QFXUBDjby6anN9p9NKfafIMHD1ZISEhzL6dJkLH1sz2fREZb2J6xKfLVvhNzLo1eckaOHOn8u0ePHurZs6cuv/xyrV+/XoMGDWrsu2uQ0NBQhYaG1tkeEhLSJD9YFTUBqqj+Z8mx7Ye3qR63loSMrZ/t+SQy2sL2jI2Z73yP0+SnkF922WXq0KGDPv/8c0lSXFycDhw44Ddz8uRJHTx40PkcT1xcnMrLy/1maq+fa+ZMnwUCAAA/L01ecv72t7/pu+++U8eOHSVJbrdbhw4dUklJiTOzdu1a1dTUKCUlxZnZuHGj3weLPB6POnfurLZt2zozRUVFfvfl8XjkdrubOhIAAGgFGlxyjh49qtLSUpWWlkqS9u7dq9LSUu3bt09Hjx7VpEmTtHnzZn311VcqKirSbbfdpiuuuELp6T9+HqVr164aMmSIxo0bp61bt+r9999Xdna2Ro4cqfj4eEnS3XffLZfLpbFjx2rXrl16/fXXNXv2bL/P00yYMEGrVq3SM888o927d2v69Onatm2bsrOzG+FhAQAArV2DS862bdvUp08f9enTR5KUm5urPn36aNq0aQoKCtKOHTv0q1/9SldddZXGjh2rvn376r333vP7LMyiRYvUpUsXDRo0SMOGDdMNN9zg9x04UVFRWrNmjfbu3au+ffvqoYce0rRp0/y+S+e6665TYWGhXnrpJfXq1Ut//vOftWzZMnXv3v1feTwAAIAlGvzB44EDB+psZ52vXr36nMdo166dCgsLzzrTs2dPvffee2edufPOO3XnnXee8/4AAMDPD3+7CgAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKDS45Gzdu1K233qr4+HgFBARo2bJlfvuNMZo2bZo6duyo8PBwpaam6rPPPvObOXjwoEaPHq3IyEhFR0dr7NixOnr0qN/Mjh07dOONNyosLEwJCQl66qmn6qxlyZIl6tKli8LCwtSjRw+tXLmyoXEAAIClGlxyjh07pl69emnevHn17n/qqaf0/PPPq6CgQFu2bNFFF12k9PR0nThxwpkZPXq0du3aJY/Ho+XLl2vjxo164IEHnP0+n09paWlKSkpSSUmJnn76aU2fPl0vvfSSM/PBBx9o1KhRGjt2rLZv367hw4dr+PDhKisra2gkAABgoeCG3mDo0KEaOnRovfuMMZo1a5amTp2q2267TZL0f//3f4qNjdWyZcs0cuRIffLJJ1q1apU+/PBD9evXT5I0Z84cDRs2TH/4wx8UHx+vRYsWqbKyUvPnz5fL5dLVV1+t0tJSPfvss04Zmj17toYMGaJJkyZJkmbOnCmPx6O5c+eqoKDggh4MAABgjwaXnLPZu3evvF6vUlNTnW1RUVFKSUlRcXGxRo4cqeLiYkVHRzsFR5JSU1MVGBioLVu26Pbbb1dxcbFuuukmuVwuZyY9PV2///3v9f3336tt27YqLi5Wbm6u3/2np6fXefvsVBUVFaqoqHCu+3w+SVJVVZWqqqr+1fiO2mOFBpp6t7d2tTlsyVMfMrZ+tueTyGgL2zM2Rb7zPVajlhyv1ytJio2N9dseGxvr7PN6vYqJifFfRHCw2rVr5zeTnJxc5xi1+9q2bSuv13vW+6lPfn6+ZsyYUWf7mjVrFBERcT4RG2Rmvxq/67Z9Zsjj8TT3EpocGVs/2/NJZLSF7RkbM9/x48fPa65RS05Ll5eX5/fqj8/nU0JCgtLS0hQZGdlo91NVVSWPx6NHtwWqoibA2V42Pb3R7qM51eYbPHiwQkJCmns5TYKMrZ/t+SQy2sL2jE2Rr/admHNp1JITFxcnSSovL1fHjh2d7eXl5erdu7czc+DAAb/bnTx5UgcPHnRuHxcXp/Lycr+Z2uvnmqndX5/Q0FCFhobW2R4SEtIkP1gVNQGqqP5nybHth7epHreWhIytn+35JDLawvaMjZnvfI/TqN+Tk5ycrLi4OBUVFTnbfD6ftmzZIrfbLUlyu906dOiQSkpKnJm1a9eqpqZGKSkpzszGjRv93nPzeDzq3Lmz2rZt68ycej+1M7X3AwAAft4aXHKOHj2q0tJSlZaWSvrxw8alpaXat2+fAgIClJOTo9/97nd66623tHPnTv36179WfHy8hg8fLknq2rWrhgwZonHjxmnr1q16//33lZ2drZEjRyo+Pl6SdPfdd8vlcmns2LHatWuXXn/9dc2ePdvvraYJEyZo1apVeuaZZ7R7925Nnz5d27ZtU3Z29r/+qAAAgFavwW9Xbdu2TTfffLNzvbZ4ZGZmauHChXr44Yd17NgxPfDAAzp06JBuuOEGrVq1SmFhYc5tFi1apOzsbA0aNEiBgYEaMWKEnn/+eWd/VFSU1qxZo6ysLPXt21cdOnTQtGnT/L5L57rrrlNhYaGmTp2qRx55RFdeeaWWLVum7t27X9ADAQAA7NLgkjNw4EAZY864PyAgQI8//rgef/zxM860a9dOhYWFZ72fnj176r333jvrzJ133qk777zz7AsGAAA/S/ztKgAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYKXg5l6AzcpCxyqk5sQpWw4321oAAPi54ZUcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKzU6CVn+vTpCggI8Lt06dLF2X/ixAllZWWpffv2atOmjUaMGKHy8nK/Y+zbt08ZGRmKiIhQTEyMJk2apJMnT/rNrF+/Xtdcc41CQ0N1xRVXaOHChY0dBQAAtGJN8krO1Vdfrf379zuXTZs2OfsmTpyot99+W0uWLNGGDRv0zTff6I477nD2V1dXKyMjQ5WVlfrggw/0xz/+UQsXLtS0adOcmb179yojI0M333yzSktLlZOTo/vvv1+rV69uijgAAKAVCm6SgwYHKy4urs72w4cP65VXXlFhYaFuueUWSdKCBQvUtWtXbd68WQMGDNCaNWv08ccf691331VsbKx69+6tmTNnavLkyZo+fbpcLpcKCgqUnJysZ555RpLUtWtXbdq0Sc8995zS09ObIhIAAGhlmqTkfPbZZ4qPj1dYWJjcbrfy8/OVmJiokpISVVVVKTU11Znt0qWLEhMTVVxcrAEDBqi4uFg9evRQbGysM5Oenq7x48dr165d6tOnj4qLi/2OUTuTk5Nz1nVVVFSooqLCue7z+SRJVVVVqqqqaoTkco4nSVWBYafvaLT7aE5OPkvy1IeMrZ/t+SQy2sL2jE2R73yP1eglJyUlRQsXLlTnzp21f/9+zZgxQzfeeKPKysrk9XrlcrkUHR3td5vY2Fh5vV5Jktfr9Ss4tftr951txufz6YcfflB4eHi9a8vPz9eMGTPqbF+zZo0iIiIuKO/ZeHo8779h5cpGv4/m5PF4mnsJTY6MrZ/t+SQy2sL2jI2Z7/jx4+c11+glZ+jQoc6/e/bsqZSUFCUlJWnx4sVnLB8/lby8POXm5jrXfT6fEhISlJaWpsjIyEa7n6qqKnk8Hg3e+V8KqTlxygL+1mj30ZycfIMHKyQkpLmX0yTI2PrZnk8ioy1sz9gU+WrfiTmXJnm76lTR0dG66qqr9Pnnn2vw4MGqrKzUoUOH/F7NKS8vdz7DExcXp61bt/odo/bsq1NnTj8jq7y8XJGRkWctUqGhoQoNDa2zPSQkpEl+sEJqTviXHMt+eJvqcWtJyNj62Z5PIqMtbM/YmPnO9zhN/j05R48e1RdffKGOHTuqb9++CgkJUVFRkbN/z5492rdvn9xutyTJ7XZr586dOnDggDPj8XgUGRmpbt26OTOnHqN2pvYYAAAAjV5y/vu//1sbNmzQV199pQ8++EC33367goKCNGrUKEVFRWns2LHKzc3VunXrVFJSonvvvVdut1sDBgyQJKWlpalbt26655579NFHH2n16tWaOnWqsrKynFdhHnzwQX355Zd6+OGHtXv3br3wwgtavHixJk6c2NhxAABAK9Xob1f97W9/06hRo/Tdd9/pkksu0Q033KDNmzfrkksukSQ999xzCgwM1IgRI1RRUaH09HS98MILzu2DgoK0fPlyjR8/Xm63WxdddJEyMzP1+OOPOzPJyclasWKFJk6cqNmzZ6tTp056+eWXOX0cAAA4Gr3kvPbaa2fdHxYWpnnz5mnevHlnnElKStLKc5yJNHDgQG3fvv2C1ggAAOzH364CAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAKwU39wJ+Ti6dssLv+ldPZjTTSgAAsB+v5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASvztqp/QV2F3n7blcLOsAwCAnwNeyQEAAFai5AAAACtRcgAAgJX4TE4zunTKijrbvnoyoxlWAgCAfXglBwAAWImSAwAArMTbVc2o7inlEqeVAwDQOCg5Lczpn9PhMzoAAFyYVl9y5s2bp6efflper1e9evXSnDlz1L9//+Ze1gU7/dWdS6cU1p2h+AAAcE6tuuS8/vrrys3NVUFBgVJSUjRr1iylp6drz549iomJae7lNYr63tI6vfhQegAAqKtVl5xnn31W48aN07333itJKigo0IoVKzR//nxNmTKlznxFRYUqKiqc64cP//j5l4MHD6qqqqrR1lVVVaXjx4/ru0qXQmpqGu24tUoCx/hd/+6RCzzQQ7sv6GZOvu++U0hIyAXeectGxtbP9nwSGW1he8amyHfkyBFJkjHmrHOttuRUVlaqpKREeXl5zrbAwEClpqaquLi43tvk5+drxowZdbYnJyc32TpbtPwOzb0CAAAu2JEjRxQVFXXG/a225PzjH/9QdXW1YmNj/bbHxsZq9+76X6HIy8tTbm6uc72mpkYHDx5U+/btFRAQ0Ghr8/l8SkhI0Ndff63IyMhGO25LYXs+iYw2sD2fREZb2J6xKfIZY3TkyBHFx8efda7VlpwLERoaqtDQUL9t0dHRTXZ/kZGRVv7A1rI9n0RGG9ieTyKjLWzP2Nj5zvYKTq1W+2WAHTp0UFBQkMrLy/22l5eXKy4urplWBQAAWopWW3JcLpf69u2roqIiZ1tNTY2KiorkdrubcWUAAKAlaNVvV+Xm5iozM1P9+vVT//79NWvWLB07dsw526q5hIaG6rHHHqvz1pgtbM8nkdEGtueTyGgL2zM2Z74Ac67zr1q4uXPnOl8G2Lt3bz3//PNKSUlp7mUBAIBm1upLDgAAQH1a7WdyAAAAzoaSAwAArETJAQAAVqLkAAAAK1FyGtm8efN06aWXKiwsTCkpKdq6dWtzL+m8bdy4Ubfeeqvi4+MVEBCgZcuW+e03xmjatGnq2LGjwsPDlZqaqs8++8xv5uDBgxo9erQiIyMVHR2tsWPH6ujRoz9hijPLz8/Xtddeq4svvlgxMTEaPny49uzZ4zdz4sQJZWVlqX379mrTpo1GjBhR5wsn9+3bp4yMDEVERCgmJkaTJk3SyZMnf8ooZ/Tiiy+qZ8+ezjeLut1uvfPOO87+1p7vdE8++aQCAgKUk5PjbGvtGadPn66AgAC/S5cuXZz9rT1frb///e/693//d7Vv317h4eHq0aOHtm3b5uxv7b9vLr300jrPY0BAgLKysiS1/uexurpajz76qJKTkxUeHq7LL79cM2fO9PuDmS3iOTRoNK+99ppxuVxm/vz5ZteuXWbcuHEmOjralJeXN/fSzsvKlSvNb3/7W/PGG28YSWbp0qV++5988kkTFRVlli1bZj766CPzq1/9yiQnJ5sffvjBmRkyZIjp1auX2bx5s3nvvffMFVdcYUaNGvUTJ6lfenq6WbBggSkrKzOlpaVm2LBhJjEx0Rw9etSZefDBB01CQoIpKioy27ZtMwMGDDDXXXeds//kyZOme/fuJjU11Wzfvt2sXLnSdOjQweTl5TVHpDreeusts2LFCvPpp5+aPXv2mEceecSEhISYsrIyY0zrz3eqrVu3mksvvdT07NnTTJgwwdne2jM+9thj5uqrrzb79+93Lt9++62zv7XnM8aYgwcPmqSkJDNmzBizZcsW8+WXX5rVq1ebzz//3Jlp7b9vDhw44PccejweI8msW7fOGNP6n8cnnnjCtG/f3ixfvtzs3bvXLFmyxLRp08bMnj3bmWkJzyElpxH179/fZGVlOderq6tNfHy8yc/Pb8ZVXZjTS05NTY2Ji4szTz/9tLPt0KFDJjQ01PzpT38yxhjz8ccfG0nmww8/dGbeeecdExAQYP7+97//ZGs/XwcOHDCSzIYNG4wxP+YJCQkxS5YscWY++eQTI8kUFxcbY34sgoGBgcbr9TozL774oomMjDQVFRU/bYDz1LZtW/Pyyy9ble/IkSPmyiuvNB6Px/zyl790So4NGR977DHTq1evevfZkM8YYyZPnmxuuOGGM+638ffNhAkTzOWXX25qamqseB4zMjLMfffd57ftjjvuMKNHjzbGtJznkLerGkllZaVKSkqUmprqbAsMDFRqaqqKi4ubcWWNY+/evfJ6vX75oqKilJKS4uQrLi5WdHS0+vXr58ykpqYqMDBQW7Zs+cnXfC6HDx+WJLVr106SVFJSoqqqKr+MXbp0UWJiol/GHj16KDY21plJT0+Xz+fTrl27fsLVn1t1dbVee+01HTt2TG6326p8WVlZysjI8Msi2fMcfvbZZ4qPj9dll12m0aNHa9++fZLsyffWW2+pX79+uvPOOxUTE6M+ffrof//3f539tv2+qays1Kuvvqr77rtPAQEBVjyP1113nYqKivTpp59Kkj766CNt2rRJQ4cOldRynsNW/WcdWpJ//OMfqq6u9vuBlKTY2Fjt3r27mVbVeLxeryTVm692n9frVUxMjN/+4OBgtWvXzplpKWpqapSTk6Prr79e3bt3l/Tj+l0uV52/TH96xvoeg9p9LcHOnTvldrt14sQJtWnTRkuXLlW3bt1UWlpqRb7XXntNf/nLX/Thhx/W2WfDc5iSkqKFCxeqc+fO2r9/v2bMmKEbb7xRZWVlVuSTpC+//FIvvviicnNz9cgjj+jDDz/Uf/3Xf8nlcikzM9O63zfLli3ToUOHNGbMGEl2/JxOmTJFPp9PXbp0UVBQkKqrq/XEE09o9OjRklrO/2ZQcvCzlJWVpbKyMm3atKm5l9LoOnfurNLSUh0+fFh//vOflZmZqQ0bNjT3shrF119/rQkTJsjj8SgsLKy5l9Mkav+fsCT17NlTKSkpSkpK0uLFixUeHt6MK2s8NTU16tevn/7nf/5HktSnTx+VlZWpoKBAmZmZzby6xvfKK69o6NChio+Pb+6lNJrFixdr0aJFKiws1NVXX63S0lLl5OQoPj6+RT2HvF3VSDp06KCgoKA6n44vLy9XXFxcM62q8dRmOFu+uLg4HThwwG//yZMndfDgwRb1GGRnZ2v58uVat26dOnXq5GyPi4tTZWWlDh065Dd/esb6HoPafS2By+XSFVdcob59+yo/P1+9evXS7NmzrchXUlKiAwcO6JprrlFwcLCCg4O1YcMGPf/88woODlZsbGyrz3i66OhoXXXVVfr888+teA4lqWPHjurWrZvftq5duzpvy9n0++avf/2r3n33Xd1///3ONhuex0mTJmnKlCkaOXKkevTooXvuuUcTJ05Ufn6+pJbzHFJyGonL5VLfvn1VVFTkbKupqVFRUZHcbnczrqxxJCcnKy4uzi+fz+fTli1bnHxut1uHDh1SSUmJM7N27VrV1NS0iD+aaoxRdna2li5dqrVr1yo5Odlvf9++fRUSEuKXcc+ePdq3b59fxp07d/r9F9Pj8SgyMrLOL+2WoqamRhUVFVbkGzRokHbu3KnS0lLn0q9fP40ePdr5d2vPeLqjR4/qiy++UMeOHa14DiXp+uuvr/P1DZ9++qmSkpIk2fH7ptaCBQsUExOjjIwMZ5sNz+Px48cVGOhfIYKCglRTUyOpBT2HjfLxZRhjfjyFPDQ01CxcuNB8/PHH5oEHHjDR0dF+n45vyY4cOWK2b99utm/fbiSZZ5991mzfvt389a9/Ncb8eDpgdHS0efPNN82OHTvMbbfdVu/pgH369DFbtmwxmzZtMldeeWWLOaVz/PjxJioqyqxfv97v1M7jx487Mw8++KBJTEw0a9euNdu2bTNut9u43W5nf+1pnWlpaaa0tNSsWrXKXHLJJS3mtM4pU6aYDRs2mL1795odO3aYKVOmmICAALNmzRpjTOvPV59Tz64ypvVnfOihh8z69evN3r17zfvvv29SU1NNhw4dzIEDB4wxrT+fMT+e/h8cHGyeeOIJ89lnn5lFixaZiIgI8+qrrzozrf33jTE/nmGbmJhoJk+eXGdfa38eMzMzzS9+8QvnFPI33njDdOjQwTz88MPOTEt4Dik5jWzOnDkmMTHRuFwu079/f7N58+bmXtJ5W7dunZFU55KZmWmM+fGUwEcffdTExsaa0NBQM2jQILNnzx6/Y3z33Xdm1KhRpk2bNiYyMtLce++95siRI82Qpq76skkyCxYscGZ++OEH85//+Z+mbdu2JiIiwtx+++1m//79fsf56quvzNChQ014eLjp0KGDeeihh0xVVdVPnKZ+9913n0lKSjIul8tccsklZtCgQU7BMab156vP6SWntWe86667TMeOHY3L5TK/+MUvzF133eX3/TGtPV+tt99+23Tv3t2EhoaaLl26mJdeeslvf2v/fWOMMatXrzaS6qzbmNb/PPp8PjNhwgSTmJhowsLCzGWXXWZ++9vf+p3e3hKewwBjTvl6QgAAAEvwmRwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWOn/AaIiI46pM6EBAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def process_clustering(name):\n",
    "    clustering = pd.read_csv('clustering/{}_cluster.tsv'.format(name), delimiter='\\t', header=None)\n",
    "    #rename heading as cluster reference and id\n",
    "    clustering.columns = [name, 'Entry']\n",
    "    clustering.drop_duplicates(subset='Entry', keep='first', inplace=True)\n",
    "    print(clustering[name].nunique())\n",
    "    clustering[name].value_counts().hist(bins=100)\n",
    "    return clustering\n",
    "\n",
    "cluster50 = process_clustering('clusterRes50')\n",
    "cluster30 = process_clustering('clusterRes30')\n",
    "\n",
    "df = pd.merge(df, cluster50, on='Entry', how='left')\n",
    "df = pd.merge(df, cluster30, on='Entry', how='left')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6b3be036",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "      <th>clusterRes50_x</th>\n",
       "      <th>clusterRes30_x</th>\n",
       "      <th>clusterRes50_y</th>\n",
       "      <th>clusterRes30_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A1AY86</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A1AY86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>A0A024SC78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "      <td>G4MM92</td>\n",
       "      <td>G4MM92</td>\n",
       "      <td>G4MM92</td>\n",
       "      <td>G4MM92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "      <td>Q9S9N9</td>\n",
       "      <td>P14721</td>\n",
       "      <td>Q9S9N9</td>\n",
       "      <td>P14721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A0A061I403</td>\n",
       "      <td>FICD_CRIGR</td>\n",
       "      <td>MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...</td>\n",
       "      <td>2.7.7.108</td>\n",
       "      <td>455</td>\n",
       "      <td>Q6AY47</td>\n",
       "      <td>Q6AY47</td>\n",
       "      <td>Q6AY47</td>\n",
       "      <td>Q6AY47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173689</th>\n",
       "      <td>P36352</td>\n",
       "      <td>POLR_PHMV</td>\n",
       "      <td>VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>178</td>\n",
       "      <td>P36352</td>\n",
       "      <td>P36352</td>\n",
       "      <td>P36352</td>\n",
       "      <td>P36352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173690</th>\n",
       "      <td>P39262</td>\n",
       "      <td>VG56_BPT4</td>\n",
       "      <td>MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>171</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173691</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173692</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173693</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q54ZD0</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q54ZD0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161738 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "1       A0A009IHW8  ABTIR_ACIB9   \n",
       "2       A0A024SC78  CUTI1_HYPJR   \n",
       "3       A0A024SH76   GUX2_HYPJR   \n",
       "4       A0A059TC02   CCR1_PETHY   \n",
       "5       A0A061I403   FICD_CRIGR   \n",
       "...            ...          ...   \n",
       "173689      P36352    POLR_PHMV   \n",
       "173690      P39262    VG56_BPT4   \n",
       "173691      Q05115   AMDA_BORBO   \n",
       "173692      Q94MV8   VG56_BPLZ5   \n",
       "173693      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence  EC number  Length  \\\n",
       "1       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.6     269   \n",
       "2       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   3.1.1.74     248   \n",
       "3       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   3.2.1.91     471   \n",
       "4       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...   1.2.1.44     333   \n",
       "5       MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...  2.7.7.108     455   \n",
       "...                                                   ...        ...     ...   \n",
       "173689  VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...   2.7.7.48     178   \n",
       "173690  MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...   3.6.1.12     171   \n",
       "173691  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...   4.1.1.76     240   \n",
       "173692  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...   3.6.1.12     172   \n",
       "173693  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...   3.1.1.29     124   \n",
       "\n",
       "       clusterRes50_x clusterRes30_x clusterRes50_y clusterRes30_y  \n",
       "1          A0A009IHW8         A1AY86     A0A009IHW8         A1AY86  \n",
       "2          A0A024SC78     A0A024SC78     A0A024SC78     A0A024SC78  \n",
       "3              G4MM92         G4MM92         G4MM92         G4MM92  \n",
       "4              Q9S9N9         P14721         Q9S9N9         P14721  \n",
       "5              Q6AY47         Q6AY47         Q6AY47         Q6AY47  \n",
       "...               ...            ...            ...            ...  \n",
       "173689         P36352         P36352         P36352         P36352  \n",
       "173690         Q94MV8         Q94MV8         Q94MV8         Q94MV8  \n",
       "173691         Q05115         Q05115         Q05115         Q05115  \n",
       "173692         Q94MV8         Q94MV8         Q94MV8         Q94MV8  \n",
       "173693         Q9J5H2         Q54ZD0         Q9J5H2         Q54ZD0  \n",
       "\n",
       "[161738 rows x 9 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[~df['EC number'].str.contains('-')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "63ca5f5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('../processed_data/protein2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "756c5423",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
