{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "06e5ee54-4659-4353-b6ad-de92dd9ab319",
   "metadata": {},
   "source": [
    "# Data download and processing from Uniprot\n",
    "\n",
    "Downloaded the data on 01 May 2024. 254,929 results filtering for \"Catalytic activity\" and Reviewed \"Swiss-Prot\".  \n",
    "\n",
    "Selected download TSV and the columns, Seqeunce (under Sequences tab), EC number (under Function).\n",
    "\n",
    "Unzipped the downloaded file.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4f961afc-aedd-41b9-8635-4c6eabd017ab",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'data/uniprot/uniprotkb_AND_reviewed_true_AND_protein_2024_05_01.tsv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpd\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mdata/uniprot/uniprotkb_AND_reviewed_true_AND_protein_2024_05_01.tsv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m df\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/parsers/readers.py:912\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m    899\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    900\u001b[0m     dialect,\n\u001b[1;32m    901\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    908\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m    909\u001b[0m )\n\u001b[1;32m    910\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 912\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/parsers/readers.py:577\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    574\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    576\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 577\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    579\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    580\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/parsers/readers.py:1407\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1404\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1406\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1407\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/parsers/readers.py:1661\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1659\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1660\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1661\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1662\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1663\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1664\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1665\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1666\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1667\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1668\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1669\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1670\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1671\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1672\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m~/miniconda3/envs/rxnfp/lib/python3.8/site-packages/pandas/io/common.py:859\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    854\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    855\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    856\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    857\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    858\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 859\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    860\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    861\u001b[0m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    862\u001b[0m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    863\u001b[0m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    864\u001b[0m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    865\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    866\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    867\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    868\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/uniprot/uniprotkb_AND_reviewed_true_AND_protein_2024_05_01.tsv'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv('data/uniprot/uniprotkb_AND_reviewed_true_AND_protein_2024_05_01.tsv', sep='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "160aa2c9-277e-427f-8e52-58d77299d79e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-; 3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A0A061I403</td>\n",
       "      <td>FICD_CRIGR</td>\n",
       "      <td>MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...</td>\n",
       "      <td>2.7.7.108; 3.1.4.-</td>\n",
       "      <td>455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254921</th>\n",
       "      <td>P83855</td>\n",
       "      <td>ACLP_MOUSE</td>\n",
       "      <td>GVYMEIGRCRXEAXRRRKEAV</td>\n",
       "      <td>4.6.1.1</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254922</th>\n",
       "      <td>P85957</td>\n",
       "      <td>PER_PSEMZ</td>\n",
       "      <td>AMAGTATVQGQGTR</td>\n",
       "      <td>1.11.1.7</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254923</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254927</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254928</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>202388 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "3       A0A024SC78  CUTI1_HYPJR   \n",
       "4       A0A024SH76   GUX2_HYPJR   \n",
       "6       A0A059TC02   CCR1_PETHY   \n",
       "8       A0A061I403   FICD_CRIGR   \n",
       "...            ...          ...   \n",
       "254921      P83855   ACLP_MOUSE   \n",
       "254922      P85957    PER_PSEMZ   \n",
       "254923      Q05115   AMDA_BORBO   \n",
       "254927      Q94MV8   VG56_BPLZ5   \n",
       "254928      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence           EC number  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.-; 3.2.2.6   \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...            3.1.1.74   \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...            3.2.1.91   \n",
       "6       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...            1.2.1.44   \n",
       "8       MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...  2.7.7.108; 3.1.4.-   \n",
       "...                                                   ...                 ...   \n",
       "254921                              GVYMEIGRCRXEAXRRRKEAV             4.6.1.1   \n",
       "254922                                     AMAGTATVQGQGTR            1.11.1.7   \n",
       "254923  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...            4.1.1.76   \n",
       "254927  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...            3.6.1.12   \n",
       "254928  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...            3.1.1.29   \n",
       "\n",
       "        Length  \n",
       "0          269  \n",
       "3          248  \n",
       "4          471  \n",
       "6          333  \n",
       "8          455  \n",
       "...        ...  \n",
       "254921      21  \n",
       "254922      14  \n",
       "254923     240  \n",
       "254927     172  \n",
       "254928     124  \n",
       "\n",
       "[202388 rows x 5 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter to only include sequneces < 512\n",
    "df = df[df['Length'] < 512]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee69eeba-0d87-4f0e-92f8-8d950cc616f3",
   "metadata": {},
   "source": [
    "# Explode out EC number so each EC has it's own row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "724377c7-9690-4cb2-ad79-e081cc7f9fb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "199220\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/gq/6ljhmvm1713fykdjqbl188pm0000gn/T/ipykernel_34484/1265130750.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['EC number'] = [ec.split(';') for ec in df['EC number']]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.-</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254921</th>\n",
       "      <td>P83855</td>\n",
       "      <td>ACLP_MOUSE</td>\n",
       "      <td>GVYMEIGRCRXEAXRRRKEAV</td>\n",
       "      <td>4.6.1.1</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254922</th>\n",
       "      <td>P85957</td>\n",
       "      <td>PER_PSEMZ</td>\n",
       "      <td>AMAGTATVQGQGTR</td>\n",
       "      <td>1.11.1.7</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254923</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254927</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254928</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>213091 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "3       A0A024SC78  CUTI1_HYPJR   \n",
       "4       A0A024SH76   GUX2_HYPJR   \n",
       "6       A0A059TC02   CCR1_PETHY   \n",
       "...            ...          ...   \n",
       "254921      P83855   ACLP_MOUSE   \n",
       "254922      P85957    PER_PSEMZ   \n",
       "254923      Q05115   AMDA_BORBO   \n",
       "254927      Q94MV8   VG56_BPLZ5   \n",
       "254928      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence EC number  Length  \n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.-     269  \n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...   3.2.2.6     269  \n",
       "3       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...  3.1.1.74     248  \n",
       "4       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...  3.2.1.91     471  \n",
       "6       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...  1.2.1.44     333  \n",
       "...                                                   ...       ...     ...  \n",
       "254921                              GVYMEIGRCRXEAXRRRKEAV   4.6.1.1      21  \n",
       "254922                                     AMAGTATVQGQGTR  1.11.1.7      14  \n",
       "254923  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...  4.1.1.76     240  \n",
       "254927  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...  3.6.1.12     172  \n",
       "254928  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...  3.1.1.29     124  \n",
       "\n",
       "[213091 rows x 5 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop rows that don't have an ec number\n",
    "df = df[df['EC number'].notna()]\n",
    "print(len(df))\n",
    "# Now expand out the ones we have left\n",
    "df['EC number'] = [ec.split(';') for ec in df['EC number']]\n",
    "df = df.explode('EC number')\n",
    "# Clean the EC numbers\n",
    "df['EC number'] = [ec.replace(' ', '') for ec in df['EC number']]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a1e43b52-f7c4-444d-a064-baaf45c77f76",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('data/uniprot/processed.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4b81d7a",
   "metadata": {},
   "source": [
    "### Added a bit here to get rid of the short ones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "e42c854f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('../processed_data/protein2EC.csv')\n",
    "df = df[df['Length'] >= 100 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "da480c10",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop_duplicates(subset=['Sequence', 'EC number'])\n",
    "df = df[~df['EC number'].str.contains('-')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "aad366c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A0A061I403</td>\n",
       "      <td>FICD_CRIGR</td>\n",
       "      <td>MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...</td>\n",
       "      <td>2.7.7.108</td>\n",
       "      <td>455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173689</th>\n",
       "      <td>P36352</td>\n",
       "      <td>POLR_PHMV</td>\n",
       "      <td>VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173690</th>\n",
       "      <td>P39262</td>\n",
       "      <td>VG56_BPT4</td>\n",
       "      <td>MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173691</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173692</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173693</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161738 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "1       A0A009IHW8  ABTIR_ACIB9   \n",
       "2       A0A024SC78  CUTI1_HYPJR   \n",
       "3       A0A024SH76   GUX2_HYPJR   \n",
       "4       A0A059TC02   CCR1_PETHY   \n",
       "5       A0A061I403   FICD_CRIGR   \n",
       "...            ...          ...   \n",
       "173689      P36352    POLR_PHMV   \n",
       "173690      P39262    VG56_BPT4   \n",
       "173691      Q05115   AMDA_BORBO   \n",
       "173692      Q94MV8   VG56_BPLZ5   \n",
       "173693      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence  EC number  Length  \n",
       "1       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.6     269  \n",
       "2       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   3.1.1.74     248  \n",
       "3       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   3.2.1.91     471  \n",
       "4       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...   1.2.1.44     333  \n",
       "5       MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...  2.7.7.108     455  \n",
       "...                                                   ...        ...     ...  \n",
       "173689  VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...   2.7.7.48     178  \n",
       "173690  MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...   3.6.1.12     171  \n",
       "173691  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...   4.1.1.76     240  \n",
       "173692  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...   3.6.1.12     172  \n",
       "173693  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...   3.1.1.29     124  \n",
       "\n",
       "[161738 rows x 5 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[['Entry', 'Entry Name', 'Sequence', 'EC number', 'Length']]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4607689a",
   "metadata": {},
   "source": [
    "### Comparison to the dataset used in CLEAN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9d4113b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P51451</td>\n",
       "      <td>2.7.10.2</td>\n",
       "      <td>MGLVSSKKPDKEKPIKEKDKGQWSPLKVSAQDKDAPPLPPLVVFNH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C7C422</td>\n",
       "      <td>3.5.2.6</td>\n",
       "      <td>MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q07457</td>\n",
       "      <td>2.3.2.27</td>\n",
       "      <td>MTAEPATKKIKLELSDPSEPLTQSDVIAFQKEALFRCINRRRVDFE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q07009</td>\n",
       "      <td>3.4.22.53</td>\n",
       "      <td>MAGIAMKLAKDREAAEGLGSHERAIKYLNQDYETLRNECLEAGALF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q9TTH8</td>\n",
       "      <td>3.4.22.54</td>\n",
       "      <td>MPTVISASVAPRTGAEPRSPGPIAQAAQGKGTEAGGGNPSGIYSAI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227357</th>\n",
       "      <td>Q97HD1</td>\n",
       "      <td>3.1.11.6</td>\n",
       "      <td>MPSKKESYESMIKELEKIVSSMENEELPLEEAMKNYEDGVKLCDKL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227358</th>\n",
       "      <td>A0A140CUL9</td>\n",
       "      <td>1.1.99.13</td>\n",
       "      <td>MAENNTYDAIVIGSGISGGWATKELTEKGLKTLMLERGRNIEHVKD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227359</th>\n",
       "      <td>W6JQJ6</td>\n",
       "      <td>1.13.12.2</td>\n",
       "      <td>MNKNNRHPADGKKPITIFGPDFPFAFDDWLEHPAGLGSIPAARHGE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227360</th>\n",
       "      <td>B3IVI7</td>\n",
       "      <td>3.5.1.30</td>\n",
       "      <td>MRIALYQGAPKPLDVPGNLQRLRHQAQLAAERGAQLLVCPEMFLTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227361</th>\n",
       "      <td>Q88QV2</td>\n",
       "      <td>3.5.1.30</td>\n",
       "      <td>MRIALYQGAPKPLDVPGNLQRLRHQAQLAADRGAQLLVCPEMFLSG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>191488 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry  EC number  \\\n",
       "0           P51451   2.7.10.2   \n",
       "1           C7C422    3.5.2.6   \n",
       "2           Q07457   2.3.2.27   \n",
       "3           Q07009  3.4.22.53   \n",
       "4           Q9TTH8  3.4.22.54   \n",
       "...            ...        ...   \n",
       "227357      Q97HD1   3.1.11.6   \n",
       "227358  A0A140CUL9  1.1.99.13   \n",
       "227359      W6JQJ6  1.13.12.2   \n",
       "227360      B3IVI7   3.5.1.30   \n",
       "227361      Q88QV2   3.5.1.30   \n",
       "\n",
       "                                                 Sequence  \n",
       "0       MGLVSSKKPDKEKPIKEKDKGQWSPLKVSAQDKDAPPLPPLVVFNH...  \n",
       "1       MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...  \n",
       "2       MTAEPATKKIKLELSDPSEPLTQSDVIAFQKEALFRCINRRRVDFE...  \n",
       "3       MAGIAMKLAKDREAAEGLGSHERAIKYLNQDYETLRNECLEAGALF...  \n",
       "4       MPTVISASVAPRTGAEPRSPGPIAQAAQGKGTEAGGGNPSGIYSAI...  \n",
       "...                                                   ...  \n",
       "227357  MPSKKESYESMIKELEKIVSSMENEELPLEEAMKNYEDGVKLCDKL...  \n",
       "227358  MAENNTYDAIVIGSGISGGWATKELTEKGLKTLMLERGRNIEHVKD...  \n",
       "227359  MNKNNRHPADGKKPITIFGPDFPFAFDDWLEHPAGLGSIPAARHGE...  \n",
       "227360  MRIALYQGAPKPLDVPGNLQRLRHQAQLAAERGAQLLVCPEMFLTG...  \n",
       "227361  MRIALYQGAPKPLDVPGNLQRLRHQAQLAADRGAQLLVCPEMFLSG...  \n",
       "\n",
       "[191488 rows x 3 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_df = pd.read_csv('/disk1/jyang4/repos/CLEAN/app/data/split100.csv', sep='\\t')\n",
    "clean_df = clean_df.drop_duplicates(subset=['Sequence', 'EC number'])\n",
    "clean_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3cdeac1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P51451</td>\n",
       "      <td>2.7.10.2</td>\n",
       "      <td>MGLVSSKKPDKEKPIKEKDKGQWSPLKVSAQDKDAPPLPPLVVFNH...</td>\n",
       "      <td>505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C7C422</td>\n",
       "      <td>3.5.2.6</td>\n",
       "      <td>MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...</td>\n",
       "      <td>270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P00917</td>\n",
       "      <td>4.2.1.1</td>\n",
       "      <td>MAHSDWGYDSPNGPZEWVKLYPIANGNNQSPIDIKTSETKHDTSLK...</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q9ULX7</td>\n",
       "      <td>4.2.1.1</td>\n",
       "      <td>MLFSALLLEVIWILAADGGQHWTYEGPHGQDHWPASYPECGNNAQS...</td>\n",
       "      <td>337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>P10506</td>\n",
       "      <td>2.7.12.2</td>\n",
       "      <td>MFKRRRNPKGLVLNPNASVKSSDNDHKEELINNQKSFESNVEAFME...</td>\n",
       "      <td>340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227341</th>\n",
       "      <td>Q9KQH9</td>\n",
       "      <td>2.3.1.179</td>\n",
       "      <td>MSKRRVVVTGMGMLSPVGNTVESSWKALLAGQSGIVNIEHFDTTNF...</td>\n",
       "      <td>414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227352</th>\n",
       "      <td>Q11WK3</td>\n",
       "      <td>6.3.5.2</td>\n",
       "      <td>MTEKILILDFGSQYTQLIARRVRELNVYCEIHPYNKAPQIDGTVKG...</td>\n",
       "      <td>509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227354</th>\n",
       "      <td>B1KNT2</td>\n",
       "      <td>4.2.1.59</td>\n",
       "      <td>MSEQLNTMDIKEIMSFLPHRYPFLLIDRVLDYTPGETLHAIKNVTI...</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227360</th>\n",
       "      <td>B3IVI7</td>\n",
       "      <td>3.5.1.30</td>\n",
       "      <td>MRIALYQGAPKPLDVPGNLQRLRHQAQLAAERGAQLLVCPEMFLTG...</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227361</th>\n",
       "      <td>Q88QV2</td>\n",
       "      <td>3.5.1.30</td>\n",
       "      <td>MRIALYQGAPKPLDVPGNLQRLRHQAQLAADRGAQLLVCPEMFLSG...</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>151801 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Entry  EC number                                           Sequence  \\\n",
       "0       P51451   2.7.10.2  MGLVSSKKPDKEKPIKEKDKGQWSPLKVSAQDKDAPPLPPLVVFNH...   \n",
       "1       C7C422    3.5.2.6  MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...   \n",
       "6       P00917    4.2.1.1  MAHSDWGYDSPNGPZEWVKLYPIANGNNQSPIDIKTSETKHDTSLK...   \n",
       "7       Q9ULX7    4.2.1.1  MLFSALLLEVIWILAADGGQHWTYEGPHGQDHWPASYPECGNNAQS...   \n",
       "11      P10506   2.7.12.2  MFKRRRNPKGLVLNPNASVKSSDNDHKEELINNQKSFESNVEAFME...   \n",
       "...        ...        ...                                                ...   \n",
       "227341  Q9KQH9  2.3.1.179  MSKRRVVVTGMGMLSPVGNTVESSWKALLAGQSGIVNIEHFDTTNF...   \n",
       "227352  Q11WK3    6.3.5.2  MTEKILILDFGSQYTQLIARRVRELNVYCEIHPYNKAPQIDGTVKG...   \n",
       "227354  B1KNT2   4.2.1.59  MSEQLNTMDIKEIMSFLPHRYPFLLIDRVLDYTPGETLHAIKNVTI...   \n",
       "227360  B3IVI7   3.5.1.30  MRIALYQGAPKPLDVPGNLQRLRHQAQLAAERGAQLLVCPEMFLTG...   \n",
       "227361  Q88QV2   3.5.1.30  MRIALYQGAPKPLDVPGNLQRLRHQAQLAADRGAQLLVCPEMFLSG...   \n",
       "\n",
       "        Length  \n",
       "0          505  \n",
       "1          270  \n",
       "6          261  \n",
       "7          337  \n",
       "11         340  \n",
       "...        ...  \n",
       "227341     414  \n",
       "227352     509  \n",
       "227354     152  \n",
       "227360     264  \n",
       "227361     264  \n",
       "\n",
       "[151801 rows x 4 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_df['Length'] = [len(seq) for seq in clean_df['Sequence']]\n",
    "clean_df = clean_df[clean_df['Length'] < 512]\n",
    "clean_df = clean_df[clean_df['Length'] >= 100]\n",
    "clean_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5250f2c8",
   "metadata": {},
   "source": [
    "### Perform clustering using mmseqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "19acec52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate a fasta file as input to mmseqs\n",
    "with open('clustering/swissprot.fasta', 'w') as f:\n",
    "    for index, row in df.iterrows():\n",
    "        f.write(f'>{row[\"Entry\"]}\\n{row[\"Sequence\"]}\\n')\n",
    "\n",
    "# !mmseqs createdb examples/swissprot.fasta swissprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "5e6cdd85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes50 clustering/mmseqs_tmp --min-seq-id 0.5 -c 0.8 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.5\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/2440689462319922810/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[161702] 0s 44ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 45ms\n",
      "Create directory clustering/mmseqs_tmp/2440689462319922810/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu clustering/mmseqs_tmp/2440689462319922810/clu_tmp --max-seqs 20 -c 0.8 --cov-mode 1 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.5 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 3.000000\n",
      "Set cluster mode GREEDY MEM\n",
      "Set cluster iterations to 3\n",
      "linclust clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 161738 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 100.00% 161.74K 1s 183ms    \n",
      "Sort kmer 0h 0m 0s 329ms\n",
      "Sort by rep. sequence 0h 0m 0s 244ms\n",
      "Time for fill: 0h 0m 0s 56ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 2s 638ms\n",
      "rescorediagonal clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 161.74K 0s 303ms    \n",
      "Time for merging to pref_rescore1: 0h 0m 0s 50ms                  ] 70.29% 113.69K eta 0s       \n",
      "Time for processing: 0h 0m 1s 368ms\n",
      "clust clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref_rescore1 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pre_clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 832ms\n",
      "\n",
      "Size of the sequence database: 161738\n",
      "Size of the alignment database: 161738\n",
      "Number of clusters: 55427\n",
      "\n",
      "Writing results 0h 0m 0s 14ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 884ms\n",
      "createsubdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/order_redundancy clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 19ms\n",
      "createsubdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/order_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 25ms\n",
      "filterdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref_filter1 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref_filter2 --filter-file clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/order_redundancy --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 100.00% 55.43K 0s 695ms    \n",
      "Time for merging to pref_filter2: 0h 0m 0s 24ms\n",
      "Time for processing: 0h 0m 1s 245ms\n",
      "rescorediagonal clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/input_step_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/input_step_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref_filter2 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 55.43K 0s 126ms    \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 28ms                  ] 5.32% 2.95K eta 2s         \n",
      "Time for processing: 0h 0m 1s 408ms\n",
      "align clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/input_step_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/input_step_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref_rescore2 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 55427 type: Aminoacid\n",
      "Target database size: 55427 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 55.43K 2s 422ms    \n",
      "Time for merging to aln: 0h 0m 0s 35ms\n",
      "114071 alignments calculated\n",
      "86352 sequence pairs passed the thresholds (0.757002 of overall calculated)\n",
      "1.557941 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 888ms\n",
      "clust clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/input_step_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/aln clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 676ms\n",
      "\n",
      "Size of the sequence database: 55427\n",
      "Size of the alignment database: 55427\n",
      "Number of clusters: 39480\n",
      "\n",
      "Writing results 0h 0m 0s 7ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 699ms\n",
      "mergeclusters clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pre_clust clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/clust --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 55.43K 0s 743ms    \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 39.48K 0s 967ms    \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 161.74K 1s 427ms     K eta 2s       \n",
      "Time for merging to clu_redundancy: 0h 0m 0s 28ms\n",
      "Time for processing: 0h 0m 1s 521ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/linclust/12313843706185258789/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_redundancy clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 18ms\n",
      "prefilter clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 39480 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 39480 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 39.48K 1s 384ms    =====>                     ] 66.39% 26.21K eta 0s       \n",
      "Index table: Masked residues: 45467\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 39.48K 0s 285ms                  ] 21.86% 8.63K eta 0s                                         ] 43.55% 17.19K eta 0s       \n",
      "Index statistics\n",
      "Entries:          5463393\n",
      "DB size:          519 MB\n",
      "Avg k-mer size:   0.085366\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t159\n",
      "    AYPWGT\t121\n",
      "    GRGHCS\t97\n",
      "    GPGGTQ\t91\n",
      "    SPGGGW\t88\n",
      "    HGGDFP\t79\n",
      "    ITVGGP\t76\n",
      "    PHEEQS\t74\n",
      "    DPAGIQ\t69\n",
      "    PARFLR\t66\n",
      "Time for index table init: 0h 0m 3s 223ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 39480\n",
      "Target db start 1 to 39480\n",
      "[=================================================================] 100.00% 39.48K 1s 87ms     \n",
      "[======================================================>          ] 83.58% 33.00K eta 0s       \n",
      "1.687239 k-mers per position\n",
      "311 DB matches per sequence\n",
      "0 overflows\n",
      "11 sequences passed prefiltering per query sequence\n",
      "11 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 24ms\n",
      "Time for processing: 0h 0m 6s 890ms\n",
      "align clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/pref_step0 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 39480 type: Aminoacid\n",
      "Target database size: 39480 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 39.48K 4s 49ms     \n",
      "Time for merging to aln_step0: 0h 0m 0s 22ms\n",
      "432164 alignments calculated\n",
      "113453 sequence pairs passed the thresholds (0.262523 of overall calculated)\n",
      "2.873683 hits per query sequence\n",
      "Time for processing: 0h 0m 4s 552ms\n",
      "clust clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/aln_step0 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_step0 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 738ms\n",
      "\n",
      "Size of the sequence database: 39480\n",
      "Size of the alignment database: 39480\n",
      "Number of clusters: 26793\n",
      "\n",
      "Writing results 0h 0m 0s 6ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 767ms\n",
      "createsubdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_step0 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 10ms\n",
      "prefilter clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step1 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step1 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 2 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 26793 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 26793 type: Aminoacid\n",
      "Index table k-mer threshold: 145 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 26.79K 1s 128ms    \n",
      "Index table: Masked residues: 37535\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 26.79K 0s 303ms    ===============>                                  ] 46.38% 12.43K eta 0s       ============>     ] 90.94% 24.37K eta 0s       \n",
      "Index statistics\n",
      "Entries:          5825903\n",
      "DB size:          521 MB\n",
      "Avg k-mer size:   0.091030\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t89\n",
      "    GQQVAR\t77\n",
      "    STYLGG\t53\n",
      "    GSELFN\t53\n",
      "    AYPWGT\t53\n",
      "    FPHNAQ\t49\n",
      "    FREPGR\t49\n",
      "    GRGHCS\t48\n",
      "    TGNKTT\t44\n",
      "    SPGGGW\t44\n",
      "Time for index table init: 0h 0m 3s 31ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 145\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 26793\n",
      "Target db start 1 to 26793\n",
      "[=================================================================] 100.00% 26.79K 1s 695ms    \n",
      "\n",
      "7.316375 k-mers per position\n",
      "487 DB matches per sequence\n",
      "0 overflows\n",
      "11 sequences passed prefiltering per query sequence\n",
      "10 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 21ms\n",
      "Time for processing: 0h 0m 7s 394ms\n",
      "align clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step1 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step1 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/pref_step1 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 26793 type: Aminoacid\n",
      "Target database size: 26793 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 26.79K 3s 595ms    \n",
      "Time for merging to aln_step1: 0h 0m 0s 20ms\n",
      "279892 alignments calculated\n",
      "28922 sequence pairs passed the thresholds (0.103333 of overall calculated)\n",
      "1.079461 hits per query sequence\n",
      "Time for processing: 0h 0m 3s 996ms\n",
      "clust clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step1 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/aln_step1 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_step1 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 685ms\n",
      "\n",
      "Size of the sequence database: 26793\n",
      "Size of the alignment database: 26793\n",
      "Number of clusters: 25786\n",
      "\n",
      "Writing results 0h 0m 0s 5ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 706ms\n",
      "createsubdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_step1 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step1 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "prefilter clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step2 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step2 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 3 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 25786 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 25786 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 25.79K 1s 144ms    \n",
      "Index table: Masked residues: 36413\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 25.79K 0s 406ms    \n",
      "Index statistics\n",
      "Entries:          7609220\n",
      "DB size:          531 MB\n",
      "Avg k-mer size:   0.118894\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t86\n",
      "    GQQVAR\t69\n",
      "    STYLGG\t53\n",
      "    GSELFN\t53\n",
      "    GGSEVS\t51\n",
      "    GRGHCS\t48\n",
      "    AYPWGT\t48\n",
      "    SGGSLR\t44\n",
      "    TGNKTT\t44\n",
      "    SPGGGW\t44\n",
      "Time for index table init: 0h 0m 3s 264ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 25786\n",
      "Target db start 1 to 25786\n",
      "[=================================================================] 100.00% 25.79K 1s 696ms    \n",
      "\n",
      "22.067970 k-mers per position\n",
      "1123 DB matches per sequence\n",
      "0 overflows\n",
      "14 sequences passed prefiltering per query sequence\n",
      "17 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 18ms\n",
      "Time for processing: 0h 0m 7s 999ms\n",
      "align clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step2 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step2 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/pref_step2 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 25786 type: Aminoacid\n",
      "Target database size: 25786 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 25.79K 3s 367ms    \n",
      "Time for merging to aln_step2: 0h 0m 0s 22ms\n",
      "334844 alignments calculated\n",
      "25960 sequence pairs passed the thresholds (0.077529 of overall calculated)\n",
      "1.006748 hits per query sequence\n",
      "Time for processing: 0h 0m 3s 830ms\n",
      "clust clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step2 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/aln_step2 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_step2 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 807ms\n",
      "\n",
      "Size of the sequence database: 25786\n",
      "Size of the alignment database: 25786\n",
      "Number of clusters: 25680\n",
      "\n",
      "Writing results 0h 0m 0s 14ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 838ms\n",
      "mergeclusters clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_redundancy clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_step0 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_step1 clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_step2 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 39.48K 0s 694ms    \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 26.79K 0s 945ms    >                                     ] 42.36% 11.35K eta 1s       \n",
      "Clustering step 3\n",
      "[=================================================================] 100.00% 25.79K 1s 194ms    \n",
      "Clustering step 4\n",
      "[=================================================================] 100.00% 25.68K 1s 417ms    ta 1s       \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 161.74K 1s 879ms        \n",
      "Time for merging to clu: 0h 0m 0s 18ms\n",
      "Time for processing: 0h 0m 1s 981ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/pref_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/aln_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/pref_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/aln_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/clu_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step1_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_tmp/3339389697974233988/input_step2_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu clustering/mmseqs_tmp/2440689462319922810/cluster.tsv --threads 384 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 37ms\n",
      "Time for processing: 0h 0m 0s 998ms\n",
      "result2repseq clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu clustering/mmseqs_tmp/2440689462319922810/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 25.68K 0s 371ms             ] 54.93% 14.11K eta 0s       \n",
      "Time for merging to clu_rep: 0h 0m 0s 35ms\n",
      "Time for processing: 0h 0m 0s 807ms\n",
      "result2flat clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu_rep clustering/mmseqs_tmp/2440689462319922810/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 71ms\n",
      "createseqfiledb clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu clustering/mmseqs_tmp/2440689462319922810/clu_seqs --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 25.68K 0s 988ms    \n",
      "Time for merging to clu_seqs: 0h 0m 0s 17ms\n",
      "Time for processing: 0h 0m 1s 527ms\n",
      "result2flat clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/input clustering/mmseqs_tmp/2440689462319922810/clu_seqs clustering/mmseqs_tmp/2440689462319922810/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 227ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 15ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering/mmseqs_tmp/2440689462319922810/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster clustering/swissprot.fasta clustering/clusterRes50 clustering/mmseqs_tmp --min-seq-id 0.5 -c 0.8 --cov-mode 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "286e6959",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster clustering/swissprot.fasta clustering/clusterRes30 clustering/mmseqs_tmp --min-seq-id 0.3 -c 0.8 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t15.6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t384\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.3\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb clustering/swissprot.fasta clustering/mmseqs_tmp/4220174760281188982/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[161702] 0s 46ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 46ms\n",
      "Create directory clustering/mmseqs_tmp/4220174760281188982/clu_tmp\n",
      "cluster clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu clustering/mmseqs_tmp/4220174760281188982/clu_tmp --max-seqs 20 -c 0.8 --cov-mode 1 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.3 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 5.000000\n",
      "Set cluster mode GREEDY MEM\n",
      "Set cluster iterations to 3\n",
      "linclust clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 161738 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 100.00% 161.74K 1s 530ms          ] 74.85% 121.06K eta 0s       \n",
      "Sort kmer 0h 0m 0s 330ms\n",
      "Sort by rep. sequence 0h 0m 0s 250ms\n",
      "Time for fill: 0h 0m 0s 65ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 3s 52ms\n",
      "rescorediagonal clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 161.74K 0s 329ms    \n",
      "Time for merging to pref_rescore1: 0h 0m 0s 57ms                  ] 4.93% 7.98K eta 17s         \n",
      "Time for processing: 0h 0m 1s 516ms\n",
      "clust clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref_rescore1 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pre_clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 914ms\n",
      "\n",
      "Size of the sequence database: 161738\n",
      "Size of the alignment database: 161738\n",
      "Number of clusters: 55427\n",
      "\n",
      "Writing results 0h 0m 0s 14ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 965ms\n",
      "createsubdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/order_redundancy clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 19ms\n",
      "createsubdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/order_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 25ms\n",
      "filterdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref_filter1 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref_filter2 --filter-file clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/order_redundancy --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 100.00% 55.43K 0s 503ms    \n",
      "Time for merging to pref_filter2: 0h 0m 0s 41ms\n",
      "Time for processing: 0h 0m 1s 141ms\n",
      "rescorediagonal clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/input_step_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/input_step_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref_filter2 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 55.43K 0s 156ms    \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 23ms                  ] 70.01% 38.80K eta 0s       \n",
      "Time for processing: 0h 0m 1s 341ms\n",
      "align clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/input_step_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/input_step_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref_rescore2 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 55427 type: Aminoacid\n",
      "Target database size: 55427 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 55.43K 3s 246ms    \n",
      "Time for merging to aln: 0h 0m 0s 26ms\n",
      "119517 alignments calculated\n",
      "115196 sequence pairs passed the thresholds (0.963846 of overall calculated)\n",
      "2.078337 hits per query sequence\n",
      "Time for processing: 0h 0m 3s 702ms\n",
      "clust clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/input_step_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/aln clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 892ms\n",
      "\n",
      "Size of the sequence database: 55427\n",
      "Size of the alignment database: 55427\n",
      "Number of clusters: 30779\n",
      "\n",
      "Writing results 0h 0m 0s 5ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 915ms\n",
      "mergeclusters clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pre_clust clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/clust --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 55.43K 0s 473ms    \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 30.78K 0s 750ms                       ] 22.85% 7.03K eta 2s       \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 161.74K 1s 226ms     ===================>                                      ] 40.23% 65.07K eta 1s       \n",
      "Time for merging to clu_redundancy: 0h 0m 0s 24ms\n",
      "Time for processing: 0h 0m 1s 330ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/linclust/1781055975559886451/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_redundancy clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 17ms\n",
      "prefilter clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 30779 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 30779 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 30.78K 1s 235ms    \n",
      "Index table: Masked residues: 38160\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 30.78K 0s 253ms    ======>           ] 81.88% 25.20K eta 0s       \n",
      "Index statistics\n",
      "Entries:          4318506\n",
      "DB size:          512 MB\n",
      "Avg k-mer size:   0.067477\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t116\n",
      "    AYPWGT\t74\n",
      "    IHDKNI\t67\n",
      "    GPGGTQ\t63\n",
      "    FREPGR\t50\n",
      "    TFPHHL\t48\n",
      "    GRGHCS\t48\n",
      "    ITVGGP\t44\n",
      "    EARPGP\t43\n",
      "    PVGVMT\t43\n",
      "Time for index table init: 0h 0m 3s 119ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 30779\n",
      "Target db start 1 to 30779\n",
      "[=================================================================] 100.00% 30.78K 1s 207ms    \n",
      "\n",
      "1.758042 k-mers per position\n",
      "249 DB matches per sequence\n",
      "0 overflows\n",
      "8 sequences passed prefiltering per query sequence\n",
      "6 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 17ms\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for processing: 0h 0m 8s 9ms\n",
      "align clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/pref_step0 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 30779 type: Aminoacid\n",
      "Target database size: 30779 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 30.78K 3s 651ms    \n",
      "Time for merging to aln_step0: 0h 0m 0s 20ms\n",
      "259896 alignments calculated\n",
      "224932 sequence pairs passed the thresholds (0.865469 of overall calculated)\n",
      "7.307970 hits per query sequence\n",
      "Time for processing: 0h 0m 4s 131ms\n",
      "clust clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/aln_step0 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_step0 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 734ms\n",
      "\n",
      "Size of the sequence database: 30779\n",
      "Size of the alignment database: 30779\n",
      "Number of clusters: 13402\n",
      "\n",
      "Writing results 0h 0m 0s 7ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 755ms\n",
      "createsubdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_step0 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "prefilter clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step1 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step1 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 3 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 13402 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 13402 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 13.40K 0s 985ms    \n",
      "Index table: Masked residues: 23792\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 13.40K 0s 197ms    \n",
      "Index statistics\n",
      "Entries:          3888095\n",
      "DB size:          510 MB\n",
      "Avg k-mer size:   0.060751\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t43\n",
      "    GPGGTL\t39\n",
      "    GLGIGK\t24\n",
      "    LVGGGL\t22\n",
      "    RDKELL\t21\n",
      "    VVGGGL\t20\n",
      "    KKGARN\t19\n",
      "    GQQVAR\t19\n",
      "    LLDPAL\t15\n",
      "    HRLPIL\t15\n",
      "Time for index table init: 0h 0m 2s 920ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 13402\n",
      "Target db start 1 to 13402\n",
      "[=================================================================] 100.00% 13.40K 0s 810ms    \n",
      "\n",
      "22.357336 k-mers per position\n",
      "643 DB matches per sequence\n",
      "0 overflows\n",
      "7 sequences passed prefiltering per query sequence\n",
      "6 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 15ms\n",
      "Time for processing: 0h 0m 7s 27ms\n",
      "align clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step1 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step1 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/pref_step1 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 13402 type: Aminoacid\n",
      "Target database size: 13402 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 13.40K 2s 635ms    \n",
      "Time for merging to aln_step1: 0h 0m 0s 20ms\n",
      "83309 alignments calculated\n",
      "37687 sequence pairs passed the thresholds (0.452376 of overall calculated)\n",
      "2.812043 hits per query sequence\n",
      "Time for processing: 0h 0m 3s 119ms\n",
      "clust clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step1 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/aln_step1 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_step1 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 718ms\n",
      "\n",
      "Size of the sequence database: 13402\n",
      "Size of the alignment database: 13402\n",
      "Number of clusters: 8895\n",
      "\n",
      "Writing results 0h 0m 0s 3ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 737ms\n",
      "createsubdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_step1 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step1 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "prefilter clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step2 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step2 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 5 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 8895 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 8895 type: Aminoacid\n",
      "Index table k-mer threshold: 118 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 8.90K 1s 22ms     \n",
      "Index table: Masked residues: 20699\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 8.90K 0s 170ms    \n",
      "Index statistics\n",
      "Entries:          2839073\n",
      "DB size:          504 MB\n",
      "Avg k-mer size:   0.044361\n",
      "Top 10 k-mers\n",
      "    IHDKNI\t41\n",
      "    VHDKNI\t21\n",
      "    RDKELL\t20\n",
      "    GPGGTL\t17\n",
      "    AVGGGL\t15\n",
      "    HRIPIL\t14\n",
      "    HRLPIL\t14\n",
      "    VHDKNL\t14\n",
      "    QCGGGP\t13\n",
      "    RVGLSR\t13\n",
      "Time for index table init: 0h 0m 2s 762ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 118\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 8895\n",
      "Target db start 1 to 8895\n",
      "[=================================================================] 100.00% 8.90K 1s 787ms    \n",
      "\n",
      "177.929010 k-mers per position\n",
      "2688 DB matches per sequence\n",
      "0 overflows\n",
      "16 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 16ms\n",
      "Time for processing: 0h 0m 7s 703ms\n",
      "align clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step2 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step2 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/pref_step2 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 8895 type: Aminoacid\n",
      "Target database size: 8895 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 8.90K 2s 142ms    \n",
      "Time for merging to aln_step2: 0h 0m 0s 20ms\n",
      "110245 alignments calculated\n",
      "12039 sequence pairs passed the thresholds (0.109202 of overall calculated)\n",
      "1.353457 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 575ms\n",
      "clust clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step2 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/aln_step2 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_step2 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 384 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 728ms\n",
      "\n",
      "Size of the sequence database: 8895\n",
      "Size of the alignment database: 8895\n",
      "Number of clusters: 7760\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 741ms\n",
      "mergeclusters clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_redundancy clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_step0 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_step1 clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_step2 --threads 384 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 30.78K 0s 648ms    \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 13.40K 0s 868ms    \n",
      "Clustering step 3\n",
      "[=================================================================] 100.00% 8.90K 1s 100ms    \n",
      "Clustering step 4\n",
      "[=================================================================] 100.00% 7.76K 1s 264ms    \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 161.74K 1s 683ms     ==========>        ] 86.35% 139.65K eta 0s       \n",
      "Time for merging to clu: 0h 0m 0s 15ms\n",
      "Time for processing: 0h 0m 1s 774ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/pref_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/aln_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/pref_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/aln_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/clu_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step1_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_tmp/12773599051404603500/input_step2_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu clustering/mmseqs_tmp/4220174760281188982/cluster.tsv --threads 384 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 20ms\n",
      "Time for processing: 0h 0m 0s 552ms\n",
      "result2repseq clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu clustering/mmseqs_tmp/4220174760281188982/clu_rep --db-load-mode 0 --compressed 0 --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 7.76K 0s 211ms              ] 77.59% 6.02K eta 0s       \n",
      "Time for merging to clu_rep: 0h 0m 0s 24ms\n",
      "Time for processing: 0h 0m 0s 609ms\n",
      "result2flat clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu_rep clustering/mmseqs_tmp/4220174760281188982/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 45ms\n",
      "createseqfiledb clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu clustering/mmseqs_tmp/4220174760281188982/clu_seqs --threads 384 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 7.76K 0s 942ms    \n",
      "Time for merging to clu_seqs: 0h 0m 0s 13ms\n",
      "Time for processing: 0h 0m 1s 392ms\n",
      "result2flat clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/input clustering/mmseqs_tmp/4220174760281188982/clu_seqs clustering/mmseqs_tmp/4220174760281188982/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 218ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 14ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering/mmseqs_tmp/4220174760281188982/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster clustering/swissprot.fasta clustering/clusterRes30 clustering/mmseqs_tmp --min-seq-id 0.3 -c 0.8 --cov-mode 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "02322b64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25680\n",
      "7760\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>EC number</th>\n",
       "      <th>Length</th>\n",
       "      <th>clusterRes50</th>\n",
       "      <th>clusterRes30</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>ABTIR_ACIB9</td>\n",
       "      <td>MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...</td>\n",
       "      <td>3.2.2.6</td>\n",
       "      <td>269</td>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>A1AY86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>CUTI1_HYPJR</td>\n",
       "      <td>MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...</td>\n",
       "      <td>3.1.1.74</td>\n",
       "      <td>248</td>\n",
       "      <td>A0A024SC78</td>\n",
       "      <td>A0A024SC78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024SH76</td>\n",
       "      <td>GUX2_HYPJR</td>\n",
       "      <td>MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...</td>\n",
       "      <td>3.2.1.91</td>\n",
       "      <td>471</td>\n",
       "      <td>G4MM92</td>\n",
       "      <td>G4MM92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CCR1_PETHY</td>\n",
       "      <td>MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...</td>\n",
       "      <td>1.2.1.44</td>\n",
       "      <td>333</td>\n",
       "      <td>Q9S9N9</td>\n",
       "      <td>P14721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A061I403</td>\n",
       "      <td>FICD_CRIGR</td>\n",
       "      <td>MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...</td>\n",
       "      <td>2.7.7.108</td>\n",
       "      <td>455</td>\n",
       "      <td>Q6AY47</td>\n",
       "      <td>Q6AY47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161733</th>\n",
       "      <td>P36352</td>\n",
       "      <td>POLR_PHMV</td>\n",
       "      <td>VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...</td>\n",
       "      <td>2.7.7.48</td>\n",
       "      <td>178</td>\n",
       "      <td>P36352</td>\n",
       "      <td>P36352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161734</th>\n",
       "      <td>P39262</td>\n",
       "      <td>VG56_BPT4</td>\n",
       "      <td>MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>171</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161735</th>\n",
       "      <td>Q05115</td>\n",
       "      <td>AMDA_BORBO</td>\n",
       "      <td>MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...</td>\n",
       "      <td>4.1.1.76</td>\n",
       "      <td>240</td>\n",
       "      <td>Q05115</td>\n",
       "      <td>Q05115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161736</th>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>VG56_BPLZ5</td>\n",
       "      <td>MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...</td>\n",
       "      <td>3.6.1.12</td>\n",
       "      <td>172</td>\n",
       "      <td>Q94MV8</td>\n",
       "      <td>Q94MV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161737</th>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>V029_FOWPN</td>\n",
       "      <td>MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...</td>\n",
       "      <td>3.1.1.29</td>\n",
       "      <td>124</td>\n",
       "      <td>Q9J5H2</td>\n",
       "      <td>Q54ZD0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161738 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry   Entry Name  \\\n",
       "0       A0A009IHW8  ABTIR_ACIB9   \n",
       "1       A0A024SC78  CUTI1_HYPJR   \n",
       "2       A0A024SH76   GUX2_HYPJR   \n",
       "3       A0A059TC02   CCR1_PETHY   \n",
       "4       A0A061I403   FICD_CRIGR   \n",
       "...            ...          ...   \n",
       "161733      P36352    POLR_PHMV   \n",
       "161734      P39262    VG56_BPT4   \n",
       "161735      Q05115   AMDA_BORBO   \n",
       "161736      Q94MV8   VG56_BPLZ5   \n",
       "161737      Q9J5H2   V029_FOWPN   \n",
       "\n",
       "                                                 Sequence  EC number  Length  \\\n",
       "0       MSLEQKKGADIISKILQIQNSIGKTTSPSTLKTKLSEISRKEQENA...    3.2.2.6     269   \n",
       "1       MRSLAILTTLLAGHAFAYPKPAPQSVNRRDWPSINEFLSELAKVMP...   3.1.1.74     248   \n",
       "2       MIVGILTTLATLATLAASVPLEERQACSSVWGQCGGQNWSGPTCCA...   3.2.1.91     471   \n",
       "3       MRSVSGQVVCVTGAGGFIASWLVKILLEKGYTVRGTVRNPDDPKNG...   1.2.1.44     333   \n",
       "4       MPMASVIAVAEPKWISVWGRFLWLTLLSMALGSLLALLLPLGAVEE...  2.7.7.108     455   \n",
       "...                                                   ...        ...     ...   \n",
       "161733  VIVGTPPISPNWPAIKDLLHLKFKTEITSSPLFCGYYLSPAGCIRN...   2.7.7.48     178   \n",
       "161734  MAHFNECAHLIEGVDKAQNEYWDILGDEKDPLQVMLDMQRFLQIRL...   3.6.1.12     171   \n",
       "161735  MQQASTPTIGMIVPPAAGLVPADGARLYPDLPFIASGLGLGSVTPE...   4.1.1.76     240   \n",
       "161736  MAHFNECAHLIEGVDKANRAYAENIMHNIDPLQVMLDMQRHLQIRL...   3.6.1.12     172   \n",
       "161737  MANVFDNSSYRDMLKMVFVIRDDLKMTKGEIVSQCCHGAISAYEKS...   3.1.1.29     124   \n",
       "\n",
       "       clusterRes50 clusterRes30  \n",
       "0        A0A009IHW8       A1AY86  \n",
       "1        A0A024SC78   A0A024SC78  \n",
       "2            G4MM92       G4MM92  \n",
       "3            Q9S9N9       P14721  \n",
       "4            Q6AY47       Q6AY47  \n",
       "...             ...          ...  \n",
       "161733       P36352       P36352  \n",
       "161734       Q94MV8       Q94MV8  \n",
       "161735       Q05115       Q05115  \n",
       "161736       Q94MV8       Q94MV8  \n",
       "161737       Q9J5H2       Q54ZD0  \n",
       "\n",
       "[161738 rows x 7 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjkAAAGdCAYAAADwjmIIAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuNSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/xnp5ZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAApl0lEQVR4nO3df1TUZaLH8Q8gA5IO+CNAVlTKEn9ruOH061oiaJw2y+NJ83apyE5euKvStbQt09wubW2WleXplrr3rG7pns0tdZUJUzPxFysplmZla20Otpnij4QRnvtHh+82gj9wh2Ae369zOMd5nofvPJ/5svTZmfkOYcYYIwAAAMuEN/cGAAAAmgIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgpVbNvYHmVFtbq6+//lpt27ZVWFhYc28HAACcB2OMjh49qqSkJIWHn/n5mou65Hz99ddKTk5u7m0AAIAL8OWXX6pz585nnL+oS07btm0l/fAgud3uoB3X7/erqKhImZmZioyMDNpxWwrb80lktIHt+SQy2sL2jE2Rr7KyUsnJyc5/x8/koi45dS9Rud3uoJecmJgYud1ua39gbc4nkdEGtueTyGgL2zM2Zb5zvdWENx4DAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWKlVc2/AZn1mrFZVzT//DPwXT2U3424AALi48EwOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALBSo0pOYWGhfv7zn6tt27aKj4/XyJEjtWfPnoA1J0+eVF5enjp06KA2bdpo1KhRqqioCFizf/9+ZWdnKyYmRvHx8ZoyZYpOnToVsGbt2rW66qqrFBUVpe7du2vhwoX19jN37lx169ZN0dHRSk9P15YtWxoTBwAAWKxRJWfdunXKy8vTpk2b5PV65ff7lZmZqePHjztrJk+erHfeeUdLly7VunXr9PXXX+v222935mtqapSdna3q6mpt3LhRv/vd77Rw4UJNnz7dWbNv3z5lZ2frxhtvVFlZmSZNmqT77rtPq1evdta8+eabKigo0OOPP66//vWv6t+/v7KysnTw4MF/5fEAAACWaNWYxatWrQq4vXDhQsXHx6u0tFQ33HCDjhw5otdff12LFy/WTTfdJElasGCBevbsqU2bNmnw4MEqKirSRx99pHfffVcJCQkaMGCAZs2apYcfflgzZsyQy+XSvHnzlJKSomeffVaS1LNnT23YsEHPPfecsrKyJEmzZ8/W+PHjdc8990iS5s2bpxUrVmj+/PmaOnXqv/zAAACA0NaoknO6I0eOSJLat28vSSotLZXf71dGRoazJjU1VV26dFFJSYkGDx6skpIS9e3bVwkJCc6arKwsTZgwQbt27dLAgQNVUlIScIy6NZMmTZIkVVdXq7S0VNOmTXPmw8PDlZGRoZKSkjPut6qqSlVVVc7tyspKSZLf75ff77/AR6G+umNFhZsGx0NdXQ5b8jSEjKHP9nwSGW1he8amyHe+x7rgklNbW6tJkybp2muvVZ8+fSRJPp9PLpdLcXFxAWsTEhLk8/mcNT8uOHXzdXNnW1NZWanvv/9e3333nWpqahpcs3v37jPuubCwUDNnzqw3XlRUpJiYmPNI3TizBtUG3F65cmXQ76M5eb3e5t5CkyNj6LM9n0RGW9ieMZj5Tpw4cV7rLrjk5OXlqby8XBs2bLjQQ/zkpk2bpoKCAud2ZWWlkpOTlZmZKbfbHbT78fv98nq9emxbuKpqw5zx8hlZQbuP5lSXb9iwYYqMjGzu7TQJMoY+2/NJZLSF7RmbIl/dKzHnckElJz8/X8uXL9f69evVuXNnZzwxMVHV1dU6fPhwwLM5FRUVSkxMdNacfhVU3dVXP15z+hVZFRUVcrvdat26tSIiIhQREdHgmrpjNCQqKkpRUVH1xiMjI5vkB6uqNkxVNf8sObb98DbV49aSkDH02Z5PIqMtbM8YzHzne5xGXV1ljFF+fr7eeustrVmzRikpKQHzaWlpioyMVHFxsTO2Z88e7d+/Xx6PR5Lk8Xi0c+fOgKugvF6v3G63evXq5az58THq1tQdw+VyKS0tLWBNbW2tiouLnTUAAODi1qhncvLy8rR48WL9+c9/Vtu2bZ330MTGxqp169aKjY1Vbm6uCgoK1L59e7ndbv3Xf/2XPB6PBg8eLEnKzMxUr169dNddd+npp5+Wz+fTo48+qry8POdZlgceeEAvvfSSHnroId17771as2aNlixZohUrVjh7KSgoUE5OjgYNGqSrr75azz//vI4fP+5cbQUAAC5ujSo5r7zyiiRpyJAhAeMLFizQ3XffLUl67rnnFB4erlGjRqmqqkpZWVl6+eWXnbURERFavny5JkyYII/Ho0suuUQ5OTl64oknnDUpKSlasWKFJk+erDlz5qhz58567bXXnMvHJemOO+7QN998o+nTp8vn82nAgAFatWpVvTcjAwCAi1OjSo4x5pxroqOjNXfuXM2dO/eMa7p27XrOK42GDBmi7du3n3VNfn6+8vPzz7knAABw8eFvVwEAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGClRpec9evX65ZbblFSUpLCwsK0bNmygPm7775bYWFhAV/Dhw8PWHPo0CGNGzdObrdbcXFxys3N1bFjxwLW7NixQ9dff72io6OVnJysp59+ut5eli5dqtTUVEVHR6tv375auXJlY+MAAABLNbrkHD9+XP3799fcuXPPuGb48OE6cOCA8/WHP/whYH7cuHHatWuXvF6vli9frvXr1+v+++935isrK5WZmamuXbuqtLRUzzzzjGbMmKFXX33VWbNx40aNHTtWubm52r59u0aOHKmRI0eqvLy8sZEAAICFWjX2G0aMGKERI0acdU1UVJQSExMbnPv444+1atUqbd26VYMGDZIkvfjii7r55pv129/+VklJSVq0aJGqq6s1f/58uVwu9e7dW2VlZZo9e7ZThubMmaPhw4drypQpkqRZs2bJ6/XqpZde0rx58xobCwAAWKbRJed8rF27VvHx8WrXrp1uuukm/frXv1aHDh0kSSUlJYqLi3MKjiRlZGQoPDxcmzdv1m233aaSkhLdcMMNcrlczpqsrCz95je/0Xfffad27dqppKREBQUFAfeblZVV7+WzH6uqqlJVVZVzu7KyUpLk9/vl9/uDEd05niRFhZsGx0NdXQ5b8jSEjKHP9nwSGW1he8amyHe+xwp6yRk+fLhuv/12paSk6LPPPtMjjzyiESNGqKSkRBEREfL5fIqPjw/cRKtWat++vXw+nyTJ5/MpJSUlYE1CQoIz165dO/l8Pmfsx2vqjtGQwsJCzZw5s954UVGRYmJiLijv2cwaVBtw27b3DHm93ubeQpMjY+izPZ9ERlvYnjGY+U6cOHFe64JecsaMGeP8u2/fvurXr58uv/xyrV27VkOHDg323TXKtGnTAp79qaysVHJysjIzM+V2u4N2P36/X16vV49tC1dVbZgzXj4jK2j30Zzq8g0bNkyRkZHNvZ0mQcbQZ3s+iYy2sD1jU+SreyXmXJrk5aofu+yyy9SxY0d9+umnGjp0qBITE3Xw4MGANadOndKhQ4ec9/EkJiaqoqIiYE3d7XOtOdN7gaQf3isUFRVVbzwyMrJJfrCqasNUVfPPkmPbD29TPW4tCRlDn+35JDLawvaMwcx3vsdp8s/J+eqrr/Ttt9+qU6dOkiSPx6PDhw+rtLTUWbNmzRrV1tYqPT3dWbN+/fqA19y8Xq969Oihdu3aOWuKi4sD7svr9crj8TR1JAAAEAIaXXKOHTumsrIylZWVSZL27dunsrIy7d+/X8eOHdOUKVO0adMmffHFFyouLtatt96q7t27Kyvrh5dqevbsqeHDh2v8+PHasmWLPvjgA+Xn52vMmDFKSkqSJN15551yuVzKzc3Vrl279Oabb2rOnDkBLzVNnDhRq1at0rPPPqvdu3drxowZ2rZtm/Lz84PwsAAAgFDX6JKzbds2DRw4UAMHDpQkFRQUaODAgZo+fboiIiK0Y8cO/eIXv9CVV16p3NxcpaWl6f333w94mWjRokVKTU3V0KFDdfPNN+u6664L+Ayc2NhYFRUVad++fUpLS9ODDz6o6dOnB3yWzjXXXKPFixfr1VdfVf/+/fXHP/5Ry5YtU58+ff6VxwMAAFii0e/JGTJkiIwxZ5xfvXr1OY/Rvn17LV68+Kxr+vXrp/fff/+sa0aPHq3Ro0ef8/4AAMDFh79dBQAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACs1uuSsX79et9xyi5KSkhQWFqZly5YFzBtjNH36dHXq1EmtW7dWRkaG9u7dG7Dm0KFDGjdunNxut+Li4pSbm6tjx44FrNmxY4euv/56RUdHKzk5WU8//XS9vSxdulSpqamKjo5W3759tXLlysbGAQAAlmp0yTl+/Lj69++vuXPnNjj/9NNP64UXXtC8efO0efNmXXLJJcrKytLJkyedNePGjdOuXbvk9Xq1fPlyrV+/Xvfff78zX1lZqczMTHXt2lWlpaV65plnNGPGDL366qvOmo0bN2rs2LHKzc3V9u3bNXLkSI0cOVLl5eWNjQQAACzUqrHfMGLECI0YMaLBOWOMnn/+eT366KO69dZbJUn/93//p4SEBC1btkxjxozRxx9/rFWrVmnr1q0aNGiQJOnFF1/UzTffrN/+9rdKSkrSokWLVF1drfnz58vlcql3794qKyvT7NmznTI0Z84cDR8+XFOmTJEkzZo1S16vVy+99JLmzZt3QQ8GAACwR6NLztns27dPPp9PGRkZzlhsbKzS09NVUlKiMWPGqKSkRHFxcU7BkaSMjAyFh4dr8+bNuu2221RSUqIbbrhBLpfLWZOVlaXf/OY3+u6779SuXTuVlJSooKAg4P6zsrLqvXz2Y1VVVaqqqnJuV1ZWSpL8fr/8fv+/Gt9Rd6yocNPgeKiry2FLnoaQMfTZnk8ioy1sz9gU+c73WEEtOT6fT5KUkJAQMJ6QkODM+Xw+xcfHB26iVSu1b98+YE1KSkq9Y9TNtWvXTj6f76z305DCwkLNnDmz3nhRUZFiYmLOJ2KjzBpUG3DbtvcMeb3e5t5CkyNj6LM9n0RGW9ieMZj5Tpw4cV7rglpyWrpp06YFPPtTWVmp5ORkZWZmyu12B+1+/H6/vF6vHtsWrqraMGe8fEZW0O6jOdXlGzZsmCIjI5t7O02CjKHP9nwSGW1he8amyFf3Ssy5BLXkJCYmSpIqKirUqVMnZ7yiokIDBgxw1hw8eDDg+06dOqVDhw4535+YmKiKioqANXW3z7Wmbr4hUVFRioqKqjceGRnZJD9YVbVhqqr5Z8mx7Ye3qR63loSMoc/2fBIZbWF7xmDmO9/jBPVzclJSUpSYmKji4mJnrLKyUps3b5bH45EkeTweHT58WKWlpc6aNWvWqLa2Vunp6c6a9evXB7zm5vV61aNHD7Vr185Z8+P7qVtTdz8AAODi1uiSc+zYMZWVlamsrEzSD282Lisr0/79+xUWFqZJkybp17/+td5++23t3LlT//Ef/6GkpCSNHDlSktSzZ08NHz5c48eP15YtW/TBBx8oPz9fY8aMUVJSkiTpzjvvlMvlUm5urnbt2qU333xTc+bMCXipaeLEiVq1apWeffZZ7d69WzNmzNC2bduUn5//rz8qAAAg5DX65apt27bpxhtvdG7XFY+cnBwtXLhQDz30kI4fP677779fhw8f1nXXXadVq1YpOjra+Z5FixYpPz9fQ4cOVXh4uEaNGqUXXnjBmY+NjVVRUZHy8vKUlpamjh07avr06QGfpXPNNddo8eLFevTRR/XII4/oiiuu0LJly9SnT58LeiAAAIBdGl1yhgwZImPMGefDwsL0xBNP6Iknnjjjmvbt22vx4sVnvZ9+/frp/fffP+ua0aNHa/To0WffMAAAuCjxt6sAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVmrV3BuwWXlUriJrT/5o5Eiz7QUAgIsNz+QAAAArUXIAAICVKDkAAMBKlBwAAGCloJecGTNmKCwsLOArNTXVmT958qTy8vLUoUMHtWnTRqNGjVJFRUXAMfbv36/s7GzFxMQoPj5eU6ZM0alTpwLWrF27VldddZWioqLUvXt3LVy4MNhRAABACGuSZ3J69+6tAwcOOF8bNmxw5iZPnqx33nlHS5cu1bp16/T111/r9ttvd+ZramqUnZ2t6upqbdy4Ub/73e+0cOFCTZ8+3Vmzb98+ZWdn68Ybb1RZWZkmTZqk++67T6tXr26KOAAAIAQ1ySXkrVq1UmJiYr3xI0eO6PXXX9fixYt10003SZIWLFignj17atOmTRo8eLCKior00Ucf6d1331VCQoIGDBigWbNm6eGHH9aMGTPkcrk0b948paSk6Nlnn5Uk9ezZUxs2bNBzzz2nrKyspogEAABCTJOUnL179yopKUnR0dHyeDwqLCxUly5dVFpaKr/fr4yMDGdtamqqunTpopKSEg0ePFglJSXq27evEhISnDVZWVmaMGGCdu3apYEDB6qkpCTgGHVrJk2adNZ9VVVVqaqqyrldWVkpSfL7/fL7/UFILud4kuQPjz59Imj30ZycfJbkaQgZQ5/t+SQy2sL2jE2R73yPFfSSk56eroULF6pHjx46cOCAZs6cqeuvv17l5eXy+XxyuVyKi4sL+J6EhAT5fD5Jks/nCyg4dfN1c2dbU1lZqe+//16tW7ducG+FhYWaOXNmvfGioiLFxMRcUN6z8fZ9IXBg5cqg30dz8nq9zb2FJkfG0Gd7PomMtrA9YzDznThx4rzWBb3kjBgxwvl3v379lJ6erq5du2rJkiVnLB8/lWnTpqmgoMC5XVlZqeTkZGVmZsrtdgftfvx+v7xer4bt/GXgJx5P+ypo99GcnHzDhikyMrK5t9MkyBj6bM8nkdEWtmdsinx1r8ScS5P/WYe4uDhdeeWV+vTTTzVs2DBVV1fr8OHDAc/mVFRUOO/hSUxM1JYtWwKOUXf11Y/XnH5FVkVFhdxu91mLVFRUlKKiouqNR0ZGNskPVmTtycCSY9kPb1M9bi0JGUOf7fkkMtrC9ozBzHe+x2nyz8k5duyYPvvsM3Xq1ElpaWmKjIxUcXGxM79nzx7t379fHo9HkuTxeLRz504dPHjQWeP1euV2u9WrVy9nzY+PUbem7hgAAABBLzn//d//rXXr1umLL77Qxo0bddtttykiIkJjx45VbGyscnNzVVBQoPfee0+lpaW655575PF4NHjwYElSZmamevXqpbvuuksffvihVq9erUcffVR5eXnOszAPPPCAPv/8cz300EPavXu3Xn75ZS1ZskSTJ08OdhwAABCigv5y1VdffaWxY8fq22+/1aWXXqrrrrtOmzZt0qWXXipJeu655xQeHq5Ro0apqqpKWVlZevnll53vj4iI0PLlyzVhwgR5PB5dcsklysnJ0RNPPOGsSUlJ0YoVKzR58mTNmTNHnTt31muvvcbl4wAAwBH0kvPGG2+cdT46Olpz587V3Llzz7ima9euWnmOK5GGDBmi7du3X9AeAQCA/fjbVQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCVKDgAAsBIlBwAAWImSAwAArETJAQAAVqLkAAAAK1FyAACAlSg5AADASpQcAABgJUoOAACwEiUHAABYiZIDAACsRMkBAABWouQAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACu1au4NXEy6TV0RcPuLp7KbaScAANiPZ3IAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYCX+dtVP6IvoO08bOdIs+wAA4GLAMzkAAMBKlBwAAGAlSg4AALAS78lpRt2mrqg39sVT2c2wEwAA7MMzOQAAwEqUHAAAYCVermpG9S8pl7isHACA4Aj5Z3Lmzp2rbt26KTo6Wunp6dqyZUtzb+lf0m3qioAvAABwYUL6mZw333xTBQUFmjdvntLT0/X8888rKytLe/bsUXx8fHNv74Kc/uxOt6mL66/hzckAAJxTSJec2bNna/z48brnnnskSfPmzdOKFSs0f/58TZ06tZl3FxwNvaR1evGh9AAAUF/Ilpzq6mqVlpZq2rRpzlh4eLgyMjJUUlLS4PdUVVWpqqrKuX3kyA/vfzl06JD8fn/Q9ub3+3XixAl9W+1SZG1t0I5bpzT87oDb3z5y7u9Jr55bb2zztKEXdP9Ovm+/VWRk5AUdo6UjY+izPZ9ERlvYnrEp8h09elSSZIw567qQLTn/+Mc/VFNTo4SEhIDxhIQE7d69u8HvKSws1MyZM+uNp6SkNMkeW5Y76o10fLYZtgEAQJAcPXpUsbGxZ5wP2ZJzIaZNm6aCggLndm1trQ4dOqQOHTooLCwsaPdTWVmp5ORkffnll3K73UE7bkthez6JjDawPZ9ERlvYnrEp8hljdPToUSUlJZ11XciWnI4dOyoiIkIVFRUB4xUVFUpMTGzwe6KiohQVFRUwFhcX11RblNvttvIHto7t+SQy2sD2fBIZbWF7xmDnO9szOHVC9hJyl8ultLQ0FRcXO2O1tbUqLi6Wx+Npxp0BAICWIGSfyZGkgoIC5eTkaNCgQbr66qv1/PPP6/jx487VVgAA4OIV0iXnjjvu0DfffKPp06fL5/NpwIABWrVqVb03I//UoqKi9Pjjj9d7acwWtueTyGgD2/NJZLSF7RmbM1+YOdf1VwAAACEoZN+TAwAAcDaUHAAAYCVKDgAAsBIlBwAAWImSE2Rz585Vt27dFB0drfT0dG3ZsqW5t3Te1q9fr1tuuUVJSUkKCwvTsmXLAuaNMZo+fbo6deqk1q1bKyMjQ3v37g1Yc+jQIY0bN05ut1txcXHKzc3VsWPHfsIUZ1ZYWKif//znatu2reLj4zVy5Ejt2bMnYM3JkyeVl5enDh06qE2bNho1alS9D5zcv3+/srOzFRMTo/j4eE2ZMkWnTp36KaOc0SuvvKJ+/fo5H7rl8Xj0l7/8xZkP9Xyne+qppxQWFqZJkyY5Y6GeccaMGQoLCwv4Sk1NdeZDPV+dv//97/r3f/93dejQQa1bt1bfvn21bds2Zz7Uf99069at3nkMCwtTXl6epNA/jzU1NXrssceUkpKi1q1b6/LLL9esWbMC/pZUiziHBkHzxhtvGJfLZebPn2927dplxo8fb+Li4kxFRUVzb+28rFy50vzqV78yf/rTn4wk89ZbbwXMP/XUUyY2NtYsW7bMfPjhh+YXv/iFSUlJMd9//72zZvjw4aZ///5m06ZN5v333zfdu3c3Y8eO/YmTNCwrK8ssWLDAlJeXm7KyMnPzzTebLl26mGPHjjlrHnjgAZOcnGyKi4vNtm3bzODBg80111zjzJ86dcr06dPHZGRkmO3bt5uVK1eajh07mmnTpjVHpHrefvtts2LFCvPJJ5+YPXv2mEceecRERkaa8vJyY0zo5/uxLVu2mG7dupl+/fqZiRMnOuOhnvHxxx83vXv3NgcOHHC+vvnmG2c+1PMZY8yhQ4dM165dzd133202b95sPv/8c7N69Wrz6aefOmtC/ffNwYMHA86h1+s1ksx7771njAn98/jkk0+aDh06mOXLl5t9+/aZpUuXmjZt2pg5c+Y4a1rCOaTkBNHVV19t8vLynNs1NTUmKSnJFBYWNuOuLszpJae2ttYkJiaaZ555xhk7fPiwiYqKMn/4wx+MMcZ89NFHRpLZunWrs+Yvf/mLCQsLM3//+99/sr2fr4MHDxpJZt26dcaYH/JERkaapUuXOms+/vhjI8mUlJQYY34oguHh4cbn8zlrXnnlFeN2u01VVdVPG+A8tWvXzrz22mtW5Tt69Ki54oorjNfrNf/2b//mlBwbMj7++OOmf//+Dc7ZkM8YYx5++GFz3XXXnXHext83EydONJdffrmpra214jxmZ2ebe++9N2Ds9ttvN+PGjTPGtJxzyMtVQVJdXa3S0lJlZGQ4Y+Hh4crIyFBJSUkz7iw49u3bJ5/PF5AvNjZW6enpTr6SkhLFxcVp0KBBzpqMjAyFh4dr8+bNP/mez+XIkSOSpPbt20uSSktL5ff7AzKmpqaqS5cuARn79u0b8IGTWVlZqqys1K5du37C3Z9bTU2N3njjDR0/flwej8eqfHl5ecrOzg7IItlzDvfu3aukpCRddtllGjdunPbv3y/Jnnxvv/22Bg0apNGjRys+Pl4DBw7U//7v/zrztv2+qa6u1u9//3vde++9CgsLs+I8XnPNNSouLtYnn3wiSfrwww+1YcMGjRgxQlLLOYch/YnHLck//vEP1dTU1Pu05YSEBO3evbuZdhU8Pp9PkhrMVzfn8/kUHx8fMN+qVSu1b9/eWdNS1NbWatKkSbr22mvVp08fST/s3+Vy1fujradnbOgxqJtrCXbu3CmPx6OTJ0+qTZs2euutt9SrVy+VlZVZke+NN97QX//6V23durXenA3nMD09XQsXLlSPHj104MABzZw5U9dff73Ky8utyCdJn3/+uV555RUVFBTokUce0datW/XLX/5SLpdLOTk51v2+WbZsmQ4fPqy7775bkh0/p1OnTlVlZaVSU1MVERGhmpoaPfnkkxo3bpyklvPfDEoOLkp5eXkqLy/Xhg0bmnsrQdejRw+VlZXpyJEj+uMf/6icnBytW7euubcVFF9++aUmTpwor9er6Ojo5t5Ok6j7f8KS1K9fP6Wnp6tr165asmSJWrdu3Yw7C57a2loNGjRI//M//yNJGjhwoMrLyzVv3jzl5OQ08+6C7/XXX9eIESOUlJTU3FsJmiVLlmjRokVavHixevfurbKyMk2aNElJSUkt6hzyclWQdOzYUREREfXeHV9RUaHExMRm2lXw1GU4W77ExEQdPHgwYP7UqVM6dOhQi3oM8vPztXz5cr333nvq3LmzM56YmKjq6modPnw4YP3pGRt6DOrmWgKXy6Xu3bsrLS1NhYWF6t+/v+bMmWNFvtLSUh08eFBXXXWVWrVqpVatWmndunV64YUX1KpVKyUkJIR8xtPFxcXpyiuv1KeffmrFOZSkTp06qVevXgFjPXv2dF6Ws+n3zd/+9je9++67uu+++5wxG87jlClTNHXqVI0ZM0Z9+/bVXXfdpcmTJ6uwsFBSyzmHlJwgcblcSktLU3FxsTNWW1ur4uJieTyeZtxZcKSkpCgxMTEgX2VlpTZv3uzk83g8Onz4sEpLS501a9asUW1trdLT03/yPZ/OGKP8/Hy99dZbWrNmjVJSUgLm09LSFBkZGZBxz5492r9/f0DGnTt3BvwP0+v1yu121/ul3VLU1taqqqrKinxDhw7Vzp07VVZW5nwNGjRI48aNc/4d6hlPd+zYMX322Wfq1KmTFedQkq699tp6H9/wySefqGvXrpLs+H1TZ8GCBYqPj1d2drYzZsN5PHHihMLDAytERESEamtrJbWgcxiUty/DGPPDJeRRUVFm4cKF5qOPPjL333+/iYuLC3h3fEt29OhRs337drN9+3YjycyePdts377d/O1vfzPG/HA5YFxcnPnzn/9sduzYYW699dYGLwccOHCg2bx5s9mwYYO54oorWswlnRMmTDCxsbFm7dq1AZd2njhxwlnzwAMPmC5dupg1a9aYbdu2GY/HYzwejzNfd1lnZmamKSsrM6tWrTKXXnppi7msc+rUqWbdunVm3759ZseOHWbq1KkmLCzMFBUVGWNCP19Dfnx1lTGhn/HBBx80a9euNfv27TMffPCBycjIMB07djQHDx40xoR+PmN+uPy/VatW5sknnzR79+41ixYtMjExMeb3v/+9sybUf98Y88MVtl26dDEPP/xwvblQP485OTnmZz/7mXMJ+Z/+9CfTsWNH89BDDzlrWsI5pOQE2Ysvvmi6dOliXC6Xufrqq82mTZuae0vn7b333jOS6n3l5OQYY364JPCxxx4zCQkJJioqygwdOtTs2bMn4BjffvutGTt2rGnTpo1xu93mnnvuMUePHm2GNPU1lE2SWbBggbPm+++/N//5n/9p2rVrZ2JiYsxtt91mDhw4EHCcL774wowYMcK0bt3adOzY0Tz44IPG7/f/xGkadu+995quXbsal8tlLr30UjN06FCn4BgT+vkacnrJCfWMd9xxh+nUqZNxuVzmZz/7mbnjjjsCPj8m1PPVeeedd0yfPn1MVFSUSU1NNa+++mrAfKj/vjHGmNWrVxtJ9fZtTOifx8rKSjNx4kTTpUsXEx0dbS677DLzq1/9KuDy9pZwDsOM+dHHEwIAAFiC9+QAAAArUXIAAICVKDkAAMBKlBwAAGAlSg4AALASJQcAAFiJkgMAAKxEyQEAAFai5AAAACtRcgAAgJUoOQAAwEqUHAAAYKX/B/w9oVWydBnYAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def process_clustering(name):\n",
    "    clustering = pd.read_csv('clustering/{}_cluster.tsv'.format(name), delimiter='\\t', header=None)\n",
    "    #rename heading as cluster reference and id\n",
    "    clustering.columns = [name, 'Entry']\n",
    "    clustering.drop_duplicates(subset='Entry', keep='first', inplace=True)\n",
    "    print(clustering[name].nunique())\n",
    "    clustering[name].value_counts().hist(bins=100)\n",
    "    return clustering\n",
    "\n",
    "cluster50 = process_clustering('clusterRes50')\n",
    "cluster30 = process_clustering('clusterRes30')\n",
    "\n",
    "df = pd.merge(df, cluster50, on='Entry', how='left')\n",
    "df = pd.merge(df, cluster30, on='Entry', how='left')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "63ca5f5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('../processed_data/protein2EC.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "756c5423",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
